<!DOCTYPE html><html><head>
            <meta content="text/html;charset=utf-8" http-equiv="Content-Type">
            <meta content="utf-8" http-equiv="encoding">
          <title>Bioinformatics
</title><meta name="viewport" content="width=device-width,initial-scale=1"><link href="/.temp_githib_css/github-markdown.css" rel="stylesheet"><link href="/.temp_highlightjs_css/default.css" rel="stylesheet"></head>
          <body class="markdown-body"><div class="toc">
<ul>
<li><a href="#HEADREF0">Introduction</a></li>
<li><a href="#HEADREF1">K-mers</a></li>
<ul>
<li><a href="#HEADREF2">K-mer Reverse Complement</a></li>
<li><a href="#HEADREF3">Find Where a K-mer Occurs</a></li>
<li><a href="#HEADREF4">Find Where a K-mer Clusters</a></li>
<li><a href="#HEADREF5">Count a Sequence's K-mers</a></li>
<li><a href="#HEADREF6">Find a Sequence's K-mer Clusters</a></li>
</ul>
</ul>
</div>
<a name="HEADREF0"></a><h1>Introduction</h1>
<a name="HEADREF1"></a><h1>K-mers</h1>
<p>A <a name="BOOKMARK0"></a><strong>k-mer</strong> is a subsequence of length k within some larger biological sequence (e.g. DNA or amino acid chain). For example, in the DNA sequence <code>GAAATC</code>, the following <a href="#BOOKMARK0">k-mer</a>'s exist:</p>
<table>
<thead>
<tr>
<th>k</th>
<th><a href="#BOOKMARK0">k-mer</a>s</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>G A A A T C</td>
</tr>
<tr>
<td>2</td>
<td>GA AA AA AT TC</td>
</tr>
<tr>
<td>3</td>
<td>GAA AAA AAT ATC</td>
</tr>
<tr>
<td>4</td>
<td>GAAA AAAT AATC</td>
</tr>
<tr>
<td>5</td>
<td>GAAAT AAATC</td>
</tr>
<tr>
<td>6</td>
<td>GAAATC</td>
</tr>
</tbody>
</table>
<p>Often times we'll need to either...</p>
<ul>
<li>search for an exact <a href="#BOOKMARK0">k-mer</a>.</li>
<li>search for an approximate <a href="#BOOKMARK0">k-mer</a> (fuzzy search).</li>
<li>find <a href="#BOOKMARK0">k-mer</a>s of interest in a sequence (e.g. repeating <a href="#BOOKMARK0">k-mer</a>s).</li>
</ul>
<p>A common use-case for <a href="#BOOKMARK0">k-mer</a> search is to find the <a name="BOOKMARK1"></a><strong>replication origin</strong> (<a name="BOOKMARK2"></a><strong>ori</strong>) of a prokaryotic organism's DNA. The <a href="#BOOKMARK1">replication origin</a> is a region within a prokaryotic cell's circular DNA where DNA replication starts. Within the <a href="#BOOKMARK2">ori</a> region, several smaller regions exist that contain DNA sequences known as <a name="BOOKMARK3"></a><strong>DnaA box</strong>es. These <a href="#BOOKMARK3">DnaA box</a>es are sequences that enzymes responsible for DNA strand replication (<a name="BOOKMARK4"></a><strong>DNA polymerase</strong>s) bind to to begin the process of replication.</p>
<p><img src="dot_d6657feb56cf0608d87e1b67b0cfae2c.svg" alt="Dot diagram"></p>
<p>Typically multiple <a href="#BOOKMARK3">DnaA box</a>es exist in the <a href="#BOOKMARK2">ori</a>.</p>
<p><img src="svgbob_2e987bde02834003f50e0411181e8f8398712af3.svg" alt="Kroki diagram output"></p>
<p>The <a href="#BOOKMARK3">DnaA box</a>es in an <a href="#BOOKMARK2">ori</a> don't have to be exactly the same. The enzyme will still bind to them if they're slightly different.</p>
<a name="HEADREF2"></a><h2>K-mer Reverse Complement</h2>
<p>Given a DNA <a href="#BOOKMARK0">k-mer</a>, calculate its reverse complement.</p>
<pre class="hljs"><code><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">reverse_complement</span><span class="hljs-params">(strand: str)</span>:</span>
    ret = <span class="hljs-string">''</span>
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(<span class="hljs-number">0</span>, len(strand)):
        base = strand[i]
        <span class="hljs-keyword">if</span> base == <span class="hljs-string">'A'</span> <span class="hljs-keyword">or</span> base == <span class="hljs-string">'a'</span>:
            base = <span class="hljs-string">'T'</span>
        <span class="hljs-keyword">elif</span> base == <span class="hljs-string">'T'</span> <span class="hljs-keyword">or</span> base == <span class="hljs-string">'t'</span>:
            base = <span class="hljs-string">'A'</span>
        <span class="hljs-keyword">elif</span> base == <span class="hljs-string">'C'</span> <span class="hljs-keyword">or</span> base == <span class="hljs-string">'c'</span>:
            base = <span class="hljs-string">'G'</span>
        <span class="hljs-keyword">elif</span> base == <span class="hljs-string">'G'</span> <span class="hljs-keyword">or</span> base == <span class="hljs-string">'g'</span>:
            base = <span class="hljs-string">'C'</span>
        <span class="hljs-keyword">else</span>:
            <span class="hljs-keyword">raise</span> Exception(<span class="hljs-string">'Unexpected base: '</span> + base)

        ret += base
    <span class="hljs-keyword">return</span> ret[::<span class="hljs-number">-1</span>]</code></pre><div style="border:1px solid black;">
<p></p>
<p>Original: TAATCCG</p>
<p>Reverse Complement: CGGATTA</p>
</div>
<p></p>
<p>Depending on the type of biological sequence, a <a href="#BOOKMARK0">k-mer</a> may have one or more alternatives. For DNA sequences specifically, a <a href="#BOOKMARK0">k-mer</a> of interest may have an alternate form. Since DNA sequences come in 2 strands, where ...</p>
<ul>
<li>each strand's direction is opposite of the other,</li>
<li>each strand position has a nucleotide that complements the nucleotide at that same position on the other stand:
<ul>
<li>A ⟷ T</li>
<li>C ⟷ G</li>
</ul>
</li>
</ul>
<p><img src="svgbob_c44b72fd8e03cfe92b63a0a3b83bfe37c5551666.svg" alt="Kroki diagram output"></p>
<p>, ... the reverse complement of that <a href="#BOOKMARK0">k-mer</a> may be just as valid as the original <a href="#BOOKMARK0">k-mer</a>. For example, if an enzyme is known to bind to a specific DNA <a href="#BOOKMARK0">k-mer</a>, it's possible that it might also bind to the reverse complement of that <a href="#BOOKMARK0">k-mer</a>.</p>
<a name="HEADREF3"></a><h2>Find Where a K-mer Occurs</h2>
<p>Given a <a href="#BOOKMARK0">k-mer</a>, find where that <a href="#BOOKMARK0">k-mer</a> occurs in some larger sequence.</p>
<pre class="hljs"><code><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">find_kmer_locations</span><span class="hljs-params">(sequence: str, kmer: str)</span> -&gt; List[int]:</span>
    k = len(kmer)
    idxes = []
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(<span class="hljs-number">0</span>, len(sequence) - k):
        <span class="hljs-keyword">if</span> sequence[i:i + k] == kmer:
            idxes.append(i)
    <span class="hljs-keyword">return</span> idxes</code></pre><div style="border:1px solid black;">
<p></p>
<p>Found AC in ACTGAACCTTACACTTAAAGGAGATGATGATTCAAAT at index [0, 5, 10, 12]</p>
</div>
<p></p>
<p>Imagine that you know of a specific <a href="#BOOKMARK0">k-mer</a> pattern that serves some function in an organism. If you see that same <a href="#BOOKMARK0">k-mer</a> pattern appearing in some other related organism, it could be a sign that that <a href="#BOOKMARK0">k-mer</a> pattern serves a similar function. For example, the same <a href="#BOOKMARK0">k-mer</a> pattern could be used by 2 related types of bacteria as a <a href="#BOOKMARK3">DnaA box</a>.</p>
<a name="HEADREF4"></a><h2>Find Where a K-mer Clusters</h2>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li><a href="#BOOKMARK0">K-mer</a> Location</li>
</ul>
</div>
<p>Given a <a href="#BOOKMARK0">k-mer</a>, find where that <a href="#BOOKMARK0">k-mer</a> clusters in some larger sequence.</p>
<pre class="hljs"><code><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">find_kmer_clusters</span><span class="hljs-params">(sequence: str, kmer: str, min_occurrence_in_cluster: int, cluster_window_size: int)</span> -&gt; List[int]:</span>
    cluster_locs = []

    locs = find_kmer_locations(sequence, kmer)
    start_i = <span class="hljs-number">0</span>
    occurrence_count = <span class="hljs-number">1</span>
    <span class="hljs-keyword">for</span> end_i <span class="hljs-keyword">in</span> range(<span class="hljs-number">1</span>, len(locs)):
        <span class="hljs-keyword">if</span> locs[end_i] - locs[start_i] &lt; cluster_window_size:  <span class="hljs-comment"># within a cluster window?</span>
            occurrence_count += <span class="hljs-number">1</span>
        <span class="hljs-keyword">else</span>:
            <span class="hljs-keyword">if</span> occurrence_count &gt;= min_occurrence_in_cluster:  <span class="hljs-comment"># did the last cluster meet the min ocurr requirement?</span>
                cluster_locs.append(locs[start_i])
            start_i = end_i
            occurrence_count = <span class="hljs-number">1</span>

    <span class="hljs-keyword">return</span> cluster_locs</code></pre><div style="border:1px solid black;">
<p></p>
<p>Found clusters of GGG (at least 3 occurrences in window of 13) in GGGACTGAACAAACAAATTTGGGAGGGCACGGGTTAAAGGAGATGATGATTCAAAGGGT at index [20]</p>
</div>
<p></p>
<p>An enzyme may need to bind to a specific region of DNA to begin doing its job. That is, it looks for a specific <a href="#BOOKMARK0">k-mer</a> pattern to bind to, where that <a href="#BOOKMARK0">k-mer</a> represents the beginning of some larger DNA region that it operates on. Since DNA is known to mutate, often times you'll find multiple copies of the same <a href="#BOOKMARK0">k-mer</a> pattern clustered together -- if one copy mutated to become unusable, the other copies are still around.</p>
<p>For example, the <a href="#BOOKMARK3">DnaA box</a> in bacteria can be found repeating multiple times in the <a href="#BOOKMARK2">ori</a> region.</p>
<a name="HEADREF5"></a><h2>Count a Sequence's K-mers</h2>
<p>Given a sequence, count how many times each unique <a href="#BOOKMARK0">k-mer</a> in that sequence occurs.</p>
<pre class="hljs"><code><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">kmer_frequency</span><span class="hljs-params">(data: str, k: int)</span> -&gt; Counter[str]:</span>
    counter = Counter()
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(<span class="hljs-number">0</span>, len(data) - k):
        pattern = data[i:i+k]
        counter[pattern] += <span class="hljs-number">1</span>
    <span class="hljs-keyword">return</span> counter</code></pre><div style="border:1px solid black;">
<p></p>
<p>4-mer frequencies for AAAACAAAAAGAAAAAAT:</p>
<ul>
<li>AAAA = 6</li>
<li>AAAC = 1</li>
<li>AACA = 1</li>
<li>ACAA = 1</li>
<li>CAAA = 1</li>
<li>AAAG = 1</li>
<li>AAGA = 1</li>
<li>AGAA = 1</li>
<li>GAAA = 1</li>
</ul>
</div>
<p></p>
<p>From past experiments, you know that a specific region of genome clusters a certain pattern. The pattern is different for each organism, but you know that it's there.</p>
<a name="HEADREF6"></a><h2>Find a Sequence's K-mer Clusters</h2>
<div style="margin:2em; background-color: #e0e0e0;">
<p><strong>↩PREREQUISITES↩</strong></p>
<ul>
<li>Find Where a <a href="#BOOKMARK0">K-mer</a> Clusters</li>
</ul>
</div>
<p>Given a sequence, find clusters of unique <a href="#BOOKMARK0">k-mer</a>s within that sequence. In other words, for each unique <a href="#BOOKMARK0">k-mer</a> that exists in the sequence, see if it clusters in the sequence.</p>
<p>An enzyme may need to bind to a specific region of DNA to begin doing its job. That is, it looks for a specific <a href="#BOOKMARK0">k-mer</a> pattern to bind to, where that <a href="#BOOKMARK0">k-mer</a> represents the beginning of some larger DNA region that it operates on. Since DNA is known to mutate, often times you'll find multiple copies of the same <a href="#BOOKMARK0">k-mer</a> pattern clustered together -- if one copy mutated to become unusable, the other copies are still around.</p>
<p>For example, the <a href="#BOOKMARK3">DnaA box</a> in bacteria can be found repeating multiple times in the <a href="#BOOKMARK2">ori</a> region. If you don't know where the <a href="#BOOKMARK2">ori</a> is, searching for clusters can give a list of potential locations.</p>

        </body></html>