<!DOCTYPE html><html><head>
            <meta content="text/html;charset=utf-8" http-equiv="Content-Type">
            <meta content="utf-8" http-equiv="encoding">
          <title>Bioinformatics
</title><meta name="viewport" content="width=device-width,initial-scale=1"><link href="/.temp_githib_css/github-markdown.css" rel="stylesheet"><link href="/.temp_highlightjs_css/default.css" rel="stylesheet"></head>
          <body class="markdown-body"><div class="toc">
<ul>
<li><a href="#HEADREF0">Introduction</a></li>
<li><a href="#HEADREF1">K-mers</a></li>
<ul>
<li><a href="#HEADREF2">Count K-mers</a></li>
</ul>
</ul>
</div>
<a name="HEADREF0"></a><h1>Introduction</h1>
<a name="HEADREF1"></a><h1>K-mers</h1>
<p>A <a name="BOOKMARK0"></a><strong>k-mer</strong> is a substring of length k within some larger biological string (e.g. DNA or amino acid chain). For example, in the DNA sequence <code>GAAATC</code>, the following <a href="#BOOKMARK0">k-mer</a>'s exist:</p>
<table>
<thead>
<tr>
<th>k</th>
<th><a href="#BOOKMARK0">k-mer</a>s</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>G A A A T C</td>
</tr>
<tr>
<td>2</td>
<td>GA AA AA AT TC</td>
</tr>
<tr>
<td>3</td>
<td>GAA AAA AAT ATC</td>
</tr>
<tr>
<td>4</td>
<td>GAAA AAAT AATC</td>
</tr>
<tr>
<td>5</td>
<td>GAAAT AAATC</td>
</tr>
<tr>
<td>6</td>
<td>GAAATC</td>
</tr>
</tbody>
</table>
<p>Often times we'll need to either...</p>
<ul>
<li>search for an exact <a href="#BOOKMARK0">k-mer</a>.</li>
<li>search for an approximate <a href="#BOOKMARK0">k-mer</a> (fuzzy search).</li>
<li>find <a href="#BOOKMARK0">k-mer</a>s of interest in a sequence (e.g. repeating <a href="#BOOKMARK0">k-mer</a>s).</li>
</ul>
<p>A common use-case for <a href="#BOOKMARK0">k-mer</a> search is to find the <a name="BOOKMARK1"></a><strong>replication origin</strong> (<a name="BOOKMARK2"></a><strong>ori</strong>) of a prokaryotic organism's DNA. The <a href="#BOOKMARK1">replication origin</a> is a region within a prokaryotic cell's circular DNA where DNA replication starts. Within the <a href="#BOOKMARK2">ori</a> region, several smaller regions exist that contain DNA sequences known as <a name="BOOKMARK3"></a><strong>DnaA box</strong>es. These <a href="#BOOKMARK3">DnaA box</a>es are sequences that enzymes responsible for DNA strand replication (<a name="BOOKMARK4"></a><strong>DNA polymerase</strong>s) bind to to begin the process of replication.</p>
<pre><code class="language-{dot}">digraph {
  Ori -&gt; DnaA_box [taillabel="1..*", arrowhead=none]
}
</code></pre>
<p>Typically multiple <a href="#BOOKMARK3">DnaA box</a>es exist in the <a href="#BOOKMARK2">ori</a>.</p>
<pre><code>                     DnaA boxes within the ori

+--------------------------------------------------------------------+
|   :   :             :   :                            :   :         |
+---+-+-+-------------+-+-+----------------------------+-+-+---------+
      |                 |                                |           
  DnaA box          DnaA box                         DnaA box        
    
</code></pre>
<p>The <a href="#BOOKMARK3">DnaA box</a>es in an <a href="#BOOKMARK2">ori</a> don't have to be exactly the same. The enzyme will still bind to them if they're slightly different.</p>
<a name="HEADREF2"></a><h2>Count K-mers</h2>
<p>Counting <a href="#BOOKMARK0">k-mer</a>s is essentially just sliding a window of size k over a string and seeing if the contents in that window match some <a href="#BOOKMARK0">k-mer</a>.</p>
<p>[0, 5, 10, 12]</p>

        </body></html>