```{title}
Biology
```

```{toc}
```

# Chemistry

`{bm} Chemistry/(Chemistry|Chemical|Chemist)/i` is defined as the study of composition, structure, and properties of matter. `{bm} Matter` is defined as anything that takes up space and has a mass (e.g. hair, water, table, etc..).

```{note}
Primary source for these notes is Khan Academy's Chemistry course, with additional information from Wikipedia and other sources.
```

## Atom

An `{bm} element` is matter that cannot be broken down any further by chemical reaction -- it's a substance made entirely out of one type of `{bm} atom`. Each element/atom has a specific set of properties that defines how it acts/reacts (e.g. weight, colour, how light reflects, etc..).

```{dot}
graph {
  rankdir="LR"
  element -- atom [headlabel="1", taillabel="1..*"];
}
```

```{note}
The term element is often used to refer to the type of atom. For example, "water is made up of the elements hydrogen and oxygen."
```

Examples of matter made up of a single type of atom:
 * Gold - Yellow coloured, reflective, malleable, and corrosion-resistant.
 * Lead - Blue/white coloured, relatively soft, malleable.
 * Carbon - Silver/gray coloured, reflective, easily oxidizes/corrodes.

```{note}
As stated above, matter made up of a single type of atom is called an element.
```

Examples of matter made of multiple types of atoms:
 * Salt water - Mixture of sodium and water.
 * Water - Water can be further broken down to the elements hydrogen and oxygen.
 * Glucose - Glucose can be further broken down to the elements carbon, hydrogen, and oxygen.

![By AG Caesar - Own work, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=67059219](240px-Atom_Diagram.svg.png)

The building blocks of atoms are referred to as `{bm} sub-atomic particle`s: protons, neutrons, and electrons. Protons and neutrons form the nucleus_atom of the atom while electrons jump around in orbit of the`{bm} nucleus/\b(nucleus|nuclei)_atom\b/i`. The configuration of an atom (number of protons/neutrons/electrons) is what defines its type and predicts how it reacts to other atoms.

For each atom, the number of ...

 * `{bm} proton`s defines the type of atom.
   * if a.protons = 1, hydrogen.
   * if a.protons = 2, helium.
   * if a.protons = 3, lithium.
   * ...
 * `{bm} neutron`s defines the extra mass an atom has.
   * if a.protons > b.protons && a.neutrons > b.neutrons, mass(a) > mass(b).
   * if a.protons < b.protons && a.neutrons > b.neutrons, mass(a) < mass(b).
 * `{bm} electron`s defines the attraction to other atoms.
   * if protons == electrons, neutral charge.
   * if protons > electrons, positively charged ion (looking to add electrons until ==).
   * if protons < electrons, negatively charged ion (looking to remove electrons until ==).

Protons are positively charged while electrons are negatively charged. Oppositely charged particles will attract while like charges repel. As such, ...
 * electrons will be attracted to protons (and vice-versa).
 * electrons will be repelled from other electrons.
 * protons will be repelled from other protons.

Neutrons may not have a charge, but they play a roll in spacing out proton atoms such that their repulsion from each other isn't strong enough to break up the nucleus_ATOM.

```{note}
If protons != electrons, the atom is referred to an ion.
```

Terminology around atoms and sub-atomic particles:

 * `{bm} atomic number` (denoted as Z) - number of protons (type) that make up an atom.
 * `{bm} mass number` (denoted as A) - number of protons and neutrons that make up an atom.
 * ion - An atom where the number of protons and electrons are different.
 * `{bm} isotope` - If two atoms have the same number of protons but a different number of neutrons, they're said to be different isotopes of the same type of atom. For example, carbon has multiple different isotopes, the most common of which has 6 neutrons (6 protons + 6 neutrons).

Atoms may form bonds with one another to form larger structures known as molecules and ionic compounds.

The periodic table below orders atoms by the atomic number (number of protons)...

![By Offnfopt - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=62296883](800px-Simple_Periodic_Table_Chart-en.svg.png)

### Electron Shell

`{bm} Electron shells` are distinct orbits/regions around the nucleus_ATOM that electrons are assigned to. There are 7 electron shells in total and each shell can hold onto a certain number of electrons. Ordered from closest to the nucleus_ATOM to farthest from from the nucleus_ATOM, these shells are...

| # | Label | Max Electrons        |
| - | ----- | -------------------- |
| 1 | K     | `{kt} 2(1^2) = 2`    |
| 2 | L     | `{kt} 2(2^2) = 8`    |
| 3 | M     | `{kt} 2(3^2) = 18`   |
| 4 | N     | `{kt} 2(4^2) = 32`   |
| 5 | O     | `{kt} 2(5^2) = 50`   |
| 6 | P     | `{kt} 2(6^2) = 72`   |
| 7 | Q     | `{kt} 2(7^2) = 98`   |

Each electron shell is made up of one or more `{bm} electron subshell`s. These electron subshells are...

| # | Label | Max electrons | Owning electron shells |
| - | ----- | ------------- | ---------------------- |
| 0 | s     | 2             | 1,2,3,4,5,6,7 (all)    |
| 1 | p     | 6             | 2,3,4,5,6,7            |
| 2 | d     | 10            | 3,4,5,6,7              |
| 3 | f     | 14            | 4,5,6,7                |
| 4 | g     | 18            | 5,6,7 ???              |


```{note}
Wikipedia says that electron subshells past f or g are theoretical? Unsure exactly what this means.
```

The outer most electron shell is called the `{bm} valence shell` and electrons within it are called `{bm} valence electron`s. Since these electrons travel the farthest from the nucleus_ATOM, the nucleus_ATOM's pull on them is the weakest. As such, valence electrons are the electrons passed between atoms and / or used in the formation of bonds.

2 atoms that have formed a bond by sharing some of their valence electrons are said to have a covalent bond.

```{note}
It's only called a valence shell if the atom isn't charged. If the atom is an ion, you can't call the outermost shell a valence shell.
```

### Notation

Atoms can be written down in 2 different forms: hyphen notation and isotope notation.


----


`{bm} Hyphen notation` is a way of describing a type of atom. It's defined as...

```
name-A
```

where ...
 * name is the name of the atom.
 * A is the mass number (number of protons and neutrons).

For example, carbon-13 is the hyphen notation for a carbon atom that's mass number is 13. Since we know that carbon has the atomic number 6 (protons=6) and the mass number of 13 (protons+neutrons=13), we know that the number of neutrons is 7 because 13-6=7.


----


`{bm} Isotope notation/(isotope notation|nuclear notation)/i` is a shorthand way of describing an atom. It's defined as...

```{kt}
_{Z}^{A}X^{C}
```

where ...
 * X is the atomic symbol (e.g. H = hydrogen, Ag = gold, Fe = iron).
 * A is the mass number (number of protons and neutrons).
 * Z is the atomic number (number of protons).
 * C is the charge.

```{kt}
_{\text{protons / atomic num}}^{\text{protons+neutrons / mass num}}X^{\text{charge / electron offset req for neutral charge}}
```

```{note}
For C, the notation is to write the sign after the digits. For example, ...
 * 3+ means that it has a positive charge of 3 (3 more protons than electrons).
 * 4- means that it has a negative charge of 4 (4 more electrons than protons).
 * 0 means that i has a neutral charge (equal number of protons and electrons).
```

For example, `{kt} _{7}^{13}N^{2-}` is an atom of nitrogen with a count of 7 protons, 9 electrons, and 6 neutrons. Since we know that...
 * the atomic number is 7 (protons=7), we know that the number of electrons is 9 because a charge of 2- means that it has 2 more electrons than protons (7+2=9).
 * the atomic number is 7 (protons=7) and the mass number is 13 (protons+neutrons=13), we know that the number of neutrons is 6 because 13-7=6.

In certain cases, some parts of the notation may be left off:
  1. Z is almost always left out (e.g. `{kt} _{7}^{13}N^{2-}` ⟶ `{kt} ^{13}N^{2-}`).
  
     Z is the number of protons (atomic number). Since the number of protons defines the type of atom, including it is redundant. The symbol already tells us what the type of atom is, which tells us what the number of protons is.

  1. A C of 0 is left out entirely (e.g. `{kt} ^{13}N^{0}` ⟶ `{kt} ^{13}N`).
  1. A C of -1 won't include the digit: (e.g. `{kt} ^{13}N^{1-}` ⟶ `{kt} ^{13}N^{-}`).
  1. A C of +1 won't include the digit: (e.g. `{kt} ^{13}N^{1+}` ⟶ `{kt} ^{13}N^{+}`).

### Atomic Mass

```{note}
The notion of atomic weight in chemistry is different from weight in physics -- read below.
```

The `{bm} atomic mass` of an atom is measured in `{bm} atomic mass unit`s (`{bm} amu/\b(amu|amus)\b/i`), where 1 amu is: `{kt} 1.66054 \cdot 10^{-24}` grams. Note that 1 amu is roughly equal to the mass of a single proton / neutron:

* 1 proton = `{kt} 1.6726231 \cdot 10^{-24}` gram
* 1 neutron = `{kt} 1.674920 \cdot 10^{-24}` gram
* 1 electron = `{kt} 9.1093897 \cdot 10^{-28}` gram

That's because the constant for 1 amu was derived by measuring the mass of carbon-12 and dividing it by 12. Carbon-12 has exactly 6 protons and 6 neutrons (12 total). Since protons and neutrons have roughly the same mass (and electrons have almost no mass), when carbon-12 is divided by 12 it roughly gives the mass of a single proton / neutron.

```{note}
Carbon-12 was chosen because it's the most common carbon isotope on earth (99% abundance).
```

The `{bm} atomic weight` / `{bm} relative atomic mass` of an atom is the weighted average of atomic mass across all its isotopes, where each isotope is weighted by how abundant it is. For example, there are 3 isotopes for the some imaginary atom type...
 * Imaginary-12 - 3 amu and 10% abundance
 * Imaginary-13 - 4 amu and 30% abundance
 * Imaginary-14 - 5 amu and 60% abundance

Its atomic weight can be calculated as `{kt} 3 \cdot 0.1 + 4 \cdot 0.3 + 5 \cdot 0.6 = 4.6` amu.

```{note}
How are abundances calculated? A chunk of the naturally occurring element is dug up and passed through a machine called a Gas Chromatograph-Mass Spectrometer. The machine isolates by weight and shows how abundant each weight is.
```

### Mole

A `{bm} mole` is the unit used to measure the number/count of the particular particle making up a larger substance (e.g. proton, atom, molecule, etc..). 1 mole is `{kt} 6.02214076 \cdot 10^{23}` particles. For example, ...
 * 1 mole of carbon-12 = `{kt} 6.02214076 \cdot 10^{23}` carbon-12 atoms = 12 grams of carbon-12 
 * 1 mole of iron-56 = `{kt} 6.02214076 \cdot 10^{23}` iron-56 atoms = 56 grams of iron-56.
 * 1 mole of gold-107 = `{kt} 6.02214076 \cdot 10^{23}` gold-107 atoms = 107 grams of gold-107.

The constant for mole is also know as the `{bm} Avogadro number/(Avogadro number|Avogadro's number)/i`. It's the number of carbon-12 atoms required to reach a mass of 12 grams.

````{note}
Notice that for each type of atom, the mass of 1 mole of name-x is equal to x grams.

Recall that...
 * both amu and grams are used to measure mass.
 * the mass of 1 proton or neutron = 1 amu.
 * the mass of 1 atom = the mass of that atoms protons and neutrons.

So if we know that an atom of gold has 107 protons and neutrons, we know that its mass is 107 amu (and vice versa).

```{img}
Atomic Mass.svg
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
Diagram visually showing the number of protons/neutrons mapping to number of amus.
```

Since we know carbon-12 has a mass of 12 amu (contains 12 protons/neutrons)...
 * 12 amu * `{kt} 6.02214076 \cdot 10^{23}` = 12 grams.

As such, name-x has a mass of x amu (contains x protons/neutrons)...
 * x amu * `{kt} 6.02214076 \cdot 10^{23}` = x grams.

For example...
 * hydrogen-1 --> 1 amu * `{kt} 6.02214076 \cdot 10^{23}` = 1 grams.
 * carbon-12 --> 12 amu * `{kt} 6.02214076 \cdot 10^{23}` = 12 grams.
 * silicon-28 --> 28 amu * `{kt} 6.02214076 \cdot 10^{23}` = 28 grams.
````

## Molecule

When atoms bind together by sharing electrons, they form a `{bm} molecule/(molecule|molecular)/i`. Each type of molecule has the same configuration of atoms -- same atoms in the same numbers, structured/shaped similarly. For example, a water molecule is made up of 2 hydrogen atoms and 1 oxygen atom binding together in a house-roof shape...

![By Dan Craggs - Own workThis vector image includes elements that have been taken or adapted from this:  Water-2D-labelled.png., Public Domain, https://commons.wikimedia.org/w/index.php?curid=7916072](200px-H2O_2D_labelled.svg.png)

```{note}
As far as I can tell, the atoms will always bind in the same way. You can't ever have a molecule that has the same types of atoms in the same numbers but with a different structure.
```

The type of bond that holds together atoms in a molecule is called `{bm} covalent bond` / `{bm} molecular bond`. In a covalent bond, 2 atoms share some or all of their electrons. The simplest example of this is hydrogen gas: 2 hydrogen atoms that are stuck together because each hydrogen atom is attracted to the other's electron.

```{img}
Covalent_bond_hydrogen.svg
By Jacek FH - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=2781099
Covalent Bond Hydrogen

scale 0.5 0.5
```

If only ...
 * 1 pair of electrons are shared (2 electrons total), it's called a `{bm} single covalent bond`.
 * 2 pairs of electrons are shared (4 electrons total), it's called a `{bm} double covalent bond`.
 * 3 pairs of electrons are shared (6 electrons total), it's called a `{bm} triple covalent bond`. 

If a molecule is made up of different types of atoms, it's called a `{bm} molecular compound`. For example...
 * water is a compound because it's comprised of an oxygen atom bonded with 2 hydrogen atoms.
 * hydrogen gas isn't a compound because it's comprised of 2 hydrogen atoms bonded together.

### Polymers

A `{bm} monomer` is a special designation for atoms/molecules that are able to join with other monomers to create even larger molecules. The process of joining is called `{bm} polymerization/(polymerization|polymerize)/i` and the resulting molecule is called a `{bm} polymer`.

```{dot}
digraph {
  rankdir=LR;
  "monomer"->"polymer" [taillabel="1..*", headlabel="1", arrowhead=none];
}
```

If the monomers that make up a polymer are all the same, the polymer is called a `{bm} homopolymer`. Otherwise, it's called a `{bm} heteropolymer` / `{bm} copolymer`.

For example, the glucose molecule is a monomer. It can combine with other glucose molecules to create the glycogen molecule, which is a polymer / homopolymer. Other examples of polymers (according to Wikipedia): amino acids and nucleotides (DNA).

```{note}
There are probably special properties to monomers that allow them to chain up. The Wikipedia page talks about a feature of monomers being a "carbon double bond" which is what allows them to form polymers.
```

Polymers are often referred to as `{bm} macromolecule`s -- molecules that have a very large number of atoms.

## Ion

An `{bm} ion/\b(ions|ion)\b/i` is a charged atom or molecule. A charged atom/molecule just means that it has an unequal number of protons and electrons:
* if protons > electrons (more protons), it's called a `{bm} positively charged/(positive charge|positive-charge|positively-charged|positively charged)/i` ion / `{bm} positive ion` / `{bm} cation`.
* if protons < electrons (more electrons), it's called a `{bm} negatively charged/(negative charge|negative-charge|negatively-charged|negatively charged)/i` ion / `{bm} negative ion` / `{bm} anion`.
* if protons == electrons, it's called a `{bm} neutral charge/(neutral charge|neutral atom)/i` (no charge).

When 2 atoms/molecules are said to be `{bm} oppositely charged/(opposite charge|opposite-charge|oppositely-charged|oppositely charged)/i`s, it means that one of them is positively charged while the other is negatively charged.

Ions are always trying to lose their charge and become neutral, either by giving up an electrons or pulling in an electrons such that the number of protons and electrons become equal. As such, ions will attract towards oppositely charged ions and repel from similarly charged ions:
* negative ions are attracted to positive ions and repelled from negative ions.
* positive ions are attracted to negative ions and repelled from positive ions.

## Ionic Compound

In an `{bm} ionic compound`, 2 oppositely charged ions create a bond by virtue of their attraction to each other. That is, the ions bond together because the other has an opposite charge, but neither ion steals/gives electrons to the other. This form of bond is called a `{bm} ionic bond`.

```{note}
The ions aren't of the same type. If they were, it wouldn't be called a compound.
```

The process for how an ionic bond forms is as follows ...

 1. 2 neutral atoms come across each other.
 1. one of the atoms takes an electron from the other,

The 2 atoms are now oppositely charged ions and as such are attracted to each other. No further electron swaps happen between them.

Since the atoms that form an ionic compound are still charged, other ions may still attract to them and vice versa. That is, an atom that's in an ionic bond is still charged. As such, other charged atoms can attract to the atoms in an ionic bond to form their own ionic bonds, there by leading to a larger structure.

For example, sodium and chlorine bind together via an ionic bond. However, even though they've bonded, other chlorine atoms atoms will attract the bonded sodium and other sodium atoms will attract to the bonded chlorine, forming chains...

```{img}
Blausen_0660_NaCl.png
Sodium atoms and chlorine atoms in ionic bonds with each other (forming a crystal lattice structure). 
By BruceBlaus - Own work, CC BY 3.0, https://commons.wikimedia.org/w/index.php?curid=33041249
scale 0.5 0.5
```

Structures like the one in the example above are called `{bm} crystal lattice`s. The smallest possible version of the above structure is referred to as a `{bm} formula unit`s. In the example above...
 * chlorine and sodium ions are interleaved in ionic bonds to form the larger crystal lattice structure.
 * the formula unit is a single sodium ion that's in an ionic bond with a chlorine ion.

```{note}
The structure in the example above is table salt.
```

## Chemical Formula

`{bm} Chemical formula`s are a way of representing molecules as text. They come in 2 forms
 * molecular formula.
 * condensed formula.

```{note}
When the chemical formula is given as text, it's almost always as a molecular formula.
```

A `{bm} molecular formula` is a string of atomic symbols where each symbol is followed by a subscript of that atom's occurrence count in the molecule. For example, a molecule of hydrazine (2 nitrogen and 4 hydrogen): `{kt} N_2H_4`.

An `{bm} empirical formula` is similar to a molecular formula, except that the subscript represents ratios of elements. For example, a molecule of hydrazine (2 nitrogen and 4 hydrogen): `{kt} NH_2`.

Both molecular formulas and empirical formulas may...
 * avoid displaying the subscript if it's 1.
   
   e.g. `{kt} H_2O` is `{kt} H_2O_1`

 * use distribution or repetition to better describe the physical structure (sometimes called the `{bm} condensed formula/(condensed formula|condensed chemical formula|condensed molecular formula|condensed empirical formula)/i`).
 
   e.g. `{kt} Al_2(SO_4)_3` is `{kt} Al_2S_3O_{12}` (distribution)

   e.g. `{kt} CH_3CH_3` is `{kt} C_2H_6` (repetition)

 * include charges.
 
   e.g. `{kt} C_6H_{11}O_6^-` needs 1 more H to be neutral
  
 * include phases.

   e.g. `{kt} 2H_2 (g) + O_2 (g) \rightarrow 2H_2O_2 (l)`, where g = gas / l = liquid / s = solid / aq = in aqueous solution

Examples...

| Name      | Molecular Formula | Empirical Formula |
| --------- | ----------------- | ----------------- |
| Water     | `{kt} H_2O`       | `{kt} H_2O`       |
| Hydrazine | `{kt} N_2H_4`     | `{kt} NH_2`       |
| Butane    | `{kt} C_4H_{10}`  | `{kt} C_2H_5`     |
| Octane    | `{kt} C_8H_{18}`  | `{kt} C_4H_9`     |

```{note}
In this basic definition of chemical formula notation, there are no set rules as to which atom type is written first. However, things quickly get more complex: see See https://chemistry.stackexchange.com/q/1239.
```

## Structural Formula

`{bm} structural formula`s are a way of representing molecules as diagrams that somewhat expose their 3D structure. In it's simplest form, it consists of atomic symbols with straight lines between them that represent bonds. For example, `{kt} NH_3` can be diagramed as...

```{chemfig}
\chemfig[][scale=2]{N(-[5]H)(-[6]H)(-[7]H)}
```

Other information such as depth can be conveyed by using a `{bm} stereo structural formula`. For example, the same `{kt} NH_3` molecule above can be diagramed as...

```{chemfig}
% Hack to get straight dashed lines working: https://tex.stackexchange.com/a/364985
\setcrambond*{1ex}{1pt}{1.5pt}
\chemfig[][scale=2]{\lewis{2:,N}(-[5]H)(<[6]H)(>:[7]H)}
```

, where the...
 * 2 dots on top of N indicate a lone electron pair.
 * straight edge indicates that the 2 atoms connected are on the same plane as the screen.
 * dash edge indicates that the atom connected is BEHIND the screen.
 * wedge edge indicates that the atom connected is IN FRONT of the screen.

```{note}
In structural formulas, edges are only drawn when if bonds are covalent bonds. If there isn't a covalent bond, there is no line drawn.
```

A better way to represent the structure for a molecule may be to use actual 3D renderings. For example, the same molecule above (`{kt} NH_3`) can be rendered either as a...

 * `{bm} stick-and-ball model/(stick-and-ball model|stick-and-ball diagram|ball-and-stick model|ball-and-stick diagram)/i`

   ```{img}
   Ammonia-3D-balls-A.png
   Stick-and-ball model of ammonia
   By Ben Mills - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=3958453
   scale 0.15 0.15
   ```

   In this type of model, sticks represent bonds while balls represent atoms.

 * `{bm} space-filling model/(space-filling model|space-filling diagram)/i`

   ```{img}
   Ammonia-3D-vdW.png
   Space-filling model of ammonia
   By Benjah-bmm27 - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=764396
   scale 0.15 0.15
   ```

   In this type of model, each atom is represented by a ball and the relative sizes of atoms are represented.

## Binary Ionic Compound Naming

Recall that, in an ionic compound, ...
1. Each ion can be either a single atom ion or a molecule ion.
1. The overall charge of the ionic compound is neutral (the net charge of all the ions = 0).

The ions that make up an ionic compound are written in such a way that...
1. cations go first.
1. anions that are atoms have their last syllable replaced with "ide."
1. ion counts / ratios aren't included.
1. ion charges aren't included unless it's ambiguous.

For example... `{kt} Na_2S` = Sodium sulfide. Note that...
* cation is listed first, followed by the anion.
* anion has its last syllable stripped and replaced with "ide."
* counts and charges aren't included.

```{note}
Why aren't counts and charges not included? Just by seeing "sodium sulfide" we know that the...
* net charge of the entire thing will be 0.
* sodium ion has a predictable charge of 1+ (cation).
* sulfur ion has a predictable charge of 2- (anion).

To determine the actual counts, you just need to do some math. How many of the ion with the lower charge (sodium) do you need to neutralize 1 instance of the ion with higher charge?

* 1x = -(-2)
* 1x = 2
* x = 2

You need 2 sodium ions to neutralize 1 sulfur ion.

What happens is x is not a whole number? multiply both counts by the coefficients required to get it to a whole number.
```

To determine the name of ...
 * an atom, map its symbol to its name via the periodic table (e.g. `{kt} Fe` = iron).
 * a molecule, map its formula to its common name via the table in the common polyatomic ions subsection (e.g. `{kt} CO_3^{2-}` = carbonate).

To determine the charges of ...
 * an atom, use the periodic table. When an atom becomes charged (ion), its charge can be predicted based on where on the periodic table it sits. Elements in group ...

   * 1 of the periodic table charges to 1+ (cation).
   * 2 of the periodic table charges to 2+ (cation).
   * 13 of the periodic table charges to 3+ (cation).
   * 14 of the periodic table charges to either 4+ (if cation) or 4- (if anion).
   * 15 of the periodic table charges to 3- (anion).
   * 16 of the periodic table charges to 2- (anion).
   * 17 of the periodic table charges to 1- (anion).
   * 18 of the periodic table doesn't charge (they remain neutral).

   Elements outside these groups have more than 1 predictable charge. For example, an iron ion can be either 2+ or 3+. Since the charges of the 2 ions that make up an ionic compound have to cancel each other out, the other ion can be used to determine which charge to pick (see examples).
 * a molecule, it'll always be a single known charge given by the formula (e.g. `{kt} CO_3^{2-}` has a 2- charge).

### Chemical Formula to Name

When converting a chemical formula to a name, use the flowchart shown below.

```{plantuml}
@startuml
(*) --> "Determine cation and anion"

if "cation type?" then
    -> [molecule] "Write name&#8203;"
    --> ===CAT_WRITE===
else
    -> [atom] "Write name&#8203;&#8203;"
    if "predictable\ncharge?" then
        --> [yes] ===CAT_WRITE===
    else
        --> [no] "Determine charge"
        --> "Write charge"
        --> ===CAT_WRITE===
    endif
endif

if "anion type?" then
    --> [atom] "Write name but replace last syllable with -ide"
    --> (*)
else
    --> [molecule] "Write name&#8203;&#8203;&#8203;"
    --> (*)
endif

@enduml
```

----

For example, `{kt} NaCl` is named sodium chloride:

```{plantuml}
@startuml
(*) -[#007F00]-> "Determine cation and anion\n\nCation=Na, charge=1+\nAnion=Cl, charge=1-" #00FF00

-[#007F00]-> if "<color:#007F00> cation type?" then
    -> [molecule] "Write name&#8203;"
    --> ===CAT_WRITE===
else
    -[#007F00]> [<color:#007F00> atom] "Write name\n\nSodium" #00FF00
    -[#007F00]-> if "<color:#007F00>predictable\n<color:#007F00>charge?" then
        -[#007F00]-> [<color:#007F00>yes (1+)] ===CAT_WRITE===
    else
        --> [no] "Determine charge"
        --> "Write charge"
        --> ===CAT_WRITE===
    endif
endif

-[#007F00]-> if "<color:#007F00>anion type?" then
    -[#007F00]-> [<color:#007F00>atom] "Write name but replace last syllable with -ide\n\nSodium Chloride" #00FF00
    -[#007F00]-> (*)
else
    --> [molecule] "Write name&#8203;&#8203;&#8203;"
    --> (*)
endif

@enduml
```

----

For example, `{kt} FeCl_3` is named iron (III) chloride:

```{plantuml}
@startuml
(*) -[#007F00]-> "Determine cation and anion\n\nCation=Fe, charge=2+/3+\nAnion=Cl, charge=1-" #00FF00

-[#007F00]-> if "<color:#007F00> cation type?" then
    -> [molecule] "Write name&#8203;"
    --> ===CAT_WRITE===
else
    -[#007F00]> [<color:#007F00> atom] "Write name\n\nIron" #00FF00
    -[#007F00]-> if "<color:#007F00>predictable\n<color:#007F00>charge?" then
        --> [yes] ===CAT_WRITE===
    else
        -[#007F00]-> [<color:#007F00>no (2+/3+)] "Determine charge" #00FF00
        -[#007F00]-> [<color:#007F00>(3+)] "Write charge\n\nIron (III)" #00FF00
        -[#007F00]-> ===CAT_WRITE===
    endif
endif

-[#007F00]-> if "<color:#007F00>anion type?" then
    -[#007F00]-> [<color:#007F00>atom] "Write name but replace last syllable with -ide\n\nIron (III) Chloride" #00FF00
    -[#007F00]-> (*)
else
    --> [molecule] "Write name&#8203;&#8203;&#8203;"
    --> (*)
endif

@enduml
```

```{note}
How was the charge of iron determined? We know that ...

1. the cation is iron, which has a predictable charge of either 2+ or 3+.
1. the anions are 3 chlorines. Chlorine has a charge of 1-.

Since the 3 chlorine anions give a 3- charge and the overall net charge for a ionic compound must be 0, iron (III) is our only option.

* -3 + x = 0
* x = 3
```

----

For example, `{kt} Na_2(CO_3)` is named sodium carbonate:

```{plantuml}
@startuml
(*) -[#007F00]-> "Determine cation and anion\n\nCation=Na, charge=1+\nAnion=CO<sub>3</sub>, charge=2-" #00FF00

-[#007F00]-> if "<color:#007F00> cation type?" then
    -> [molecule] "Write name&#8203;"
    --> ===CAT_WRITE===
else
    -[#007F00]> [<color:#007F00> atom] "Write name\n\nSodium" #00FF00
    -[#007F00]-> if "<color:#007F00>predictable\n<color:#007F00>charge?" then
        -[#007F00]-> [<color:#007F00>yes (1+)] ===CAT_WRITE===
    else
        --> [no] "Determine charge"
        --> "Write charge"
        --> ===CAT_WRITE===
    endif
endif

-[#007F00]-> if "<color:#007F00>anion type?" then
    --> [atom] "Write name but replace last syllable with -ide"
    --> (*)
else
    -[#007F00]-> [<color:#007F00>molecule] "Write name\n\nSodium Carbonate" #00FF00
    -[#007F00]-> (*)
endif

@enduml
```

----

For example, `{kt} (NH_4)(NO_2)` is named ammonium nitrate:

```{plantuml}
@startuml
(*) -[#007F00]-> "Determine cation and anion\n\nCation=NH<sub>4</sub>, charge=1+\nAnion=NO<sub>2</sub>, charge=1-" #00FF00

-[#007F00]-> if "<color:#007F00> cation type?" then
    -[#007F00]> [<color:#007F00>molecule] "Write name\n\nAmmonium" #00FF00
    -[#007F00]-> ===CAT_WRITE===
else
    -> [atom] "Write name"
    --> if "predictable\ncharge?" then
        --> ===CAT_WRITE===
    else
        --> [no] "Determine charge"
        --> "Write charge"
        --> ===CAT_WRITE===
    endif
endif

-[#007F00]-> if "<color:#007F00>anion type?" then
    --> [atom] "Write name but replace last syllable with -ide"
    --> (*)
else
    -[#007F00]-> [<color:#007F00>molecule] "Write name\n\nAmmonium Nitrate" #00FF00
    -[#007F00]-> (*)
endif

@enduml
```

### Name to Chemical Formula

When converting a chemical formula to a name, use the flowchart shown below.

```{plantuml}
@startuml
(*) --> "Determine cation and anion"
--> "Determine ratios"
note right: Confused as to what\nthis means? see note\nbelow.

if "cation type?" then
    --> [molecule] "Write formula\nin parenthesis"
    --> ===CAT_WRITE===
else
    --> [atom] "Write symbol"
    --> ===CAT_WRITE===
endif

if "ratio > 1?" then
    --> [yes] "Write ratio"
    --> ===CAT_WRITE_RATIO===
else
    --> [no] ===CAT_WRITE_RATIO===
endif

if "anion type?" then
    --> [molecule] "Write formula\nin parenthesis&#8203;"
    --> ===AN_WRITE===
else
    --> [atom] "Write symbol&#8203;"
    --> ===AN_WRITE===
endif

if "ratio > 1?" then
    --> [yes] "Write ratio&#8203;"
    --> (*)
else
    --> [no] (*)
endif

@enduml
```

````{note}
Ratio refers to the ratio of cations to anions needed to satisfy a neutral charge. In other words, the number of cations to anions needed to create 1 instance (formula unit?) of the ionic compound.

```java
x = 1 / cation.charge
y = 1 / anion.charge
if (!is_whole_number(x)) {
  x = x * cation.charge
  y = y * cation.charge
}
if (!is_whole_number(y)) {
  x = x * anion.charge
  y = y * anion.charge
}
```

Ratio will be x:y.

For example, cation=3+ and anion=1-...

```java
x = 1 / cation.charge         //      x=1/3
y = 1 / anion.charge          //      x=1/3 y=1/1
if (!is_whole_number(x)) {    //true
  x = x * cation.charge       //      x=3/3 y=1/1
  y = y * cation.charge       //      x=3/3 y=3/1
}                             //
if (!is_whole_number(y)) {    //false
  x = x * anion.charge
  y = y * anion.charge
}
```

The ratio here is 1:3 -- the compound requires 1 cations and 3 anions.

For example, cation=2+ and anion=3-...

```java
x = 1 / cation.charge         //      x=1/2
y = 1 / anion.charge          //      x=1/2 y=1/3
if (!is_whole_number(x)) {    //true
  x = x * cation.charge       //      x=2/2 y=1/3
  y = y * cation.charge       //      x=2/2 y=2/3
}                             //
if (!is_whole_number(y)) {    //true
  x = x * anion.charge        //      x=6/2 y=2/3
  y = y * anion.charge        //      x=6/2 y=6/3
}
```

The ratio here is 3:2 -- the compound requires 3 cations and 2 anions.
````

----

For example, the chemical formula for sodium chloride is `{kt} NaCl`:

```{plantuml}
@startuml
(*) -[#007F00]-> "Determine cation and anion\n\nCation=Na, charge=1+\nAnion=Cl, charge=1-" #00FF00
-[#007F00]-> "Determine ratios\n\n1:1" #00FF00

-[#007F00]-> if "<color:#007F00>cation type?" then
    --> [molecule] "Write formula\nin parenthesis"
    --> ===CAT_WRITE===
else
    -[#007F00]-> [<color:#007F00>atom] "Write symbol\n\nNa" #00FF00
    -[#007F00]-> ===CAT_WRITE===
endif

-[#007F00]-> if "<color:#007F00>ratio > 1?" then
    --> [yes] "Write ratio"
    --> ===CAT_WRITE_RATIO===
else
    -[#007F00]-> [<color:#007F00>no] ===CAT_WRITE_RATIO===
endif

-[#007F00]-> if "<color:#007F00>anion type?" then
    --> [molecule] "Write formula\nin parenthesis&#8203;"
    --> ===AN_WRITE===
else
    -[#007F00]-> [<color:#007F00>atom] "Write symbol&#8203;\n\nNaCl" #00FF00
    -[#007F00]-> ===AN_WRITE===
endif

-[#007F00]-> if "<color:#007F00>ratio > 1?" then
    --> [yes] "Write ratio&#8203;"
    --> (*)
else
    -[#007F00]-> [<color:#007F00>no] (*)
endif

@enduml
```

----

For example, the chemical formula for iron (III) oxide is `{kt} FeCl_3`:

```{plantuml}
@startuml
(*) -[#007F00]-> "Determine cation and anion\n\nCation=Fe, charge=3+\nAnion=Cl, charge=1-" #00FF00
-[#007F00]-> "Determine ratios\n\n1:3" #00FF00

-[#007F00]-> if "<color:#007F00>cation type?" then
    --> [molecule] "Write formula\nin parenthesis"
    --> ===CAT_WRITE===
else
    -[#007F00]-> [<color:#007F00>atom] "Write symbol\n\nFe" #00FF00
    -[#007F00]-> ===CAT_WRITE===
endif

-[#007F00]-> if "<color:#007F00>ratio > 1?" then
    --> [yes] "Write ratio"
    --> ===CAT_WRITE_RATIO===
else
    -[#007F00]-> [<color:#007F00>no] ===CAT_WRITE_RATIO===
endif

-[#007F00]-> if "<color:#007F00>anion type?" then
    --> [molecule] "Write formula\nin parenthesis&#8203;"
    --> ===AN_WRITE===
else
    -[#007F00]-> [<color:#007F00>atom] "Write symbol&#8203;\n\nFeCl" #00FF00
    -[#007F00]-> ===AN_WRITE===
endif

-[#007F00]-> if "<color:#007F00>ratio > 1?" then
    -[#007F00]-> [<color:#007F00>yes] "Write ratio&#8203;\n\nFeCl<sub>3</sub>" #00FF00
    -[#007F00]-> (*)
else
    --> [no] (*)
endif

@enduml
```

----

For example, the chemical formula for sodium carbonate is `{kt} Na_2(CO_3)`:

```{plantuml}
@startuml
(*) -[#007F00]-> "Determine cation and anion\n\nCation=Na, charge=1+\nAnion=CO<sub>3</sub>, charge=2-" #00FF00
-[#007F00]-> "Determine ratios\n\n2:1" #00FF00

-[#007F00]-> if "<color:#007F00>cation type?" then
    --> [molecule] "Write formula\nin parenthesis"
    --> ===CAT_WRITE===
else
    -[#007F00]-> [<color:#007F00>atom] "Write symbol\n\nNa" #00FF00
    -[#007F00]-> ===CAT_WRITE===
endif

-[#007F00]-> if "<color:#007F00>ratio > 1?" then
    -[#007F00]-> [<color:#007F00>yes] "Write ratio\n\nNa<sub>2</sub>" #00FF00
    -[#007F00]-> ===CAT_WRITE_RATIO===
else
    --> [no] ===CAT_WRITE_RATIO===
endif

-[#007F00]-> if "<color:#007F00>anion type?" then
    -[#007F00]-> [<color:#007F00>molecule] "Write formula\nin parenthesis&#8203;\n\nNa<sub>2</sub>(CO<sub>3</sub>)" #00FF00
    -[#007F00]-> ===AN_WRITE===
else
    --> [atom] "Write symbol&#8203;"
    --> ===AN_WRITE===
endif

-[#007F00]-> if "<color:#007F00>ratio > 1?" then
    --> [<color:#007F00>yes] "Write ratio&#8203;"
    --> (*)
else
    -[#007F00]-> [no] (*)
endif

@enduml
```

----

For example, the chemical formula for ammonium nitrate is `{kt} (NH_4)(NO_2)`:

```{plantuml}
@startuml
(*) -[#007F00]-> "Determine cation and anion\n\nCation=NH<sub>4</sub>, charge=1+\nAnion=CO<sub>3</sub>, charge=1-" #00FF00
-[#007F00]-> "Determine ratios\n\n1:1" #00FF00

-[#007F00]-> if "<color:#007F00>cation type?" then
    -[#007F00]-> [<color:#007F00>molecule] "Write formula\nin parenthesis\n\n(NH<sub>4</sub>)" #00FF00
    -[#007F00]-> ===CAT_WRITE===
else
    --> [atom] "Write symbol"
    --> ===CAT_WRITE===
endif

-[#007F00]-> if "<color:#007F00>ratio > 1?" then
    --> [yes] "Write ratio"
    --> ===CAT_WRITE_RATIO===
else
    -[#007F00]-> [<color:#007F00>no] ===CAT_WRITE_RATIO===
endif

-[#007F00]-> if "<color:#007F00>anion type?" then
    -[#007F00]-> [<color:#007F00>molecule] "Write formula\nin parenthesis&#8203;\n\n(NH<sub>4</sub>)(CO<sub>3</sub>)" #00FF00
    -[#007F00]-> ===AN_WRITE===
else
    --> [atom] "Write symbol&#8203;"
    --> ===AN_WRITE===
endif

-[#007F00]-> if "<color:#007F00>ratio > 1?" then
    --> [<color:#007F00>yes] "Write ratio&#8203;"
    --> (*)
else
    -[#007F00]-> [no] (*)
endif

@enduml
```


In cases like this where the ratio's 1:1, parenthesis can be omitted if desired: `{kt} NH_4NO_2`.

In cases where the ratio isn't 1:1, parenthesis can be removed by multiplying out. For example, ammonium has a 1+ charge while carbonate has a 2- charge -- you need 2 ammonium to cancel the charge of 1 carbonate: `{kt} {(NH_4)}_2CO_3`. Multiplying the the components of the ammonium out by its subscript results in the simplified chemical formula: `{kt} N_2H_8CO_3`.

### Common Polyatomic Ion Names

The following table lists the shorthand names / chemical formulas of many common polyatomic ions (charged molecules).

| Chemical Formula    | Name                        | Informal/Common Names |
| ------------------- | --------------------------- | --------------------- |
| `{kt} NH_4^{+}`     | `{bm} Ammonium`             |                       |
| `{kt} NO_2^{-}`     | `{bm} Nitrite`              |                       |
| `{kt} NO_3^{-}`     | `{bm} Nitrate`              |                       |
| `{kt} SO_3^{2-}`    | `{bm} Sulfite`              |                       |
| `{kt} SO_4^{2-}`    | `{bm} Sulfate`              |                       |
| `{kt} HSO_4^{-}`    | `{bm} Hydrogen sulfate`     | `{bm} Bisulfate`      |
| `{kt} OH^{-}`       | `{bm} Hydroxide`            |                       |
| `{kt} CN^{-}`       | `{bm} Cyanide`              |                       |
| `{kt} PO_4^{3-}`    | `{bm} Phosphate`            |                       |
| `{kt} HPO_4^{2-}`   | `{bm} Hydrogen phosphate`   |                       |
| `{kt} H_2PO_4^{-}`  | `{bm} Dihydrogen phosphate` |                       |
| `{kt} SCN^{-}`      | `{bm} Thiocyanate`          |                       |
| `{kt} CO_3^{2-}`    | `{bm} Carbonate`            |                       |
| `{kt} HCO_3^{-}`    | `{bm} Hydrogen carbonate`   | `{bm} Bicarbonate`    |
| `{kt} ClO^{-}`      | `{bm} Hypochlorite`         |                       |
| `{kt} ClO_2^{-}`    | `{bm} Chlorite`             |                       |
| `{kt} ClO_3^{-}`    | `{bm} Chlorate`             |                       |
| `{kt} ClO_4^{-}`    | `{bm} Perchlorate`          |                       |
| `{kt} C_2H_3O_2^-` (`{kt} CH_3COO^-`) | `{bm} Acetate` |                  |
| `{kt} MnO_4^-`      | `{bm} Permanganate`         |                       |
| `{kt} Cr_2O_7^{2-}` | `{bm} Dichromate`           |                       |
| `{kt} CrO_4^{2-}`   | `{bm} Chromate`             |                       |
| `{kt} O_2^{2-}`     | `{bm} Peroxide`             |                       |
| `{kt} C_2O_4^{2-}`  | `{bm} Oxalate`              |                       |


## Chemical Equation

A `{bm} chemical reaction` is when energy is used to make molecules and ionic compounds reorganize their atoms such that they form different molecules / ionic compounds. For example, 2 hydrogen pair molecules and a oxygen pair molecule may react such that they form 2 water molecules:

```{chemfig}
\parbox{5cm}{\centering \chemfig{H-[4]H} \\ \chemfig{H-[4]H} \\ \chemfig{O-[4]O}}
```

results in...

```{chemfig}
\chemfig{O(-[5]H)(-[7]H)}
\chemfig{O(-[5]H)(-[7]H)}
```

The...

* inputs of a chemical reaction are called `{bm} reactant/(reactant)_CHEM/i`s.
* outputs of a chemical reaction are called `{bm} product/(product)_CHEM/i`s.

For the example above, the reactant_CHEMs are `{kt} 2H_2` and `{kt} O_2` while the product_CHEM is `{kt} 2H_2O`.

A chemical reaction that takes places can be written out as a `{bm} chemical equation`, where reactant_CHEMs are placed on one-side, product_CHEMs are placed on the other, and an arrow indicates which side the inputs are and which side the outputs are. For example, the chemical equation for the example above is `{kt} 2H_2 + O_2 \rightarrow 2H_2O`.

The type of arrow used in a chemical equation determines the type of reaction it is:
 * `{kt} \rightarrow` - An `{bm} irreversible reaction/(irreversible chemical reaction|irreversible reaction)/i` is a chemical reaction that it'll likely only ever occur one-way. It's a reaction that requires much more energy to reverse the reverse than it did to originally perform, so much so that it'll almost never occur naturally (without human intervention).

   The example above is an irreversible reaction: `{kt} 2H_2 + O_2 \rightarrow 2H_2O`.

  * `{kt} \rightleftharpoons` - A `{bm} reversible reaction/(reversible chemical reaction|reversible reaction)/i` is a chemical reaction that's likely to occur either way.
  
    An example of a reversible reaction: `{kt} HCO^-_3 + H^+ \rightleftharpoons H_2CO_3`.

```{note}
Don't get confused. It's called an irreversible reaction but it actually it's irreversible -- it's just means that it's very hard to reverse it.
```

### Balance

`{bm} Balancing a chemical equation/(balance the chemical equation|balance a chemical equation|balancing the chemical equation|balancing a chemical equation|balance chemical equation|balanced chemical equation|unbalanced chemical equation|chemical equation balancing)/i` means finding the ratios of reactant_CHEMs and product_CHEMs in a a chemical equation. In other words, how much of the reactant_CHEMs and product_CHEMs are needed to have an equal count of elements on both sides of the chemical equation.

For example, the chemical equation `{kt} H_2 + O_2 \rightarrow H_2O` is unbalanced because the number of hydrogen and oxygen elements between the left-hand side and the right-hand side are NOT equal:
* left-hand side has 2 hydrogen and 2 oxygen atoms.
* right-hand side has 2 hydrogen but only 1 oxygen atom.

To balance it, find the coefficients for each item in the chemical equation: `{kt} xH_2 + yO_2 = zH_2O` -- solve for x, y, and z. There are 2 ways do this...

* visual inspection / trial-and-error
* algebra

These methods are detailed in the sub-sections below.

#### Inspection

The first method is to use trial-and-error until the element counts match up between the sides. The high-level algorithm for this is to...
1. pick a set of coefficients.

   x=1, y=1, z=1  →  `{kt} 1H_2 + 1O_2 = 1H_2O` 
  
1. for each item on the LHS, add up the number of elements.

   `{kt} 1H_2 + 1O_2` contains `{kt} 2H` and `{kt} 2O`.

1. for each item on the RHS, add up the number of elements.

   `{kt} 1H_2O` contains `{kt} 2H` and `{kt} 1O`.

1. if the element counts between LHS and RHS don't match, go to 1 (new set of coefficients).

   `{kt} 2H = 2H` -- ✔️ matches!

   `{kt} 2O \neq 1O` -- ❌ no match, try again with different coefficients
   

In the example above, when ...
* ❌ x=1, y=1, z=1 → `{kt} 1H_2 + 1O_2 \neq 1H_2O` because `{kt} 2H = 2H` but `{kt} 2O \neq O`.
* ❌ x=1, y=1, z=2 → `{kt} 1H_2 + 1O_2 \neq 2H_2O` because `{kt} 2H \neq 4H` but `{kt} O = O`.
* ✔️ x=2, y=1, z=2 → `{kt} 2H_2 + 1O_2 = 2H_2O` because `{kt} 4H = 4H` and `{kt} O = O`.

```{chemfig}
\parbox{5cm}{\centering \chemfig{H-[4]H} \\ \chemfig{H-[4]H} \\ \chemfig{O-[4]O}}
```

results in...

```{chemfig}
\chemfig{O(-[5]H)(-[7]H)}
\chemfig{O(-[5]H)(-[7]H)}
```

#### Algebra

The second method is to use algebra to figure out the element counts. That is, if enough relationships exist between the reactant_CHEMs and the product_CHEMs, those relationships can be mapped out as a system equations (solvable via algebra).

`{kt} xH_2 + yO_2 \rightarrow zH_2O`.

Write out all properties that should be equal __if the chemical equation were balanced__. For example, if the equation were balanced, the number of oxygen atoms should be equal between the reactant_CHEM side and the product_CHEM side.

|                         | `{kt} H_2` | `{kt} O_2` | `{kt} H_2O` | Equation / Relationship    |
| ----------------------- | ---------- | ---------- | ----------- | -------------------------- |
| atom count for `{kt} H` | 2          | 0          | 2           | `{kt} x(2) + y(0) = z(2)`  |
| atom count for `{kt} O` | 0          | 2          | 1           | `{kt} x(0) + y(2) = z(1)`  |
| atom count for all      | 2          | 2          | 3           | `{kt} x(2) + y(2) = z(3)`  |
| charges                 | 0          | 0          | 0           | `{kt} x(0) + y(0) = z(0)`  |

Of those relationships, how many are usable? For the above example, on closer inspection it becomes apparent that...

 * the 3rd equation (atom count for all) isn't usable because it represents existing relationships.
 
   This 3rd equation is just the first two equations added up. If you take the first two equations, add the left-hand sides together and the right-hand sides together, you'll end up with the 3rd equation.

   `{kt} (x(2) + y(0)) + (x(0) + y(2)) = z(2) + z(1)` ⟶ `{kt} x(2) + y(2) = z(3)`

   An equation representing relationships that are already expressed by other equations in the system is redundant -- it doesn't contribute to solving the system.

 * the 4th equation (charges) isn't usable because it doesn't convey a relationship.
   
   The sum of charges on the reactant_CHEMs side and the product_CHEMs side are equal. But, each of the reactant_CHEMs / product_CHEMs have a 0 charge. The equation simplifies to `{kt} 0 = 0` -- it isn't conveying a relationship between any of the variables [x, y, z].

These non-usable equations are discarded.

|                         | `{kt} H_2` | `{kt} O_2` | `{kt} H_2O` | Equation / Relationship              |
| ----------------------- | ---------- | ---------- | ----------- | ------------------------------------ |
| atom count for `{kt} H` | 2          | 0          | 2           | `{kt} x(2) + y(0) = z(2)`            |
| atom count for `{kt} O` | 0          | 2          | 1           | `{kt} x(0) + y(2) = z(1)`            |
| ~~atom count for all~~  | ~~2~~      | ~~2~~      | ~~3~~       | `{kt} \xcancel{x(2) + y(2) = z(3)}`  |
| ~~charges~~             | ~~0~~      | ~~0~~      | ~~0~~       | `{kt} \xcancel{x(0) + y(0) = z(0)}`  |

Of the usable equations, if equations.length >= variables.length - 1, its solvable. The example above has 3 variables ([x, y, z]) and exactly 2 equations, so it's solvable. Isolate a variable in each of the equations, then set the last variable to 1 and solve.

Isolate `{kt} x` in equation 1:
 * `{kt} x(2) + y(0) = z(2)`
 * `{kt} 2x = 2z`
 * `{kt} x = z`

Isolate `{kt} y` in equation 2:
 * `{kt} x(0) + y(2) = z(1)`
 * `{kt} 2y = z`
 * `{kt} y = \frac{z}{2}`

Set `{kt} z` to 1 and solve via algebra:
 * `{kt} x = z` ⟶ `{kt} x = 1`
 * `{kt} y = \frac{z}{2}` ⟶ `{kt} y = \frac{1}{2}`
 * `{kt} z = 1`

```{note}
We can set the remaining variable to 1 because we're dealing with ratios. The ratios of reactant_CHEMs and product_CHEMs will all be relative to each other -- when we set a variable to 1, the other variables will get scaled accordingly.
```

In the example above, the balanced chemical equation comes out to `{kt} 1H_2 + \frac{1}{2}O_2 = 1H_2O`. This is correct in that it provides the ratios of reactant_CHEMs and product_CHEMs needed, but not the overall counts of each. To get the overall counts, multiply each item by y's divisor (2):

→ `{kt} 2H_2 + 1O_2 = 2H_2O` ✔️

```{chemfig}
\parbox{5cm}{\centering \chemfig{H-[4]H} \\ \chemfig{H-[4]H} \\ \chemfig{O-[4]O}}
```

results in...

```{chemfig}
\chemfig{O(-[5]H)(-[7]H)}
\chemfig{O(-[5]H)(-[7]H)}
```

```{note}
In most cases, it's totally fine to have the ratios (fractions) rather than the counts (whole numbers).
```

What happens if there aren't enough equations available to solve the system? It means that there isn't a single distinct solution. For example, there is no single solution for something like  `{kt} H_2 + O \rightarrow H_2O_2 + O_2`.

|                         | `{kt} H_2` | `{kt} O` | `{kt} H_2O_2` | `{kt} O_2` | Equation / Relationship           |
| ----------------------- | ---------- | -------- | ------------- | ---------- | --------------------------------- |
| atom count for `{kt} H` | 2          | 0        | 2             | 0          | `{kt} w(2) + x(0) = y(2) + z(0)`  |
| atom count for `{kt} O` | 0          | 1        | 2             | 2          | `{kt} w(0) + x(1) = y(2) + z(2)`  |
| ~~atom count for all~~  | ~~2~~      | ~~1~~    | ~~4~~         | ~~2~~      | `{kt} \xcancel{w(2) + x(1) = y(4) + z(2)}`  |
| ~~charges~~             | ~~0~~      | ~~0~~    | ~~0~~         | ~~0~~      | `{kt} \xcancel{w(0) + x(0) = y(0) + z(0)}`  |

```{note}
Last 2 eq struck out because they're not usable / useful in solving the system. Reasoning for this is discussed earlier on in this section.
```

There are 4 variables to solve but only 2 equations in the system -- at least 3 equations are needed. Attempting to solve without a 3rd equation results in...

Isolate `{kt} w` in equation 1:
 * `{kt} w(2) + x(0) = y(2) + z(0)`
 * `{kt} w(2) = y(2)`
 * `{kt} w = y`

Isolate `{kt} x` in equation 2:
 * `{kt} w(0) + x(1) = y(2) + z(2)`
 * `{kt} x = 2y + 2z`

Set `{kt} y` to 1 and solve as far as possible via algebra:
 * `{kt} w = y` ⟶ `{kt} w = 1`
 * `{kt} x = 2y + 2z` ⟶ `{kt} x = 2(1) + 2z` ⟶ `{kt} x = 2z + 2`
 * `{kt} y = 1`

There is no distinct solution...

`{kt} 1H_2 + (2z + 2)O \rightarrow 1H_2O_2 + zO_2`

Any of the following balanced chemical equations are possible...
* `{kt} H_2 + 4O \rightarrow H_2O_2 + O_2` is a valid answer.
* `{kt} H_2 + 6O \rightarrow H_2O_2 + 2O_2` is a valid answer.
* `{kt} H_2 + 8O \rightarrow H_2O_2 + 3O_2` is a valid answer.
* etc..

### Stoichiometry

`{bm} Stoichiometry/(stoichiometry|stoichiometric)/i` is the process of using the coefficients in a balanced chemical equation to calculate the quantities of reactant_CHEMs and product_CHEMs. In other words, given that you have some amount of a reactant_CHEMs/product_CHEMs, use the balanced chemical equation to determine the amounts of the other reactant_CHEMs/product_CHEMs.

For example, imagine the balanced chemical equation is `{kt} 2H_2 + O_2 \rightarrow 2H_2O`. If you have 3g of the reactant_CHEM `{kt} O_2`, ...
* how much of the other reactant_CHEMs (`{kt} H_2`) do you need?
* how much of the product_CHEMs (`{kt} H_2O`) are produced?

The high-level algorithm for this is...

```{plantuml}
@startuml
(*) --> "Balance chemical equation\n\n(if not already done so)"
--> ===SOLVE===
(*) --> "Convert known reactant/product\nquantities to moles (particle counts)\n\n(e.g. 6g of carbon = 0.5 moles)"
--> ===SOLVE===
--> "Determine moles for remaining reactants/products"
--> "Convert counts (moles) back to original quantity type"
@enduml
```

```{note}
The quantity type doesn't have to be grams. It often is grams but it could be something else like volume. If it were moles, the conversion to/from wouldn't be necessary.
```

1. Balance the chemical equation:

   `{kt} 2H_2 + O_2 \rightarrow 2H_2O`
   
   The coefficients in the balanced chemical equation are called the `{bm} stoichiometric coefficient`s. Think of it as a ratio. For the example, the ratio is 2:1:2 -- for every 2 instances of `{kt} H_2`, you'll need 1 instances of `{kt} O_2` and you'll get back 2 instances of `{kt} H_2O`.

   Stoichiometric coefficients may also be referred to as `{bm} mole ratio`, `{bm} stoichiometric factor`, or `{bm} stoichiometric ratio`.

2. Convert known quantity (3g of `{kt} O_2`) to moles:

   ```{note}
   Why moles? Recall that...
   * Grams are a unit of mass.
   * Moles are a unit of counts (e.g. the number of molecules).

   ----

   atom_count(1 gram of oxygen) != atom_count(1 gram of hydrogen) -- they have different weights. We need to convert from grams to moles because we're dealing with counts when we balance a chemical equation, not mass. Moles are a unit of counts.
   
   1 instance of `{kt} O_2` binds with 2 instances of `{kt} H_2` to produce 2 instances `{kt} H_2O`. 
   ```

   We have 3g of `{kt} O_2`. We know that 1mole of `{kt} O_2` = 32g.
   
   Therefore, 3g of `{kt} O_2` = 96g mole of `{kt} O_2`.
  
3. Use the stoichiometric ratio from step 1 to determine the moles of other reactant_CHEMs and product_CHEMs:

   `{kt} 2H_2 + O_2 \rightarrow 2H_2O` has the stoichiometric ratio 2:1:2.

   Since the amount of `{kt} O_2` is 96 moles, we know that the amount of...
   * `{kt} H_2` needs to be 2 times that... 192 moles of `{kt} H_2`.
   * `{kt} H_2O` will be be 2 times that... 192 moles of `{kt} H_2O`.

4. Convert quantities back to grams:

   * Since 1 mole of `{kt} H_2` = 2g, 192 moles of `{kt} H_2` = 2 / 192 = 96g.
   * Since 1 mole of `{kt} H_2O` = 18g, 192 moles of `{kt} H_2O` = 18 / 192 = 10.666g.

#### Limiting Reactant

Often times, more than 1 reactant_CHEM quantity is given for a stoichiometric calculation. The reactant_CHEM(s) that ...

 * gets used up entirely during the reaction are called the `{bm} limiting reactant/(limiting reactant|limiting reagent)/i`(s).
 * have some amount remaining after the reaction are called `{bm} excess reactant/(excess reactant|excess reagent)/i`(s).

For example, imagine the balanced chemical equation is `{kt} 2H_2 + O_2 \rightarrow 2H_2O`. If you have 3g of the reactant_CHEM `{kt} O_2` but also 50g of `{kt} H_2`, which reactant_CHEM would be the limiting reactant?

The way to solve this problem is to perform stoichiometric calculations for both reactant_CHEM quantities to see which one runs out first. When you do the calculations for...

| Quantity of reactant_CHEM | `{kt} H_2`     | `{kt} O_2`      | `{kt} H_2O` |
| ------------------------- | -------------- | --------------- | ----------- |
| 3g of `{kt} O_2`          | 0.37g <= 50g ✔️ | 3g <= 3g ✔️      | 3.37g       |
| 50g of `{kt} H_2`         | 50g <= 50g ✔️   | 396.86g <= 3g ❌ | 446.86g     |

As such, the limiting reactant is `{kt} O_2` and the excess reactant is `{kt} H_2` -- the 3g of `{kt} O_2` will need only 0.37g out of 50g of `{kt} H_2` available, leaving 49.63g of `{kt} H_2` remaining.

#### Theoretical Yield

The `{bm} theoretical yield` is the quantity of product_CHEMs produced if the reaction were to happen perfectly. For example, imagine the balanced chemical equation is `{kt} 2H_2 + O_2 \rightarrow 2H_2O`.  If you have 3g of the reactant_CHEM `{kt} O_2` but also 50g of `{kt} H_2`, the limiting reactant ends up being `{kt} O_2`...

| Quantity of reactant_CHEM | `{kt} H_2`     | `{kt} O_2`      | `{kt} H_2O` |
| ------------------------- | -------------- | --------------- | ----------- |
| 3g of `{kt} O_2`          | 0.37g <= 50g ✔️ | 3g <= 3g ✔️      | 3.37g       |
| 50g of `{kt} H_2`         | 50g <= 50g ✔️   | 396.86g <= 3g ❌ | 446.86g     |

Since the limiting reactant is `{kt} O_2`, so the theoretical yield is 3.37g of `{kt} H_2O`. That is, if the reaction were to happen perfectly, 3.37g of `{kt} H_2O` would be produced.

In reality, chemical reactions don't happen perfectly. The actual quantity of product_CHEMs varies based on several factors:
 * stability of the reactant_CHEMs.
 * purity of the reactant_CHEMs.
 * environmental factors (e.g. humidity).
 * `{bm} side reactions` -- reactant_CHEMs reacting in unpredictable ways that give unexpected product_CHEMs.
 * etc...

The `{bm} percent yield` is the ratio of of actual yield vs theoretical yield, calculated as `{kt} \frac{\text{actual yield}}{\text{theoretical yield}} \cdot 100`. In the example above, if only 3g of `{kt} H_2O` were actually produced, the percent yield would be `{kt} \frac{3}{3.37} \cdot 100 = 89`.

### Gravimetry

`{bm} Gravimetric/(gravimetric|gravimetry)/i` analysis is the process of using...
 1. phase changes (e.g. solid, gas, etc..)
 2. changes in mass

... to measure the quantity / concentration of some substance in a solution. When performing gravimetry, the substance of interest is sometimes referred to as a `{bm} analyte`.

Gravimetry comes in 2 forms:

* `{bm} volatilization gravimetry/(volatilization gravimetry|volatilization gravimetric)/i` decomposes the analyte such that a parts of the analyte breaks off as a gas.

  This only works if ...
  * decomposition only happens on the analyte -- no substance in the solution other than analyte decomposes.
  * decomposition is consistent -- specific parts must stay in solution while other parts escape as gas.
  
  Since we know the weight before and after decomposition, the stoichiometric ratio for the decomposition can be used to determine the particle counts for parts that escaped as gas. It then becomes trivial to figure out how much of the original analyte there was. For example, the analyte breaks into 4 pieces, 1 of which escape as a gas, the count of gas particles escaped is the count of the original analyte.

  ```
  Original analyte:
   -----
  |     |
  |  a  |
  |     |
   -----

  After heating decomposes analyte to 4 different substances:
   ---    ---
  | b |  | c |
   ---    ---

   ---    ---
  | d |  | e |
   ---    ---

  One of the 4 substances escapes as gas:
         ---
        | c | <-- escaped solution because its a gas
         ---

   ---
  | b |
   ---

   ---    ---
  | d |  | e |
   ---    ---

  Since we weighed the solution before and after, and we know the reaction taking place, stoichiometry lets us know
  how many particles of c there were. The number of c particles is the number of a particles (the original analyte),
  so we now know the concentration of the original analyte.
  ```

* `{bm} precipitation gravimetry/(precipitation gravimetry|precipitation gravimetric)/i` separates analyte in a solution by incorporating it into a insoluble solid.

  A reactant_CHEM (called the `{bm} precipitant` / `{bm} precipitating agent`) is added to the solution that will react with the analyte to produce a solid product_CHEM. The solid product_CHEM then gets filtered / collected (easy to do since it's a solid in a solution), dried, and measured. Since the balanced chemical equation for the reaction is known, stoichiometry can be used to determine the concentration of analyte.

  ```
  Original analyte:
   ---    ---
  | a |  | a |
   ---    ---

   ---    ---
  | a |  | a |
   ---    ---

  Add precipitant in excess:
   ---    ---           ---    ---    ---
  | a |  | a |         | p |  | p |  | p |
   ---    ---           ---    ---    ---

   ---    ---           ---    ---    ---
  | a |  | a |         | p |  | p |  | p |
   ---    ---           ---    ---    ---

  Reaction happens between analyte and precipitant to form a solid
   ----    ----           ---
  | ap |  | ap |         | p |
   ----    ----           ---

   ----    ----           ---
  | ap |  | ap |         | p |
   ----    ----           ---

  Since ap is solid, it can be collected, dried, and weighed. Since we know the reaction taking place between a and p
  (reactant) along with the weight of ap (product), stoichiometry lets us know how many particles of ap there were.
  The number of ap particles is the number of a particles (the original analyte), so we now know the concentration of
  the original analyte.
  ```

```{note}
This is a high-level overview of gravimetry. The Khan academy articles aren't doing a very good job of explaining the topic. I've explained as much as I've been able to understand but I think there are still large parts missing. This section needs cleanup and examples.
```

### Software Model

The software model for balancing chemical equations and stoichiometry is straight-forward.

 1. Parse the chemical equation.

    Parsing is performed using ANTLR's grammar syntax. An in-memory DOM model is constructed from the grammar. Each atomic element that gets parsed is directly mapped to a data structure that contains its details: symbols, names, atomic weights, atomic masses, isotopes, etc.. This data was extracted from from [CIAAW](https://ciaaw.org) using a browser plugin called [CopyTables](https://chrome.google.com/webstore/detail/copytables/ekdpkppgmlalfkphpibadldikjimijon?hl=en).

    Example of equation parsing...

    ```{define-block}
    ceparse
    ceparse_macro/
    cetools_code/
    ```

    ```{ceparse}
    2H2 + O2 -> 2H2O
    ```

    Example of bond information lookup...

    ```{define-block}
    cebondinfo
    cebondinfo_macro/
    cetools_code/
    ```

    ```{cebondinfo}
    H2O
    ```

 2. Use the algebra method to balance the chemical equation.

     Chemical equation balancing is done using the algebra method. The actual implementation of solving is delegated to the EJML library, where a matrix is populated with the coefficients and solved. As noted in the section on balancing equations, the algebra method doesn't always work -- if there aren't enough equations then solving via algebra isn't possible.

     ```{note}
     Charges aren't supported. That is, this solver doesn't generate an equation based on charges. Charges aren't supported by the parsing / DOM model.
     ```

     Example of balancing...

     ```{define-block}
     cebalance
     cebalance_macro/
     cetools_code/
     ```

     ```{cebalance}
     C3H8 + O2 -> CO2 + H2O
     ```

 3. Use the stoichiometry ratio from step2 to determine the amounts of reactant_CHEMs and product_CHEMs.

    Stoichiometry is done using the extracted CIAAW data from step 1. The reactant_CHEM/product_CHEM that's known has its mass converted from grams to moles, the stoichiometry ratio is applied to find the moles fro the remaining reactant_CHEMs and product_CHEMs, then they're all converted back from moles to grams. 

    Example of stoichiometry...

    ```{define-block}
    cestoichiometry
    cestoichiometry_macro/
    cetools_code/
    ```

    ```{cestoichiometry}
    Zn + CuCl2 -> ZnCl2 + Cu
    CuCl2
    13
    ```

## pH

`{bm} pH/\b(pH)\b/` stands for *potential of hydrogen* and it's the measure of positively charged hydrogen ions in a solution. The more...
* the more `{bm} acidic/(acidity|acidic|acid)/i` something is, the more positively charged hydrogen ions it has.
* the more `{bm} basic/(basicity|basic)/i` `{bm} /(bases|base)_pH/i` (`{bm} alkaline`) something is, the more reactive it is to positively charged hydrogen ions (it wants to give off electrons to those hydrogen ions).

![By Heinrich-Boll-Stiftung - https://www.flickr.com/photos/boellstiftung/35805740223, CC BY-SA 2.0, https://commons.wikimedia.org/w/index.php?curid=66098876](800px-PH_Scale-_Acidic_vs._Basic_(Alkaline).png)

pH is scaled logarithmically from 1 to 14. Each notch on the scale moves the acidity/basicity by a factory of 10. Going...
* back a notch (-1) increases acidity / decreases basicity by a factor of 10.
* forward a notch (+1) decreases acidity / increases basicity by a factory of 10.

For example, going from  7 to 4 increases acidity by 1000x times / decreases basicity by 1000x.

The closer to...
* 1 something is, the more acidic it is (more positively charged hydrogen ions) and the more sour/sharp it tastes.
* 14 something is, the more basic it is (more stuff that can react with positively charged hydrogen ions) and the more bitter it tastes.
* 7 something is, the more neutral it is (not reactive).

```{note}
https://www.quora.com/Why-is-pure-water-considered-neutral -- Since pH is defined as the negative log of the hydrogen ion concentration, the pH of pure water is 7 or neutral. Pure water is neutral because the number of positive hydrogen ions produced is equal to the number of negative.
```

# Biology

`{bm} Biology/(Biology|Biological|Biologist)/i` is defined as the study of life / living things. A `{bm} living/\b(living|life)\b/i` thing is defined as anything that converts energy from one form to another, while using that energy to grow, change, and reproduce.

```{note}
Primary source for these notes is Khan Academy's HS Biology course, with additional information from Wikipedia and other sources.
```

## Scientific Method

The `{bm} scientific method` is the standard guideline for discovery and experimentation in the sciences (chemistry, physics, biology, etc..) The `{bm-ri} basic` steps are...

1. Observe.
2. Ask a question about the observation.
3. Make a guess that answers the question (hypothesis).
4. Test the guess to see if it's correct (experiment).
5. Refine and iterate.

The last step (refine and iterate) just means that you do it all over again but make changes based on the things you learned from your experiment. For example, ...

* do additional experiments to dig into some aspect deeper.
* if the hypothesis wasn't supported by the experiment, maybe come up with a new hypothesis.

### Hypothesis

The scientific method revolves around making an observation and coming up with a testable explanation for that observation -- called a `{bm} hypothesis`. If the explanation isn't *testable*, you can't consider it a hypothesis. For example, a good hypothesis may be that increased sun exposure leads to an increased risk of skin cancer because it's something you can test. A bad explanation may be that exposure to centaurs increase the risk of skin cancer because centaurs don't exist (and as such the hypothesis can't be tested).

```{note}
The material mentions that for a hypothesis to be testable, you should be able to come up with an experiment that shows that its false -- it's falsifiable. How you word your hypothesis is typically what determines if it's testable/falsifiable -- when you read the hypothesis, what defines a failure?
```

```{note}
A hypothesis and a theory are different things. Hypothesis is a potential answer for a specific problem. A theory provides a potential framework for a much broader class of problems based on supporting evidence. The example given by the material: "The toaster won't toast because the electrical outlet is broken" is a hypothesis, whereas "Electrical appliances need a source of electricity in order to run" is closer to a theory.
```

### Experiment

Once you have a hypothesis, you design an `{bm} experiment` to test it. In the case of our sun exposure leads to increased risk of skin cancer hypothesis, an experiment may be to expose skin cells to UV rays in amounts equivalent to that given off by the sun and then check to see if those cells have been damaged (compared to a control group of skin cells that you haven't exposed to UV rays).

What makes a good experiment?

* `{bm} Control group` - A control group contains a set of entities that don't get treatment, but those entities are equal to the entities in the `{bm} experimental group` (as much as possible) and exposed to the same conditions as the entities in the experimental group (as much as possible). This helps eliminate outside factors/variables from tainting the results of an experiment -- it's a `{bm} controlled experiment` . For example, in the skin cancer example experiment described above, a control group may be a set of skin cells that sit along side the experimental group skin cells that get treated with UV rays -- the cells in the control group shouldn't show signs of damage while the cells in the experimental group should.
* `{bm} Random selection` - The entities being used for an experiment must be randomly selected (both for the control group and the experimental group). Random selection helps minimize skewed results introduced from bias during the selection process. For example, the individual responsible for selecting candidates for drug trials may choose to disproportionately select people of a certain race (e.g. racism).
* `{bm} Double blind` - The people conducting the experiment may implicitly/explicitly taint the experiment via their own bias. For example, if the experiment involves an interviewer asking a set of questions, the interviewer may subconsciously change the pronunciations of words if he/she knows that the interviewee is in the control group, there by effecting the answer given by the interviewee.
* **Statistical significance** - The results between the control group and the experimental group needs to be large enough to support your hypothesis. A small difference could mean that outcome was due to variance in the selected groups (there's a branch of statistics that deals with this).
* **Replicable** - Others should be able to repeat the same experiment and (hopefully) come to the same result. If they don't, it could be that something was wrong with your experiment or the entities being tested had some underlying difference that changed the result.

```{note}
There's always at least one control group in any experiment to provide a baseline. There's no limit to the number of experimental groups -- each group may have a slightly different type/amount of treatment applied.
```

```{note}
Because things are so wishy-washy/not-exactly in biology, it's typical for an experiments to be repeated multiple times and to have a large sample size -- the larger our sample sizes and the more times we conduct the experiment, the more we can be confident of our result. What do I mean by wishy-washy? Genetic variation between samples may result in different types/levels of responses. For example, people with a certain gene may respond quicker to certain drugs than people who don't produce that gene.
```

Other terminology around the scientific method...

* `{bm} treatment` - The application of the test given to the experimental group(s) but not the control group.
* `{bm} independent variable` - An independent variable is `{bm-ri} basically` the treatment you apply -- you can think of it as the input knob you control for your experiment. For example, in the skin cancer experiment, the independent variable would be the amount of artificial UV rays you apply to the skin cells.
* `{bm} dependent variable` - A dependent variable is a response you measure after applying the treatment -- you can think of it as the output from your experiment. For example, in the skin cancer experiment, the dependent variable might be a measurement of how different the cells are functioning (e.g. a cell with damaged DNA may produce different kinds / amounts of tRNA molecules -- something you can measure).
* **data** - This is exactly what you think it is -- measurements/observations made during the experiment.
* `{bm} placebo effect/(placebo effect|placebo)/i` - When experimenting on people, there's a phenomenon called the placebo effect: if someone takes something that shouldn't help them but are under the impression it will help them, it often times will help them, just from the psychological effect alone.

```{note}
You can have more than one independent variable if you follow specific guidelines and are experienced enough, but the general rule of thumb is to have only 1 independent variable just because it makes things much simpler to analyze/interpret.
```

## Carbohydrate Molecule

`{bm} Carbohydrate/(carbohydrate|sugar)/i`s (also called `{bm} saccharide`s) are molecules that consist of a mix of carbon, hydrogen, and oxygen atoms. In biological systems, carbohydrates are often associated with...
* being a source of energy
* providing a structural role (for plants / certain plants).

```{note}
It was never explained what 'structural role' actually means.
```

The term `{bm} monosaccharide` is just means a carbohydrate that's a monomer (e.g. glucose). Similarly, the term `{bm} polysaccharide` means a carbohydrate built from other monosaccharides (e.g. glycogen is made of chained glucose).

```{dot}
digraph {
  rankdir=LR;
  "monosaccharide"->"polysaccharide" [taillabel="1..*", headlabel="1", arrowhead=none];
}
```

## Protein Molecule

`{bm} Protein`s are molecules that consist of monomers called `{bm} amino acid`s. The amino acids get chained together into a polymer called a `{bm} polypeptide` chain, and one or more polypeptide chains fold to a 3D structure and combine to become a protein. The 3D structure / shape of the protein (how its folded) is what gives it its abilities.

In biological systems, proteins are often associated with that facilitating some biological function. For example, the protein protease is responsible for breaking down food.

```{dot}
digraph {
  rankdir=LR;
  "amino acid"->"polypeptide" [taillabel="1..*", headlabel="1", arrowhead=none];
  "polypeptide"->"protein" [taillabel="1..*", headlabel="1", arrowhead=none];
}
```

![By No machine-readable author provided. DrKjaergaard assumed (based on copyright claims). - No machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=1967109](320px-Protein_folding.png)

```{note}
The ribosome is what's responsible for folding? Not able to get a clear answer on this.
```

The `{bm-ri} basic` structure of an amino acid is as follows. The R is a placeholder that, when set, defines what type of amino acid it is...

![By GYassineMrabetTalk✉This W3C-unspecified vector image was created with Inkscape. - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=2551977](320px-AminoAcidball.svg.png)

## Lipid Molecule

`{bm} Lipid`s are molecules that are somewhat not water soluble -- meaning that they have parts that resist water but maybe also parts that are attracted to water. In biological systems, lipids are often associated with...
* energy storage (fats)
* cellular membranes (phospholipids)

![Public Domain, https://commons.wikimedia.org/w/index.php?curid=1019610](296px-Trimyristin-3D-vdW.png)

```{note}
Lipids are not always fats. All fats are lipids but not all lipids are fats.
```

## Water Molecule

`{bm} Water` is essential to life -- it has unique properties that almost all biological processes depend on.

Recall that...
1. a water molecule consists of 2 hydrogen atoms connected to an oxygen atom via covalent bonds. A covalent bond is a pair of electrons that both atoms share, thus bonding the atoms together.
1. the position of an electron is based on probability. Electrons aren't fixed in a certain position or neatly orbiting around a nucleus_atom as certain diagrams show. Rather, they're constantly buzzing/hopping around the nucleus_atom. Depending on their environment, they may be more likely to be at certain locations vs other locations.

Oxygen atoms are extremely `{bm} electronegative`, meaning that oxygen has the propensity to pull the buzzing/hopping electrons more around itself than the atoms it's bound to. As such, in a water molecule, the electrons will spend more time solely around the oxygen atom than they do the hydrogen atom or a position that binds the hydrogen and oxygen together. This is what gives the oxygen atom in a water molecule a `{bm} weakly negative` charge (as indicated by δ-) while the hydrogen atoms have a `{bm} weakly positive` charge (as indicated by δ+). These types of charged molecules are called `{bm} polar molecule`s.

![By OpenStax College - Anatomy & Physiology, Connexions Web site. http://cnx.org/content/col11496/1.6/, Jun 19, 2013., CC BY 3.0, https://commons.wikimedia.org/w/index.php?curid=30131144](209_Polar_Covalent_Bonds_in_a_Water_Molecule.jpg)

```{note}
Notice the shape of the water molecule in the diagram(s) above. Electron pairs are repelled from each other. They're also responsible for binding. That's what gives molecules their shapes/structure.
```

This weakly negative / weakly positive charge is what gives water several of the unique properties that biological properties depend on. Water molecules have a tendency to gravitate towards each other because the weakly negative oxygen atoms and the weakly positive hydrogen atoms of different water molecules attract. This attraction is called a `{bm} hydrogen bond`. Hydrogen bonds are weaker than covalent bonds in that the bonds aren't really solid -- water molecules can easily break off and go past each other.

![By User Qwerter at Czech wikipedia: Qwerter. Transferred from cs.wikipedia to Commons by sevela.p. Translated to english by by Michal Maňas (User:snek01). Vectorized by Magasjukur2 - File:3D model hydrogen bonds in water.jpg, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=14929959](242px-3D_model_hydrogen_bonds_in_water.svg.png)

```{note}
The above paragraph is just giving the mechanics for how/why water is a liquid. Water is the only molecule that's liquid at room temperature? This can't be right -- see https://chemistry.stackexchange.com/q/76346. Why can't biological processes work in one of these other molecules just as they do in water? Maybe because they stay liquid at a shorter temperature range (e.g. 15-25C instead of 1-99C?
```

The weak attraction between water molecules is also what makes water a `{bm} solvent`. So long as they're polar molecules, other molecules can travel inside of water using the same attraction from weakly negative / weakly positive charges -- they gravitate and float around water molecules just as other water molecules do. For example, the cytoplasm of a cell is a solvent (mostly water). It works because other molecules in the cytoplasm (e.g. cellular machinery) can float around / travel around using the weakly negative / weakly positive charges.

Water is called a `{bm} universal solvent` because it can dissolve more molecules than other other liquid. Note that the term universal doesn't mean that it can dissolve everything, just that it can dissolve more things than the others.

The properties that make water conducive for biological processes to operate:

* Fluid under normal conditions - under normal temperatures, pressures, etc.. there's `{bm} cohesion` between other water molecules (fluid) / `{bm} adhesion` between water molecules and other polar molecules / ions (solvent).
* Universal solvent - Water dissolves more things than any other liquid (more things doesn't mean everything).
* High heat capacity - Because it takes a lot of energy to heat up water, its temperature can remain relatively stable as the environment around it changes (temperature regulation).
* High heat of vaporization - As water evaporates, it traps heat from whatever its on (e.g. human body) and releases it into the environment (temperature regulation / cooling).
* Less dense as a solid - As water freezes it forms crystals that space the molecules away from each other. As such, it becomes less dense than liquid water and ends up floats on top. This means bodies of water can freeze over but life can continue operating in the liquid underneath the frozen ice.

Other terminology related to water:
* Molecules that are charged are called polar molecules (e.g. the water molecule is a polar molecule because of its weakly positive / weakly negative charge).
* Molecules / ions that are charged and mixed with a solvent are called `{bm} solute`s.
* Molecules that are attracted to water are called `{bm} hydrophilic`.
* Molecules that are repelled from water are called `{bm} hydrophobic`.
* Molecules that have parts that are hydrophilic as well as parts that are hydrophobic are called `{bm} amphipathic`

## Cell

`{bm} Cell`s are the `{bm-ri} basic` unit of living things / the building blocks of life. They're tiny structures that encapsulate information and machinery that allows them to replicate/reproduce and perform other important functions (e.g. appendages to move around).

```{note}
Viruses are not cells but they may also be considered living because they reproduce in a roundabout way: the require machinery in the host cell to reproduce.
```

There are 2 types of cells: eukaryotic and prokaryotic. There main differences between them are that...

1. the guts of eukaryotes are organized into organelles (membrane-bound compartments) where each one is responsible for some functionality, while prokaryotes have no organelles at all (guts are free floating).
1. the DNA in eukaryotes into multiple independent segments (chromosomes), while prokaryotes have a single circular chain.

Other differences between eukaryotes and prokaryotes ...

```{csv}
!!{ "firstLineHeader": true }
, Eukaryotes, Prokaryotes
Size, 10 to 100 micrometers (μm), 0.1 to 5 micrometers (μm)
Complexity, More complex, More simple
Sub-compartments (organelles), Yes, No
DNA layout, Multiple stands, Single circular strand
Single-cell organisms, Yes (e.g. amoeba), Yes (e.g. bacteria and archaea)
Multi-cell organisms, Yes (e.g. animals and fungus), No
```

```{note}
Archaea is an organism that looks like bacteria but they're totally different.
```

### Features

Different cell species vary in features. The subsections below detail common cell features (not exhaustive).

Some features are only present in certain cell speicies (e.g. only some cells have a flagellum tail) while other features are present in all cells but in different amounts (e.g. every cell has cytosol but larger cells have more cytosol).

#### Cytoplasm

The `{bm} cytoplasm` (both eukaryotic and prokaryotic) is the insides/guts of a cell. `{bm} Cytosol` refers to just the fluid, while cytoplasm refers to fluid as well as everything else inside the cell.

Every cell has cytoplasm.

```{img}
Anima_cell_notext.svg
Eukaryote with cell ribosomes highlighted
By No machine-readable author provided. Chb assumed (based on copyright claims). - No machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=688296
scale 0.5 0.5
arrow 1 0.5 0.7 0.5
expand 2 1 0 0
text 0.5 0.5 cytoplasm
text 0.5 0.6 (everything inside)
```

#### Membrane

The `{bm} plasma membrane/(Plasma membrane|Membrane)/i` (present in both eukaryotic and prokaryotic cells) is the thing encapsulating the cytoplasm. It's what keeps the guys of the cell inside and controls the movement of substances coming into / going out of the cytoplasm.

Every cell has a membrane encapsulating its cytoplasm. Membranes in general follow the fluid mosaic model.

The term membrane can refer to either the plasma membrane or the membrane of a eukaryotic cell's organelle. How you should interpret it depends on the context in which its used.

```{img}
Anima_cell_notext.svg
Eukaryote with cell ribosomes highlighted
By No machine-readable author provided. Chb assumed (based on copyright claims). - No machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=688296
scale 0.5 0.5
arrow 1 0.5 0.81 0.6
expand 2 1 0 0
text 0.5 0.5 cell membrane
text 0.5 0.6 (holds everything in)
```

```{img}
Average_prokaryote_cell-_unlabled.svg
Prokaryotic with capsule highlighted
By Mariana Ruiz Villarreal LadyofHats - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=7356226
scale 0.5 0.5
arrow 1 0.5 0.8 0.5
expand 2 1 0 0
text 0.5 0.5 cell membrane
text 0.5 0.6 (inner layer)
```

Facts about cell membranes:

 * Membranes have other molecules embedded in them for identification, moving stuff in/out, etc.. (see fluid mosaic model).
 * Membranes are semipermeable. Uncharged substances such as carbon dioxide and oxygen move across fine, while charged ions or large molecules require either
   * membrane proteins such as a carrier protein or a channel protein (see fluid mosaic model).
   * membrane wrapper (see golgi).

#### Cell Wall

The `{bm} cell wall` (present in both eukaryotic and prokaryotic cells) is a stiff layer around the membrane meant for protection. *Not all cells have a cell wall* -- for example, animal cells don't but plant cells do. Technically, the cell wall (if it exists) isn't considered to be part of the cell. The membrane and everything in it is.

The material states that cell walls...
1. provide an extra layer of protection.
1. help maintain shape.
1. help prevent dehydration.

Almost all prokaryotes have cell walls. Only some eukaryotes have cell walls (e.g. fungi and plants). The material says that cell walls for most bacteria are made up of a molecule called `{bm} peptidoglycan`, but it can be different for other cells. For example, [this link](https://www.quora.com/Do-Eukaryotes-have-cell-walls) says that plant cells have cell walls made up of cellulose.

```{img}
Average_prokaryote_cell-_unlabled.svg
Prokaryotic with capsule highlighted
By Mariana Ruiz Villarreal LadyofHats - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=7356226
scale 0.5 0.5
arrow 1 0.5 0.85 0.5
expand 2 1 0 0
text 0.5 0.5 cell wall
text 0.5 0.6 (middle layer)
```

#### Capsule

The `{bm} Capsule` (present in prokaryotic cell only) is the outermost layer of some types of cells (typically bacteria cells). Capsules are made up of carbohydrates and there mainly to help the cell stick itself to the environment.

```{img}
Average_prokaryote_cell-_unlabled.svg
Prokaryotic with capsule highlighted
By Mariana Ruiz Villarreal LadyofHats - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=7356226
scale 0.5 0.5
arrow 1 0.5 0.9 0.5
expand 2 1 0 0
text 0.5 0.5 capsule
text 0.5 0.6 (outter-most layer)
```

```{note}
Although eukaryotic cells don't have capsules, they do have carbohydrates on their outside. Those carbohydrates aren't organized as a capsule though: https://www.quora.com/Do-some-eukaryotic-cells-have-capsules-or-is-it-just-prokaryotes-Are-there-exceptions-of-eukaryotes-having-capsules. Is this talking about the same carbohydrates that are embedded in the membrane (glycolipids / glycoproteins).
```

#### Ribosome

`{bm} Ribosome` (present in both eukaryotic and prokaryotic cells) are tiny molecular machines inside the cytoplasm that take in mRNA molecules (portions of DNA that have been written out) and produce proteins. Ribosomes themselves are structures made of proteins and RNA.

Ribosomes can either be floating around in the cytoplasm (called `{bm} free ribosome`) or be embedded in the membrane of endoplasmic reticulum.

```{img}
Anima_cell_notext.svg
Eukaryote with cell ribosomes highlighted
By No machine-readable author provided. Chb assumed (based on copyright claims). - No machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=688296
crop 0.32 0.16 0.6 0.34
arrow 0.9 0.5 0.5 0.6
arrow 0.9 0.5 0.65 0.35
expand 1.3 1 0 0
text 0.7 0.5 ribosomes
text 0.7 0.65 (little blue dots)
```

#### Appendage

Some cells have appendages that help them move (or stay put). There are different types of appendages...
  * `{bm} Flagellum` are tails that extend from the cell (e.g. tail on a sperm cell). There can be more than 1 flagellum.
  * `{bm} Cilia/\b(Cil|Cils|Cilia)\b/i` are much smaller hair-like appendages used to help move the cell itself or things in the vicinity of the cell.
  * `{bm} Fimbriae` are much smaller hair-like appendages used to help attach to host cells and surfaces (e.g. bacteria cells).
  * `{bm} Pili` are much smaller hair-like appendages used to help transfer DNA between cells and/or to help move (e.g. bacteria cells).
  * `{bm} Pseudopodia/\b(Pseudopod|Pseudopodia)\b/i` are much larger leg-like appendages used to crawl (e.g. amoeba cells).

  ```{img}
  Anima_cell_notext.svg
  Eukaryote with cell flagellum highlighted
  By No machine-readable author provided. Chb assumed (based on copyright claims). - No machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=688296
  scale 0.5 0.5
  arrow 1 0.5 0.3 0.84
  expand 2 1 0 0
  text 0.5 0.5 flagellum
  ```

### Fluid Mosaic Model

The `{bm} fluid mosaic model` is the accepted model for how cell membranes work. The model says that a cell membrane is composed of a phospholipid bilayer with proteins, lipids, and carbohydrates floating around on either side or embedded in between.

```{note}
The description above is the rational for the name 'fluid mosaic model'. It's fluid and there's a mosaic of different things embedded or attached to it.
```

![By LadyofHats Mariana Ruiz - Own work. Image renamed from File:Cell membrane detailed diagram.svg, Public Domain, https://commons.wikimedia.org/w/index.php?curid=6027169](800px-Cell_membrane_detailed_diagram_en.svg.png)

A `{bm} phospholipid` is a amphipathic lipid molecule that involves a phosphate group. The...
* bulbus phosphate group at the top is hydrophilic -- it has a charge and as such is attracted to water (either the cytoplasm or the fluid outside the cell).
* long fatty acid tails are hydrophobic -- they have no obvious charge and as such aren't attracted to water.

```{note}
For a refresher on how hydrophobic / hydrophilic molecules work, see the section on Water. Specifically: adhesion / weakly negative / weakly positive.
```

![By OpenStax - https://cnx.org/contents/FPtK1zmh@8.25:fEI3C8Ot@10/Preface, CC BY 4.0, https://commons.wikimedia.org/w/index.php?curid=30131167](0301_Phospholipid_Structure.jpg)

As such, phospholipids have a natural tendency to form as a `{bm} phospholipid bilayer` (2 layers attached together, called a `{bm} liposome`) or a ball (called a `{bm} micelle`). The hydrophilic heads are going to point towards the water causing the hydrophobic tails to point at each other.

![By Mariana Ruiz Villarreal ,LadyofHats - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=3032610](Phospholipids_aqueous_solution_structures.svg.png)

```{note}
If the phospholipids have small tails, they may form a micelle (a small, single-layered sphere), while if they have bulkier tails, they may form a liposome.
```

How fluid a phospholipid bilayer is depends on the types of phospholipid molecules that make it up and the temperature. Phospholipid molecules have 2 fatty acid tails. The fatty acid tails can be either...

* both saturated (straight tails)
* one saturated (straight tail) and one unsaturated (bent tail).

At cooler temperatures, phospholipids that have 2 saturated fatty acid tails (straight tails) tend to get more rigid / dense because they can more easily pack together. Phospholipids with unsaturated fatty acid tails (bent tails) don't end up getting as rigid / dense, allowing the membrane to stay fluid at lower temperatures. Cholesterol embedded in the phospholipid bilayer also helps it stay more fluid at lower temperatures.

```{note}
Phospholipid bilayers have the consistency of oil-based salad dressing. It may seem weak but it's strong enough to act as a separator between the environment inside and the environment outside. Water from one-side can move to the other but does so very rarely -- a single molecule may sneak through the layer every now and then. Aquaporins are proteins embedded in the phospholipid bilayer that allow water to rapidly pass (when needed).
```

Examples of molecules that can be embedded in or attached to the phospholipid bilayer include...

* `{bm} Glycolipid`s / `{bm} Glycoprotein`s - These are lipids and proteins with carbohydrates attached to them and are typically found pointing outside of the cell. These carbohydrates are how a cell identifies if a neighbouring cell is foreign or not.
* `{bm} Integral protein/(integral protein|integral enzyme)/i`s - Proteins that are partially hydrophobic. Either one end of the protein is anchored inside the membrane or the protein is fully going through the membrane.
  * `{bm} Transmembrane protein/(transmembrane protein|transmembrane enzyme)/i`s - Proteins that have one end inside the cytoplasm and the other outside of the cell. Typical use case for these types of proteins is moving substances from inside the cell (cytoplasm) to outside the cell and vice-versa: `{bm} channel protein`s and `{bm} carrier protein`s.
* `{bm} Peripheral protein/(peripheral protein|peripheral enzyme)/i`s - Proteins that are found on the surface of the protein (either on the cytoplasm side or the outside side). Unlike integral proteins, they don't have any portion going inside the membrane. As such, they more freely move around compared to integral proteins.
* `{bm} Cholesterol` - Sits inside of the membrane to help ensure that the membrane doesn't become too stiff.

```{note}
See first diagram in this section for an example of each of the molecules listed above.
```

The term `{bm} facilitated diffusion` refers to the movement of molecules across the membrane via proteins embedded in the membrane (e.g. channel proteins and/or carrier proteins). These molecules wouldn't be able to cross the membrane by themselves. For example, the sodium potassium pump (carrier protein) helps sodium and potassium ions move across the cell membrane by opening/closing its gates.

### History of Modern Cell Theory

The first record of a cell was in 1665 when `{bm} Robert Hooke` published a book called The Micrographia. The book contains drawings of observations he made while looking at various dead organisms through a rudimentary microscope.

A few years later, a Dutch lenscrafter by the name of `{bm} Antonie Van Leeuwenhoek/\b(Antonie Van Leeuwenhoek|Anthony Van Leeuwenhoek)\b/i` decided to use his expertise to craft a better microscope to better observe living cells / organisms. For example, he was able to observe sperm and Protists (unicellular organisms while he dubbed `{bm} animalcule`s).

In the 1830s, `{bm} Matthias Schleiden` and `{bm} Theodore Schwann` began laying the groundwork for modern cell theory. They came up with the idea that...
1. all life is composed of one or more cells.
2. a cell is the `{bm-ri} basic` unit of life.

They also suspected that cells come from other cells, but didn't know for sure if that was the only way they were produced. It was `{bm} Robert Remak` that in the mid-1800s established that...

3. all cells come from other cells.

`{bm} /\b(Rudolph Virchoi|Rudolph Virchow)\b/i`

```{note}
The credit for this sometimes goes to Rudolph Virchoi but it's been established that he was a plagiarist.
```

It's still an open question as to how the first / initial cell came to be. The current working theory is that, 3.5 billion years ago, phospholipids (the molecules that form the membrane of a cell) naturally form bilayers and connect in a circle. A membrane may have naturally encapsulated a set of arbitrary self-replicating molecules (e.g. protein or RNA) and that's how the first cell began growing and splitting off.

```{note}
There are an estimated 37 trillion cells in the human body.
```

## Eukaryotic Cell

```{img}
Animal_cell_structure_en.svg
ANIMAL CELL
By Mariana Ruiz Villarreal, LadyofHats - Own work (Source: Typical prokaryotic cell, Chapter 4: Mutagenicity of alkyl N-acetoxybenzohydroxamates, Concept 1: Common Features of All Cells, Cells - Structure and Function), Public Domain, https://commons.wikimedia.org/w/index.php?curid=3648821
text 0.45 0.05 ANIMAL CELL

bg_color #00000000

fg_color #ff00ffff
rect 0.25 0.012 0.17 0.23
fg_color #0000ffff
rect 0.01 0.55 0.24 0.08
fg_color #007f7fff
rect 0.01 0.44 0.24 0.075
fg_color #00ff00ff
rect 0.81 0.355 0.18 0.04
fg_color #7f7f00ff
rect 0.81 0.285 0.18 0.04
fg_color #000080ff
rect 0.81 0.205 0.18 0.04
```

```{img}
Plant_cell_structure-en.svg
Eukaryotic plant cell
By LadyofHats - Self-made using Adobe Illustrator. (The original edited was also made by me, LadyofHats), Public Domain, https://commons.wikimedia.org/w/index.php?curid=844682
text 0.45 0.1 PLANT CELL

bg_color #00000000

fg_color #ff0000ff
rect 0.01 0.25 0.2 0.135
fg_color #800080ff
rect 0.01 0.395 0.2 0.11
fg_color #ff00ffff
rect 0.79 0.75 0.17 0.16
fg_color #0000ffff
rect 0.46 0.87 0.13 0.11
fg_color #0000ffff
rect 0.86 0.34 0.135 0.11
fg_color #00ff00ff
rect 0.01 0.51 0.2 0.08
fg_color #0000ffff
rect 0.01 0.6 0.2 0.06
fg_color #007f7fff
rect 0.18 0.8 0.27 0.15
```

`{bm} Eukaryotic/(Eukaryote|Eukaryotic)/i` cells are typically larger and have membrane-bound sub-compartments, called organelle, that hold in the guts of different regions of the cell. For example, their DNA is encapsulated in a organelle called the nucleus_cell.

Eukaryotes have their DNA broken up into multiple strands called chromosomes. They can either be single-cellular organisms (e.g. amoeba) or multi-cellular organisms (e.g. human). Single-cellular organism that are eukaryotic are called `{bm} protist`s.

The following are descriptions for some of the organelles shown in the diagram above.

* <span style="color:#ff00ffff">**`{bm-ri} Nucleus`**</span> - See nucleus_cell section.
* <span style="color:#0000ffff">**`{bm-ri} Endoplasmic Reticulum`**</span> - See endoplasmic reticulum section.
* <span style="color:#007f7fff">**`{bm-ri} Golgi`**</span> - See golgi section.
* <span style="color:#00ff00ff">**`{bm-ri} Mitochondria`**</span> - See mitochondria section.
* <span style="color:#7f7f00ff">**`{bm-ri} Lysosome`**</span> (mostly animal) - See lysosome section.
* <span style="color:#0000ffff">**`{bm-ri} Peroxisome`**</span> - See peroxisome section. 
* <span style="color:#ff0000ff">**`{bm-ri} Chloroplast`**</span> (plant / algae) - See chloroplast section.
* <span style="color:#800080ff">**`{bm-ri} Vacuole`**</span> (mostly plant / algae) - See vacuole section. 

### Organelle

Eukaryotic cells have membrane-bound sub-compartments, called `{bm} organelle`s, that house different functional regions of the cell. The following subsections detail common organelles.

#### Nucleus

`{bm} Nucleus/\b(Nucleus|Nuclei)_cell\b/i` is an organelle that contains DNA (genetic information required for the functioning and replication). Both prokaryotic and eukaryotic cells have DNA, but only eukaryotic cells have a nucleus_cell. In prokaryotic cells, the DNA flows around freely instead of being encapsulated in a nucleus_cell.

```{img}
Diagram_human_cell_nucleus.svg
 comprehensive diagram of a human cell nucleus.
By Mariana Ruiz LadyofHats - I did it myself with adobe ilustrator using the information found here [1], [2] ,[3], [4] and [5], Public Domain, https://commons.wikimedia.org/w/index.php?curid=736389
scale 0.75 0.75
```

Most eukaryotic cells contain a single nucleus_cell, but some contain can have 0 and others can have more than one. An example of 0 is blood cells -- mature blood cells don't have any DNA, therefore no nucleus_cell. An example of more than 1 is the organism Oxytricha trifillax -- it contains 2 nuclei_cell, each containing different DNA (its DNA is fragmented across 2 nuclei_cell).

#### Endoplasmic Reticulum

`{bm} Endoplasmic Reticulum` is layered membrane (organelle?) that surrounds the nucleus_cell and is directly connected to pores on the nucleus_cell. Large portions of the endoplasmic reticulum's membrane have ribosomes attached. The parts that have ribosomes attached are called `{bm} rough endoplasmic reticulum` while the parts that don't are called `{bm} smooth endoplasmic reticulum`.

```{note}
It's called rough endoplasmic reticulum because the ribosomes make the surface look rough.
```

```{img}
0313_Endoplasmic_Reticulum.jpg
a) The ER is a winding network of thin membranous sacs found in close association with the cell nucleus. The smooth and rough endoplasmic reticula are very different in appearance and function (source: mouse tissue). (b) Rough ER is studded with numerous ribosomes, which are sites of protein synthesis (source: mouse tissue). EM × 110,000. (c) Smooth ER synthesizes phospholipids, steroid hormones, regulates the concentration of cellular Ca++, metabolizes some carbohydrates, and breaks down certain toxins (source: mouse tissue). EM × 110,510. (Micrographs provided by the Regents of University of Michigan Medical School © 2012)
By OpenStax - https://cnx.org/contents/FPtK1zmh@8.25:fEI3C8Ot@10/Preface, CC BY 4.0, https://commons.wikimedia.org/w/index.php?curid=30131197
scale 0.5 0.5
```

Recall that ribosomes are what translate mRNA to proteins. Since the endoplasmic reticulum is directly connected to the nucleus_cell (via pores on the nucleus_cell), it provides a fairly straight-forward path for protein generation: mRNA produced in the nucleus_cell...
1. travels to the endoplasmic reticulum via the connected pores,
2. then travels to the membrane of the endoplasmic reticulum where it ends up hitting ribosomes embedded in the (thereby producing proteins).

#### Golgi

`{bm} Golgi/(Golgi Apparatus|Golgi Complex|Golgi Body|Golgi)/i`</span> are layered membrane (organelle?) that look similar to rough endoplasmic reticulum but aren't attached to the nucleus_cell. Golgi package molecules (e.g. proteins) for travel to either another part of the cell or outside of the cell. They do this by pinching off parts of their membrane to wrap around the molecule.

They're also responsible for building lysosomes (cell digestion machines).

```{note}
The terms golgi, golgi apparatus, golgi complex, and golgi body all refer to the same thing.
```

```{img}
Golgi_apparatus_(borderless_version)-en.svg
Golgi apparatus
By Kelvinsong - Own work, CC BY 3.0, https://commons.wikimedia.org/w/index.php?curid=23090802
scale 0.15 0.15
```

#### Mitochondria
`{bm} Mitochondria/(Mitochondria|Mitochondrion)/i`</span> are organelles responsible for cellular respiration: the process of producing Adenosine Triphosphate (ATP) from molecules such as sugars. ATP is a chemical that provides energy to drive various biological processes (e.g. muscle contractions). As such, mitochondria are often referred to as "the power house of the cell."

```{img}
Animal_mitochondrion_diagram_en.svg
Mitochondria
By Mariana Ruiz Villarreal LadyofHats - the diagram i made myself using adobe illustrator. as a source for the information i used the diagrams found here:[1], [2], [3], [4], [5], [6] and [7]., Public Domain, https://commons.wikimedia.org/w/index.php?curid=8152599
scale 0.75 0.75

bg_color #00000000

fg_color #ff00ffff
rect 0.13 0.12 0.3 0.07
fg_color #ff00ffff
rect 0.7 0.75 0.3 0.12

fg_color #008080ff
rect 0.3 0.18 0.15 0.07

fg_color #808000ff
rect 0.13 0.35 0.13 0.07
```

```{note}
Mitochondria exist in both animal and plant cells. 
```

The major parts of chloroplast are...

 * <span style="color:#ff00ffff">`{bm} mitochondrial envelope`</span> - 2 membrane layers that have a gap between them. Holds the guts in.
 * <span style="color:#008080ff">`{bm} matrix`</span> - Internal fluid of mitochondria. Contains the guts (ribosomes, mitochondrial DNA, and more).
 * <span style="color:#808000ff">`{bm} cristae`</span> - Long caverns that encapsulate and connect different parts of the matrix.

Mitochondria have their own independent DNA (different from the DNA in the nucleus_cell). It's speculated that at some point in the past they may have been independent single-cell organisms that formed a symbiotic relationship with a larger cell by `{bm-ri} living` in it, eventually becoming part of the cell (endosymbiosis).

Unlike how normal offspring DNA gets produced by mixing DNA from both parents, mitochondrial DNA comes entirely from the mother's side.


#### Lysosome

`{bm} Lysosome`s are organelles (animal cells only) that help break down waste `{bm-ri} products` / foreign substances by containing various enzymes and maintaining an acidic pH. Lysosomes are more often found in animals cells than plant and algae cells.

```{note}
According to the material, the evidence that they've been found in plant cells is recent.
```

```{img}
Lysosome.jpg
Structure of Lysosome
By lumoreno - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=17380475
crop 0.05 0.05 0.9 0.9
scale 0.5 0.5
```

#### Peroxisome

`{bm} Peroxisome`s are organelles that are similar to Lysosomes -- both are small organelles that break down unwanted substances. The difference is that peroxisomes have different types of enzymes that require oxygen (oxidative enzymes).

```{note}
The material says that peroxisomes make hydrogen peroxide: Similarly, structures called peroxisomes carry out chemical reactions called oxidation reactions and produce hydrogen peroxide, both of which would damage the cell if they weren’t safely stored away in their own “room.”
```

```{img}
Peroxisome.svg
Basic structure of a peroxisome, showing the crystallized enzyme core as found in rat liver cells.
By Qef - Own work by uploader, based on the arrangement of a bitmap equivalent by Anthony Atkielski (Agateller), Public Domain, https://commons.wikimedia.org/w/index.php?curid=7072127
scale 0.5 0.5
```

#### Chloroplast

`{bm} Chloroplast`s are organelle (only plant and algae cells) responsible for photosynthesis. Photosynthesis is the process of taking in light and using it to build sugars from water and carbon dioxide. Those sugars are then used by the mitochondria to produce energy in a process called cellular respiration.

```{img}
Chloroplast_structure.svg
Ultrastructure of a chloroplast.
By Kelvinsong - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=26147364
scale 0.5 0.5

bg_color #00000000

fg_color #ff00ffff
rect 0.025 0.13 0.3 0.15
fg_color #808000ff
rect 0.24 0.015 0.3 0.15
fg_color #008080ff
rect 0.85 0.2 0.15 0.07
fg_color #ff0000ff
rect 0.035 0.35 0.15 0.07
```

The major parts of chloroplast are...

 * <span style="color:#ff00ffff">`{bm} chloroplast envelope`</span> - 2 membrane layers that have a gap between them. Holds the guts in.
 * <span style="color:#008080ff">`{bm} stroma`</span> - Internal fluid of chloroplast (chloroplast's version of cytosol).
 * <span style="color:#808000ff">`{bm} thylakoid`</span> - Membrane-bound discs. The membrane contains light-harvesting substances while the space inside (also called `{bm} lumen`) is hollow.
 * <span style="color:#ff0000ff">`{bm} grana/(grana|granum)/i`</span> - Stack of thylakoids. These stacks are interconnected. The singular of grana is granum.

 `{bm} Chlorophyll` is a pigment / compound found in chloroplast that absorbs light and uses it to produce carbohydrates. It's found in the thylakoid membrane as well as the stroma, and it only absorbs red and blue light (while reflecting green).

 Like mitochondria, chloroplast have their own independent DNA (different from the DNA in the nucleus_cell). It's speculated that at some point in the past they may have been independent single-cell organisms that formed a symbiotic relationship with a larger cell by `{bm-ri} living` in it, eventually becoming part of the cell (endosymbiosis). A descendant of that organism may be `{bm} cyanobacterium/(cyanobacterium|cyanobacteria)/i`, which has a similar ability to generate sugars from light (see [Wikipedia](https://en.wikipedia.org/w/index.php?title=Chloroplast&oldid=918130624#Parent_group:_Cyanobacteria)).

#### Vacuole

`{bm} Vacuole`s are organelle (mostly plant and algae cells) responsible for storage (water, food, waste?) and enzymes that help break things down. Vacuoles are typically found in plant and algae cells, but may also exist in animal cells. The ones in plants / algae tend to be much larger.

Vacuoles are often responsible for a plant's shape. For example, a well watered plant will be upright and spry because its vacuoles are full. A plant that isn't as well watered may be sagging down or wilting because the vacuoles are less full

```{img}
Turgor_pressure_on_plant_cells_diagram.svg
By LadyofHats - did it myself based on [1], [2] ,[3] and [4]., Public Domain, https://commons.wikimedia.org/w/index.php?curid=1685428
In biology, turgor pressure or turgidity is the pressure of the cell contents against the cell wall, in plant cells, determined by the water content of the vacuole, resulting from osmotic pressure. 
scale 0.7 0.7
```

### Chromosome

The genome of eukaryotes are split into linear strands of DNA. These linear DNA strands come in 2 forms...

1. `{bm} Chromatin` is the normal state, where the DNA is loosely floating around with structural proteins called `{bm} histone`s.
1. `{bm} Chromatid` is the state where it's going through replication. The DNA is packed using even more structural proteins called  `{bm} scaffold protein`s and the entire thing has a copy of itself attached. The original and the copy are individually referred to as `{bm} sister chromatid`s. Together they're referred to as a `{bm} sister chromatid pair`.

Chromatin and chromatid are mutually exclusive. As long as it has a copy attached to it, it's referred to as chromatid. Otherwise, it's referred to as chromatin.

The term `{bm} chromosome` can refer either to a chromatin or a sister chromatid pair, but never a single sister chromatid.

```{img}
Chromatin_Structures.png
By Original uploader was Richard Wheeler at en.wikipedia - Transferred from en.wikipedia to Commons by sevela.p., CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=4017531
The major chromatin structures.
scale 0.5 0.5
```

Each chromosome codes for the same set of genes. Since eukaryotes reproduce sexually (mother and father required), they contain genetic information from both the mother and father. The way this manifests itself is that most eukaryotes will come with multiple versions of the same chromosome. In...

 * humans, that's 2n chromosomes (2 chromosomes that are alternate versions of each other, called diploid).
 * louts plants, that's >2n chromosomes (2 or more chromosomes that are alternate versions of each other, called polyploid).

These chromosomes are referred to as `{bm} homologous chromosomes/(homologous chromosome|homologous chromosome pair|homologous pair of chromosomes|chromosomes of a homologous pair)/i`: 2 or more chromosomes that code for the same set of genes but aren't exactly the same (different versions of the same genes). In other words, the chromosomes are homologous.

```{note}
An example straight from the material of homologous pair of chromosomes working: As a real example, let's consider a gene on chromosome 9 that determines blood type (A, B, AB, or O). It's possible for a person to have two identical copies of this gene, one on each homologous chromosome—for example, you may have a double dose of the gene version for type A. On the other hand, you may have two different gene versions on your two homologous chromosomes, such as one for type A and one for type B (giving AB blood).

A different version of the same gene is called an allele.
```

When creating cells for reproduction (gamete cells), the cells only keep 1 copy from each each homologous chromosome pair. These cells are referred to as haploid cells because they only carrying n chromosomes.

```{note}
To see how gametes are formed, see the meiosis section.
```

To recap, a cell can have a number of chromosome copies:

 * `{bm} diploid` cells carry 2n chromosomes (2 copies of each chromosome). 1 copy comes from the mother and the other comes from the father, and the copies aren't identical -- they code for different versions of the same genes. Each of the 2 copies are said to be a homologous chromosome pair. Examples include animal somatic cells and animal germ cells.
 * `{bm} haploid` cells carry n chromosomes (1 copy of each chromosome). Examples include animal gamete cells and most unicellular eukaryotes.
 * `{bm} polyploid` cells carry An chromosomes, where A > 2 (more than 2 copies of each chromosome). Examples include certain plant species (e.g. lotus plants).

```{note}
An organism has an extra / missing copy of a chromosome (e.g. 2 expected but 3 present) is called aneuploid.
```

The following diagram shows a karyotype of a human's diploid homologous chromosome pairs 1 2 and 3:

```{img}
DNA_human_male_chromosomes.gif
By National Human Genome Research Institute, http://www.genome.gov/Images/EdKit/bio1c_large.gif, Public Domain, https://commons.wikimedia.org/w/index.php?curid=2132905
DNA, human male chromosomes
scale 0.5 0.5
crop 0 0 0.5 0.25
```

Certain eukaryotic species (some mammals/snakes/insects/etc..) have an extra pair of chromosomes that aren't alternate versions of each other but instead are totally different and used to determine the sex of the offspring. This extra pair is called `{bm} sex chromosome/(sex chromosome|sex determining chromosome|sex-determining chromosome)/i`s / `{bm} XY chromosome/(XY chromosome|X\/Y chromosome|X-Y chromosome)/i`s, and it determines the sex of the organism. The X and the Y refer to the chromosome types that can appear in the pair -- XX results in a female, while XY results in a male.

The chromosomes that make up the homologous chromosomes are sometimes referred to as `{bm} autosome/(autosome|autosomal chromosome)/i`s while the ones that make up sex-determining chromosomes (non-homologous chromosomes) are called `{bm} allosome/(allosome|allosomal chromosome)/i`s. For example, in humans there are 44 autosomes (22 homologous chromosome pairs) and 2 allosomes (1 sex-determining chromosome pair).

```{img}
DNA_human_male_chromosomes.gif
By National Human Genome Research Institute, http://www.genome.gov/Images/EdKit/bio1c_large.gif, Public Domain, https://commons.wikimedia.org/w/index.php?curid=2132905
DNA, human male chromosomes
scale 0.3 0.3

fg_color #ff0000ff
bg_color #00000000
rect 0.85 0.78 0.14 0.215
```

An organism has an extra / missing copy of a chromosome (e.g. 2 expected but 3 present) is called `{bm} aneuploid`. Depending on the species and circumstances, aneuploidy may lead to death, disease, or possibly no adverse effects at all. For example, allosomal aneuploidy in certain female mammals rarely leads to adverse effects because of X-linked inactivation.


### Mitosis (Cell Cycle)

The `{bm} cell cycle` is the sequence of events a cell goes through from when it's created (divides off) up to when it divides itself. In eukaryotic cells, the cell cycle has 2 major phases:

 1. Interphase: The majority of a cell's life is spent in interphase, where it's growing and going about its business.
 1. Mitotic phase: The portion of a cell's life where it divides into 2 new cells. 

Each phase has a set of inner phases it goes through.

```{img}
Animal_cell_cycle-en_UNCLUTTERED.svg
By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076
Animal cell cycle
scale 0.35 0.35
```

#### Interphase

The majority of a cell's life is spent in `{bm} interphase`, where it's growing and going about its business.

```{img}
Animal_cell_cycle-en_UNCLUTTERED.svg
By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076
Animal cell cycle
scale 0.35 0.35
crop 0 0 0.446 1
```

1. `{bm} G1 phase` / `{bm} Gap 1 phase` /  `{bm} Gap I phase`
  
   ```{img}
   G1 Phase.svg
   By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076
   G1 phase showing the cell growing
   scale 0.1 0.1
   ```

   The cell grows.
   
   Prior to entering the next phase, the cell can go into the `{bm} G0 phase` / `{bm} resting phase`,  where it's essentially pausing division. One reason for this may be `{bm} contact inhibition`: when cells start touching up on other cells, it's a signal that it's becoming too crowded and that they should stop dividing. 

1. `{bm} S phase` / `{bm} Synthesis phase`
  
   ```{img}
   S Phase.svg
   By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076
   S Phase showing chromatin and centrosome replication
   scale 0.1 0.1
   ```

   The chromatin in the nucleus_cell gets replicated to become a sister chromatid pair (2 copies that are attached). The centrosome also gets duplicated.

1. `{bm} G2 phase` / `{bm} Gap 2 phase` /  `{bm} Gap II phase`

   ```{img}
   G2 Phase.svg
   By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076
   G2 phase showing the cell growing even more
   scale 0.1 0.1
   ```

   The cell grows more.

#### Mitotic Phase

The `{bm} mitotic phase/(Mitotic phase|Mitotic|Mitosis)/i` is portion of a cell's life where it divides into 2 new daughter cells.

```{img}
Animal_cell_cycle-en_UNCLUTTERED.svg
By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076
Animal cell cycle
scale 0.35 0.35
crop 0.446 0 0.554 1
```

1. `{bm} Prophase` / `{bm} Prometaphase`

   ```{img}
   Prophase.svg
   By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076
   Prophase showing the sister chromatid condensing and nucleus membrane breaking down
   scale 0.1 0.1
   ```

   The sister chromatid pairs start to condense into an X shape, where the only part they remain attached at is their centromeres. The membrane of the nucleus_cell goes away and the 2 centrosomes that were originally attached to that membrane get moved to opposite ends of the cell.

   Microtubules are attached from the centrosomes to the kinetochore of the sister chromatid pairs. `{bm} Kinetochore`s are proteins located at the centromere of the sister chromatids.

1. `{bm} Metaphase`

   ```{img}
   Metaphase.svg
   By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076
   Metaphase showing the sister chromatid lining up
   scale 0.1 0.1
   ```

   The sister chromatid pairs line up in the middle of the cell.

1. `{bm} Anaphase`

   ```{img}
   Anaphase.svg
   By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076
   Anaphase showing the sister chromatid breaking apart
   scale 0.1 0.1
   ```

   As the cell splits apart, the microtubules also split apart the sister chromatid pairs. Once split, they're individually referred to as chromatin again.

   ```{note}
   I believe the labeling on the above diagram to be incorrect. Once split, they're no longer referred to as sister chromatids / chromatids. They're referred to as chromatin.
   ```

1. `{bm} Telophase` and `{bm} Cytokinesis`

   ```{img}
   Telophase and Cytokinesis.svg
   By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076
   Telophase / cytokinesis showing nucleus membranes re-forming and cell finally splitting into 2
   scale 0.1 0.1
   ```

   During telophase, nucleus_cell membranes re-form around the newly split chromatin. The chromatin also start to un-condense.

   During cytokinesis, the middle of the cell starts to push together and pinch off, forming 2 separate cells. Cytokinesis technically happens in parallel with mitosis (it isn't a part of mitosis), but it starts near the end of mitosis (around telophase).

### Meiosis

`{bm-ambiguous} Add the suffix _GAMETE if referring to egg or ova cells, or _NORM/\b(egg)/i`
`{bm-ambiguous} Add the suffix _GAMETE if referring to egg or ova cells, or _NORM/\b(ova)/i` 
`{bm-ignore} (egg)_NORM/i`
`{bm-ignore} (ova)_NORM/i`

`{bm} Meiosis` is a type of eukaryotic cell division that results in 4 daughter cells, where each daughter cell has half the number of chromosomes as the parent cell.

The source cell is known as a `{bm} germ` cell -- a cell that can either go through mitosis or meiosis. `{bm} Somatic` cells, on the other hand, are general body cells and can only go through mitosis.

The resulting daughter cells are known as `{bm} gamete` cells -- cells that have half the genetic information from the original parent (haploid -- only 1 chromosome of from each of the parent's homologous chromosome pairs), and when they merge they mix that genetic material to create the new genetic material for the offspring. Male gamete cells are called `{bm} sperm`, while female gamete cells are called `{bm} ova/\b(ova)_GAMETE/i` or `{bm} egg/\b(egg)_GAMETE/i`s.

```{note}
Gamete cells no longer have the ability to divide (this is a terminal operation)
```

Meiosis has 2 major phases:
 1. Meiosis I phase: Germ cell mixes genetic information in itself and divides into 2.
 1. Meiosis II phase: Each of the 2 cells created above get divided again. But, this time the resulting daughter cells only have half the generic information (gamete cells). 

```{img}
Meiosis_Stages.svg
By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
Meiosis Stages

scale 0.35 0.35
```

#### Meiosis I

During `{bm} meiosis I`, a germ cell divides in such a way that, rather than exactly duplicating each chromosome, it duplicates and mixes between each homologous chromosome pair. That is, chromosome pairs that are different versions of each other end up swapping segments (chromosomal crossover).

```{note}
Remember that most cells have a diploid number of chromosomes (2n), where each chromosome is a different version of another chromosome in that same cell (homologous chromosome pairs -- e.g. code for different versions of the same gene). One comes from the mother and one comes from the father.
```

```{img}
Meiosis_Stages.svg
By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
Meiosis Stages

scale 0.35 0.35
crop 0 0 0.5260 1.0
```

 1. `{bm} Prophase I`

    ```{img}
    Meiosis_Stages.svg
    By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
    Meiosis Stages

    scale 0.5 0.5
    crop 0 0 0.15 1.0
    ```

    Each chromatin duplicates to become a sister chromatid pair (2 chromatids that are attached to each other). Then, each sister chromatid pair condenses and forms an X shape where the only part that the sister chromatids remain attached at is their centromeres.
    
    The membrane of the nucleus_cell goes away and the 2 centrosomes that were originally attached to that membrane get moved to opposite ends of the cell. Microtubules are attached from the centrosomes to the kinetochore of the sister chromatid pairs.

    While in their condensed form, sister chromatid pairs go through a process called `{bm} chromosomal crossover`: A chromatid from a sister chromatid pair swaps segments of itself with a chromatid from another sister chromatid pair. The sister chromatid pairs doing the swapping must be homologs of each other (contains alternative versions of the same genes).

    ```{note}
    Remember that the source cell comes with 2n chromosomes, where each chromosome has a sibling that has alternate versions of the same genes. One sibling came from the father and the other sibling came from the mother. The homolog refers to this sibling -- each sibling contains different versions of the same set of genes.
    ```

    ```{img}
    Chromosome Crossover.svg
    This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
    Chromosome Crossover Diagram

    scale 1.45 1.45
    ```

    The end result is that the sister chromatids are no longer exact copies of each other.
   
    The point at which a segment swap happens is called the `{bm} chiasma/(chiasmata|chiasma)/i`. The chiasma is a well defined point on the chromatids (it isn't random).
 
 1. `{bm} Metaphase I`

    ```{img}
    Meiosis_Stages.svg
    By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
    Meiosis Stages

    scale 0.5 0.5
    crop 0.15 0 0.1255 1.0
    ```

    The sister chromatid pairs line up in the middle of the cell.

 1. `{bm} Anaphase I`

    ```{img}
    Meiosis_Stages.svg
    By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
    Meiosis Stages

    scale 0.5 0.5
    crop 0.265 0 0.1355 1.0
    ```

    As the cell splits apart, microtubules move apart the homologous pairs of sister chromatid pairs to opposite ends of the cell. Each side gets 1 of the of sister chromatid pairs.

    The term _homologous pairs of sister chromatid pairs_ here is confusing:
     * _sister chromatid_ means 1 of 2 chromatids that are attached to each other at their centromeres.
     * _sister chromatid pair_ means 2 _sister chromatids_ that are attached to each other at their centromeres.
     * _pair of sister chromatid pairs_ means 2 _sister chromatid pairs_.
     * _homologous pair of sister chromatid pairs_ refers to a 2 _sister chromatid pairs_ that were derived from 2 homologous chromatin/chromosomes.

    ```{note}
    What does homologous chromatin mean? Remember that the source cell comes with 2n chromatin, where each chromatin has a sibling that has alternate versions of the same genes. One sibling came from the father and the other sibling came from the mother. Homologous chromatin/chromosomes refers to those 2 siblings.
    ```

 1. `{bm} Telophase I`

    ```{img}
    Meiosis_Stages.svg
    By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
    Meiosis Stages

    scale 0.5 0.5
    crop 0.4005 0 0.1255 1.0
    ```

    Nucleus_cell membranes re-form around the newly moved apart sister chromatid pairs. The sister chromatid pairs also start to un-condense but still remain attached at their centromeres. As such, the resulting daughter cells are said to have a haploid number of chromosomes -- each sister chromatid pair is technically considered to be 1 chromosome until it splits apart.

    During cytokinesis, the middle of the cell starts to push together and pinch off, forming 2 separate cells. Cytokinesis technically happens in parallel with telophase I (it isn't a part of telophase I), but it starts around the same time.

#### Meiosis II

During `{bm} meiosis II`, each daughter cell from meiosis I will divide again, but this time the sister chromatid pairs split and each resulting daughter cell keeps 1. In other words, each daughter cell from meiosis I will divide again to become 2 gamete cells.

The steps in meiosis II are almost exactly the same as the steps in the miotic phase -- the sister chromatid pairs are essentially being split apart into 2 new cells. Similar to how there's a rest period between mitosis iterations (called interphase), there may be a rest period between meiosis I and meiosis II called `{bm} interphase II`.

```{img}
Meiosis_Stages.svg
By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
Meiosis Stages

scale 0.35 0.35
crop 0.5260 0 0.4740 1.0
```

 1. `{bm} Prophase II`

    ```{img}
    Meiosis_Stages.svg
    By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
    Meiosis Stages

    scale 0.5 0.5
    crop 0.5260 0 0.1255 1.0
    ```

    The sister chromatid pairs start to condense into an X shape, where the only part they remain attached at is the centromere. The membrane of the nucleus_cell goes away and the 2 centrosomes that were originally attached to that membrane get moved to opposite ends of the cell.

    Microtubules are attached from the centrosomes to the kinetochore of the sister chromatid pairs.

 1. `{bm} Metaphase II`

    ```{img}
    Meiosis_Stages.svg
    By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
    Meiosis Stages

    scale 0.5 0.5
    crop 0.6515 0 0.12 1.0
    ```

    The sister chromatid pairs line up in the middle of the cell.

 1. `{bm} Anaphase II`

    ```{img}
    Meiosis_Stages.svg
    By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
    Meiosis Stages

    scale 0.5 0.5
    crop 0.77 0 0.1 1.0
    ```

    As the cell splits apart, the microtubules also split apart the sister chromatid pairs. Once split, they're individually referred to as chromatin again.

 1. `{bm} Telophase II`

    ```{img}
    Meiosis_Stages.svg
    By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece & Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204
    Meiosis Stages

    scale 0.5 0.5
    crop 0.8770 0 0.12 1.0
    ```

    During telophase, nucleus_cell membranes re-form around the newly split chromatin. The chromatin also start to un-condense. The resulting daughter cells are gametes at this point (e.g. sperm cells). They no longer divide. 

    During cytokinesis, the middle of the cell starts to push together and pinch off, forming 2 separate cells. Cytokinesis technically happens in parallel with telophase II (it isn't a part of telophase II), but it starts around the same time.

### Cancer

`{bm} Cancer` refers to the uncontrolled division of cells (mitosis) in a multicellular organism, typically brought on by one or more mutations.

Normally, a cell has internal and external cues / regulators that signal when it should stop dividing (e.g. contact inhibition). When a mutation occurs that causes one or more of these cues to be ignored, the cell goes through a form a programmed suicide called apoptosis. If the genetic mutations are so severe that apoptosis no longer occurs, that's when cancer occurs.

A group of cells that have been dividing unimpeded is known as a `{bm} neoplasm` / `{bm} tumor`. If the cells...
 * eventually stop dividing on their own, that mass of cells is called a `{bm} benign neoplasm` / `{bm} benign tumor`.
 * continually divide without ever stopping, that mass of cells is called a `{bm} malignant neoplasm` / `{bm} malignant tumor`.

Mutations of two types of cell cycle regulators can promote the development of cancer:
 * Positive regulators, which normally promote cell growth, may become hyperactive (oncogenic).
 * Negative regulators (tumor suppressors), which prevent the formation of tumors, may become inactivated.

Tumor cells may continue to mutate at a more rapid pace than other cells. If one of those mutations results in tumor cells breaking off from the original mass of tumor cells and floating around the body, the cancer is said to have `{bm} metastasize`d. 

### Fertilization

`{bm} Fertilization/(fertilization|fertilize)/i` is the act of bringing together a male gamete (e.g. sperm) to a female gamete (e.g. egg_GAMETE) (cell types produced through meiosis). In humans, a fertilized egg_GAMETE is referred to as a ...

 * `{bm} zygote` upon creation. At this point the male gamete and the female gamete have fused together to become a single cell, but their nuclei_cell will still be separate (will eventually fuse).

   ```{img}
   Zygote1.jpg
   Zygote with 2 separate nuclei
   By Nina Sesina - https://commons.wikimedia.org/wiki/File:Zygote.tif, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=67459911

   scale 0.1 0.1
   ```

 * `{bm} morula` upon dividing to become 16 cells (around a few days).

 * `{bm} blastocyst` upon dividing to become 200-300 cells (around 5 to 9 days). At this point the group of cells form a hollowed out sphere where a mass is growing on inner top portion.

   ```{img}
   Human_blastocyst.jpg
   Blastocyst
   By Mr. J. Conaghan - http://stemcells.nih.gov/info/scireport/pages/chapter3.aspx, Public Domain, https://commons.wikimedia.org/w/index.php?curid=32289210

   scale 0.65 0.65
   ```

 * `{bm} embryo` after around 2 weeks. At this point the major internal organs and overall shape of the human are starting to emerge (e.g. the beginning of legs, arms, eyes, etc..).

   ```{img}
   Embryo_7_weeks_after_conception.jpg
   Embryo
   By Nina Sesina - https://commons.wikimedia.org/wiki/File:Zygote.tif, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=67459911

   scale 0.1 0.1
   ```

 * `{bm} fetus/(fetal|fetus|foetus|foeti)/i` after around 8-10 weeks. At this point the features formed in the embryonic stage grow and cells begin to differentiate for their actual function / purpose (e.g. neuron cells, kidney cells, liver cells, etc..).

   ```{img}
   Fetus_3_months.jpg
   Fetus
   By National Museum of Health and Medicine - http://nmhm.washingtondc.museum/exhibits/single_cell/imgs/14_Fetus_3_months.jpg, Public Domain, https://commons.wikimedia.org/w/index.php?curid=9998241

   scale 0.4 0.4
   ```

The table below is a visualization of the phases. Note that the table starts count from the `{bm} gestational age`: the point of last menstruation. The durations in the descriptions above are from the point of fertilization.

```{img}
Prenatal_development_table.svg
Prenatal Development Table
By Mikael Häggström.When using this image in external works, it may be cited as:Häggström, Mikael (2014). "Medical gallery of Mikael Häggström 2014". WikiJournal of Medicine 1 (2). DOI:10.15347/wjm/2014.008. ISSN 2002-4436. Public Domain.orBy Mikael Häggström, used with permission. - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=6843176

scale 0.5 0.5
```

### Apoptosis

`{bm} Apoptosis` is a form of programmed cell death in multicellular organisms. For example, apoptosis can be triggered by the cell...

* detecting a faulty mutation within its own DNA, there by avoiding a potential problem (e.g. cancer).
* to sculpt body parts as part of normal development (e.g. cells in feet die during embryonic / fetal development to create gaps between toes).

In apoptosis, the membrane of the cell begins to shrink and pinch inward. It breaks up its DNA and organelle and encircles them into miniature components that eventually bud off from the cell and float away. Cells from the immune system will then come and ingest those buds and/or other cells can reuse the components in those buds for their own purposes without suffering any damage.

```{img}
Apoptotic_cell_disassembly.png
Apoptotic Cell Disassembly
By Aaron Smith, Michael AF Parkes, Georgia K Atkin-Smith, Rochelle Tixeira, Ivan KH Poon - Wikiversity:Draft:WikiJournal of Medicine/Cell disassembly during cell death, CC BY 4.0, https://commons.wikimedia.org/w/index.php?curid=59865845

scale 0.5 0.5
```

In contrast, `{bm} necrosis` is a form of cell death that isn't programmed -- the cell swells up and explodes. It isn't desirable and often occurs in extreme cases (e.g. exposed to a chemical toxin, mechanical damage, etc..). The guts of the cell, rather than being neatly encircled and budding off, will burst outwards and potentially damage other cells in the vicinity. 

```{img}
Structural_changes_of_cells_undergoing_necrosis_or_apoptosis.png
Apoptosis vs Necrosis
By National institute on alcohol abuse and alcoholism (NIAAA) - File:Structural changes of cells undergoing necrosis or apoptosis.gif; (pubs.niaaa.nih.gov), Public Domain, https://commons.wikimedia.org/w/index.php?curid=24184862

scale 0.5 0.5
```

## Prokaryotic Cell

```{img}
Average_prokaryote_cell-_en.svg
Prokaryotic cell
By Mariana Ruiz Villarreal, LadyofHats - Own work (Source: Typical prokaryotic cell, Chapter 4: Mutagenicity of alkyl N-acetoxybenzohydroxamates, Concept 1: Common Features of All Cells, Cells - Structure and Function), Public Domain, https://commons.wikimedia.org/w/index.php?curid=3648821
```

`{bm} Prokaryotic/(Prokaryote|Prokaryotic)/i` cells: These cells are typically smaller and don't have organelles. For example, their DNA is free-floating in the cell (it's free floating but stays mostly in the center area called the  `{bm} nucleoid`).

Prokaryotes have a single circular-strand of DNA. They can only be single-cellular organisms (e.g. bacteria).

## Enzyme

An `{bm} enzyme` is a molecule that takes in a specific set of input molecules and transforms them into a specific set of output molecules. The transformation takes the inputs and either ...
 * assembles them into new larger molecules (N to 1).
 * separates them into smaller molecules (1 to N).
 * transfers fragments between them (N to N).

Enzymes facilitate these transformations by lowering the `{bm} activation energy` (`{kt} E_A`) required for the chemical reactions to take place.  Normally this excess energy would come in the form of heat, but enzymes use different mechanisms such as...
* temporarily contorting input molecules
* temporarily pulling away some charge from adjacent atoms of input molecules

... such that other atoms can get close enough to bond.

```{note}
How does heat provide activation energy? More heat = more molecules moving faster = more things bumping into each other faster. 2 molecules may have atoms that want to bond but neighbouring atoms on those molecules may be repelling away with stronger force. Increased speed means the repelling is less effective. 
```

An enzyme is almost always a protein molecule but can also be a RNA-like molecule called a `{bm} ribozyme`.

The general terminology for enzymes are as follows:

 * `{bm} Substrate` / `{bm} Reactant/(reactant)_ENZYME/i`s refers to the input molecules of an enzyme.
 * `{bm} Product/(product)_ENZYME/i` refers to the output molecules of an enzyme.
 * `{bm} Metabolite` refers to either a substrate or product_ENZYME.
 * `{bm} Metabolism` is the total sum of all the chemical reactions that take place inside of a cell. These reactions happen via enzymes.
 * `{bm} Active site`s are the physical locations that substrates float into on an enzyme.
 * `{bm} Catalysis` / `{bm} Catalyze` is the term for increasing the rate / probability of a chemical reaction. Enzymes perform catalysis.
 * `{bm} Catalyst`s are entities that speed up chemical reactions but don't change in the process (reaction happens but catalyst remains). Enzymes are catalysts.

Enzymes have a limited set of substrate types that they accept. A substrate will bind to the active site of the enzyme only if it fits into the active site. For example, the following diagram shows 2 substrates binding to an enzyme, the enzyme facilitating their their assembly, then releasing back out.

```{img}
Anabolism.svg
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
Anabolism process.
scale 0.5 0.5
```

It was previously thought that enzymes had a “lock-and-key” model, similar to how puzzle pieces fit together. Later on it was found out that an enzymes actually `{bm} induce fit/(induced fit|induce fit)/i` by changing shape slightly when they bind with substrates, such that they can better hold on to those substrates.

Examples of enzymes and what they do:

 * protease → breaks up protein molecules to smaller protein molecules.
 * sucrase → breaks up sucrose molecules to simpler sugar molecules.
 * RNA polymerase → scans over DNA... splitting it, generating mRNA, and stitching it back up.

```{img}
Label_RNA_pol_II.png
RNA Polymerase II
By JWSchmidt - http://www.rcsb.org/pdb/static.do?p=general_information/news_publications/newsletters/2003q2/mom.htmlhttps://en.wikiversity.org/wiki/File:Label_RNA_pol_II.png, Public Domain, https://commons.wikimedia.org/w/index.php?curid=32033313
scale 0.5 0.5

```

### Metabolic Pathway

A `{bm} metabolic pathway` is a network/graph of enzymes that produces a final resulting molecule. Each enzyme produces molecules that feed into other enzymes in the pathway, eventually forming the final molecule. The term `{bm} intermediate/(intermediate)_ENZYME/` refers to an output of one enzyme that’s used as an input by another.

For example, the following graph is the metabolic pathway for gamma-hydroxybutyric acid...

```{img}
GHB_metabolic_pathway.svg
By Anypodetos - Own work, vectorised version of File:GHB metab path.png by User:Meodipt, Public Domain, https://commons.wikimedia.org/w/index.php?curid=8988213
Example metabolic pathway
scale 0.5 0.5
```

### Anabolism

Metabolism can be broken down into 2 categories: anabolism (building-up) and catabolism (breaking-down).

The process that builds up a molecule from smaller molecules is called `{bm} anabolism/(anabolic|anabolism)/i`. An enzyme takes in the molecules and creates bonds between them via an `{bm} endergonic reaction`s: energy is stored as bonds between the smaller molecules thereby forming the larger molecule.

```{note}
A good way to remember the reaction types... In ENDergonic reactions, the energy ENDs up in a bond. In EXergonic reactions, the energy EXplodes out thereby breaking the bond.
```

```{img}
Anabolism.svg
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
Anabolism process.
scale 0.5 0.5
```

An example of anabolism is photosynthesis: plants will bond carbon dioxide gas (`{kt} CO_2`) with water (`{kt} H_2O`) using energy from the sun, creating sugar (`{kt} C_6H_{12}O_6`)

`{kt} 6CO_2 + 6H_2O + energy \to C_6H_{12}O_6 + 6O_2`

### Catabolism

Metabolism can be broken down into 2 categories: anabolism (building-up) and catabolism (breaking-down).

The process that breaks down a large molecule into smaller molecules is called `{bm} catabolism/(catabolic|catabolism)/i`. An enzyme takes in a larger molecule breaks up some of its bonds via `{bm} exergonic reaction`s: energy used as bonds in the molecule are release thereby breaking it up into smaller molecules.

```{note}
A good way to remember the reaction types... In ENDergonic reactions, the energy ENDs up in a bond. In EXergonic reactions, the energy EXplodes out thereby breaking the bond.
```

```{img}
Catabolism.svg
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
Catabolism process.
scale 0.5 0.5
```

An example of catabolism is cellular respiration: cells will break down the bonds in glucose (`{kt} C_6H_{12}O_6 + 6O_2`) to release energy, splitting into carbon dioxide (`{kt} CO_2`) and water (`{kt} H_2O`)

`{kt} C_6H_{12}O_6 + 6O_2 \to 6CO_2 + 6H_2O + energy`

## Molecular Genetics

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

TODO: Fill me in after learning enough chemistry

### Nucleotides

`{bm} Nucleic Acid` is a molecule (heteropolymer) that's built up from other molecules called `{bm} nucleotide`s (monomers). Nucleic acid comes in 2 flavours: DNA and RNA. Each nucleotide consists of a sugar molecule (ribose in RNA / deoxyribose in DNA) attached to a phosphate group and a nitrogen-containing base_nucleotide.

```{note}
It's called nucleic acid because it has some acidic properties to it and DNA is found in the nucleus_cell of a eukaryotic cell. But DNA also in prokaryotic cells and some organelles -- those don't have a nucleus_cell.
```

```{img}
0322_DNA_Nucleotides.jpg
By OpenStax - https://cnx.org/contents/FPtK1zmh@8.25:fEI3C8Ot@10/Preface, CC BY 4.0, https://commons.wikimedia.org/w/index.php?curid=30131206
DNA Nucleotides
scale 0.75 0.75
```

There are 5 `{bm} base/(bases|base)_nucleotide/i`.:
 * A (`{bm} adenine`)
 * C (`{bm} cytosine`)
 * G (`{bm} guanine`)
 * T (`{bm} thymine`) / U (`{bm} uracil`)

```{note}
T only appears in DNA. In RNA, T is replaced by U.
```

```{note}
The base_nucleotide is what distinguishes the types of nucleotides from each other. The term nucleotide and base_nucleotide are often used interchangeably.
```

 Two nucleotides connected together are called a `{bm} base pair`. The rules to base pairs are:
 * A only ever binds to T/U (e.g. AT or TA in DNA, AU or UA in RNA)
 * G only ever binds to C (e.g. GC or CG)

### DNA

`{bm} Deoxyribonucleic acid` (`{bm} DNA`) is a nucleic acid molecule that contains the instructions needed for the growth/functioning/maintenance of an organism. Depending on the type of organism, DNA is located in different parts fo the cell.
 * For prokaryotes, DNA is free-floating in the cytoplasm.
 * For eukaryotes, DNA is isolated in the nucleus_cell (`{bm} nuclear DNA`). Their mitochondria and chloroplast also have their own DNA.

```{img}
1032px-DNA_Structure+Key+Labelled.pn_NoBB.png
By Zephyris - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=15027555
DNA
scale 0.3 0.3
```

DNA is composed of two strands of nucleotides that connect at various points in between. The order these nucleotides appear in defines the genetic information/instructions of the organism. For example, a string/sequence of DNA bases_nucleotide: ATATTTTCGATATCCACCA.

DNA strands can be made up of 4 different nucleotide types (bases_nucleotide): 
 * A (adenine)
 * C (cytosine)
 * G (guanine)
 * T (thymine)

The two nucleotides that make up a connection are called a base pair. In DNA, the rules to base pairs are...
 * A only ever binds to T (e.g. AT or TA)
 * G only ever binds to C (e.g. GC or CG)

Terminology specific to DNA:
 * `{bm} genome` - The entire set of DNA for an organism. For eukaryotes, this refers to nuclear DNA, not the independent DNA carried by organelles such as mitochondria.
 * `{bm} gene/\b(genes|gene)\b/i` - A section of an organism's DNA that contains instructions for some functionality, typically for building some protein.
 * `{bm} allele` - A version of a gene (different coding at the some position of DNA).

### RNA

`{bm} Ribonucleic acid` (`{bm} RNA/(RNA)/`) is a nucleic acid molecule used in various ways to facilitate building proteins. It can also act as an enzyme (ribozyme) or contain the genetic information for some viruses.

RNA is commonly composed of a single strand that folds over onto itself.

```{img}
Pre-mRNA-1ysv-tubes.png
By Vossman - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=7115139
RNA
scale 0.1 0.1
```

RNA strands can be made up of 4 different nucleotide types (bases_nucleotide): 
* A (adenine)
* C (cytosine)
* G (guanine)
* U (uracil)

The two nucleotides that make up a connection are called a base pair. In RNA, the rules to base pairs are...
 * A only ever binds to U (e.g. AU or UA)
 * G only ever binds to C (e.g. GC or CG)

```{note}
The rules are similar to DNA, except T is replaced by U. DNA can't have U and RNA can't have T.
```

Unlike DNA, RNA is transient (lasts for minutes) and comes in multiple flavours:
 * `{bm} mRNA/(mRNA)/` or `{bm} messenger RNA` - Carries information from DNA to the ribosomes detailing a protein to build. Ribosomes are the parts of cells that synthesize proteins.
 * `{bm} tRNA/(tRNA)/` or `{bm} transfer RNA` - Brings to the ribosome the amino acids required to build the protein that the mRNA is requesting.
 * `{bm} rRNA/(rRNA)/` or `{bm} ribosomal RNA` - Makes up a large portion of the ribosome (60%). The rest of the ribosome is protein.

## Classical Genetics

`{bm} Classical genetics` is study of the probability of which genes get passed down from parents to offspring and the probability of which genes express themselves into some physically observable trait(s). It's focused exclusively on sexually reproducing eukaryotic organisms. The term...
 * `{bm} genotype` refers to the alleles present for a set of one or more genes.
 * `{bm} phenotype` refers to the observable trait(s) expressed by a genotype.

Essentially, classical genetics is the idea that the coding / alleles for a specific set of genes (genotype) results in some change in the organism that's observable (phenotype), and the probability that an organism can end up with a specific genotype/phenotype can be calculated from the genotype of its parents. For example, a certain set of alleles (genotype) may be responsible for blue eyes (phenotype).

If a single gene is responsible for contributing to multiple unrelated phenotypes, it's referred to as `{bm} pleiotropy/(pleiotropy|pleiotropic)/i`. For example, a specific gene is responsible for generating melanin. If the allele for that gene has a specific mutation in it, melanin `{bm-ri} production` stops / is drastically reduced. A lack of melanin results in albinism: a condition that effects the color of eyes, hair, and skin (3 separate phenotypes).

If multiple genes are responsible for contribution to a single phenotype, it's referred to as `{bm} polygene/(polygenic|polygene)/i`. For example, many genes are responsible for contributing to a person's height (around 400).

```{note}
Don't be fooled by the name classical genetics. The term classical doesn't mean that it's deprecated. 
```

Eukaryotic organisms that sexually reproduce hold homologous pairs of chromosomes -- they have multiple copies of chromosomes, each of which codes for the same genes but different versions of those genes (alleles). In humans, the number of chromosomes that make up a single set of homologous chromosomes is 2 (diploid). That means that a human has 2 alleles for each gene -- 1 allele is randomly chosen from the father and the other allele is randomly chosen from the mother.

```{note}
This isn't the case for XY chromosomes. See sex-linked genes section for more information on how XY chromosomes (which are not homologous chromosomes) are treated.
```

Depending on which allele combinations are present, different observable traits may be expressed. For example, imagine the color of a flower is determined by a single gene. If a flower had both a red allele and a white allele for that gene, the observable trait would be that it'd have a pink color.

```{dot}
digraph G {
	subgraph clusterparent1 {
		red1 [label="Red"];
    red2 [label="Red"];
		label = "Male parent color alleles";
	}
	subgraph clusterparent2 {
		white1 [label="White"];
    white2 [label="White"];
		label = "Female parent color alleles";
	}
	subgraph clusterchild1 {
		red3 [label="Red"];
    white3 [label="White"];
		label = "Child color alleles";
	}

  red1->red3 [label="randomly chosen from male parent"];
  white2->white3 [label="randomly chosen from female parent"];
}
```

If an organism has...
 * the same allele for each gene, it's said to be `{bm} homozygous`.
 * different alleles for each gene, it's said to be `{bm} heterozygous`.

In the above example, the...
 * male parent's genotype is 2 red alleles (homozygous) while the phenotype is red.
 * female parent's genotype is 2 white alleles (homozygous) while the phenotype is white.
 * child's genotype is 1 red allele and 1 white allele (heterozygous) while the phenotype is pink.

For each gene, offspring get a random allele from each parent. The odds of which alleles the offspring ends up with can be visualized using a `{bm} Punnett square` diagram: a table where the alleles for a gene are written across the top axis (male parent) and down the left axis (female parent), and each `{bm-ri} cell` maps to the alleles in the top/left coordinate that it's in.

For example, a gene that controls the color of a flower has 2 alleles: red and white. The male parent has 2 red alleles while the female parent has both a red allele and a white allele. The Punnett square describing the probability offspring's alleles:

```{csv}
 , R , R
R, RR, RR
W, WR, WR
```

```{note}
Unsure what the convention for Punnett squares is -- I'm always putting the allele on the left axis (female parent) first.
```

The above diagram visualizes that the probability of the offspring having...
 * 2 red alleles is `{kt} P(RR)=\frac{2}{4}`.
 * 2 white alleles is `{kt} P(WW)=\frac{0}{4}`.
 * a red allele and a white allele is `{kt} P(RW|WR)=\frac{2}{4}`.

Punnett squares can be extended to cover multiple genes so long as those genes are independently assorted. That is, each gene in the list must be on a different chromosome. If they were on the same chromosome, the chance of you getting gene A may be dependent on you getting gene B (see linked genes).

For example, a gene that controls...
* the color of a flower has 2 alleles: red and white.
* whether the flower has thorns has 2 alleles: true and false.

Both parents have 1 red allele and 1 white allele for color, and 1 true allele and 1 false allele for thorns. The Punnett square describing the odds of the offspring:

```{csv}
      , [R][T]  , [R][F]  , [W][T]  , [W][F]
[R][T], [RR][TT], [RR][RF], [RW][TT], [RW][TF]
[R][F], [RR][FT], [RR][FF], [RW][FT], [RW][FF]
[W][T], [WR][TT], [WR][TF], [WW][TT], [WW][TF]
[W][F], [WR][FT], [WR][FF], [WW][FT], [WW][FF]
```

### Laws

Classical genetics was started in the 1800s by a scientist monk named `{bm} Gregor Mendel/(Gregor Mendel|Gregor Johann Mendel)/` (prior to the discovery of DNA) with his model called `{bm} Mendelian inheritance/(Mendelian inheritance|Mendelian genetics)/i`. Gregor Mendel wasn't aware of DNA and genes (hadn't been discovered yet) and instead created his model based on some abstract idea of "inheritable characteristics." Classical genetics corrects and extends the model introduced by Gregor Mendel by taking DNA / chromosomes / genes / alleles / etc.. into account.

Mendel originally had 3 laws:

 1. `{bm} law of segregation`
 2. `{bm} law of independent assortment`
 3. `{bm} law of dominance`

The subsections below describe these laws in their modern / corrected form.

#### Segregation

```{note}
Segregation was originally discovered by Gregor Mendel as Mendelian inheritance's first law / law of segregation.
```

Segregation is the idea every sexually reproducing organism has 2 alleles for each gene, but when it comes time to reproduce only 1 allele is kept by the reproductive cell (gamete). As such, the resulting offspring gets 1 allele from its father and 1 allele from its mother.

```{note}
Unsure how this extends to organisms that are polyploid (have more than 2 homologous chromosomes / more than 2 alleles for each gene).
```

```{dot}
digraph G {
	subgraph clusterparent1 {
		blue1 [label="Blue"];
    brown1 [label="Brown"];
		label = "Male parent eye color alleles";
	}
	subgraph clusterparent2 {
		brown2 [label="Brown"];
    brown3 [label="Brown"];
		label = "Female parent eye color alleles";
	}
	subgraph clusterchild1 {
		blue2 [label="Blue"];
    brown4 [label="Brown"];
		label = "Child eye color alleles";
	}

  blue1->blue2 [label="randomly chosen from male parent"];
  brown3->brown4 [label="randomly chosen from female parent"];
}
```

The reason for this is that sexually reproducing eukaryotic species (e.g. humans) carry pairs of homologous chromosomes. That is, chromosomes come in pairs where each chromosome in the pair has the same set of genes as the other but different versions of those genes (alleles). One chromosome comes from the mother and the other from the father.

During meiosis, each gamete cell gets 1 chromosome from each homologous pair (1 set of alleles). When gametes meet to form an offspring, the chromosomes from each gamete match up to to their homolog to create that offspring's set of homologous chromosome pairs.

```{note}
There are cases where a gamete gets 2 chromosomes instead of 1. Depending on the chromosome, it may end up being fatal or cause disease (or be benign). See aneuploid section.
```

#### Independent Assortment

```{note}
Independent assortment was originally discovered by Gregor Mendel as Mendelian inheritance's second law / law of independent assortment. It states that inherited attributes are passed down from parent to child independently of each other. For example, the odds that a baby ends up with green eyes isn't linked to the odds that it'll have dark hair. It turns out that this isn't entirely correctly.
```

The high-level algorithm for how a parent passes down its genetic information is as follows:

 1. For each homologous chromosome pair, randomly swap segments between chromosomes (chromosomal crossover).

    ```{img}
    Independent Assortment (chromosomal crossover).svg
    This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
    Homologous chromosomes segment swap diagram
    ```

 2. For each homologous chromosome pair, randomly pick 1 chromosome.
 3. The picked chromosomes from parent A combine with the picked chromosomes from parent B to form the homologous chromosome pairs of the new child.

    ```{img}
    Independent Assortment (chromosomal selection).svg
    This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
    Diagram showing a random chromosome being selected (after segment swap) from each parent's homologous chromosome pair to make up the child's homologous chromosome pair.
    ```

Essentially, what the above says / shows is that alleles get passed down from parent to child in groups, where each group is a segment of one of the chromosomes. In the example above, the segments passed down are...

 * CHILD_C1 = PARENT1_C2[1,3] + PARENT1_C1[4,28]
 * CHILD_C2 = PARENT2_C2[1,5] + PARENT2_C1[6,20] + PARENT2_C2[21,28]

Genes located close to each other are more likely to get grouped together as part of the same segment being passed down -- it's very unlikely that a chromosomal crossover boundary will be at a point which splits them. These genes are referred to as `{bm} linked gene`s. The phenotypes they express are often referred to as `{bm} linked trait`s. For example, if diagram above were for a plant, and gene 11 controlled height while gene 12 controlled color, there would likely be a correlation between the height and color across that plant species.

```{note}
See section on meiosis for more in depth description of how this steps happen.
```

 ```{note}
Unsure how this extends to organisms that are polyploid (have more than 2 homologous chromosomes / more than 2 alleles for each gene).
```

#### Dominance

```{note}
Dominance was originally discovered by Gregor Mendel as Mendelian inheritance's third law / law of dominance. It states in a cross of parents that are pure for contrasting traits, only one form of the trait will appear in the next generation. Offspring that are hybrid for a trait will have only the dominant trait in the phenotype while the recessive trait remains dormant. It turns out that this isn't entirely correct (it's close) -- more scenarios have come up: incomplete dominance, co-dominance, etc..
```

In an organism that is heterozygous for some gene, some alleles may take precedence in expressing themselves over others. How these alleles are expressed define the type of dominance they have over other alleles.

Alleles that ...
 * take precedence over others are referred to as `{bm} dominant allele`s. The phenotypes they express are often referred to as `{bm} dominant trait`s.
 * surrender precedence to others are referred to as `{bm} recessive allele`s. The phenotypes they express are often referred to as `{bm} recessive trait`s.

For example, given the dominant allele D and the recessive allele d, ...
 * [Dd] results in the expression of the phenotype for D -- D takes precedence / d surrenders precedence.
 * [DD] results in the expression of the phenotype for D -- no other types to contend with.
 * [dd] results in the expression of the phenotype for d -- no other types to surrender to.

In cases where both alleles are dominant alleles, there's several different ways phenotype may be expressed. For example, given the dominant alleles A and B, if the phenotype expressed by ...
 * [AA] is the same as [AB]/[BA] but not [BB], A is said to have `{bm} complete dominance/(complete dominance|completely dominant)/i` -- one allele's phenotype is expressed while others are hidden.
 * [AB]/[BA] is a blend of [AA] and [BB], it's said to have `{bm} incomplete dominance/(incomplete dominance|incompletely dominant)/i` -- more than one allele is dominant, resulting in a blended phenotype expression.
 * [AB]/[BA] is the phenotype expressed by both [AA] and [BB], it's said to have `{bm} co-dominance/(co-dominance|co-dominant)/i` -- more than one allele is dominant, resulting in a both phenotypes being expressed.

For example, in the diagram below the offspring ends up with red color allele and a white color allele.

```{dot}
digraph G {
	subgraph clusterparent1 {
		red1 [label="Red"];
    red2 [label="Red"];
		label = "Male parent color alleles";
	}
	subgraph clusterparent2 {
		white1 [label="White"];
    white2 [label="White"];
		label = "Female parent color alleles";
	}
	subgraph clusterchild1 {
		red3 [label="Red"];
    white3 [label="White"];
		label = "Child color alleles";
	}

  red1->red3 [label="randomly chosen from male parent"];
  white2->white3 [label="randomly chosen from female parent"];
}
```

 * In a complete dominance scenario, the child's color may end up being totally red because the red allele is the only one that gets expressed.

   ```{img}
   Complete dominance.svg
   This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
   Diagram showing a red flower offspring.
   ```

 * In a incomplete dominance scenario, the child's color may end up being pink because both the red and the white allele express together and blend.

   ```{img}
   Incomplete dominance.svg
   This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
   Diagram showing a pink flower offspring.
   ```

 * In a co-dominance scenario, the child may end up having blotches of red / white as its colors because both the red and the white allele express but are discrete.

   ```{img}
   Co-dominance.svg
   This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
   Diagram showing a blotchy flower offspring.
   ```

   ```{note}
   Another likely scenario for co-dominance is that some petals will be red while others are white.
   ```

Rarely ever is there only 2 alleles for a gene. A real-life example of dominance is blood type. 3 alleles are present: A, B, and O. A and B have incomplete dominance while O is recessive. If the alleles for a person's blood type is...
 * [AB]/[BA], their blood type (phenotype) is AB.
 * [AA], their blood type (phenotype) is A.
 * [BB], their blood type (phenotype) is B.
 * [AO]/[OA], their blood type (phenotype) is A.
 * [BO]/[OB], their blood type (phenotype) is B.
 * [OO], their blood type (phenotype) is O.

If the alleles present for a gene are...
 * are both dominant, it's said to be `{bm} homozygous dominant`. In the example above, [AB]/[BA]/[AA]/[BB] are all homozygous dominant (both alleles are dominant alleles).
 * are both recessive, it's said to be `{bm} homozygous recessive`. In the example above, [OO] is heterozygous recessive (both alleles are recessive alleles).

### Lethality

`{bm} Lethal allele/(lethal allele|lethal gene|\blethals\b)/i`s are alleles that cause the death of an organism that carries them. These alleles may cause death during development (e.g. gestation in the womb) or possibly after having fully matured.

Lethal alleles may fall into one of many categories. A ...

 * `{bm} recessive lethal allele/(recessive lethal|recessive lethal allele)/i` is a condition where 2 of 2 alleles for a gene need to be the lethal allele for it to cause death (needs to be homozygous). For example, Tay-Sachs disease appears and leads to death only if 2 copies of the lethal allele are present.

   ```{note}
   If only 1 copy of a recessive lethal allele is present, the organism doesn't die but may live on in a diseased form.
   ```

 * `{bm} dominant lethal allele/(dominant lethal|dominant lethal allele)/i` is a condition where only 1 of 2 alleles for a gene needs to be the lethal allele for it to cause death (can be heterozygous or homozygous). For example, Huntington's disease leads to death and only requires a single allele copy of an allele to be present.

   ```{note}
   If 2 copies of a dominant lethal allele are present, the organism still dies.
   ```
   
 * `{bm} conditional lethal allele/(conditional lethal|conditional lethal allele)/i` is a condition where an allele can only be fatal in response to some environmental factor. For example, favism is a disease that causes the organism carrying it to die when fava beans are consumed.

```{note}
Don't get confused. Dominant lethals / recessive lethals have nothing to do with dominant alleles / recessive alleles. The only thing they state is how many copies of the allele are needed for it to be lethal (1 / heterozygous or 2 / homozygous). Whether that lethal allele is a dominant allele / recessive allele isn't a requirement.
```

### Expressivity

`{bm} Expressivity` is the concept that, even if a known genotype is responsible for a phenotype, organisms having that genotype will show a variable expression of that phenotype. For example, in certain genetic disorders, the same genotype may result in stronger or weaker forms of that disorder (maybe even to the point of not developing at all).

 * `{bm} Narrow expressivity/(narrow expressivity|narrow expression|narrowly expressed)/i` refers to low variability in the expression of the phenotype. That is, the phenotype is likely to be expressed to the same degrees across organisms.

   Example of expression rates for narrow expressivity:

   ```{csv}
   0.99, 0.95, 0.98, 0.99
   0.91, 0.94, 0.95, 0.93
   ```

 * `{bm} Variable expressivity/(variable expressivity|variable expression|variably expressed)/i` refers to high variability in the expression of the phenotype. That is, the phenotype is likely to be expressed to different degrees across organisms. 

   Example of expression rates for variable expressivity:

   ```{csv}
   0.21, 0.55, 0.19, 0.78
   0.91, 0.43, 0.67, 0.31
   ```

`{bm} Penetrance` is the concept that, even if a known genotype is responsible for a phenotype, organisms having that genotype may not express the related phenotype. or example, in certain genetic disorders, the same genotype may not result in the disorder.

 * `{bm} Complete penetrance/(complete penetrance|completely penetrant)/i` refers to the phenotype always being expressed when the genotype is present.

   Example of expression rates for complete penetrance:

   ```{csv}
   1.00, 1.00, 1.00, 1.00
   1.00, 1.00, 1.00, 1.00
   ```

 * `{bm} Incomplete penetrance/(incomplete penetrance|incompletely penetrant)/i` refers to the phenotype maybe being expressed when the genotype is present.

   Example of expression rates for incomplete penetrance:

   ```{csv}
   1.00, 0.00, 0.00, 1.00
   1.00, 1.00, 1.00, 0.00
   ```

```{note}
It was never really expressed why this happens. My only guess is environmental factors or maybe some type of intrinsic built-in genetic randomness (e.g. not enough of the specific allele for a disease randomly collided with whatever other molecule(s) were required to express the phenotype).
```

### Sex-linkage

In humans / mammals, sex is determined by the XY chromosomes. In...

* females, the XY chromosome pair is a homologous chromosome pair -- X and X chromosomes.
* males, the XY chromosome pair is NOT a homologous chromosome pair -- X and Y chromosomes.

```{note}
How different are the X and Y chromosomes? Each codes for a completely different set of genes.

 * X chromosome codes for ~1500 genes.
 * Y chromosome codes for ~78 genes.

For example, the SRY gene contributes to development of testicles and is only found on the Y chromosome.
```

Genes on the XY chromosomes are called `{bm} sex-linked gene/(sex-linked disorder|sex-linked disease|sex-linked gene|sex-linked allele|sex-linked condition)/i`s.

#### X-linkage

Since males only have a single copy of both X and Y chromosomes, standard dominant allele / recessive allele rules DO NOT apply to these chromosomes in males. For example, regardless of if an allele on the X chromosome is a dominant allele or recessive allele, it will always express in males. There is no second X chromosome to provide a second allele for that gene.

`{bm} /(X-linked disorder|X-linked disease|X-linked gene|X-linked allele|X-linked condition)/i`Since a human / mammal must have at least a single X chromosome, ...
 * recessive alleles on the X chromosome are called `{bm} X-linked recessive allele/(X-linked recessive allele|recessive X-linked allele)/i`s.
 * recessive traits caused by alleles on the X chromosome are called `{bm} X-linked recessive trait/(X-linked recessive trait|recessive X-linked trait)/i`s.

For example, hemophilia is caused by an x-linked recessive allele. If there's a 2nd X chromosome that has a dominant allele for that gene, it'll suppress the hemophilia allele from expressing. For...
 * females (X and X), there IS a 2nd X chromosome. If that 2nd X chromosome contains the same recessive hemophilia allele (homozygous), the hemophilia phenotype will express.
 * males (X and Y), there ISN'T a 2nd X chromosome. If the allele on the single X chromosome is the hemophilia allele, the hemophilia phenotype will express.

If the probability of having the hemophilia allele on one X chromosome were `{kt} \frac{1}{7000}`, the probability of expressing the hemophilia phenotype / trait for...
 * females would be `{kt} \frac{1}{7000} \cdot \frac{1}{7000} = \frac{1}{49000000}`.
 
   ```{note}
   Since there's 2 X chromosomes and it's a recessive allele, both chromosomes need that allele (homozygous).
   ```

 * males would be `{kt} \frac{1}{7000}`.
 
   ```{note}
   Since there's only 1 X chromosome, only that chromosome needs the allele.
   ```

As such, X-linked recessive traits are much more common in men than they are women -- women need 2 copies of the recessive allele(s) while men only need 1.

```{note}
To find out the exact mechanism around X-linked recessive traits, see the section on x-linked inactivation below.
```

#### X-inactivation

The reason why males can function with only 1 copy of an X chromosome is because of `{bm} X-linked inactivation/(x-linked inactivation|x-inactivation)/i`. A single X chromosome is all that's needed for a human / certain mammals to operate normally. If more than 1 copy of an X chromosome is present in a cell, the cell chooses 1 at random to use while the others are made inactive. Inactive X chromosomes are compacted into small dense structures called `{bm} Barr bodies/(Barr bodies|Barr body)/i` that prevent most of their genes from expressing.

The above applies to both females (XX) and males/females with cases of allosomal aneuploidy (e.g. XXX or XXXY).

```{note}
How is it possible for there to be more than 2 X chromosomes? For example, the XX pair for a female may fail to separate during meiosis, meaning one of the egg_GAMETEs will contain 2 X chromosomes while the other will contain 0 X chromosomes.

Because of x-linked inactivation, allosomal aneuploidy isn't fatal but may lead to mild forms of disease. For example, Klinefelter syndrome (XXY) leads to infertility and may lead to learning disabilities / low testosterone.
```

```{note}
The process of X-linked inactivation is called `{bm} lyonization`.
```

How is it that x-linked recessive traits happen if only 1 X chromosome is ever active? Aren't 2 copies of the recessive allele needed for the phenotype to be expressed? It turns out that if some cells choose the X chromosome that contains the non recessive allele, those cells overpower the cells with the recessive allele.

For example, color blindness is an x-linked recessive trait. All copies of the X chromosome have to have the color-blind allele for human to be color blind. If one of the X chromosomes doesn't, some of the eye cells will be able to see color. The cells that do see color will relay color information back to the brain (the person will see color).

#### Punnett Square Examples

The following are examples showing the probability that offspring end up with the recessive X-linked disease known as hemophilia.


----


Given a hemophiliac mother and a hemophiliac father, the odds that ...

```{csv}
                , H (X chromosome), N/A (Y chromosome)
H (X chromosome), H (X) H (X) ✓   , H (X) N/A (Y) ✓
H (X chromosome), H (X) H (X) ✓   , H (X) N/A (Y) ✓
```

 * the child will have hemophilia are 100%.


----


Given a carrier mother and a hemophiliac father, the odds that ...

```{csv}
                , H (X chromosome), N/A (Y chromosome)
H (X chromosome), H (X) H (X) ✓   , H (X) N/A (Y) ✓
h (X chromosome), h (X) H (X)     , h (X) N/A (Y)
```

 * the child will be a female with hemophilia is 25%.
 * the child will be a male with hemophilia is 25%.
 * a daughter will have hemophilia is 50%.

   ```{note}
   In other words, out of all possible female children (XX) what are the chances of hemophilia. Punnett square below shows 2 female children, only one of which has HH (2 hemophilia alleles).
   ```

 * a son will have hemophilia is 50%.

   ```{note}
   In other words, out of all possible male children (XY) what are the chances of hemophilia. Punnett square below shows 2 male children, only one of which has HH (2 hemophilia alleles).
   ```


----


Given a carrier mother and a non-hemophiliac father, the odds that ...

```{csv}
                , h (X chromosome), N/A (Y chromosome)
H (X chromosome), H (X) h (X)     , H (X) N/A (Y) ✓
h (X chromosome), h (X) H (X)     , h (X) N/A (Y)
```

 * the child will be a female with hemophilia is 0%.
 * the child will be a male with hemophilia is 25%.
 * a daughter will have hemophilia is 0%.

   ```{note}
   In other words, out of all possible female children (XX) what are the chances of hemophilia. Punnett square below shows 2 female children, none of which has HH (2 hemophilia alleles).
   ```

 * a son will have hemophilia is 50%.

   ```{note}
   In other words, out of all possible male children (XY) what are the chances of hemophilia. Punnett square below shows 2 male children, only one of which has H (1 hemophilia alleles).
   ```

### Pedigree

A `{bm} pedigree` chart is a hierarchy diagram that shows the appearance of a specific phenotype in an organism and its ancestors.

```{img}
Pedigree.svg
Example pedigree chart.
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
```

Each individual in the diagram is identified by a shape:
 * Squares represent males.
 * Circles represent females.

The male (square) always comes first in a mating couple.

```{note}
Is a different shape necessary? The species that pedigree charts are used for (dogs, horses, humans, etc..) always have 2 sexes: male and female. Just starting off each mating couple with the same sex is enough to convey sex.
```

```{img}
Pedigree with sex labeling.svg
Example pedigree chart with explicitly labeled sexes.
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
```

If the shape is filled in, it means that the specific phenotype was expressed in that organism.

```{img}
Pedigree with sex and phenotype labeling.svg
Example pedigree chart with explicitly labeled sexes and phenotype expression.
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
```

For each mating pair, the ...
 * horizontal line connecting a male and a female is called a `{bm} marriage line/(marriage line|mating line)/i`.
 * vertical line that goes down from the marriage line is known as the `{bm} line of descent`.
 * horizontal line connected to the bottom of that line of descent is known as a `{bm} sibling line`.

```{img}
Pedigree with sex and phenotype labeling.svg
Example pedigree chart close-up with highlighted marriage line, line of descent, and sibling line.
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.

crop 0.35 0 0.56 0.3
scale 2 2

bg_color #00000000
stroke 4

fg_color #ff0000
rect 0.12 0.15 0.08 0.17
text 0.3 0.2 marriage line

fg_color #007f7f
rect 0.15 0.15 0.03 0.7
text 0.3 0.35 line of descent

fg_color #ff00ff
rect 0 0.7 1 0.15
text 0.3 0.5 sibling line
```

Pedigree charts can be used to infer / deduce the odds that a child will end up with the specific phenotype that the chart is for. For example, if the phenotype is caused by a single allele (pleiotropic) and you know that the allele needs to be homozygous recessive for the phenotype to be expressed (recessive allele), you can infer the genotypes of the parents...

```{img}
Pedigree with sex and phenotype labeling.svg
Example pedigree chart mating pair close-up genotype possibilities.
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.

crop 0.35 0 0.18 0.2
scale 2.5 2.5
expand 1.5 1.5 0.25 0.5

bg_color #00000000
font_size 25
text 0.0 0.25 [TT] / [Tt]
text 0.65 0.25 [tt]
```

```{note}
Since it's a recessive allele, both alleles need to be that recessive allele for the phenotype to express. That's why we know the female is [tt] (expresses phenotype) and the male is either [TT] or [Tt] (doesn't express phenotype).
```

From there you can use Punnett squares to visualize child phenotype outcomes in the different scenarios...

```{csv}
 , T , T
t, tT, tT
t, tT, tT
```

0% chance that the child ends up with [tt] in this scenario, meaning the phenotype will never express in any of the children.

```{csv}
 , T , t
t, tT, tt ✓
t, tT, tt ✓
```

50% chance that the child ends up with [tt] in this scenario, meaning the phenotype has a 50/50 chance of showing up in each child.

In the example above, if the 2 parents already had a few children and those children were present in the pedigree chart, you may be able to narrow the male down to just one of the genotype possibilities...

```{img}
Pedigree with sex and phenotype labeling.svg
Example pedigree chart mating pair and children close-up genotype possibilities.
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.

crop 0.25 0 0.75 0.55
scale 2.5 2.5
```

Since we can see children exhibiting the phenotype, we know that that the [TT] genotype of the male can't be the case. If it were, no children would end up with the phenotype. The only other option is the [Tt] genotype.

### Software Model

```{plantuml}
@startuml

interface ChromosomePair {
  Chromosome getOne()
  Chromosome getTwo()
  Chromosome pickRandom()
  List<Allele> getAllelesByGene(Gene gene)
}

class HomologousChromosomePair {
  void chromosomalCrossover()
}

note bottom of HomologousChromosomePair
  chromosomes must have the same set
  of genes in the same order.
end note

class SexChromosomePair {
}

note bottom of SexChromosomePair
  chromosomes can have different genes
end note

interface Chromosome {
  List<Gene> getGenes()
  List<Allele> getAlleles()
  Allele getAlleleByGene(Gene gene)
  void setAllele(Allele allele)
}

class Allele {
  Gene gene
  int dominance
  Set<PhenoType> phenotypes
}

interface Gene {
  Set<Allele> getAlleles()
}

interface Phenotype {
  EqualExpressionPattern getEqualExpressionPattern()
}

enum EqualExpressionPattern {
  IncompleteDominance,
  CoDominance
}

Phenotype -- "1" EqualExpressionPattern: how phenotype expresses if it meets another phenotype of same dominance
Gene "1" -- "1..*" Allele: "multiple versions (alleles) of a gene"
Allele "1" -- "0..*" Phenotype: "one allele can contributes to many phenotypes"
Chromosome "1" -- "1..*" Allele: "each chromosome holds alleles for a set of genes"
Chromosome "2" -- ChromosomePair 
ChromosomePair <|-- HomologousChromosomePair
ChromosomePair <|-- SexChromosomePair

@enduml
```

The above is a very basic software model that represents the concepts around classical genetics. It's not entirely correct but it lays down a good enough foundation to build out something more elaborate.

The idea of dominant alleles and recessive alleles are handled by `Allele.dominance`. It's used to rank each allele for a gene.

The idea of homozygous and heterozygous is handled by `HomologousChromosomePair`. The 2 chromosomes it holds on to are homologous (have the same set of genes). As such, for each gene you'll 2 alleles -- 1 on each chromosome. If the alleles are the...
 * same, they're homozygous.
 * different, they're heterozygous.

The law of segregation is handled by `ChromosomePair.pickRandom()`.

The law of dominance, incomplete dominance, and co-dominance is codified using `EqualExpressionPattern`. That is, if 2 alleles have the same `Allele.dominance`, any matching phenotypes between them will be expressed using `Phenotype.EqualExpressionPattern`.

The idea of linked genes is handled by `HomologousChromosomePair.chromosomalCrossover()`. It swaps segments of the 2 chromosomes it holds on to.

## Adenosine Triphosphate

`{bm} /\b(ATP)(?:s{0,1})\b/` `{bm} Adenosine Triphosphate` (ATP) is a molecule that provides energy to drive various biological processes (e.g. muscle contractions). The third phosphoral group at the very end has a high-energy bond. When broken, energy is released and the resulting molecules are the broken up phosphoral group and `{bm} /\b(ADP)(?:s{0,1})\b/` `{bm} Adenosine Diphosphate` (ADP).

```{note}
High-energy bonds are actually a thing: A chemical bond whose hydrolysis results in the generation of 30kJ (7kcal) of energy or, if coupled to an energetically unfavourable reaction, can drive that reaction forward. (https://www.genscript.com/molecular-biology-glossary/1364/high-energy-bond)
```

```{img}
ADP_ATP_cycle.png
By Muessig - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=27614630
ADP ATP cycle
scale 0.25 0.25
```

ATP is produced in the mitochondria. Similar to how the mitochondria is referred to as the powerhouse of the cell, ATP is often referred to as the energy currency of the cell / energy store for the cell.

## Transport

There are 2 different types of mechanism used to transport molecules in and out of a cell: passive transport and active transport.

`{bm} Passive transport` is when molecules naturally move towards the gradient. In this context, gradient refers to the natural direction in which things are supposed to go -- no explicit energy is needed to move/push it along, it just moves in that direction by virtue of some implicit property.

 * A `{bm} concentration gradient` is when the concentration of a molecule evens out in a volume just by virtue of the molecules randomly bouncing around (diffusion). For example, gas pumped into a vacuum will end up filling the vacuum evenly (more-or-less) -- traveling from areas of high concentration to areas of low concentration.
 
   ```{img}
   Concentration Gradient.svg
   This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
   Concentration Gradient
   ```

   ```{note}
   All this is saying is that if you have a group of items clustered in a single area, each traveling at a different speed / direction, will spread out over time.
   ```

 * An `{bm} electrical gradient` is when molecules flow in some direction because their electrical charges attract. For example, a negatively charged molecule will gravitate towards a positively charged molecule. Similarly, a negative molecule will gravitate away from from other negative molecules / a positive molecule will gravitate away from other positive molecules.

   ```{img}
   Electrical Gradient.svg
   This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
   Electrical Gradient
   ```

 * An `{bm} electrochemical gradient` is a combination of both a concentration gradient and an electrical gradient.

`{bm} Active transport` is when molecules use energy (e.g. ATP) to move against their gradient. It's the opposite of passive transport -- energy is explicitly used to drive a molecule to where it naturally / normally wouldn't go. An example of active transport is the "sodium potassium pump" enzyme: ATP is used to force open/close the ends of the enzyme, which allow sodium and potassium to be exchanged across the cell membrane.

Note that the active transport in the example above is the opening/closing of the enzyme ends, not the exchange of sodium and potassium. Energy (ATP) is being used to shape-shift the enzyme to open/close (active transport) while the sodium and potassium are passively entering and exiting the gates (passive transport via facilitated diffusion).

## Osmosis

`{bm} Osmosis/(osmosis|osmotic)/i` is the passive transport of solvent molecules (typically water), across a semipermeable membrane, from areas where solutes are less concentrated to areas where solutes are more concentrated.

For example, imagine you have a semipermeabl membrane that allows water molecules (solvent) to pass but not sodium (solute). That membrane is separating 2 regions, where the ...

* left region is a solution of 25% sodium (solute) and 75% water (solvent).
* right region is a solution of 75% sodium (solute) and 25% water (solvent).

There will be a net movement of some water molecules from the left region (lower solute concentration) to the right region (higher solute concentration).

```{img}
0307_Osmosis.jpg
By OpenStax - https://cnx.org/contents/FPtK1zmh@8.25:fEI3C8Ot@10/Preface, CC BY 4.0, https://commons.wikimedia.org/w/index.php?curid=30131189
Osmosis is the diffusion of water through a semipermeable membrane down its concentration gradient. If a membrane is permeable to water, though not to a solute, water will equalize its own concentration by diffusing to the side of lower water concentration (and thus the side of higher solute concentration). In the beaker on the left, the solution on the right side of the membrane is hypertonic.
scale 0.65 0.65
```

```{note}
Another diagram that may make more sense conceptually: https://commons.wikimedia.org/wiki/File:Osmosis_diagram.svg
```

There are 2 reasons why osmosis happens. The first is that the semipermeable membrane will only allow certain types of molecules to pass through. If the semipermeable membrane is gated by ...

* size, the solute molecules may be too large to pass through. 
* charge, the solute molecules may not have the required charge to pass through.

The higher the concentration of solute molecules, the less likely it is for the solvent molecules to reach a pore in the semipermeable membrane. The side with the lower concentration of solute molecules is more likely to have a solvent molecule reach a pore than the other way around.

```{img}
Semipermeable membrane gated by size.svg
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
Semipermeable membrane with pores big enough for water but not sodium.
```

The second reason is that, depending on the charge of solvent and charge of solute, the solvent may be attracted to the solute. More solute = more chance that a solvent gets attracted to it instead of crossing a pore in the membrane. For example, if the solvent is water and the solute is sodium, the weakly negative charge of the oxygen atom in a water molecule may get attracted to the positive charge of the sodium ion.

```{img}
Semipermeable membrane but solvent attracted to solute.svg
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
Semipermeable membrane but solvent attracted to solute.
```

`{bm} Tonicity` is the amount of pressure applied to a semipermeable membrane due to osmosis. In other words, it's the amount of water that flows in or out of a cell due to the type of solution it's put in. A ...

* `{bm} hypertonic` solution is one that has more solutes than the cell, meaning the inside of the cell loses water (cell shrivels).
* `{bm} hypotonic` solution is one that has less solutes than the cell, meaning the inside of the cell gains water (cell swells).
* `{bm} isotonic` solution is one that has a roughly equal amount of solutes, meaning the cell's water content remains the same.

```{img}
Tonicity.svg
This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License.
Different types of tonicity.
```

```{note}
The prefix is referring to the amount of solute in the solution relative to the cell. Hyper = more. Hypo = less. Iso = same.
```

The following is a micrograph of the red blood cells in solutions of different tonicity. Notice how they shrivel in a hypertonic solution (lose water) and expand in a hypotonic solution (gain water).

```{img}
Human_Erythrocytes_OsmoticPressure_PhaseContrast_Plain.svg
By Zephyris - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=18401754
Human erythrocytes (red blood cells) viewed by phase contrast light microscopy. Three conditions are shown: hypertonic conditions (where the erythrocytes contract and appear "spiky"), isotonic conditions (where the erythrocytes appear normal) and hypotonic conditions (where the etrythrocytes expand and become more round).
scale 0.55 0.55
```

## Photosynthesis

`{bm} Photosynthesis` is the process by which certain organisms convert light energy (photons) to chemical energy (sugars). These organisms are called `{bm} Photoautotroph`s, and they include ...

* plants and algae (eukaryotic), which perform photosynthesis via their chloroplast.
* cyanobacterium, which are bacteria that can generate their own food.

```{note}
Another way to think of photosynthesis is that it uses light energy (PHOTOsynthesis) to synthesize (photoSYNTHESIS) sugars.
```

```{note}
Chloroplast and cyanobacterium share a similar structure. It's speculated that they have the same parent organism: that parent formed an endosymbiotic relationship with a larger cell and eventually became the chloroplast organelle.
```

The overall chemical reaction for this is `{kt} 6CO_2 + 6H_2O + energy \to C_6H_{12}O_6 + 6O_2`. Carbon dioxide gas (`{kt} CO_2`) bonds with water (`{kt} H_2O`) using energy from the sun (photons), creating glucose (`{kt} C_6H_{12}O_6`).

This reaction happens in 2 steps:

1. `{bm}Light-dependent reaction`s: Energy molecules are created from water and photos, with oxygen being a byproduct.

   This occurs in a thylakoid membrane.

   ```{img}
   Chloroplast_structure.svg
   Ultrastructure of a chloroplast.
   By Kelvinsong - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=26147364
   scale 0.3 0.3
 
   bg_color #00000000

   fg_color #808000ff
   rect 0.24 0.015 0.3 0.15
   ```
2. `{bm} Calvin cycle`: It's a cyclical process that requires multiple iterations (3?) to produce a single glucose molecule. Each cycle, ATP and NADPH are used for energy (producing ADP and NADP+ respectively), while the carbon dioxide (`{kt} CO_2`) is used as a source of carbons for the resulting glucose.

   This occurs in the stroma. 

   ```{img}
   Chloroplast_structure.svg
   Ultrastructure of a chloroplast.
   By Kelvinsong - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=26147364
   scale 0.3 0.3
 
   bg_color #00000000

   fg_color #008080ff
   rect 0.85 0.2 0.15 0.07
   ```

The following workflow diagram provides a ultra-simple high-level overview of the processes that take place. Note that this doesn't specify how many of each molecule get input / output, nor does it provide a complete set of a input / output molecules for each reaction.

```{dot}
digraph G {
  compound=true;
  Photon [label="Photons\n(sunlight)"]
  H2O [label="H2O\n(water)"]
  CO2 [label="CO2\n(carbon dioxide)"]
  O2 [label="O2\n(oxygen)"]
  ATP
  ADP
  LDR [label="Light-dependent reactions\n(happens in thylakoid membrane)", shape=box]
  CALVIN [label="Calvin cycle\n(happens in stroma)", shape=box]
  CARB [label="Glucose"]
  Photon -> LDR
  H2O ->  LDR
  LDR -> ATP
  LDR -> O2
  ATP -> CALVIN
  CO2 -> CALVIN
  CALVIN -> CARB
  CALVIN -> ADP
}
```

## Cellular Respiration

`{bm} Cellular respiration` is the process by which certain organisms convert glucose (sugar) to energy. These organisms include ...

* eukaryotic cells, which perform cellular respiration via their mitochondria.
* bacteria which share their ancestry with mitochondria.

``` {note}
Remember that all eukaryotes have mitochondria -- both plant and animal cells. Unsure if all bacteria can perform cellular respiration?
```

The overall chemical reaction for this is `{kt} C_6H_{12}O_6 + 6O_2 \to 6CO_2 + 6H_2O + energy`. Glucose (`{kt} C_6H_{12}O_6`) and oxygen (`{kt} O_2`) break down into carbon dioxide gas (`{kt} CO_2`), water (`{kt} H_2O`), and energy (roughly 38 ATP molecules and some heat).

```{note}
The number of ATPs actually generated is variable and dependent on many factors, but 38 is the generally agreed upon number.
```

The reaction happens in 3 steps:

1. `{bm} Glycolysis`: The carbon backbone of the glucose molecule is split, creating 2 `{bm} Pyruvate` molecules along with water and several other molecules. This is an anaerobic process (no oxygen needed) that nets 2 ATPs. This happens in the cytoplasm of cells.

   ```{img}
   Anima_cell_notext.svg
   Eukaryote with cell ribosomes highlighted
   By No machine-readable author provided. Chb assumed (based on copyright claims). - No  machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=688296
   scale 0.5 0.5
   arrow 1 0.5 0.7 0.5
   expand 2 1 0 0
   text 0.5 0.5 cytoplasm
   text 0.5 0.6 (everything inside)
   ```

1. `{bm} Krebs cycle/(krebs cycle|citric acid cycle|tricarboxylic acid cycle)/i`: The pyruvate molecules get further sliced and diced with other molecules.  This is an aerobic process (requires oxygen) that nets 2 ATPs. This happens in the matrix of the mitochondria.

   ```{img}
   Animal_mitochondrion_diagram_en.svg
   Mitochondria
   By Mariana Ruiz Villarreal LadyofHats - the diagram i made myself using adobe illustrator. as a source for the information i used the diagrams found here:[1], [2], [3], [4], [5], [6] and [7]., Public Domain, https://commons.wikimedia.org/w/index.php?curid=8152599
   
   scale 0.52 0.52
   
   bg_color #00000000

   fg_color #ff00ffff
   rect 0.7 0.75 0.3 0.12 
   ```

1. `{bm} Oxidative phosphorylation`: Produces around 34 ATPs. This is an aerobic process (requires oxygen) that nets roughly 34 ATPs (bulk of conversions). This happens in the `{bm} electron transport chain` section of the mitochondria (inner membrane).

   ```{img}
   Animal_mitochondrion_diagram_en.svg
   Mitochondria
   By Mariana Ruiz Villarreal LadyofHats - the diagram i made myself using adobe illustrator. as a source for the information i used the diagrams found here:[1], [2], [3], [4], [5], [6] and [7]., Public Domain, https://commons.wikimedia.org/w/index.php?curid=8152599
   
   scale 0.55 0.55
   
   bg_color #00000000
   
   fg_color #008080ff
   rect 0.3 0.18 0.15 0.07
   ```

```{note}
The material says that, technically step 3 doesn't have to happen after step 2 but it usually does.
```

Because the Krebs cycle and the oxidative phosphorylation are aerobic processes (require oxygen), if no oxygen is present the output of glycolysis goes through a process called `{bm} fermentation/(ferment|fermentation)/i`. Fermentation is an anaerobic process (no oxygen required). Depending on the organism, it'll end up producing either...

* `{bm} alcohol` via `{bm} alcohol fermentation` (e.g. yeast).
* `{bm} lactic acid` via `{bm} lactic acid fermentation` (e.g. humans and mammals).

Fermentation does produce ATP, but much less so than Kerbs cycle + oxidative phosphorylation. 

For example, if a human is vigorously running, that human may not have enough oxygen available to trigger the Krebs cycle / oxidative phosphorylation (steps 2 and 3). As such, the output from glycolysis (step 1) will end up going through lactic acid fermentation instead.

The following workflow diagram provides a ultra-simple high-level overview of the processes that take place. Note that this doesn't specify how many of each molecule get input / output, nor does it provide a complete set of a input / output molecules for each reaction.

```{dot}
digraph G {
  compound=true;
  O2 [label="O2\n(oxygen)"]
  GLUCOSE [label="Glucose"]
  PY [label="Pyruvate"]
  ATP2 [label="ATP"]
  ATP3 [label="ATP"]
  ATP4 [label="ATP"]
  UNKNOWN [label="?????"]
  GLYC [label="Glycolysis\n(happens in cell cytoplasm)", shape=box]
  KREBS [label="Krebs cycle\n(happens in mitochondria matrix)", shape=box]
  OP [label="Oxidative Phosphorylation\n(happens in mitochondria inner membrane)", shape=box] 
  FERM [label="Fermentation", shape=box]
  FERM_RES [label="Lactic acid or alcohol\n(depends on the organism)"]
  O2 -> GLYC
  GLUCOSE -> GLYC
  GLYC -> PY
  GLYC -> ATP2
  PY -> FERM [label="not enough oxygen available"]
  PY -> KREBS [label="oxygen available"]
  KREBS -> UNKNOWN
  KREBS -> ATP3
  UNKNOWN -> OP
  OP -> ATP4
  FERM -> FERM_RES
}
```

## Sodium Potassium Pump

The `{bm} sodium potassium pump` is an transmembrane enzyme that allows the exchange of sodium ions and potassium ions across the cell membrane by opening and closing its ends.

```{img}
Scheme_sodium-potassium_pump-en.svg
By LadyofHats Mariana Ruiz Villarreal - Own work. Image renamed from Image:Sodium-Potassium_pump.svg, Public Domain, https://commons.wikimedia.org/w/index.php?curid=3981038
Example of primary active transport, where energy from hydrolysis of ATP is directly coupled to the movement of a specific substance across a membrane independent of any other species.
```

Only 1 end of the enzyme is open at a time. When the...

* intracellular end is open, the enzyme has an affinity for 3 sodium ions followed by an affinity for ATP.
* extracellular end is open, the enzyme has an affinity for 2 potassium ions.

Since both potassium and sodium have a positive charge and an an unequal number are being exchanged each cycle (3 sodium out vs 2 potassium in), the intracellular space will be more positive than the extracellular space.

This charge difference is further reinforced by membrane channel proteins which allow potassium ions to flow across the membrane (potassium channels). Since there's a higher concentration of potassium ions inside the cell, those potassium ions have a higher chance of flowing through the channel to the outside. Some percentage may be impeded by the slightly more positive charge on the outside, but overall more will make it to the outside than stay on the inside.

This charge difference is referred to as the `{bm} resting membrane potential` for a cell.

## Microscopy

`{bm} Microscope`s are devices used to magnify (zoom in) on objects, such that you can see things that you normally would be too small to see on your own. The term microscope comes from the words...
* micro, meaning small.
* scope, meaning to look at.

A picture taken through a microscope is called a `{bm} micrograph`. The distinguishing factors for most microscopes or the amount of magnification and the resolution of the output image.

There are different types of microscopes:
* `{bm} Simple microscope`s have just one lense (e.g. magnifying glass).
* `{bm} Compound microscope`s have 2 ore more lenses (e.g. traditional microscopes used in labs).
  * `{bm} Light microscope`s / `{bm} Brightfield microscope`s work by shining a light through the specimen (which can be alive) and into the lens system that eventually reaches your eye. They're the type of microscopes typically used in student labs.
    * `{bm} Fluorescence microscope`s are microscopes that require the specimen being viewed is treated in some way to make it glow, allowing you to highlight parts of the specimen.
    * `{bm} Confocal microscope`s are a special form of fluorescence microscopes that uses a laser to excite a sample and get it to glow fluorescent.
* `{bm} Electron microscope`s use beams of electrons rather than beams of light. They produce very magnified very high resolutions images. But, unlike light microscopes, the specimens being observed are dead because electron microscopes operate in a vacuum.
  * `{bm} scanning electron microscope`s (SEM) work by scanning a beam of electrons over the surface of the specimen to produce an image of the 3D surface.
  * `{bm} transmission electron microscope`s (TEM) work by slicing up the specimen into ultra thin slices and passing an electron bean through those slices to produce an image of the insides.

# Other Terminology

Terminology that's relevant but doesn't fit in any other section goes here.

`{bm-ignore} condense`
`{bm} Density/(density|dense)/i` - The mass per unit volume of a substance.

`{bm} Specific heat capacity/(heat capacity|heat capacities)/i` - The amount of heat needed to raise the temperature of one gram of a substance by one degree Celsius.

`{bm} Heat of vaporization` - The amount of energy needed to change one gram of a liquid substance to a gas at constant temperature.

`{bm} Endosymbiosis/(endosymbiosis|endosymbiotic)/i` - A form of symbiosis where one organism lives inside of of the other (e.g. gut bacteria lives in our colon). The prefix *endo* means *within*.

`{bm} Diffusion/(diffusion|diffuse)/i` - A physical process where molecules of a material move from an area of high concentration (where there are many molecules) to an area of low concentration (where there are fewer molecules) until it has reached equilibrium (molecules evenly spread). [See more](https://simple.wikipedia.org/wiki/Diffusion).

`{bm} Equilibrium` - A state in which opposing forces / influences are balanced. In the context of a concentration gradient, it means the state at which a substance is equally distributed throughout the volume that it's in (roughly).

`{bm} Permeability/(permeable|permeability)/i` - The state or quality of a material or membrane that causes it to allow liquids/gases to pass through it.

`{bm} Semipermeable/(semipermeable|selectively permeable)/i` - The state or quality of a material or membrane that causes it to allow certain types of molecules to pass through it.

`{bm} Intracellular` - The fluid inside of the cell, which is technically on the inside of the cell membrane (cytoplasm).

`{bm} Extracellular` - The fluid outside of the cell.

```{note}
Technically unsure at which layer the extracellular region begins. Is it outside of the cell membrane? cell wall? cell capsule? I'm pretty sure any fluid outside of the cell membrane qualifies as extracellular, while any fluid inside of the cell membrane qualifies as intracellular (cytoplasm).
```

`{bm} Aerobic` - A biological process that requires oxygen.

`{bm} Anaerobic` - A biological process that doesn't require oxygen.

`{bm} Fission` - The act of dividing or splitting something into two or more parts.

`{bm} Tetrad` - A group / set of 4.

`{bm} Homologous/(Homologous|Homology|Homolog)/i` - Having the same relation, relative position, or structure. Particularly in biology, it s the existence of a shared ancestry between a pair of structures or genes.

`{bm} Karyotype` - Micrograph image of diploid set of chromosomes, grouped in pairs.

`{bm} Asexual` - When offspring is created using the genetic material from 1 parent. The offspring are essentially copies of the parent in terms of their genetic material (clone). Examples of asexual reproduction include:
  * `{bm} Binary fission` - Process of cloning a prokaryotic cell and certain organelle inside of eukaryotic cell (e.g. mitochondria).
  * Mitosis - Process of cloning a eukaryotic cell.
  * `{bm} Budding` - When a growth on a plant or animal breaks off, that broken piece is the offspring.
  * `{bm} Fragmentation` - When an organism breaks into two or more pieces, each fragment grows back into a whole.
  * `{bm} Parthenogenesis` - When an animal lays an egg_GAMETE but that egg_GAMETE already has all the genetic information needed to develop (it doesn't need to be fertilized).

`{bm} Sexual` - When offspring is created by fusing genetic material from 2 parents. The offspring has a mixture of genetic material from both parents. An example of sexual reproduction is when a gamete cells merge to create the offspring.

```{note}
In some cases, the genetic material being fused in sexual reproduction may be from the same parent. Answer to a question on the site... it is still sexual, because sexual reproduction means fusion of male and female gametes, doesn't matter if they're from the same plant.
```

`{bm} Model system` - A system with a reduced set of parameters/complexity that makes it easy for a researcher to investigate a particular scientific question. For example, Gregor Mendel used pea pods to research the theory of classical genetics / inheritance (simple, grows and matures quickly, inbreeding okay)

`{bm} Monatomic` - Made up of a single atom.

`{bm} Polyatomic` - Made up of > 1 atoms.

`{bm} Compound` - In the context of chemistry, a compound is a bond between atoms of different types. For example, hydrogen gas (`{kt}H_2`) is not a compound, but water (`{kt}H_2O`) is. See ...
  * ionic compound.
  * molecular compound.



`{bm-ignore} baseline`
`{bm-ambiguous} Add the suffix _pH or _nucleotide/((?!based)(?:bases|base))/i`

`{bm-ignore} byproduct|reproduction|reproduce|reproductive`
`{bm-ambiguous} Add the suffix _ENZYME for enzyme output or _CHEM for chem reaction output/(product)/i`
`{bm-ambiguous} Add the suffix _ENZYME for enzyme output/(intermediate)/i`

`{bm-ambiguous} Add the suffix _ENZYME for enzyme input or _CHEM for chem reaction input/(reactant)/i`

`{bm-ambiguous} Add the suffix _atom or _cell/\b(nucleus|nuclei)\b/i`