<!DOCTYPE html><html><head>
            <meta content="text/html;charset=utf-8" http-equiv="Content-Type">
            <meta content="utf-8" http-equiv="encoding">
          <title>Biology
</title><link href="/.datadir_65c4776c20884922953aa28351b977e2/katex.min.css" rel="stylesheet"><meta name="viewport" content="width=device-width,initial-scale=1"><link href="/.temp_githib_css/github-markdown.css" rel="stylesheet"><link href="/.temp_highlightjs_css/default.css" rel="stylesheet"></head>
          <body class="markdown-body"><div class="toc">
<ul>
<li><a href="#HEADREF0">Chemistry</a></li>
<ul>
<li><a href="#HEADREF1">Atom</a></li>
<ul>
<li><a href="#HEADREF2">Atomic Mass</a></li>
<li><a href="#HEADREF3">Atomic Count</a></li>
</ul>
<li><a href="#HEADREF4">Molecule</a></li>
<li><a href="#HEADREF5">Ion</a></li>
<li><a href="#HEADREF6">pH</a></li>
</ul>
<li><a href="#HEADREF7">Biology</a></li>
<ul>
<li><a href="#HEADREF8">Scientific Method</a></li>
<ul>
<li><a href="#HEADREF9">Hypothesis</a></li>
<li><a href="#HEADREF10">Experiment</a></li>
</ul>
<li><a href="#HEADREF11">Carbohydrate Molecule</a></li>
<li><a href="#HEADREF12">Protein Molecule</a></li>
<li><a href="#HEADREF13">Lipid Molecule</a></li>
<li><a href="#HEADREF14">Water Molecule</a></li>
<li><a href="#HEADREF15">Cell</a></li>
<ul>
<li><a href="#HEADREF16">Features</a></li>
<ul>
<li><a href="#HEADREF17">Cytoplasm</a></li>
<li><a href="#HEADREF18">Membrane</a></li>
<li><a href="#HEADREF19">Cell Wall</a></li>
<li><a href="#HEADREF20">Capsule</a></li>
<li><a href="#HEADREF21">Ribosome</a></li>
<li><a href="#HEADREF22">Appendage</a></li>
</ul>
<li><a href="#HEADREF23">Fluid Mosaic Model</a></li>
<li><a href="#HEADREF24">History of Modern Cell Theory</a></li>
</ul>
<li><a href="#HEADREF25">Eukaryotic Cell</a></li>
<ul>
<li><a href="#HEADREF26">Organelle</a></li>
<ul>
<li><a href="#HEADREF27">Nucleus</a></li>
<li><a href="#HEADREF28">Endoplasmic Reticulum</a></li>
<li><a href="#HEADREF29">Golgi</a></li>
<li><a href="#HEADREF30">Mitochondria</a></li>
<li><a href="#HEADREF31">Lysosome</a></li>
<li><a href="#HEADREF32">Peroxisome</a></li>
<li><a href="#HEADREF33">Chloroplast</a></li>
<li><a href="#HEADREF34">Vacuole</a></li>
</ul>
<li><a href="#HEADREF35">Chromosome</a></li>
<li><a href="#HEADREF36">Mitosis (Cell Cycle)</a></li>
<ul>
<li><a href="#HEADREF37">Interphase</a></li>
<li><a href="#HEADREF38">Mitotic Phase</a></li>
</ul>
<li><a href="#HEADREF39">Meiosis</a></li>
<ul>
<li><a href="#HEADREF40">Meiosis I</a></li>
<li><a href="#HEADREF41">Meiosis II</a></li>
</ul>
<li><a href="#HEADREF42">Cancer</a></li>
<li><a href="#HEADREF43">Fertilization</a></li>
<li><a href="#HEADREF44">Apoptosis</a></li>
</ul>
<li><a href="#HEADREF45">Prokaryotic Cell</a></li>
<li><a href="#HEADREF46">Enzyme</a></li>
<ul>
<li><a href="#HEADREF47">Metabolic Pathway</a></li>
<li><a href="#HEADREF48">Anabolism</a></li>
<li><a href="#HEADREF49">Catabolism</a></li>
</ul>
<li><a href="#HEADREF50">Molecular Genetics</a></li>
<ul>
<li><a href="#HEADREF51">Nucleotides</a></li>
<li><a href="#HEADREF52">DNA</a></li>
<li><a href="#HEADREF53">RNA</a></li>
</ul>
<li><a href="#HEADREF54">Classical Genetics</a></li>
<ul>
<li><a href="#HEADREF55">Laws</a></li>
<ul>
<li><a href="#HEADREF56">Segregation</a></li>
<li><a href="#HEADREF57">Independent Assortment</a></li>
<li><a href="#HEADREF58">Dominance</a></li>
</ul>
<li><a href="#HEADREF59">Lethality</a></li>
<li><a href="#HEADREF60">Expressivity</a></li>
<li><a href="#HEADREF61">Sex-linkage</a></li>
<ul>
<li><a href="#HEADREF62">X-linkage</a></li>
<li><a href="#HEADREF63">X-inactivation</a></li>
<li><a href="#HEADREF64">Punnett Square Examples</a></li>
</ul>
<li><a href="#HEADREF65">Pedigree</a></li>
</ul>
<li><a href="#HEADREF66">Adenosine Triphosphate</a></li>
<li><a href="#HEADREF67">Transport</a></li>
<li><a href="#HEADREF68">Osmosis</a></li>
<li><a href="#HEADREF69">Photosynthesis</a></li>
<li><a href="#HEADREF70">Cellular Respiration</a></li>
<li><a href="#HEADREF71">Sodium Potassium Pump</a></li>
<li><a href="#HEADREF72">Microscopy</a></li>
</ul>
<li><a href="#HEADREF73">Other Terminology</a></li>
</ul>
</div>
<a name="HEADREF0"></a><h1>Chemistry</h1>
<p><a name="BOOKMARK0"></a><strong>Chemistry</strong> is defined as the study of composition, structure, and properties of <a href="#BOOKMARK1">matter</a>. <a name="BOOKMARK1"></a><strong>Matter</strong> is defined as anything that takes up space and has a mass (e.g. hair, <a href="#BOOKMARK59">water</a>, table, etc..).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Primary source for these notes is Khan Academy's <a href="#BOOKMARK0">Chemistry</a> course, with additional information from Wikipedia and other sources.</p>
</div>
<a name="HEADREF1"></a><h2>Atom</h2>
<p>An <a name="BOOKMARK2"></a><strong>element</strong> is <a href="#BOOKMARK1">matter</a> that cannot be broken down any further by <a href="#BOOKMARK0">chemical</a> reaction -- it's a substance made entirely out of one type of <a name="BOOKMARK3"></a><strong>atom</strong>. Each <a href="#BOOKMARK2">element</a>/<a href="#BOOKMARK3">atom</a> has a specific set of properties that defines how it acts/reacts (e.g. weight, colour, how light reflects, etc..).</p>
<p><img src="/.datafile_5e5ffc315d2062f3cb5a7af5464d156e/diagram.svg" alt="Graphviz Dot Diagram"></p><p>Examples of <a href="#BOOKMARK1">matter</a> made up of a single type of <a href="#BOOKMARK3">atom</a>:</p>
<ul>
<li>Gold - Yellow coloured, reflective, malleable, and corrosion-resistant.</li>
<li>Lead - Blue/white coloured, relatively soft, malleable.</li>
<li>Carbon Silver/gray coloured, reflective, easily oxidizes/corrodes.</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>As stated above, <a href="#BOOKMARK1">matter</a> made up of a single type of <a href="#BOOKMARK3">atom</a> is called an <a href="#BOOKMARK2">element</a>.</p>
</div>
<p>Examples of <a href="#BOOKMARK1">matter</a> made of multiple types of <a href="#BOOKMARK3">atom</a>s:</p>
<ul>
<li>Salt <a href="#BOOKMARK59">water</a> - Mixture of sodium and <a href="#BOOKMARK59">water</a>.</li>
<li><a href="#BOOKMARK59">Water</a> - <a href="#BOOKMARK59">Water</a> can be further broken down to the <a href="#BOOKMARK2">element</a>s hydrogen and oxygen.</li>
<li>Glucose - Glucose can be further broken down to the <a href="#BOOKMARK2">element</a>s carbon, hydrogen, and oxygen.</li>
</ul>
<p><img src="240px-Atom_Diagram.svg.png" alt="By AG Caesar - Own work, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=67059219"></p>
<p>The building blocks of <a href="#BOOKMARK3">atom</a>s are referred to as <a name="BOOKMARK4"></a><strong>sub-atomic particle</strong>s: <a href="#BOOKMARK6">proton</a>s, <a href="#BOOKMARK7">neutron</a>s, and <a href="#BOOKMARK8">electron</a>s. <a href="#BOOKMARK6">Proton</a>s and <a href="#BOOKMARK7">neutron</a>s form the <a href="#BOOKMARK5">nucleus</a> of the <a href="#BOOKMARK3">atom</a> while <a href="#BOOKMARK8">electron</a>s jump around in orbit of the<a name="BOOKMARK5"></a><strong>nucleus</strong>.</p>
<p>For each <a href="#BOOKMARK3">atom</a>, the number of ...</p>
<ul>
<li><a name="BOOKMARK6"></a><strong>proton</strong>s defines the type of <a href="#BOOKMARK3">atom</a>. For example, hydrogen has 1 <a href="#BOOKMARK6">proton</a>, helium as 2, lithium has 3, etc..</li>
<li><a name="BOOKMARK7"></a><strong>neutron</strong>s defines the extra <a href="#BOOKMARK11">atomic mass</a> of the <a href="#BOOKMARK3">atom</a>. For example, carbon-12 and carbon-13 have the same number of <a href="#BOOKMARK6">proton</a>s but carbon-13 has 1 more <a href="#BOOKMARK7">neutron</a> and thus a greater <a href="#BOOKMARK11">atomic mass</a>.</li>
<li><a name="BOOKMARK8"></a><strong>electron</strong>s defines the attraction to other <a href="#BOOKMARK3">atom</a>s. For example, an <a href="#BOOKMARK3">atom</a> is always looking to have <a href="#BOOKMARK8">electron</a> count = <a href="#BOOKMARK6">proton</a> count. It attracts to other <a href="#BOOKMARK3">atom</a>s in an effort to steal / give away <a href="#BOOKMARK8">electron</a>s.</li>
</ul>
<p><a href="#BOOKMARK3">Atom</a>s with the same number of <a href="#BOOKMARK6">proton</a>s (same type) but a different number of <a href="#BOOKMARK7">neutron</a>s (different <a href="#BOOKMARK11">atomic mass</a>) are called <a name="BOOKMARK9"></a><strong>isotope</strong>s. For example, carbon-12 and carbon-13 are both <a href="#BOOKMARK9">isotope</a>s of carbon -- they have the same <a href="#BOOKMARK0">chemical</a> properties but different mass.</p>
<p>Essentially, the configuration of an <a href="#BOOKMARK3">atom</a> (number of <a href="#BOOKMARK6">proton</a>s/<a href="#BOOKMARK7">neutron</a>s/<a href="#BOOKMARK8">electron</a>s) is what predicts its properties and how it reacts to with other <a href="#BOOKMARK3">atom</a>s.</p>
<p>The periodic table below orders <a href="#BOOKMARK3">atom</a>s by the number of <a href="#BOOKMARK6">proton</a>s (also called the <a name="BOOKMARK10"></a><strong>atomic number</strong>)...</p>
<p><img src="800px-Simple_Periodic_Table_Chart-en.svg.png" alt="By Offnfopt - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=62296883"></p>
<a name="HEADREF2"></a><h3>Atomic Mass</h3>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The notion of <a href="#BOOKMARK14">atomic weight</a> in <a href="#BOOKMARK0">chemistry</a> is different from weight in physics -- read below.</p>
</div>
<p>The <a name="BOOKMARK11"></a><strong>atomic mass</strong> of an <a href="#BOOKMARK3">atom</a> is measured in <a name="BOOKMARK12"></a><strong>atomic mass unit</strong>s (<a name="BOOKMARK13"></a><strong>amu</strong>), where 1 <a href="#BOOKMARK13">amu</a> is: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1.66054</mn><mo>⋅</mo><mn>1</mn><msup><mn>0</mn><mrow><mo>−</mo><mn>24</mn></mrow></msup></mrow><annotation encoding="application/x-tex">1.66054 \cdot 10^{-24}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord">.</span><span class="mord">6</span><span class="mord">6</span><span class="mord">0</span><span class="mord">5</span><span class="mord">4</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">−</span><span class="mord mtight">2</span><span class="mord mtight">4</span></span></span></span></span></span></span></span></span></span></span></span> grams. Note that 1 <a href="#BOOKMARK13">amu</a> is roughly equal to the mass of a single <a href="#BOOKMARK6">proton</a> / <a href="#BOOKMARK7">neutron</a>:</p>
<ul>
<li>1 <a href="#BOOKMARK6">proton</a> = <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1.6726231</mn><mo>⋅</mo><mn>1</mn><msup><mn>0</mn><mrow><mo>−</mo><mn>24</mn></mrow></msup></mrow><annotation encoding="application/x-tex">1.6726231 \cdot 10^{-24}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord">.</span><span class="mord">6</span><span class="mord">7</span><span class="mord">2</span><span class="mord">6</span><span class="mord">2</span><span class="mord">3</span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">−</span><span class="mord mtight">2</span><span class="mord mtight">4</span></span></span></span></span></span></span></span></span></span></span></span> gram</li>
<li>1 <a href="#BOOKMARK7">neutron</a> = <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1.674920</mn><mo>⋅</mo><mn>1</mn><msup><mn>0</mn><mrow><mo>−</mo><mn>24</mn></mrow></msup></mrow><annotation encoding="application/x-tex">1.674920 \cdot 10^{-24}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord">.</span><span class="mord">6</span><span class="mord">7</span><span class="mord">4</span><span class="mord">9</span><span class="mord">2</span><span class="mord">0</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">−</span><span class="mord mtight">2</span><span class="mord mtight">4</span></span></span></span></span></span></span></span></span></span></span></span> gram</li>
<li>1 <a href="#BOOKMARK8">electron</a> = <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>9.1093897</mn><mo>⋅</mo><mn>1</mn><msup><mn>0</mn><mrow><mo>−</mo><mn>28</mn></mrow></msup></mrow><annotation encoding="application/x-tex">9.1093897 \cdot 10^{-28}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">9</span><span class="mord">.</span><span class="mord">1</span><span class="mord">0</span><span class="mord">9</span><span class="mord">3</span><span class="mord">8</span><span class="mord">9</span><span class="mord">7</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">−</span><span class="mord mtight">2</span><span class="mord mtight">8</span></span></span></span></span></span></span></span></span></span></span></span> gram</li>
</ul>
<p>That's because the constant for 1 <a href="#BOOKMARK13">amu</a> was derived by measuring the mass of carbon-12 and dividing it by 12. Carbon-12 has exactly 6 <a href="#BOOKMARK6">proton</a>s and 6 <a href="#BOOKMARK7">neutron</a>s (12 total). Since <a href="#BOOKMARK6">proton</a>s and <a href="#BOOKMARK7">neutron</a>s have roughly the same mass (and <a href="#BOOKMARK8">electron</a>s have almost no mass), when carbon-12 is divided by 12 it roughly gives the mass of a single <a href="#BOOKMARK6">proton</a> / <a href="#BOOKMARK7">neutron</a>.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Carbon-12 was chosen because it's the most common carbon <a href="#BOOKMARK9">isotope</a> on earth (99% abundance).</p>
</div>
<p>The <a name="BOOKMARK14"></a><strong>atomic weight</strong> / <a name="BOOKMARK15"></a><strong>relative atomic mass</strong> of an <a href="#BOOKMARK3">atom</a> is the weighted average of <a href="#BOOKMARK11">atomic mass</a> across all its <a href="#BOOKMARK9">isotope</a>s, where each <a href="#BOOKMARK9">isotope</a> is weighted by how abundant it is. For example, there are 3 <a href="#BOOKMARK9">isotope</a>s for the some imaginary <a href="#BOOKMARK3">atom</a> type...</p>
<ul>
<li>Imaginary-12 - 3 <a href="#BOOKMARK13">amu</a> and 10% abundance</li>
<li>Imaginary-13 - 4 <a href="#BOOKMARK13">amu</a> and 30% abundance</li>
<li>Imaginary-14 - 5 <a href="#BOOKMARK13">amu</a> and 60% abundance</li>
</ul>
<p>Its <a href="#BOOKMARK14">atomic weight</a> can be calculated as <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>3</mn><mo>⋅</mo><mn>0.1</mn><mo>+</mo><mn>4</mn><mo>⋅</mo><mn>0.3</mn><mo>+</mo><mn>5</mn><mo>⋅</mo><mn>0.6</mn><mo>=</mo><mn>4.6</mn></mrow><annotation encoding="application/x-tex">3 \cdot 0.1 + 4 \cdot 0.3 + 5 \cdot 0.6 = 4.6</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">3</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.72777em;vertical-align:-0.08333em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">1</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">4</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.72777em;vertical-align:-0.08333em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">3</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">5</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">0</span><span class="mord">.</span><span class="mord">6</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">4</span><span class="mord">.</span><span class="mord">6</span></span></span></span> <a href="#BOOKMARK13">amu</a>.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>How are abundances calculated? A chunk of the naturally occurring <a href="#BOOKMARK2">element</a> is dug up and passed through a machine called a Gas Chromatograph-Mass Spectrometer. The machine isolates by weight and shows how abundant each weight is.</p>
</div>
<a name="HEADREF3"></a><h3>Atomic Count</h3>
<p>A <a name="BOOKMARK16"></a><strong>mole</strong> is the unit used to measure the number/count of the particular particle making up a larger substance (e.g. <a href="#BOOKMARK6">proton</a>, <a href="#BOOKMARK3">atom</a>, <a href="#BOOKMARK18">molecule</a>, etc..). 1 <a href="#BOOKMARK16">mole</a> is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>6.02214076</mn><mo>⋅</mo><mn>1</mn><msup><mn>0</mn><mn>23</mn></msup></mrow><annotation encoding="application/x-tex">6.02214076 \cdot 10^{23}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">6</span><span class="mord">.</span><span class="mord">0</span><span class="mord">2</span><span class="mord">2</span><span class="mord">1</span><span class="mord">4</span><span class="mord">0</span><span class="mord">7</span><span class="mord">6</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span><span class="mord mtight">3</span></span></span></span></span></span></span></span></span></span></span></span> particles. For example, ...</p>
<ul>
<li>1 <a href="#BOOKMARK16">mole</a> of carbon-12 = <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>6.02214076</mn><mo>⋅</mo><mn>1</mn><msup><mn>0</mn><mn>23</mn></msup></mrow><annotation encoding="application/x-tex">6.02214076 \cdot 10^{23}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">6</span><span class="mord">.</span><span class="mord">0</span><span class="mord">2</span><span class="mord">2</span><span class="mord">1</span><span class="mord">4</span><span class="mord">0</span><span class="mord">7</span><span class="mord">6</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span><span class="mord mtight">3</span></span></span></span></span></span></span></span></span></span></span></span> carbon-12 <a href="#BOOKMARK3">atom</a>s = 12 grams of carbon-12</li>
<li>1 <a href="#BOOKMARK16">mole</a> of iron-56 = <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>6.02214076</mn><mo>⋅</mo><mn>1</mn><msup><mn>0</mn><mn>23</mn></msup></mrow><annotation encoding="application/x-tex">6.02214076 \cdot 10^{23}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">6</span><span class="mord">.</span><span class="mord">0</span><span class="mord">2</span><span class="mord">2</span><span class="mord">1</span><span class="mord">4</span><span class="mord">0</span><span class="mord">7</span><span class="mord">6</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span><span class="mord mtight">3</span></span></span></span></span></span></span></span></span></span></span></span> iron-56 <a href="#BOOKMARK3">atom</a>s = 56 grams of iron-56.</li>
<li>1 <a href="#BOOKMARK16">mole</a> of <a href="#BOOKMARK59">water</a> = <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>6.02214076</mn><mo>⋅</mo><mn>1</mn><msup><mn>0</mn><mn>23</mn></msup></mrow><annotation encoding="application/x-tex">6.02214076 \cdot 10^{23}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.64444em;vertical-align:0em;"></span><span class="mord">6</span><span class="mord">.</span><span class="mord">0</span><span class="mord">2</span><span class="mord">2</span><span class="mord">1</span><span class="mord">4</span><span class="mord">0</span><span class="mord">7</span><span class="mord">6</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.8141079999999999em;vertical-align:0em;"></span><span class="mord">1</span><span class="mord"><span class="mord">0</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141079999999999em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span><span class="mord mtight">3</span></span></span></span></span></span></span></span></span></span></span></span> <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>s = 18.02 grams.</li>
</ul>
<p>The constant for <a href="#BOOKMARK16">mole</a> is also know as the <a name="BOOKMARK17"></a><strong>Avogadro number</strong>. It's the number of carbon-12 <a href="#BOOKMARK3">atom</a>s required to reach a mass of 12 grams.</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<p>TODO: FIX TO EXPLAIN THE RELATIONSHIP BETWEEN <a href="#BOOKMARK3">ATOM</a> <a href="#BOOKMARK13">AMU</a> TO GRAMS (e.g. 1 <a href="#BOOKMARK16">mole</a> of iron-56=56g, 1 <a href="#BOOKMARK16">mole</a> of carbon-12=12g)</p>
<a name="HEADREF4"></a><h2>Molecule</h2>
<p>When <a href="#BOOKMARK3">atom</a>s bind together, they form a <a name="BOOKMARK18"></a><strong>molecule</strong>. Each type of <a href="#BOOKMARK18">molecule</a> has the same configuration of <a href="#BOOKMARK3">atom</a>s -- same <a href="#BOOKMARK3">atom</a>s in the same numbers, structured/shaped similarly. For example, a <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a> is made up of 2 hydrogen <a href="#BOOKMARK3">atom</a>s and 1 oxygen <a href="#BOOKMARK3">atom</a> binding together in a house-roof shape...</p>
<p><img src="200px-H2O_2D_labelled.svg.png" alt="By Dan Craggs - Own workThis vector image includes elements that have been taken or adapted from this:  Water-2D-labelled.png., Public Domain, https://commons.wikimedia.org/w/index.php?curid=7916072"></p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>As far as I can tell, the <a href="#BOOKMARK3">atom</a>s will always bind in the same way. You can't ever have a <a href="#BOOKMARK18">molecule</a> that has the same types of <a href="#BOOKMARK3">atom</a>s in the same numbers but with a different structure.</p>
</div>
<p>A <a name="BOOKMARK19"></a><strong>monomer</strong> is a special designation for <a href="#BOOKMARK3">atom</a>s/<a href="#BOOKMARK18">molecule</a>s that are able to join with other <a href="#BOOKMARK19">monomer</a>s to create even larger <a href="#BOOKMARK18">molecule</a>s. The process of joining is called <a name="BOOKMARK20"></a><strong>polymerization</strong> and the resulting <a href="#BOOKMARK18">molecule</a> is called a <a name="BOOKMARK21"></a><strong>polymer</strong>.</p>
<p><img src="/.datafile_b796c4d8f40b54b5e3c03483795ce7d0/diagram.svg" alt="Graphviz Dot Diagram"></p><p>If the <a href="#BOOKMARK19">monomer</a>s that make up a <a href="#BOOKMARK21">polymer</a> are all the same, the <a href="#BOOKMARK21">polymer</a> is called a <a name="BOOKMARK22"></a><strong>homopolymer</strong>. Otherwise, it's called a <a name="BOOKMARK23"></a><strong>heteropolymer</strong> / <a name="BOOKMARK24"></a><strong>copolymer</strong>.</p>
<p>For example, the glucose <a href="#BOOKMARK18">molecule</a> is a <a href="#BOOKMARK19">monomer</a>. It can combine with other glucose <a href="#BOOKMARK18">molecule</a>s to create the glycogen <a href="#BOOKMARK18">molecule</a>, which is a <a href="#BOOKMARK21">polymer</a> / <a href="#BOOKMARK22">homopolymer</a>. Other examples of <a href="#BOOKMARK21">polymer</a>s (according to Wikipedia): <a href="#BOOKMARK56">amino acid</a>s and <a href="#BOOKMARK232">nucleotide</a>s (<a href="#BOOKMARK241">DNA</a>).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>There are probably special properties to <a href="#BOOKMARK19">monomer</a>s that allow them to chain up. The Wikipedia page talks about a feature of <a href="#BOOKMARK19">monomer</a>s being a "carbon double bond" which is what allows them to form <a href="#BOOKMARK21">polymer</a>s.</p>
</div>
<p><a href="#BOOKMARK21">Polymer</a>s are often referred to as <a name="BOOKMARK25"></a><strong>macromolecule</strong>s -- <a href="#BOOKMARK18">molecule</a>s that have a very large number of <a href="#BOOKMARK3">atom</a>s.</p>
<a name="HEADREF5"></a><h2>Ion</h2>
<p>An <a name="BOOKMARK26"></a><strong>ion</strong> is a charged <a href="#BOOKMARK3">atom</a> or <a href="#BOOKMARK18">molecule</a>. A charged <a href="#BOOKMARK3">atom</a>/<a href="#BOOKMARK18">molecule</a> just means that it has an unequal number of <a href="#BOOKMARK6">proton</a>s and <a href="#BOOKMARK8">electron</a>s:</p>
<ul>
<li>if <a href="#BOOKMARK6">proton</a>s &gt; <a href="#BOOKMARK8">electron</a>s (more <a href="#BOOKMARK6">proton</a>s), it's called a <a name="BOOKMARK27"></a><strong>positively charged</strong> <a href="#BOOKMARK26">ion</a> / <a name="BOOKMARK28"></a><strong>positive ion</strong>.</li>
<li>if <a href="#BOOKMARK6">proton</a>s &lt; <a href="#BOOKMARK8">electron</a>s (more <a href="#BOOKMARK8">electron</a>s), it's called a <a name="BOOKMARK29"></a><strong>negatively charged</strong> <a href="#BOOKMARK26">ion</a> / <a name="BOOKMARK30"></a><strong>negative ion</strong> / <a name="BOOKMARK31"></a><strong>ANION</strong>.</li>
<li>if <a href="#BOOKMARK6">proton</a>s == <a href="#BOOKMARK8">electron</a>s, it's a neutral (no charge).</li>
</ul>
<p><a href="#BOOKMARK26">Ions</a> are always trying to lose their charge and become neutral, either by giving up an <a href="#BOOKMARK8">electron</a>s or pulling in an <a href="#BOOKMARK8">electron</a>s such that the number of <a href="#BOOKMARK6">proton</a>s and <a href="#BOOKMARK8">electron</a>s become equal. As such, <a href="#BOOKMARK26">ions</a> will attract towards oppositely charged <a href="#BOOKMARK26">ions</a> and repel from similarly charged <a href="#BOOKMARK26">ions</a>:</p>
<ul>
<li><a href="#BOOKMARK30">negative ion</a>s are attracted to <a href="#BOOKMARK28">positive ion</a>s and repelled from <a href="#BOOKMARK30">negative ion</a>s.</li>
<li><a href="#BOOKMARK28">positive ion</a>s are attracted to <a href="#BOOKMARK30">negative ion</a>s and repelled from <a href="#BOOKMARK28">positive ion</a>s.</li>
</ul>
<a name="HEADREF6"></a><h2>pH</h2>
<p><a name="BOOKMARK32"></a><strong>pH</strong> stands for <em>potential of hydrogen</em> and it's the measure of <a href="#BOOKMARK27">positively charged</a> hydrogen <a href="#BOOKMARK26">ions</a> in a solution. The more...</p>
<ul>
<li>the more <a name="BOOKMARK33"></a><strong>acidic</strong> something is, the more <a href="#BOOKMARK27">positively charged</a> hydrogen <a href="#BOOKMARK26">ions</a> it has.</li>
<li>the more <a name="BOOKMARK34"></a><strong>basic</strong> <a name="BOOKMARK35"></a><strong></strong> (<a name="BOOKMARK36"></a><strong>alkaline</strong>) something is, the more reactive it is to <a href="#BOOKMARK27">positively charged</a> hydrogen <a href="#BOOKMARK26">ions</a> (it wants to give off <a href="#BOOKMARK8">electron</a>s to those hydrogen <a href="#BOOKMARK26">ions</a>).</li>
</ul>
<p><img src="800px-PH_Scale-_Acidic_vs._Basic_(Alkaline).png" alt="By Heinrich-Boll-Stiftung - https://www.flickr.com/photos/boellstiftung/35805740223, CC BY-SA 2.0, https://commons.wikimedia.org/w/index.php?curid=66098876"></p>
<p><a href="#BOOKMARK32">pH</a> is scaled logarithmically from 1 to 14. Each notch on the scale moves the <a href="#BOOKMARK33">acidity</a>/<a href="#BOOKMARK34">basicity</a> by a factory of 10. Going...</p>
<ul>
<li>back a notch (-1) increases <a href="#BOOKMARK33">acidity</a> / decreases <a href="#BOOKMARK34">basicity</a> by a factor of 10.</li>
<li>forward a notch (+1) decreases <a href="#BOOKMARK33">acidity</a> / increases <a href="#BOOKMARK34">basicity</a> by a factory of 10.</li>
</ul>
<p>For example, going from  7 to 4 increases <a href="#BOOKMARK33">acidity</a> by 1000x times / decreases <a href="#BOOKMARK34">basicity</a> by 1000x.</p>
<p>The closer to...</p>
<ul>
<li>1 something is, the more <a href="#BOOKMARK33">acidic</a> it is (more <a href="#BOOKMARK27">positively charged</a> hydrogen <a href="#BOOKMARK26">ions</a>) and the more sour/sharp it tastes.</li>
<li>14 something is, the more <a href="#BOOKMARK34">basic</a> it is (more stuff that can react with <a href="#BOOKMARK27">positively charged</a> hydrogen <a href="#BOOKMARK26">ions</a>) and the more bitter it tastes.</li>
<li>7 something is, the more neutral it is (not reactive).</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>https://www.quora.com/Why-is-pure-<a href="#BOOKMARK59">water</a>-considered-neutral -- Since <a href="#BOOKMARK32">pH</a> is defined as the negative log of the hydrogen <a href="#BOOKMARK26">ion</a> concentration, the <a href="#BOOKMARK32">pH</a> of pure <a href="#BOOKMARK59">water</a> is 7 or neutral. Pure <a href="#BOOKMARK59">water</a> is neutral because the number of positive hydrogen <a href="#BOOKMARK26">ions</a> produced is equal to the number of negative.</p>
</div>
<a name="HEADREF7"></a><h1>Biology</h1>
<p><a name="BOOKMARK37"></a><strong>Biology</strong> is defined as the study of <a href="#BOOKMARK38">life</a> / <a href="#BOOKMARK38">living</a> things. A <a name="BOOKMARK38"></a><strong>living</strong> thing is defined as anything that converts energy from one form to another, while using that energy to grow, change, and reproduce.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Primary source for these notes is Khan Academy's HS <a href="#BOOKMARK37">Biology</a> course, with additional information from Wikipedia and other sources.</p>
</div>
<a name="HEADREF8"></a><h2>Scientific Method</h2>
<p>The <a name="BOOKMARK39"></a><strong>scientific method</strong> is the standard guideline for discovery and <a href="#BOOKMARK41">experiment</a>ation in the sciences (<a href="#BOOKMARK0">chemistry</a>, physics, <a href="#BOOKMARK37">biology</a>, etc..) The basic steps are...</p>
<ol>
<li>Observe.</li>
<li>Ask a question about the observation.</li>
<li>Make a guess that answers the question (<a href="#BOOKMARK40">hypothesis</a>).</li>
<li>Test the guess to see if it's correct (<a href="#BOOKMARK41">experiment</a>).</li>
<li>Refine and iterate.</li>
</ol>
<p>The last step (refine and iterate) just means that you do it all over again but make changes based on the things you learned from your <a href="#BOOKMARK41">experiment</a>. For example, ...</p>
<ul>
<li>do additional <a href="#BOOKMARK41">experiment</a>s to dig into some aspect deeper.</li>
<li>if the <a href="#BOOKMARK40">hypothesis</a> wasn't supported by the <a href="#BOOKMARK41">experiment</a>, maybe come up with a new <a href="#BOOKMARK40">hypothesis</a>.</li>
</ul>
<a name="HEADREF9"></a><h3>Hypothesis</h3>
<p>The <a href="#BOOKMARK39">scientific method</a> revolves around making an observation and coming up with a testable explanation for that observation -- called a <a name="BOOKMARK40"></a><strong>hypothesis</strong>. If the explanation isn't <em>testable</em>, you can't consider it a <a href="#BOOKMARK40">hypothesis</a>. For example, a good <a href="#BOOKMARK40">hypothesis</a> may be that increased sun exposure leads to an increased risk of skin <a href="#BOOKMARK193">cancer</a> because it's something you can test. A bad explanation may be that exposure to centaurs increase the risk of skin <a href="#BOOKMARK193">cancer</a> because centaurs don't exist (and as such the <a href="#BOOKMARK40">hypothesis</a> can't be tested).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The material mentions that for a <a href="#BOOKMARK40">hypothesis</a> to be testable, you should be able to come up with an <a href="#BOOKMARK41">experiment</a> that shows that its false -- it's falsifiable. How you word your <a href="#BOOKMARK40">hypothesis</a> is typically what determines if it's testable/falsifiable -- when you read the <a href="#BOOKMARK40">hypothesis</a>, what defines a failure?</p>
</div>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>A <a href="#BOOKMARK40">hypothesis</a> and a theory are different things. <a href="#BOOKMARK40">Hypothesis</a> is a potential answer for a specific problem. A theory provides a potential framework for a much broader class of problems based on supporting evidence. The example given by the material: "The toaster won't toast because the electrical outlet is broken" is a <a href="#BOOKMARK40">hypothesis</a>, whereas "Electrical appliances need a source of electricity in order to run" is closer to a theory.</p>
</div>
<a name="HEADREF10"></a><h3>Experiment</h3>
<p>Once you have a <a href="#BOOKMARK40">hypothesis</a>, you design an <a name="BOOKMARK41"></a><strong>experiment</strong> to test it. In the case of our sun exposure leads to increased risk of skin <a href="#BOOKMARK193">cancer</a> <a href="#BOOKMARK40">hypothesis</a>, an <a href="#BOOKMARK41">experiment</a> may be to expose skin <a href="#BOOKMARK74">cell</a>s to UV rays in amounts equivalent to that given off by the sun and then check to see if those <a href="#BOOKMARK74">cell</a>s have been damaged (compared to a <a href="#BOOKMARK42">control group</a> of skin <a href="#BOOKMARK74">cell</a>s that you haven't exposed to UV rays).</p>
<p>What makes a good <a href="#BOOKMARK41">experiment</a>?</p>
<ul>
<li><a name="BOOKMARK42"></a><strong>Control group</strong> - A <a href="#BOOKMARK42">control group</a> contains a set of entities that don't get <a href="#BOOKMARK47">treatment</a>, but those entities are equal to the entities in the <a name="BOOKMARK43"></a><strong>experimental group</strong> (as much as possible) and exposed to the same conditions as the entities in the <a href="#BOOKMARK43">experimental group</a> (as much as possible). This helps eliminate outside factors/variables from tainting the results of an <a href="#BOOKMARK41">experiment</a> -- it's a <a name="BOOKMARK44"></a><strong>controlled experiment</strong> . For example, in the skin <a href="#BOOKMARK193">cancer</a> example <a href="#BOOKMARK41">experiment</a> described above, a <a href="#BOOKMARK42">control group</a> may be a set of skin <a href="#BOOKMARK74">cell</a>s that sit along side the <a href="#BOOKMARK43">experimental group</a> skin <a href="#BOOKMARK74">cell</a>s that get treated with UV rays -- the <a href="#BOOKMARK74">cell</a>s in the <a href="#BOOKMARK42">control group</a> shouldn't show signs of damage while the <a href="#BOOKMARK74">cell</a>s in the <a href="#BOOKMARK43">experimental group</a> should.</li>
<li><a name="BOOKMARK45"></a><strong>Random selection</strong> - The entities being used for an <a href="#BOOKMARK41">experiment</a> must be randomly selected (both for the <a href="#BOOKMARK42">control group</a> and the <a href="#BOOKMARK43">experimental group</a>). <a href="#BOOKMARK45">Random selection</a> helps minimize skewed results introduced from bias during the selection process. For example, the individual responsible for selecting candidates for drug trials may choose to disproportionately select people of a certain race (e.g. racism).</li>
<li><a name="BOOKMARK46"></a><strong>Double blind</strong> - The people conducting the <a href="#BOOKMARK41">experiment</a> may implicitly/explicitly taint the <a href="#BOOKMARK41">experiment</a> via their own bias. For example, if the <a href="#BOOKMARK41">experiment</a> involves an interviewer asking a set of questions, the interviewer may subconsciously change the pronunciations of words if he/she knows that the interviewee is in the <a href="#BOOKMARK42">control group</a>, there by effecting the answer given by the interviewee.</li>
<li><strong>Statistical significance</strong> - The results between the <a href="#BOOKMARK42">control group</a> and the <a href="#BOOKMARK43">experimental group</a> needs to be large enough to support your <a href="#BOOKMARK40">hypothesis</a>. A small difference could mean that outcome was due to variance in the selected groups (there's a branch of statistics that deals with this).</li>
<li><strong>Replicable</strong> - Others should be able to repeat the same <a href="#BOOKMARK41">experiment</a> and (hopefully) come to the same result. If they don't, it could be that something was wrong with your <a href="#BOOKMARK41">experiment</a> or the entities being tested had some underlying difference that changed the result.</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>There's always at least one <a href="#BOOKMARK42">control group</a> in any <a href="#BOOKMARK41">experiment</a> to provide a baseline. There's no limit to the number of <a href="#BOOKMARK43">experimental group</a>s -- each group may have a slightly different type/amount of <a href="#BOOKMARK47">treatment</a> applied.</p>
</div>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Because things are so wishy-washy/not-exactly in <a href="#BOOKMARK37">biology</a>, it's typical for an <a href="#BOOKMARK41">experiment</a>s to be repeated multiple times and to have a large sample size -- the larger our sample sizes and the more times we conduct the <a href="#BOOKMARK41">experiment</a>, the more we can be confident of our result. What do I mean by wishy-washy? Genetic variation between samples may result in different types/levels of responses. For example, people with a certain <a href="#BOOKMARK244">gene</a> may respond quicker to certain drugs than people who don't produce that <a href="#BOOKMARK244">gene</a>.</p>
</div>
<p>Other terminology around the <a href="#BOOKMARK39">scientific method</a>...</p>
<ul>
<li><a name="BOOKMARK47"></a><strong>treatment</strong> - The application of the test given to the <a href="#BOOKMARK43">experimental group</a>(s) but not the <a href="#BOOKMARK42">control group</a>.</li>
<li><a name="BOOKMARK48"></a><strong>independent variable</strong> - An <a href="#BOOKMARK48">independent variable</a> is basically the <a href="#BOOKMARK47">treatment</a> you apply -- you can think of it as the input knob you control for your <a href="#BOOKMARK41">experiment</a>. For example, in the skin <a href="#BOOKMARK193">cancer</a> <a href="#BOOKMARK41">experiment</a>, the <a href="#BOOKMARK48">independent variable</a> would be the amount of artificial UV rays you apply to the skin <a href="#BOOKMARK74">cell</a>s.</li>
<li><a name="BOOKMARK49"></a><strong>dependent variable</strong> - A <a href="#BOOKMARK49">dependent variable</a> is a response you measure after applying the <a href="#BOOKMARK47">treatment</a> -- you can think of it as the output from your <a href="#BOOKMARK41">experiment</a>. For example, in the skin <a href="#BOOKMARK193">cancer</a> <a href="#BOOKMARK41">experiment</a>, the <a href="#BOOKMARK49">dependent variable</a> might be a measurement of how different the <a href="#BOOKMARK74">cell</a>s are functioning (e.g. a <a href="#BOOKMARK74">cell</a> with damaged <a href="#BOOKMARK241">DNA</a> may produce different kinds / amounts of <a href="#BOOKMARK250">tRNA</a> <a href="#BOOKMARK18">molecule</a>s -- something you can measure).</li>
<li><strong>data</strong> - This is exactly what you think it is -- measurements/observations made during the <a href="#BOOKMARK41">experiment</a>.</li>
<li><a name="BOOKMARK50"></a><strong>placebo effect</strong> - When <a href="#BOOKMARK41">experiment</a>ing on people, there's a phenomenon called the <a href="#BOOKMARK50">placebo effect</a>: if someone takes something that shouldn't help them but are under the impression it will help them, it often times will help them, just from the psychological effect alone.</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>You can have more than one <a href="#BOOKMARK48">independent variable</a> if you follow specific guidelines and are experienced enough, but the general rule of thumb is to have only 1 <a href="#BOOKMARK48">independent variable</a> just because it makes things much simpler to analyze/interpret.</p>
</div>
<a name="HEADREF11"></a><h2>Carbohydrate Molecule</h2>
<p><a name="BOOKMARK51"></a><strong>Carbohydrate</strong>s (also called <a name="BOOKMARK52"></a><strong>saccharide</strong>s) are <a href="#BOOKMARK18">molecule</a>s that consist of a mix of carbon, hydrogen, and oxygen <a href="#BOOKMARK3">atom</a>s. In <a href="#BOOKMARK37">biological</a> systems, <a href="#BOOKMARK51">carbohydrate</a>s are often associated with...</p>
<ul>
<li>being a source of energy</li>
<li>providing a structural role (for plants / certain plants).</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>It was never explained what 'structural role' actually means.</p>
</div>
<p>The term <a name="BOOKMARK53"></a><strong>monosaccharide</strong> is just means a <a href="#BOOKMARK51">carbohydrate</a> that's a <a href="#BOOKMARK19">monomer</a> (e.g. glucose). Similarly, the term <a name="BOOKMARK54"></a><strong>polysaccharide</strong> means a <a href="#BOOKMARK51">carbohydrate</a> built from other <a href="#BOOKMARK53">monosaccharide</a>s (e.g. glycogen is made of chained glucose).</p>
<p><img src="/.datafile_3309dc1fa4329d35f76d8967744e0ce8/diagram.svg" alt="Graphviz Dot Diagram"></p><a name="HEADREF12"></a><h2>Protein Molecule</h2>
<p><a name="BOOKMARK55"></a><strong>Protein</strong>s are <a href="#BOOKMARK18">molecule</a>s that consist of <a href="#BOOKMARK19">monomer</a>s called <a name="BOOKMARK56"></a><strong>amino acid</strong>s. The <a href="#BOOKMARK56">amino acid</a>s get chained together into a <a href="#BOOKMARK21">polymer</a> called a <a name="BOOKMARK57"></a><strong>polypeptide</strong> chain, and one or more <a href="#BOOKMARK57">polypeptide</a> chains fold to a 3D structure and combine to become a <a href="#BOOKMARK55">protein</a>. The 3D structure / shape of the <a href="#BOOKMARK55">protein</a> (how its folded) is what gives it its abilities.</p>
<p>In <a href="#BOOKMARK37">biological</a> systems, <a href="#BOOKMARK55">protein</a>s are often associated with that facilitating some <a href="#BOOKMARK37">biological</a> function. For example, the <a href="#BOOKMARK55">protein</a> protease is responsible for breaking down food.</p>
<p><img src="/.datafile_420e0100031cd8640addc82b4dcb49b4/diagram.svg" alt="Graphviz Dot Diagram"></p><p><img src="320px-Protein_folding.png" alt="By No machine-readable author provided. DrKjaergaard assumed (based on copyright claims). - No machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=1967109"></p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The <a href="#BOOKMARK81">ribosome</a> is what's responsible for folding? Not able to get a clear answer on this.</p>
</div>
<p>The basic structure of an <a href="#BOOKMARK56">amino acid</a> is as follows. The R is a placeholder that, when set, defines what type of <a href="#BOOKMARK56">amino acid</a> it is...</p>
<p><img src="320px-AminoAcidball.svg.png" alt="By GYassineMrabetTalk✉This W3C-unspecified vector image was created with Inkscape. - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=2551977"></p>
<a name="HEADREF13"></a><h2>Lipid Molecule</h2>
<p><a name="BOOKMARK58"></a><strong>Lipid</strong>s are <a href="#BOOKMARK18">molecule</a>s that are somewhat not <a href="#BOOKMARK59">water</a> soluble -- meaning that they have parts that resist <a href="#BOOKMARK59">water</a> but maybe also parts that are attracted to <a href="#BOOKMARK59">water</a>. In <a href="#BOOKMARK37">biological</a> systems, <a href="#BOOKMARK58">lipid</a>s are often associated with...</p>
<ul>
<li>energy storage (fats)</li>
<li><a href="#BOOKMARK74">cell</a>ular <a href="#BOOKMARK77">membrane</a>s (<a href="#BOOKMARK89">phospholipid</a>s)</li>
</ul>
<p><img src="296px-Trimyristin-3D-vdW.png" alt="Public Domain, https://commons.wikimedia.org/w/index.php?curid=1019610"></p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p><a href="#BOOKMARK58">Lipid</a>s are not always fats. All fats are <a href="#BOOKMARK58">lipid</a>s but not all <a href="#BOOKMARK58">lipid</a>s are fats.</p>
</div>
<a name="HEADREF14"></a><h2>Water Molecule</h2>
<p><a name="BOOKMARK59"></a><strong>Water</strong> is essential to <a href="#BOOKMARK38">life</a> -- it has unique properties that almost all <a href="#BOOKMARK37">biological</a> processes depend on.</p>
<p>Recall that...</p>
<ol>
<li>a <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a> consists of 2 hydrogen <a href="#BOOKMARK3">atom</a>s connected to an oxygen <a href="#BOOKMARK3">atom</a> via <a href="#BOOKMARK60">covalent bond</a>s. A <a name="BOOKMARK60"></a><strong>covalent bond</strong> is a pair of <a href="#BOOKMARK8">electron</a>s that both <a href="#BOOKMARK3">atom</a>s share, thus bonding the <a href="#BOOKMARK3">atom</a>s together.</li>
<li>the position of an <a href="#BOOKMARK8">electron</a> is based on probability. <a href="#BOOKMARK8">Electron</a>s aren't fixed in a certain position or neatly orbiting around a <a href="#BOOKMARK5">nucleus</a> as certain diagrams show. Rather, they're constantly buzzing/hopping around the <a href="#BOOKMARK5">nucleus</a>. Depending on their environment, they may be more likely to be at certain locations vs other locations.</li>
</ol>
<p>Oxygen <a href="#BOOKMARK3">atom</a>s are extremely <a name="BOOKMARK61"></a><strong>electronegative</strong>, meaning that oxygen has the propensity to pull the buzzing/hopping <a href="#BOOKMARK8">electron</a>s more around itself than the <a href="#BOOKMARK3">atom</a>s it's bound to. As such, in a <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>, the <a href="#BOOKMARK8">electron</a>s will spend more time solely around the oxygen <a href="#BOOKMARK3">atom</a> than they do the hydrogen <a href="#BOOKMARK3">atom</a> or a position that binds the hydrogen and oxygen together. This is what gives the oxygen <a href="#BOOKMARK3">atom</a> in a <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a> a <a name="BOOKMARK62"></a><strong>weakly negative</strong> charge (as indicated by δ-) while the hydrogen <a href="#BOOKMARK3">atom</a>s have a <a name="BOOKMARK63"></a><strong>weakly positive</strong> charge (as indicated by δ+). These types of charged <a href="#BOOKMARK18">molecule</a>s are called <a name="BOOKMARK64"></a><strong>polar molecule</strong>s.</p>
<p><img src="209_Polar_Covalent_Bonds_in_a_Water_Molecule.jpg" alt="By OpenStax College - Anatomy &amp; Physiology, Connexions Web site. http://cnx.org/content/col11496/1.6/, Jun 19, 2013., CC BY 3.0, https://commons.wikimedia.org/w/index.php?curid=30131144"></p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Notice the shape of the <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a> in the diagram(s) above. <a href="#BOOKMARK8">Electron</a> pairs are repelled from each other. They're also responsible for binding. That's what gives <a href="#BOOKMARK18">molecule</a>s their shapes/structure.</p>
</div>
<p>This <a href="#BOOKMARK62">weakly negative</a> / <a href="#BOOKMARK63">weakly positive</a> charge is what gives <a href="#BOOKMARK59">water</a> several of the unique properties that <a href="#BOOKMARK37">biological</a> properties depend on. <a href="#BOOKMARK59">Water</a> <a href="#BOOKMARK18">molecule</a>s have a tendency to gravitate towards each other because the <a href="#BOOKMARK62">weakly negative</a> oxygen <a href="#BOOKMARK3">atom</a>s and the <a href="#BOOKMARK63">weakly positive</a> hydrogen <a href="#BOOKMARK3">atom</a>s of different <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>s attract. This attraction is called a <a name="BOOKMARK65"></a><strong>hydrogen bond</strong>. <a href="#BOOKMARK65">Hydrogen bond</a>s are weaker than <a href="#BOOKMARK60">covalent bond</a>s in that the bonds aren't really solid -- <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>s can easily break off and go past each other.</p>
<p><img src="242px-3D_model_hydrogen_bonds_in_water.svg.png" alt="By User Qwerter at Czech wikipedia: Qwerter. Transferred from cs.wikipedia to Commons by sevela.p. Translated to english by by Michal Maňas (User:snek01). Vectorized by Magasjukur2 - File:3D model hydrogen bonds in water.jpg, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=14929959"></p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The above paragraph is just giving the mechanics for how/why <a href="#BOOKMARK59">water</a> is a liquid. <a href="#BOOKMARK59">Water</a> is the only <a href="#BOOKMARK18">molecule</a> that's liquid at room temperature? This can't be right -- see https://<a href="#BOOKMARK0">chemistry</a>.stackexchange.com/q/76346. Why can't <a href="#BOOKMARK37">biological</a> processes work in one of these other <a href="#BOOKMARK18">molecule</a>s just as they do in <a href="#BOOKMARK59">water</a>? Maybe because they stay liquid at a shorter temperature range (e.g. 15-25C instead of 1-99C?</p>
</div>
<p>The weak attraction between <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>s is also what makes <a href="#BOOKMARK59">water</a> a <a name="BOOKMARK66"></a><strong>solvent</strong>. So long as they're <a href="#BOOKMARK64">polar molecule</a>s, other <a href="#BOOKMARK18">molecule</a>s can travel inside of <a href="#BOOKMARK59">water</a> using the same attraction from <a href="#BOOKMARK62">weakly negative</a> / <a href="#BOOKMARK63">weakly positive</a> charges -- they gravitate and float around <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>s just as other <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>s do. For example, the <a href="#BOOKMARK75">cytoplasm</a> of a <a href="#BOOKMARK74">cell</a> is a <a href="#BOOKMARK66">solvent</a> (mostly <a href="#BOOKMARK59">water</a>). It works because other <a href="#BOOKMARK18">molecule</a>s in the <a href="#BOOKMARK75">cytoplasm</a> (e.g. <a href="#BOOKMARK74">cell</a>ular machinery) can float around / travel around using the <a href="#BOOKMARK62">weakly negative</a> / <a href="#BOOKMARK63">weakly positive</a> charges.</p>
<p><a href="#BOOKMARK59">Water</a> is called a <a name="BOOKMARK67"></a><strong>universal solvent</strong> because it can dissolve more <a href="#BOOKMARK18">molecule</a>s than other other liquid. Note that the term universal doesn't mean that it can dissolve everything, just that it can dissolve more things than the others.</p>
<p>The properties that make <a href="#BOOKMARK59">water</a> conducive for <a href="#BOOKMARK37">biological</a> processes to operate:</p>
<ul>
<li>Fluid under normal conditions - under normal temperatures, pressures, etc.. there's <a name="BOOKMARK68"></a><strong>cohesion</strong> between other <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>s (fluid) / <a name="BOOKMARK69"></a><strong>adhesion</strong> between <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>s and other <a href="#BOOKMARK64">polar molecule</a>s / <a href="#BOOKMARK26">ions</a> (<a href="#BOOKMARK66">solvent</a>).</li>
<li><a href="#BOOKMARK67">Universal solvent</a> - <a href="#BOOKMARK59">Water</a> dissolves more things than any other liquid (more things doesn't mean everything).</li>
<li>High <a href="#BOOKMARK343">heat capacity</a> - Because it takes a lot of energy to heat up <a href="#BOOKMARK59">water</a>, its temperature can remain relatively stable as the environment around it changes (temperature regulation).</li>
<li>High <a href="#BOOKMARK344">heat of vaporization</a> - As <a href="#BOOKMARK59">water</a> evaporates, it traps heat from whatever its on (e.g. human body) and releases it into the environment (temperature regulation / cooling).</li>
<li>Less <a href="#BOOKMARK342">dense</a> as a solid - As <a href="#BOOKMARK59">water</a> freezes it forms crystals that space the <a href="#BOOKMARK18">molecule</a>s away from each other. As such, it becomes less <a href="#BOOKMARK342">dense</a> than liquid <a href="#BOOKMARK59">water</a> and ends up floats on top. This means bodies of <a href="#BOOKMARK59">water</a> can freeze over but <a href="#BOOKMARK38">life</a> can continue operating in the liquid underneath the frozen ice.</li>
</ul>
<p>Other terminology related to <a href="#BOOKMARK59">water</a>:</p>
<ul>
<li><a href="#BOOKMARK18">Molecule</a>s that are charged are called <a href="#BOOKMARK64">polar molecule</a>s (e.g. the <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a> is a <a href="#BOOKMARK64">polar molecule</a> because of its <a href="#BOOKMARK63">weakly positive</a> / <a href="#BOOKMARK62">weakly negative</a> charge).</li>
<li><a href="#BOOKMARK18">Molecule</a>s / <a href="#BOOKMARK26">ions</a> that are charged and mixed with a <a href="#BOOKMARK66">solvent</a> are called <a name="BOOKMARK70"></a><strong>solute</strong>s.</li>
<li><a href="#BOOKMARK18">Molecule</a>s that are attracted to <a href="#BOOKMARK59">water</a> are called <a name="BOOKMARK71"></a><strong>hydrophilic</strong>.</li>
<li><a href="#BOOKMARK18">Molecule</a>s that are repelled from <a href="#BOOKMARK59">water</a> are called <a name="BOOKMARK72"></a><strong>hydrophobic</strong>.</li>
<li><a href="#BOOKMARK18">Molecule</a>s that have parts that are <a href="#BOOKMARK71">hydrophilic</a> as well as parts that are <a href="#BOOKMARK72">hydrophobic</a> are called <a name="BOOKMARK73"></a><strong>amphipathic</strong></li>
</ul>
<a name="HEADREF15"></a><h2>Cell</h2>
<p><a name="BOOKMARK74"></a><strong>Cell</strong>s are the basic unit of <a href="#BOOKMARK38">living</a> things / the building blocks of <a href="#BOOKMARK38">life</a>. They're tiny structures that encapsulate information and machinery that allows them to replicate/reproduce and perform other important functions (e.g. appendages to move around).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Viruses are not <a href="#BOOKMARK74">cell</a>s but they may also be considered <a href="#BOOKMARK38">living</a> because they reproduce in a roundabout way: the require machinery in the host <a href="#BOOKMARK74">cell</a> to reproduce.</p>
</div>
<p>There are 2 types of <a href="#BOOKMARK74">cell</a>s: <a href="#BOOKMARK109">eukaryotic</a> and <a href="#BOOKMARK210">prokaryotic</a>. There main differences between them are that...</p>
<ol>
<li>the guts of <a href="#BOOKMARK109">eukaryote</a>s are organized into <a href="#BOOKMARK111">organelle</a>s (<a href="#BOOKMARK77">membrane</a>-bound compartments) where each one is responsible for some functionality, while <a href="#BOOKMARK210">prokaryote</a>s have no <a href="#BOOKMARK111">organelle</a>s at all (guts are free floating).</li>
<li>the <a href="#BOOKMARK241">DNA</a> in <a href="#BOOKMARK109">eukaryote</a>s into multiple independent segments (<a href="#BOOKMARK138">chromosome</a>s), while <a href="#BOOKMARK210">prokaryote</a>s have a single circular chain.</li>
</ol>
<p>Other differences between <a href="#BOOKMARK109">eukaryote</a>s and <a href="#BOOKMARK210">prokaryote</a>s ...</p>
<table><tbody><tr><th></th><th> Eukaryotes</th><th> Prokaryotes</th></tr><tr><td>Size</td><td> 10 to 100 micrometers (μm)</td><td> 0.1 to 5 micrometers (μm)</td></tr><tr><td>Complexity</td><td> More complex</td><td> More simple</td></tr><tr><td>Sub-compartments (organelles)</td><td> Yes</td><td> No</td></tr><tr><td>DNA layout</td><td> Multiple stands</td><td> Single circular strand</td></tr><tr><td>Single-cell organisms</td><td> Yes (e.g. amoeba)</td><td> Yes (e.g. bacteria and archaea)</td></tr><tr><td>Multi-cell organisms</td><td> Yes (e.g. animals and fungus)</td><td> No</td></tr></tbody></table><div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Archaea is an organism that looks like bacteria but they're totally different.</p>
</div>
<a name="HEADREF16"></a><h3>Features</h3>
<p>Different <a href="#BOOKMARK74">cell</a> species vary in features. The subsections below detail common <a href="#BOOKMARK74">cell</a> features (not exhaustive).</p>
<p>Some features are only present in certain <a href="#BOOKMARK74">cell</a> speicies (e.g. only some <a href="#BOOKMARK74">cell</a>s have a <a href="#BOOKMARK83">flagellum</a> tail) while other features are present in all <a href="#BOOKMARK74">cell</a>s but in different amounts (e.g. every <a href="#BOOKMARK74">cell</a> has <a href="#BOOKMARK76">cytosol</a> but larger <a href="#BOOKMARK74">cell</a>s have more <a href="#BOOKMARK76">cytosol</a>).</p>
<a name="HEADREF17"></a><h4>Cytoplasm</h4>
<p>The <a name="BOOKMARK75"></a><strong>cytoplasm</strong> (both <a href="#BOOKMARK109">eukaryotic</a> and <a href="#BOOKMARK210">prokaryotic</a>) is the insides/guts of a <a href="#BOOKMARK74">cell</a>. <a name="BOOKMARK76"></a><strong>Cytosol</strong> refers to just the fluid, while <a href="#BOOKMARK75">cytoplasm</a> refers to fluid as well as everything else inside the <a href="#BOOKMARK74">cell</a>.</p>
<p>Every <a href="#BOOKMARK74">cell</a> has <a href="#BOOKMARK75">cytoplasm</a>.</p>
<p><img src="/.datafile_9541606f126b39872505c6514907e068/data.svg" alt="Eukaryote with cell ribosomes highlighted" title="By No machine-readable author provided. Chb assumed (based on copyright claims). - No machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=688296"></p><a name="HEADREF18"></a><h4>Membrane</h4>
<p>The <a name="BOOKMARK77"></a><strong>plasma membrane</strong> (present in both <a href="#BOOKMARK109">eukaryotic</a> and <a href="#BOOKMARK210">prokaryotic</a> <a href="#BOOKMARK74">cell</a>s) is the thing encapsulating the <a href="#BOOKMARK75">cytoplasm</a>. It's what keeps the guys of the <a href="#BOOKMARK74">cell</a> inside and controls the movement of substances coming into / going out of the <a href="#BOOKMARK75">cytoplasm</a>.</p>
<p>Every <a href="#BOOKMARK74">cell</a> has a <a href="#BOOKMARK77">membrane</a> encapsulating its <a href="#BOOKMARK75">cytoplasm</a>. <a href="#BOOKMARK77">Membrane</a>s in general follow the <a href="#BOOKMARK88">fluid mosaic model</a>.</p>
<p>The term <a href="#BOOKMARK77">membrane</a> can refer to either the <a href="#BOOKMARK77">plasma membrane</a> or the <a href="#BOOKMARK77">membrane</a> of a <a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a>'s <a href="#BOOKMARK111">organelle</a>. How you should interpret it depends on the context in which its used.</p>
<p><img src="/.datafile_e8a840891af175db6acdfb48d221a7f5/data.svg" alt="Eukaryote with cell ribosomes highlighted" title="By No machine-readable author provided. Chb assumed (based on copyright claims). - No machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=688296"></p><p><img src="/.datafile_82fd0d4f0e5976b6860319c80b1a45ac/data.svg" alt="Prokaryotic with capsule highlighted" title="By Mariana Ruiz Villarreal LadyofHats - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=7356226"></p><p>Facts about <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK77">membrane</a>s:</p>
<ul>
<li><a href="#BOOKMARK77">Membrane</a>s have other <a href="#BOOKMARK18">molecule</a>s embedded in them for identification, moving stuff in/out, etc.. (see <a href="#BOOKMARK88">fluid mosaic model</a>).</li>
<li><a href="#BOOKMARK77">Membrane</a>s are <a href="#BOOKMARK349">semipermeable</a>. Uncharged substances such as carbon dioxide and oxygen move across fine, while charged <a href="#BOOKMARK26">ions</a> or large <a href="#BOOKMARK18">molecule</a>s require either
<ul>
<li><a href="#BOOKMARK77">membrane</a> <a href="#BOOKMARK55">protein</a>s such as a <a href="#BOOKMARK98">carrier protein</a> or a <a href="#BOOKMARK97">channel protein</a> (see <a href="#BOOKMARK88">fluid mosaic model</a>).</li>
<li><a href="#BOOKMARK77">membrane</a> wrapper (see <a href="#BOOKMARK116">golgi</a>).</li>
</ul>
</li>
</ul>
<a name="HEADREF19"></a><h4>Cell Wall</h4>
<p>The <a name="BOOKMARK78"></a><strong>cell wall</strong> (present in both <a href="#BOOKMARK109">eukaryotic</a> and <a href="#BOOKMARK210">prokaryotic</a> <a href="#BOOKMARK74">cell</a>s) is a stiff layer around the <a href="#BOOKMARK77">membrane</a> meant for protection. <em>Not all <a href="#BOOKMARK74">cell</a>s have a <a href="#BOOKMARK78">cell wall</a></em> -- for example, animal <a href="#BOOKMARK74">cell</a>s don't but plant <a href="#BOOKMARK74">cell</a>s do. Technically, the <a href="#BOOKMARK78">cell wall</a> (if it exists) isn't considered to be part of the <a href="#BOOKMARK74">cell</a>. The <a href="#BOOKMARK77">membrane</a> and everything in it is.</p>
<p>The material states that <a href="#BOOKMARK78">cell wall</a>s...</p>
<ol>
<li>provide an extra layer of protection.</li>
<li>help maintain shape.</li>
<li>help prevent dehydration.</li>
</ol>
<p>Almost all <a href="#BOOKMARK210">prokaryote</a>s have <a href="#BOOKMARK78">cell wall</a>s. Only some <a href="#BOOKMARK109">eukaryote</a>s have <a href="#BOOKMARK78">cell wall</a>s (e.g. fungi and plants). The material says that <a href="#BOOKMARK78">cell wall</a>s for most bacteria are made up of a <a href="#BOOKMARK18">molecule</a> called <a name="BOOKMARK79"></a><strong>peptidoglycan</strong>, but it can be different for other <a href="#BOOKMARK74">cell</a>s. For example, <a href="https://www.quora.com/Do-Eukaryotes-have-cell-walls">this link</a> says that plant <a href="#BOOKMARK74">cell</a>s have <a href="#BOOKMARK78">cell wall</a>s made up of <a href="#BOOKMARK74">cell</a>ulose.</p>
<p><img src="/.datafile_691032114de14246a7d9efa1b7050c6c/data.svg" alt="Prokaryotic with capsule highlighted" title="By Mariana Ruiz Villarreal LadyofHats - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=7356226"></p><a name="HEADREF20"></a><h4>Capsule</h4>
<p>The <a name="BOOKMARK80"></a><strong>Capsule</strong> (present in <a href="#BOOKMARK210">prokaryotic</a> <a href="#BOOKMARK74">cell</a> only) is the outermost layer of some types of <a href="#BOOKMARK74">cell</a>s (typically bacteria <a href="#BOOKMARK74">cell</a>s). <a href="#BOOKMARK80">Capsule</a>s are made up of <a href="#BOOKMARK51">carbohydrate</a>s and there mainly to help the <a href="#BOOKMARK74">cell</a> stick itself to the environment.</p>
<p><img src="/.datafile_a0e053364e06b2611796278d25649b80/data.svg" alt="Prokaryotic with capsule highlighted" title="By Mariana Ruiz Villarreal LadyofHats - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=7356226"></p><div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Although <a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a>s don't have <a href="#BOOKMARK80">capsule</a>s, they do have <a href="#BOOKMARK51">carbohydrate</a>s on their outside. Those <a href="#BOOKMARK51">carbohydrate</a>s aren't organized as a <a href="#BOOKMARK80">capsule</a> though: https://www.quora.com/Do-some-<a href="#BOOKMARK109">eukaryotic</a>-<a href="#BOOKMARK74">cell</a>s-have-<a href="#BOOKMARK80">capsule</a>s-or-is-it-just-<a href="#BOOKMARK210">prokaryote</a>s-Are-there-exceptions-of-<a href="#BOOKMARK109">eukaryote</a>s-having-<a href="#BOOKMARK80">capsule</a>s. Is this talking about the same <a href="#BOOKMARK51">carbohydrate</a>s that are embedded in the <a href="#BOOKMARK77">membrane</a> (<a href="#BOOKMARK93">glycolipid</a>s / <a href="#BOOKMARK94">glycoprotein</a>s).</p>
</div>
<a name="HEADREF21"></a><h4>Ribosome</h4>
<p><a name="BOOKMARK81"></a><strong>Ribosome</strong> (present in both <a href="#BOOKMARK109">eukaryotic</a> and <a href="#BOOKMARK210">prokaryotic</a> <a href="#BOOKMARK74">cell</a>s) are tiny <a href="#BOOKMARK16">mole</a>cular machines inside the <a href="#BOOKMARK75">cytoplasm</a> that take in <a href="#BOOKMARK248">mRNA</a> <a href="#BOOKMARK18">molecule</a>s (portions of <a href="#BOOKMARK241">DNA</a> that have been written out) and produce <a href="#BOOKMARK55">protein</a>s. <a href="#BOOKMARK81">Ribosome</a>s themselves are structures made of <a href="#BOOKMARK55">protein</a>s and <a href="#BOOKMARK247">RNA</a>.</p>
<p><a href="#BOOKMARK81">Ribosome</a>s can either be floating around in the <a href="#BOOKMARK75">cytoplasm</a> (called <a name="BOOKMARK82"></a><strong>free ribosome</strong>) or be embedded in the <a href="#BOOKMARK77">membrane</a> of <a href="#BOOKMARK113">endoplasmic reticulum</a>.</p>
<p><img src="/.datafile_e57f11cb45a1025a8fe77f5678b628dd/data.svg" alt="Eukaryote with cell ribosomes highlighted" title="By No machine-readable author provided. Chb assumed (based on copyright claims). - No machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=688296"></p><a name="HEADREF22"></a><h4>Appendage</h4>
<p>Some <a href="#BOOKMARK74">cell</a>s have appendages that help them move (or stay put). There are different types of appendages...</p>
<ul>
<li><a name="BOOKMARK83"></a><strong>Flagellum</strong> are tails that extend from the <a href="#BOOKMARK74">cell</a> (e.g. tail on a <a href="#BOOKMARK177">sperm</a> <a href="#BOOKMARK74">cell</a>). There can be more than 1 <a href="#BOOKMARK83">flagellum</a>.</li>
<li><a name="BOOKMARK84"></a><strong>Cilia</strong> are much smaller hair-like appendages used to help move the <a href="#BOOKMARK74">cell</a> itself or things in the vicinity of the <a href="#BOOKMARK74">cell</a>.</li>
<li><a name="BOOKMARK85"></a><strong>Fimbriae</strong> are much smaller hair-like appendages used to help attach to host <a href="#BOOKMARK74">cell</a>s and surfaces (e.g. bacteria <a href="#BOOKMARK74">cell</a>s).</li>
<li><a name="BOOKMARK86"></a><strong>Pili</strong> are much smaller hair-like appendages used to help transfer <a href="#BOOKMARK241">DNA</a> between <a href="#BOOKMARK74">cell</a>s and/or to help move (e.g. bacteria <a href="#BOOKMARK74">cell</a>s).</li>
<li><a name="BOOKMARK87"></a><strong>Pseudopodia</strong> are much larger leg-like appendages used to crawl (e.g. amoeba <a href="#BOOKMARK74">cell</a>s).</li>
</ul>
<p><img src="/.datafile_53a566249623143a5d791c2ef0ae68d2/data.svg" alt="Eukaryote with cell flagellum highlighted" title="By No machine-readable author provided. Chb assumed (based on copyright claims). - No machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=688296"></p><a name="HEADREF23"></a><h3>Fluid Mosaic Model</h3>
<p>The <a name="BOOKMARK88"></a><strong>fluid mosaic model</strong> is the accepted model for how <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK77">membrane</a>s work. The model says that a <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK77">membrane</a> is composed of a <a href="#BOOKMARK90">phospholipid bilayer</a> with <a href="#BOOKMARK55">protein</a>s, <a href="#BOOKMARK58">lipid</a>s, and <a href="#BOOKMARK51">carbohydrate</a>s floating around on either side or embedded in between.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The description above is the rational for the name '<a href="#BOOKMARK88">fluid mosaic model</a>'. It's fluid and there's a mosaic of different things embedded or attached to it.</p>
</div>
<p><img src="800px-Cell_membrane_detailed_diagram_en.svg.png" alt="By LadyofHats Mariana Ruiz - Own work. Image renamed from File:Cell membrane detailed diagram.svg, Public Domain, https://commons.wikimedia.org/w/index.php?curid=6027169"></p>
<p>A <a name="BOOKMARK89"></a><strong>phospholipid</strong> is a <a href="#BOOKMARK73">amphipathic</a> <a href="#BOOKMARK58">lipid</a> <a href="#BOOKMARK18">molecule</a> that involves a phosphate group. The...</p>
<ul>
<li>bulbus phosphate group at the top is <a href="#BOOKMARK71">hydrophilic</a> -- it has a charge and as such is attracted to <a href="#BOOKMARK59">water</a> (either the <a href="#BOOKMARK75">cytoplasm</a> or the fluid outside the <a href="#BOOKMARK74">cell</a>).</li>
<li>long fatty <a href="#BOOKMARK33">acid</a> tails are <a href="#BOOKMARK72">hydrophobic</a> -- they have no obvious charge and as such aren't attracted to <a href="#BOOKMARK59">water</a>.</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>For a refresher on how <a href="#BOOKMARK72">hydrophobic</a> / <a href="#BOOKMARK71">hydrophilic</a> <a href="#BOOKMARK18">molecule</a>s work, see the section on <a href="#BOOKMARK59">Water</a>. Specifically: <a href="#BOOKMARK69">adhesion</a> / <a href="#BOOKMARK62">weakly negative</a> / <a href="#BOOKMARK63">weakly positive</a>.</p>
</div>
<p><img src="0301_Phospholipid_Structure.jpg" alt="By OpenStax - https://cnx.org/contents/FPtK1zmh@8.25:fEI3C8Ot@10/Preface, CC BY 4.0, https://commons.wikimedia.org/w/index.php?curid=30131167"></p>
<p>As such, <a href="#BOOKMARK89">phospholipid</a>s have a natural tendency to form as a <a name="BOOKMARK90"></a><strong>phospholipid bilayer</strong> (2 layers attached together, called a <a name="BOOKMARK91"></a><strong>liposome</strong>) or a ball (called a <a name="BOOKMARK92"></a><strong>micelle</strong>). The <a href="#BOOKMARK71">hydrophilic</a> heads are going to point towards the <a href="#BOOKMARK59">water</a> causing the <a href="#BOOKMARK72">hydrophobic</a> tails to point at each other.</p>
<p><img src="Phospholipids_aqueous_solution_structures.svg.png" alt="By Mariana Ruiz Villarreal ,LadyofHats - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=3032610"></p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>If the <a href="#BOOKMARK89">phospholipid</a>s have small tails, they may form a <a href="#BOOKMARK92">micelle</a> (a small, single-layered sphere), while if they have bulkier tails, they may form a <a href="#BOOKMARK91">liposome</a>.</p>
</div>
<p>How fluid a <a href="#BOOKMARK90">phospholipid bilayer</a> is depends on the types of <a href="#BOOKMARK89">phospholipid</a> <a href="#BOOKMARK18">molecule</a>s that make it up and the temperature. <a href="#BOOKMARK89">Phospholipid</a> <a href="#BOOKMARK18">molecule</a>s have 2 fatty <a href="#BOOKMARK33">acid</a> tails. The fatty <a href="#BOOKMARK33">acid</a> tails can be either...</p>
<ul>
<li>both saturated (straight tails)</li>
<li>one saturated (straight tail) and one unsaturated (bent tail).</li>
</ul>
<p>At cooler temperatures, <a href="#BOOKMARK89">phospholipid</a>s that have 2 saturated fatty <a href="#BOOKMARK33">acid</a> tails (straight tails) tend to get more rigid / <a href="#BOOKMARK342">dense</a> because they can more easily pack together. <a href="#BOOKMARK89">Phospholipid</a>s with unsaturated fatty <a href="#BOOKMARK33">acid</a> tails (bent tails) don't end up getting as rigid / <a href="#BOOKMARK342">dense</a>, allowing the <a href="#BOOKMARK77">membrane</a> to stay fluid at lower temperatures. <a href="#BOOKMARK100">Cholesterol</a> embedded in the <a href="#BOOKMARK90">phospholipid bilayer</a> also helps it stay more fluid at lower temperatures.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p><a href="#BOOKMARK90">Phospholipid bilayer</a>s have the consistency of oil-based salad dressing. It may seem weak but it's strong enough to act as a separator between the environment inside and the environment outside. <a href="#BOOKMARK59">Water</a> from one-side can move to the other but does so very rarely -- a single <a href="#BOOKMARK18">molecule</a> may sneak through the layer every now and then. Aquaporins are <a href="#BOOKMARK55">protein</a>s embedded in the <a href="#BOOKMARK90">phospholipid bilayer</a> that allow <a href="#BOOKMARK59">water</a> to rapidly pass (when needed).</p>
</div>
<p>Examples of <a href="#BOOKMARK18">molecule</a>s that can be embedded in or attached to the <a href="#BOOKMARK90">phospholipid bilayer</a> include...</p>
<ul>
<li><a name="BOOKMARK93"></a><strong>Glycolipid</strong>s / <a name="BOOKMARK94"></a><strong>Glycoprotein</strong>s - These are <a href="#BOOKMARK58">lipid</a>s and <a href="#BOOKMARK55">protein</a>s with <a href="#BOOKMARK51">carbohydrate</a>s attached to them and are typically found pointing outside of the <a href="#BOOKMARK74">cell</a>. These <a href="#BOOKMARK51">carbohydrate</a>s are how a <a href="#BOOKMARK74">cell</a> identifies if a neighbouring <a href="#BOOKMARK74">cell</a> is foreign or not.</li>
<li><a name="BOOKMARK95"></a><strong>Integral protein</strong>s - <a href="#BOOKMARK55">Protein</a>s that are partially <a href="#BOOKMARK72">hydrophobic</a>. Either one end of the <a href="#BOOKMARK55">protein</a> is anchored inside the <a href="#BOOKMARK77">membrane</a> or the <a href="#BOOKMARK55">protein</a> is fully going through the <a href="#BOOKMARK77">membrane</a>.
<ul>
<li><a name="BOOKMARK96"></a><strong>Transmembrane protein</strong>s - <a href="#BOOKMARK55">Protein</a>s that have one end inside the <a href="#BOOKMARK75">cytoplasm</a> and the other outside of the <a href="#BOOKMARK74">cell</a>. Typical use case for these types of <a href="#BOOKMARK55">protein</a>s is moving substances from inside the <a href="#BOOKMARK74">cell</a> (<a href="#BOOKMARK75">cytoplasm</a>) to outside the <a href="#BOOKMARK74">cell</a> and vice-versa: <a name="BOOKMARK97"></a><strong>channel protein</strong>s and <a name="BOOKMARK98"></a><strong>carrier protein</strong>s.</li>
</ul>
</li>
<li><a name="BOOKMARK99"></a><strong>Peripheral protein</strong>s - <a href="#BOOKMARK55">Protein</a>s that are found on the surface of the <a href="#BOOKMARK55">protein</a> (either on the <a href="#BOOKMARK75">cytoplasm</a> side or the outside side). Unlike <a href="#BOOKMARK95">integral protein</a>s, they don't have any portion going inside the <a href="#BOOKMARK77">membrane</a>. As such, they more freely move around compared to <a href="#BOOKMARK95">integral protein</a>s.</li>
<li><a name="BOOKMARK100"></a><strong>Cholesterol</strong> - Sits inside of the <a href="#BOOKMARK77">membrane</a> to help ensure that the <a href="#BOOKMARK77">membrane</a> doesn't become too stiff.</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>See first diagram in this section for an example of each of the <a href="#BOOKMARK18">molecule</a>s listed above.</p>
</div>
<p>The term <a name="BOOKMARK101"></a><strong>facilitated diffusion</strong> refers to the movement of <a href="#BOOKMARK18">molecule</a>s across the <a href="#BOOKMARK77">membrane</a> via <a href="#BOOKMARK55">protein</a>s embedded in the <a href="#BOOKMARK77">membrane</a> (e.g. <a href="#BOOKMARK97">channel protein</a>s and/or <a href="#BOOKMARK98">carrier protein</a>s). These <a href="#BOOKMARK18">molecule</a>s wouldn't be able to cross the <a href="#BOOKMARK77">membrane</a> by themselves. For example, the <a href="#BOOKMARK328">sodium potassium pump</a> (<a href="#BOOKMARK98">carrier protein</a>) helps sodium and potassium <a href="#BOOKMARK26">ions</a> move across the <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK77">membrane</a> by opening/closing its gates.</p>
<a name="HEADREF24"></a><h3>History of Modern Cell Theory</h3>
<p>The first record of a <a href="#BOOKMARK74">cell</a> was in 1665 when <a name="BOOKMARK102"></a><strong>Robert Hooke</strong> published a book called The <a href="#BOOKMARK331">Micrograph</a>ia. The book contains drawings of observations he made while looking at various dead organisms through a rudimentary <a href="#BOOKMARK330">microscope</a>.</p>
<p>A few years later, a Dutch lenscrafter by the name of <a name="BOOKMARK103"></a><strong>Antonie Van Leeuwenhoek</strong> decided to use his expertise to craft a better <a href="#BOOKMARK330">microscope</a> to better observe <a href="#BOOKMARK38">living</a> <a href="#BOOKMARK74">cell</a>s / organisms. For example, he was able to observe <a href="#BOOKMARK177">sperm</a> and <a href="#BOOKMARK110">Protist</a>s (uni<a href="#BOOKMARK74">cell</a>ular organisms while he dubbed <a name="BOOKMARK104"></a><strong>animalcule</strong>s).</p>
<p>In the 1830s, <a name="BOOKMARK105"></a><strong>Matthias Schleiden</strong> and <a name="BOOKMARK106"></a><strong>Theodore Schwann</strong> began laying the groundwork for modern <a href="#BOOKMARK74">cell</a> theory. They came up with the idea that...</p>
<ol>
<li>all <a href="#BOOKMARK38">life</a> is composed of one or more <a href="#BOOKMARK74">cell</a>s.</li>
<li>a <a href="#BOOKMARK74">cell</a> is the basic unit of <a href="#BOOKMARK38">life</a>.</li>
</ol>
<p>They also suspected that <a href="#BOOKMARK74">cell</a>s come from other <a href="#BOOKMARK74">cell</a>s, but didn't know for sure if that was the only way they were produced. It was <a name="BOOKMARK107"></a><strong>Robert Remak</strong> that in the mid-1800s established that...</p>
<ol start="3">
<li>all <a href="#BOOKMARK74">cell</a>s come from other <a href="#BOOKMARK74">cell</a>s.</li>
</ol>
<p><a name="BOOKMARK108"></a><strong></strong></p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The credit for this sometimes goes to <a href="#BOOKMARK108">Rudolph Virchoi</a> but it's been established that he was a plagiarist.</p>
</div>
<p>It's still an open question as to how the first / initial <a href="#BOOKMARK74">cell</a> came to be. The current working theory is that, 3.5 billion years ago, <a href="#BOOKMARK89">phospholipid</a>s (the <a href="#BOOKMARK18">molecule</a>s that form the <a href="#BOOKMARK77">membrane</a> of a <a href="#BOOKMARK74">cell</a>) naturally form bilayers and connect in a circle. A <a href="#BOOKMARK77">membrane</a> may have naturally encapsulated a set of arbitrary self-replicating <a href="#BOOKMARK18">molecule</a>s (e.g. <a href="#BOOKMARK55">protein</a> or <a href="#BOOKMARK247">RNA</a>) and that's how the first <a href="#BOOKMARK74">cell</a> began growing and splitting off.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>There are an estimated 37 trillion <a href="#BOOKMARK74">cell</a>s in the human body.</p>
</div>
<a name="HEADREF25"></a><h2>Eukaryotic Cell</h2>
<p><img src="/.datafile_b3c197175c0982d0be33d8ab2e009467/data.svg" alt="ANIMAL CELL" title="By Mariana Ruiz Villarreal, LadyofHats - Own work (Source: Typical prokaryotic cell, Chapter 4: Mutagenicity of alkyl N-acetoxybenzohydroxamates, Concept 1: Common Features of All Cells, Cells - Structure and Function), Public Domain, https://commons.wikimedia.org/w/index.php?curid=3648821"></p><p><img src="/.datafile_fc75023772983e80e042364c989c9e88/data.svg" alt="Eukaryotic plant cell" title="By LadyofHats - Self-made using Adobe Illustrator. (The original edited was also made by me, LadyofHats), Public Domain, https://commons.wikimedia.org/w/index.php?curid=844682"></p><p><a name="BOOKMARK109"></a><strong>Eukaryotic</strong> <a href="#BOOKMARK74">cell</a>s are typically larger and have <a href="#BOOKMARK77">membrane</a>-bound sub-compartments, called <a href="#BOOKMARK111">organelle</a>, that hold in the guts of different regions of the <a href="#BOOKMARK74">cell</a>. For example, their <a href="#BOOKMARK241">DNA</a> is encapsulated in a <a href="#BOOKMARK111">organelle</a> called the <a href="#BOOKMARK112">nucleus</a>.</p>
<p><a href="#BOOKMARK109">Eukaryote</a>s have their <a href="#BOOKMARK241">DNA</a> broken up into multiple strands called <a href="#BOOKMARK138">chromosome</a>s. They can either be single-<a href="#BOOKMARK74">cell</a>ular organisms (e.g. amoeba) or multi-<a href="#BOOKMARK74">cell</a>ular organisms (e.g. human). Single-<a href="#BOOKMARK74">cell</a>ular organism that are <a href="#BOOKMARK109">eukaryotic</a> are called <a name="BOOKMARK110"></a><strong>protist</strong>s.</p>
<p>The following are descriptions for some of the <a href="#BOOKMARK111">organelle</a>s shown in the diagram above.</p>
<ul>
<li><span style="color:#ff00ffff"><strong>Nucleus</strong></span> - See <a href="#BOOKMARK112">nucleus</a> section.</li>
<li><span style="color:#0000ffff"><strong>Endoplasmic Reticulum</strong></span> - See <a href="#BOOKMARK113">endoplasmic reticulum</a> section.</li>
<li><span style="color:#007f7fff"><strong>Golgi</strong></span> - See <a href="#BOOKMARK116">golgi</a> section.</li>
<li><span style="color:#00ff00ff"><strong>Mitochondria</strong></span> - See <a href="#BOOKMARK117">mitochondria</a> section.</li>
<li><span style="color:#7f7f00ff"><strong>Lysosome</strong></span> (mostly animal) - See <a href="#BOOKMARK121">lysosome</a> section.</li>
<li><span style="color:#0000ffff"><strong>Peroxisome</strong></span> - See <a href="#BOOKMARK122">peroxisome</a> section.</li>
<li><span style="color:#ff0000ff"><strong>Chloroplast</strong></span> (plant / algae) - See <a href="#BOOKMARK123">chloroplast</a> section.</li>
<li><span style="color:#800080ff"><strong>Vacuole</strong></span> (mostly plant / algae) - See <a href="#BOOKMARK131">vacuole</a> section.</li>
</ul>
<a name="HEADREF26"></a><h3>Organelle</h3>
<p><a href="#BOOKMARK109">Eukaryotic</a> <a href="#BOOKMARK74">cell</a>s have <a href="#BOOKMARK77">membrane</a>-bound sub-compartments, called <a name="BOOKMARK111"></a><strong>organelle</strong>s, that house different functional regions of the <a href="#BOOKMARK74">cell</a>. The following subsections detail common <a href="#BOOKMARK111">organelle</a>s.</p>
<a name="HEADREF27"></a><h4>Nucleus</h4>
<p><a name="BOOKMARK112"></a><strong>Nucleus</strong> is an <a href="#BOOKMARK111">organelle</a> that contains <a href="#BOOKMARK241">DNA</a> (genetic information required for the functioning and replication). Both <a href="#BOOKMARK210">prokaryotic</a> and <a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a>s have <a href="#BOOKMARK241">DNA</a>, but only <a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a>s have a <a href="#BOOKMARK112">nucleus</a>. In <a href="#BOOKMARK210">prokaryotic</a> <a href="#BOOKMARK74">cell</a>s, the <a href="#BOOKMARK241">DNA</a> flows around freely instead of being encapsulated in a <a href="#BOOKMARK112">nucleus</a>.</p>
<p><img src="/.datafile_a6a5bcaa2d640eae684a9c0d1ee4ce4d/data.svg" alt=" comprehensive diagram of a human cell nucleus." title="By Mariana Ruiz LadyofHats - I did it myself with adobe ilustrator using the information found here [1], [2] ,[3], [4] and [5], Public Domain, https://commons.wikimedia.org/w/index.php?curid=736389"></p><p>Most <a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a>s contain a single <a href="#BOOKMARK112">nucleus</a>, but some contain can have 0 and others can have more than one. An example of 0 is blood <a href="#BOOKMARK74">cell</a>s -- mature blood <a href="#BOOKMARK74">cell</a>s don't have any <a href="#BOOKMARK241">DNA</a>, therefore no <a href="#BOOKMARK112">nucleus</a>. An example of more than 1 is the organism Oxytricha trifillax -- it contains 2 <a href="#BOOKMARK112">nuclei</a>, each containing different <a href="#BOOKMARK241">DNA</a> (its <a href="#BOOKMARK241">DNA</a> is fragmented across 2 <a href="#BOOKMARK112">nuclei</a>).</p>
<a name="HEADREF28"></a><h4>Endoplasmic Reticulum</h4>
<p><a name="BOOKMARK113"></a><strong>Endoplasmic Reticulum</strong> is layered <a href="#BOOKMARK77">membrane</a> (<a href="#BOOKMARK111">organelle</a>?) that surrounds the <a href="#BOOKMARK112">nucleus</a> and is directly connected to pores on the <a href="#BOOKMARK112">nucleus</a>. Large portions of the <a href="#BOOKMARK113">endoplasmic reticulum</a>'s <a href="#BOOKMARK77">membrane</a> have <a href="#BOOKMARK81">ribosome</a>s attached. The parts that have <a href="#BOOKMARK81">ribosome</a>s attached are called <a name="BOOKMARK114"></a><strong>rough endoplasmic reticulum</strong> while the parts that don't are called <a name="BOOKMARK115"></a><strong>smooth endoplasmic reticulum</strong>.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>It's called <a href="#BOOKMARK114">rough endoplasmic reticulum</a> because the <a href="#BOOKMARK81">ribosome</a>s make the surface look rough.</p>
</div>
<p><img src="/.datafile_b270c13d0bb642adf2ffa7a1a1e77982/data.svg" alt="a) The ER is a winding network of thin membranous sacs found in close association with the cell nucleus. The smooth and rough endoplasmic reticula are very different in appearance and function (source: mouse tissue). (b) Rough ER is studded with numerous ribosomes, which are sites of protein synthesis (source: mouse tissue). EM × 110,000. (c) Smooth ER synthesizes phospholipids, steroid hormones, regulates the concentration of cellular Ca++, metabolizes some carbohydrates, and breaks down certain toxins (source: mouse tissue). EM × 110,510. (Micrographs provided by the Regents of University of Michigan Medical School © 2012)" title="By OpenStax - https://cnx.org/contents/FPtK1zmh@8.25:fEI3C8Ot@10/Preface, CC BY 4.0, https://commons.wikimedia.org/w/index.php?curid=30131197"></p><p>Recall that <a href="#BOOKMARK81">ribosome</a>s are what translate <a href="#BOOKMARK248">mRNA</a> to <a href="#BOOKMARK55">protein</a>s. Since the <a href="#BOOKMARK113">endoplasmic reticulum</a> is directly connected to the <a href="#BOOKMARK112">nucleus</a> (via pores on the <a href="#BOOKMARK112">nucleus</a>), it provides a fairly straight-forward path for <a href="#BOOKMARK55">protein</a> generation: <a href="#BOOKMARK248">mRNA</a> produced in the <a href="#BOOKMARK112">nucleus</a>...</p>
<ol>
<li>travels to the <a href="#BOOKMARK113">endoplasmic reticulum</a> via the connected pores,</li>
<li>then travels to the <a href="#BOOKMARK77">membrane</a> of the <a href="#BOOKMARK113">endoplasmic reticulum</a> where it ends up hitting <a href="#BOOKMARK81">ribosome</a>s embedded in the (thereby producing <a href="#BOOKMARK55">protein</a>s).</li>
</ol>
<a name="HEADREF29"></a><h4>Golgi</h4>
<p><a name="BOOKMARK116"></a><strong>Golgi</strong> are layered <a href="#BOOKMARK77">membrane</a> (<a href="#BOOKMARK111">organelle</a>?) that look similar to <a href="#BOOKMARK114">rough endoplasmic reticulum</a> but aren't attached to the <a href="#BOOKMARK112">nucleus</a>. <a href="#BOOKMARK116">Golgi</a> package <a href="#BOOKMARK18">molecule</a>s (e.g. <a href="#BOOKMARK55">protein</a>s) for travel to either another part of the <a href="#BOOKMARK74">cell</a> or outside of the <a href="#BOOKMARK74">cell</a>. They do this by pinching off parts of their <a href="#BOOKMARK77">membrane</a> to wrap around the <a href="#BOOKMARK18">molecule</a>.</p>
<p>They're also responsible for building <a href="#BOOKMARK121">lysosome</a>s (<a href="#BOOKMARK74">cell</a> digestion machines).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The terms <a href="#BOOKMARK116">golgi</a>, <a href="#BOOKMARK116">golgi apparatus</a>, <a href="#BOOKMARK116">golgi complex</a>, and <a href="#BOOKMARK116">golgi body</a> all refer to the same thing.</p>
</div>
<p><img src="/.datafile_162098445c030efc6a8aba4af4e331a3/data.svg" alt="Golgi apparatus" title="By Kelvinsong - Own work, CC BY 3.0, https://commons.wikimedia.org/w/index.php?curid=23090802"></p><a name="HEADREF30"></a><h4>Mitochondria</h4>
<p><a name="BOOKMARK117"></a><strong>Mitochondria</strong> are <a href="#BOOKMARK111">organelle</a>s responsible for <a href="#BOOKMARK317">cellular respiration</a>: the process of producing <a href="#BOOKMARK300">Adenosine Triphosphate</a> (<a href="#BOOKMARK299">ATP</a>) from <a href="#BOOKMARK18">molecule</a>s such as <a href="#BOOKMARK51">sugar</a>s. <a href="#BOOKMARK299">ATP</a> is a <a href="#BOOKMARK0">chemical</a> that provides energy to drive various <a href="#BOOKMARK37">biological</a> processes (e.g. muscle contractions). As such, <a href="#BOOKMARK117">mitochondria</a> are often referred to as "the power house of the <a href="#BOOKMARK74">cell</a>."</p>
<p><img src="/.datafile_cd73c8df606552a6d66a124d3acfca95/data.svg" alt="Mitochondria" title="By Mariana Ruiz Villarreal LadyofHats - the diagram i made myself using adobe illustrator. as a source for the information i used the diagrams found here:[1], [2], [3], [4], [5], [6] and [7]., Public Domain, https://commons.wikimedia.org/w/index.php?curid=8152599"></p><div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p><a href="#BOOKMARK117">Mitochondria</a> exist in both animal and plant <a href="#BOOKMARK74">cell</a>s.</p>
</div>
<p>The major parts of <a href="#BOOKMARK123">chloroplast</a> are...</p>
<ul>
<li><span style="color:#ff00ffff"><a name="BOOKMARK118"></a><strong>mitochondrial envelope</strong></span> - 2 <a href="#BOOKMARK77">membrane</a> layers that have a gap between them. Holds the guts in.</li>
<li><span style="color:#008080ff"><a name="BOOKMARK119"></a><strong>matrix</strong></span> - Internal fluid of <a href="#BOOKMARK117">mitochondria</a>. Contains the guts (<a href="#BOOKMARK81">ribosome</a>s, <a href="#BOOKMARK117">mitochondria</a>l <a href="#BOOKMARK241">DNA</a>, and more).</li>
<li><span style="color:#808000ff"><a name="BOOKMARK120"></a><strong>cristae</strong></span> - Long caverns that encapsulate and connect different parts of the <a href="#BOOKMARK119">matrix</a>.</li>
</ul>
<p><a href="#BOOKMARK117">Mitochondria</a> have their own independent <a href="#BOOKMARK241">DNA</a> (different from the <a href="#BOOKMARK241">DNA</a> in the <a href="#BOOKMARK112">nucleus</a>). It's speculated that at some point in the past they may have been independent single-<a href="#BOOKMARK74">cell</a> organisms that formed a symbiotic relationship with a larger <a href="#BOOKMARK74">cell</a> by living in it, eventually becoming part of the <a href="#BOOKMARK74">cell</a> (<a href="#BOOKMARK345">endosymbiosis</a>).</p>
<p>Unlike how normal offspring <a href="#BOOKMARK241">DNA</a> gets produced by mixing <a href="#BOOKMARK241">DNA</a> from both parents, <a href="#BOOKMARK117">mitochondria</a>l <a href="#BOOKMARK241">DNA</a> comes entirely from the mother's side.</p>
<a name="HEADREF31"></a><h4>Lysosome</h4>
<p><a name="BOOKMARK121"></a><strong>Lysosome</strong>s are <a href="#BOOKMARK111">organelle</a>s (animal <a href="#BOOKMARK74">cell</a>s only) that help break down waste products / foreign substances by containing various <a href="#BOOKMARK212">enzyme</a>s and maintaining an <a href="#BOOKMARK33">acidic</a> <a href="#BOOKMARK32">pH</a>. <a href="#BOOKMARK121">Lysosome</a>s are more often found in animals <a href="#BOOKMARK74">cell</a>s than plant and algae <a href="#BOOKMARK74">cell</a>s.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>According to the material, the evidence that they've been found in plant <a href="#BOOKMARK74">cell</a>s is recent.</p>
</div>
<p><img src="/.datafile_5a9c03d8b3f1bf15156e514b1f45ec4c/data.svg" alt="Structure of Lysosome" title="By lumoreno - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=17380475"></p><a name="HEADREF32"></a><h4>Peroxisome</h4>
<p><a name="BOOKMARK122"></a><strong>Peroxisome</strong>s are <a href="#BOOKMARK111">organelle</a>s that are similar to <a href="#BOOKMARK121">Lysosome</a>s -- both are small <a href="#BOOKMARK111">organelle</a>s that break down unwanted substances. The difference is that <a href="#BOOKMARK122">peroxisome</a>s have different types of <a href="#BOOKMARK212">enzyme</a>s that require oxygen (oxidative <a href="#BOOKMARK212">enzyme</a>s).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The material says that <a href="#BOOKMARK122">peroxisome</a>s make hydrogen peroxide: Similarly, structures called <a href="#BOOKMARK122">peroxisome</a>s carry out <a href="#BOOKMARK0">chemical</a> reactions called oxidation reactions and produce hydrogen peroxide, both of which would damage the <a href="#BOOKMARK74">cell</a> if they weren’t safely stored away in their own “room.”</p>
</div>
<p><img src="/.datafile_a3e02114612860a0511d60117a2dac9c/data.svg" alt="Basic structure of a peroxisome, showing the crystallized enzyme core as found in rat liver cells." title="By Qef - Own work by uploader, based on the arrangement of a bitmap equivalent by Anthony Atkielski (Agateller), Public Domain, https://commons.wikimedia.org/w/index.php?curid=7072127"></p><a name="HEADREF33"></a><h4>Chloroplast</h4>
<p><a name="BOOKMARK123"></a><strong>Chloroplast</strong>s are <a href="#BOOKMARK111">organelle</a> (only plant and algae <a href="#BOOKMARK74">cell</a>s) responsible for <a href="#BOOKMARK313">photosynthesis</a>. <a href="#BOOKMARK313">Photosynthesis</a> is the process of taking in light and using it to build <a href="#BOOKMARK51">sugar</a>s from <a href="#BOOKMARK59">water</a> and carbon dioxide. Those <a href="#BOOKMARK51">sugar</a>s are then used by the <a href="#BOOKMARK117">mitochondria</a> to produce energy in a process called <a href="#BOOKMARK317">cellular respiration</a>.</p>
<p><img src="/.datafile_36b0217c4861c2333f781d947ee33287/data.svg" alt="Ultrastructure of a chloroplast." title="By Kelvinsong - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=26147364"></p><p>The major parts of <a href="#BOOKMARK123">chloroplast</a> are...</p>
<ul>
<li><span style="color:#ff00ffff"><a name="BOOKMARK124"></a><strong>chloroplast envelope</strong></span> - 2 <a href="#BOOKMARK77">membrane</a> layers that have a gap between them. Holds the guts in.</li>
<li><span style="color:#008080ff"><a name="BOOKMARK125"></a><strong>stroma</strong></span> - Internal fluid of <a href="#BOOKMARK123">chloroplast</a> (<a href="#BOOKMARK123">chloroplast</a>'s version of <a href="#BOOKMARK76">cytosol</a>).</li>
<li><span style="color:#808000ff"><a name="BOOKMARK126"></a><strong>thylakoid</strong></span> - <a href="#BOOKMARK77">Membrane</a>-bound discs. The <a href="#BOOKMARK77">membrane</a> contains light-harvesting substances while the space inside (also called <a name="BOOKMARK127"></a><strong>lumen</strong>) is hollow.</li>
<li><span style="color:#ff0000ff"><a name="BOOKMARK128"></a><strong>grana</strong></span> - Stack of <a href="#BOOKMARK126">thylakoid</a>s. These stacks are interconnected. The singular of <a href="#BOOKMARK128">grana</a> is <a href="#BOOKMARK128">granum</a>.</li>
</ul>
<p><a name="BOOKMARK129"></a><strong>Chlorophyll</strong> is a pigment / compound found in <a href="#BOOKMARK123">chloroplast</a> that absorbs light and uses it to produce <a href="#BOOKMARK51">carbohydrate</a>s. It's found in the <a href="#BOOKMARK126">thylakoid</a> <a href="#BOOKMARK77">membrane</a> as well as the <a href="#BOOKMARK125">stroma</a>, and it only absorbs red and blue light (while reflecting green).</p>
<p>Like <a href="#BOOKMARK117">mitochondria</a>, <a href="#BOOKMARK123">chloroplast</a> have their own independent <a href="#BOOKMARK241">DNA</a> (different from the <a href="#BOOKMARK241">DNA</a> in the <a href="#BOOKMARK112">nucleus</a>). It's speculated that at some point in the past they may have been independent single-<a href="#BOOKMARK74">cell</a> organisms that formed a symbiotic relationship with a larger <a href="#BOOKMARK74">cell</a> by living in it, eventually becoming part of the <a href="#BOOKMARK74">cell</a> (<a href="#BOOKMARK345">endosymbiosis</a>). A descendant of that organism may be <a name="BOOKMARK130"></a><strong>cyanobacterium</strong>, which has a similar ability to generate <a href="#BOOKMARK51">sugar</a>s from light (see <a href="https://en.wikipedia.org/w/index.php?title=Chloroplast&amp;oldid=918130624#Parent_group:_Cyanobacteria">Wikipedia</a>).</p>
<a name="HEADREF34"></a><h4>Vacuole</h4>
<p><a name="BOOKMARK131"></a><strong>Vacuole</strong>s are <a href="#BOOKMARK111">organelle</a> (mostly plant and algae <a href="#BOOKMARK74">cell</a>s) responsible for storage (<a href="#BOOKMARK59">water</a>, food, waste?) and <a href="#BOOKMARK212">enzyme</a>s that help break things down. <a href="#BOOKMARK131">Vacuole</a>s are typically found in plant and algae <a href="#BOOKMARK74">cell</a>s, but may also exist in animal <a href="#BOOKMARK74">cell</a>s. The ones in plants / algae tend to be much larger.</p>
<p><a href="#BOOKMARK131">Vacuole</a>s are often responsible for a plant's shape. For example, a well <a href="#BOOKMARK59">water</a>ed plant will be upright and spry because its <a href="#BOOKMARK131">vacuole</a>s are full. A plant that isn't as well <a href="#BOOKMARK59">water</a>ed may be sagging down or wilting because the <a href="#BOOKMARK131">vacuole</a>s are less full</p>
<p><img src="/.datafile_dc4ef0ca430c98a807cf410e893a8aeb/data.svg" alt="By LadyofHats - did it myself based on [1], [2] ,[3] and [4]., Public Domain, https://commons.wikimedia.org/w/index.php?curid=1685428" title="In biology, turgor pressure or turgidity is the pressure of the cell contents against the cell wall, in plant cells, determined by the water content of the vacuole, resulting from osmotic pressure. "></p><a name="HEADREF35"></a><h3>Chromosome</h3>
<p>The <a href="#BOOKMARK243">genome</a> of <a href="#BOOKMARK109">eukaryote</a>s are split into linear strands of <a href="#BOOKMARK241">DNA</a>. These linear <a href="#BOOKMARK241">DNA</a> strands come in 2 forms...</p>
<ol>
<li><a name="BOOKMARK132"></a><strong>Chromatin</strong> is the normal state, where the <a href="#BOOKMARK241">DNA</a> is loosely floating around with structural <a href="#BOOKMARK55">protein</a>s called <a name="BOOKMARK133"></a><strong>histone</strong>s.</li>
<li><a name="BOOKMARK134"></a><strong>Chromatid</strong> is the state where it's going through replication. The <a href="#BOOKMARK241">DNA</a> is packed using even more structural <a href="#BOOKMARK55">protein</a>s called  <a name="BOOKMARK135"></a><strong>scaffold protein</strong>s and the entire thing has a copy of itself attached. The original and the copy are individually referred to as <a name="BOOKMARK136"></a><strong>sister chromatid</strong>s. Together they're referred to as a <a name="BOOKMARK137"></a><strong>sister chromatid pair</strong>.</li>
</ol>
<p><a href="#BOOKMARK132">Chromatin</a> and <a href="#BOOKMARK134">chromatid</a> are mutually exclusive. As long as it has a copy attached to it, it's referred to as <a href="#BOOKMARK134">chromatid</a>. Otherwise, it's referred to as <a href="#BOOKMARK132">chromatin</a>.</p>
<p>The term <a name="BOOKMARK138"></a><strong>chromosome</strong> can refer either to a <a href="#BOOKMARK132">chromatin</a> or a <a href="#BOOKMARK137">sister chromatid pair</a>, but never a single <a href="#BOOKMARK136">sister chromatid</a>.</p>
<p><img src="/.datafile_03dfe853480c7f88b2ce5b3f7c550341/data.svg" alt="By Original uploader was Richard Wheeler at en.wikipedia - Transferred from en.wikipedia to Commons by sevela.p., CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=4017531" title="The major chromatin structures."></p><p>Each <a href="#BOOKMARK138">chromosome</a> codes for the same set of <a href="#BOOKMARK244">genes</a>. Since <a href="#BOOKMARK109">eukaryote</a>s reproduce <a href="#BOOKMARK363">sexual</a>ly (mother and father required), they contain genetic information from both the mother and father. The way this manifests itself is that most <a href="#BOOKMARK109">eukaryote</a>s will come with multiple versions of the same <a href="#BOOKMARK138">chromosome</a>. In...</p>
<ul>
<li>humans, that's 2n <a href="#BOOKMARK138">chromosome</a>s (2 <a href="#BOOKMARK138">chromosome</a>s that are alternate versions of each other, called <a href="#BOOKMARK140">diploid</a>).</li>
<li>louts plants, that's &gt;2n <a href="#BOOKMARK138">chromosome</a>s (2 or more <a href="#BOOKMARK138">chromosome</a>s that are alternate versions of each other, called <a href="#BOOKMARK142">polyploid</a>).</li>
</ul>
<p>These <a href="#BOOKMARK138">chromosome</a>s are referred to as <a name="BOOKMARK139"></a><strong>homologous chromosomes</strong>: 2 or more <a href="#BOOKMARK138">chromosome</a>s that code for the same set of <a href="#BOOKMARK244">genes</a> but aren't exactly the same (different versions of the same <a href="#BOOKMARK244">genes</a>). In other words, the <a href="#BOOKMARK138">chromosome</a>s are <a href="#BOOKMARK356">homologous</a>.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>An example straight from the material of <a href="#BOOKMARK139">homologous pair of chromosomes</a> working: As a real example, let's consider a <a href="#BOOKMARK244">gene</a> on <a href="#BOOKMARK138">chromosome</a> 9 that determines blood type (A, B, AB, or O). It's possible for a person to have two identical copies of this <a href="#BOOKMARK244">gene</a>, one on each <a href="#BOOKMARK139">homologous chromosome</a>—for example, you may have a double dose of the <a href="#BOOKMARK244">gene</a> version for type A. On the other hand, you may have two different <a href="#BOOKMARK244">gene</a> versions on your two <a href="#BOOKMARK139">homologous chromosome</a>s, such as one for type A and one for type B (giving AB blood).</p>
<p>A different version of the same <a href="#BOOKMARK244">gene</a> is called an <a href="#BOOKMARK245">allele</a>.</p>
</div>
<p>When creating <a href="#BOOKMARK74">cell</a>s for reproduction (<a href="#BOOKMARK176">gamete</a> <a href="#BOOKMARK74">cell</a>s), the <a href="#BOOKMARK74">cell</a>s only keep 1 copy from each each <a href="#BOOKMARK139">homologous chromosome</a> pair. These <a href="#BOOKMARK74">cell</a>s are referred to as <a href="#BOOKMARK141">haploid</a> <a href="#BOOKMARK74">cell</a>s because they only carrying n <a href="#BOOKMARK138">chromosome</a>s.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>To see how <a href="#BOOKMARK176">gamete</a>s are formed, see the <a href="#BOOKMARK173">meiosis</a> section.</p>
</div>
<p>To recap, a <a href="#BOOKMARK74">cell</a> can have a number of <a href="#BOOKMARK138">chromosome</a> copies:</p>
<ul>
<li><a name="BOOKMARK140"></a><strong>diploid</strong> <a href="#BOOKMARK74">cell</a>s carry 2n <a href="#BOOKMARK138">chromosome</a>s (2 copies of each <a href="#BOOKMARK138">chromosome</a>). 1 copy comes from the mother and the other comes from the father, and the copies aren't identical -- they code for different versions of the same <a href="#BOOKMARK244">genes</a>. Each of the 2 copies are said to be a <a href="#BOOKMARK139">homologous chromosome</a> pair. Examples include animal <a href="#BOOKMARK175">somatic</a> <a href="#BOOKMARK74">cell</a>s and animal <a href="#BOOKMARK174">germ</a> <a href="#BOOKMARK74">cell</a>s.</li>
<li><a name="BOOKMARK141"></a><strong>haploid</strong> <a href="#BOOKMARK74">cell</a>s carry n <a href="#BOOKMARK138">chromosome</a>s (1 copy of each <a href="#BOOKMARK138">chromosome</a>). Examples include animal <a href="#BOOKMARK176">gamete</a> <a href="#BOOKMARK74">cell</a>s and most uni<a href="#BOOKMARK74">cell</a>ular <a href="#BOOKMARK109">eukaryote</a>s.</li>
<li><a name="BOOKMARK142"></a><strong>polyploid</strong> <a href="#BOOKMARK74">cell</a>s carry An <a href="#BOOKMARK138">chromosome</a>s, where A &gt; 2 (more than 2 copies of each <a href="#BOOKMARK138">chromosome</a>). Examples include certain plant species (e.g. lotus plants).</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>An organism has an extra / missing copy of a <a href="#BOOKMARK138">chromosome</a> (e.g. 2 expected but 3 present) is called <a href="#BOOKMARK147">aneuploid</a>.</p>
</div>
<p>The following diagram shows a <a href="#BOOKMARK357">karyotype</a> of a human's <a href="#BOOKMARK140">diploid</a> <a href="#BOOKMARK139">homologous chromosome</a> pairs 1 2 and 3:</p>
<p><img src="/.datafile_43d6aa8f79bfc500fc4665ca1512a890/data.svg" alt="By National Human Genome Research Institute, http://www.genome.gov/Images/EdKit/bio1c_large.gif, Public Domain, https://commons.wikimedia.org/w/index.php?curid=2132905" title="DNA, human male chromosomes"></p><p>Certain <a href="#BOOKMARK109">eukaryotic</a> species (some mammals/snakes/insects/etc..) have an extra pair of <a href="#BOOKMARK138">chromosome</a>s that aren't alternate versions of each other but instead are totally different and used to determine the sex of the offspring. This extra pair is called <a name="BOOKMARK143"></a><strong>sex chromosome</strong>s / <a name="BOOKMARK144"></a><strong>XY chromosome</strong>s, and it determines the sex of the organism. The X and the Y refer to the <a href="#BOOKMARK138">chromosome</a> types that can appear in the pair -- XX results in a female, while XY results in a male.</p>
<p>The <a href="#BOOKMARK138">chromosome</a>s that make up the <a href="#BOOKMARK139">homologous chromosome</a>s are sometimes referred to as <a name="BOOKMARK145"></a><strong>autosome</strong>s while the ones that make up <a href="#BOOKMARK143">sex-determining chromosome</a>s (non-<a href="#BOOKMARK139">homologous chromosome</a>s) are called <a name="BOOKMARK146"></a><strong>allosome</strong>s. For example, in humans there are 44 <a href="#BOOKMARK145">autosome</a>s (22 <a href="#BOOKMARK139">homologous chromosome</a> pairs) and 2 <a href="#BOOKMARK146">allosome</a>s (1 <a href="#BOOKMARK143">sex-determining chromosome</a> pair).</p>
<p><img src="/.datafile_ace8ddb89647c7b4cb7af60217b9b3fe/data.svg" alt="By National Human Genome Research Institute, http://www.genome.gov/Images/EdKit/bio1c_large.gif, Public Domain, https://commons.wikimedia.org/w/index.php?curid=2132905" title="DNA, human male chromosomes"></p><p>An organism has an extra / missing copy of a <a href="#BOOKMARK138">chromosome</a> (e.g. 2 expected but 3 present) is called <a name="BOOKMARK147"></a><strong>aneuploid</strong>. Depending on the species and circumstances, <a href="#BOOKMARK147">aneuploid</a>y may lead to death, disease, or possibly no adverse effects at all. For example, allosomal <a href="#BOOKMARK147">aneuploid</a>y in certain female mammals rarely leads to adverse effects because of <a href="#BOOKMARK292">X-linked inactivation</a>.</p>
<a name="HEADREF36"></a><h3>Mitosis (Cell Cycle)</h3>
<p>The <a name="BOOKMARK148"></a><strong>cell cycle</strong> is the sequence of events a <a href="#BOOKMARK74">cell</a> goes through from when it's created (divides off) up to when it divides itself. In <a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a>s, the <a href="#BOOKMARK148">cell cycle</a> has 2 major phases:</p>
<ol>
<li><a href="#BOOKMARK149">Interphase</a>: The majority of a <a href="#BOOKMARK74">cell</a>'s <a href="#BOOKMARK38">life</a> is spent in <a href="#BOOKMARK149">interphase</a>, where it's growing and going about its business.</li>
<li><a href="#BOOKMARK161">Mitotic phase</a>: The portion of a <a href="#BOOKMARK74">cell</a>'s <a href="#BOOKMARK38">life</a> where it divides into 2 new <a href="#BOOKMARK74">cell</a>s.</li>
</ol>
<p>Each phase has a set of inner phases it goes through.</p>
<p><img src="/.datafile_d9e0dafdeaa913660dc6826e813d3ceb/data.svg" alt="By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076" title="Animal cell cycle"></p><a name="HEADREF37"></a><h4>Interphase</h4>
<p>The majority of a <a href="#BOOKMARK74">cell</a>'s <a href="#BOOKMARK38">life</a> is spent in <a name="BOOKMARK149"></a><strong>interphase</strong>, where it's growing and going about its business.</p>
<p><img src="/.datafile_a121f753368b1cbe73bf9e455c4172ab/data.svg" alt="By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076" title="Animal cell cycle"></p><ol>
<li>
<p><a name="BOOKMARK150"></a><strong>G1 phase</strong> / <a name="BOOKMARK151"></a><strong>Gap 1 phase</strong> /  <a name="BOOKMARK152"></a><strong>Gap I phase</strong></p>
<p><img src="/.datafile_2de435d3396ca4108452eef48ccc6c14/data.svg" alt="By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076" title="G1 phase showing the cell growing"></p><p>The <a href="#BOOKMARK74">cell</a> grows.</p>
<p>Prior to entering the next phase, the <a href="#BOOKMARK74">cell</a> can go into the <a name="BOOKMARK153"></a><strong>G0 phase</strong> / <a name="BOOKMARK154"></a><strong>resting phase</strong>,  where it's essentially pausing division. One reason for this may be <a name="BOOKMARK155"></a><strong>contact inhibition</strong>: when <a href="#BOOKMARK74">cell</a>s start touching up on other <a href="#BOOKMARK74">cell</a>s, it's a signal that it's becoming too crowded and that they should stop dividing.</p>
</li>
<li>
<p><a name="BOOKMARK156"></a><strong>S phase</strong> / <a name="BOOKMARK157"></a><strong>Synthesis phase</strong></p>
<p><img src="/.datafile_6cf48c8da3bf28f70543188c65231903/data.svg" alt="By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076" title="S Phase showing chromatin and centrosome replication"></p><p>The <a href="#BOOKMARK132">chromatin</a> in the <a href="#BOOKMARK112">nucleus</a> gets replicated to become a <a href="#BOOKMARK137">sister chromatid pair</a> (2 copies that are attached). The centrosome also gets duplicated.</p>
</li>
<li>
<p><a name="BOOKMARK158"></a><strong>G2 phase</strong> / <a name="BOOKMARK159"></a><strong>Gap 2 phase</strong> /  <a name="BOOKMARK160"></a><strong>Gap II phase</strong></p>
<p><img src="/.datafile_e2fb3bf0947c0c4d1f5999a1a9ae4207/data.svg" alt="By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076" title="G2 phase showing the cell growing even more"></p><p>The <a href="#BOOKMARK74">cell</a> grows more.</p>
</li>
</ol>
<a name="HEADREF38"></a><h4>Mitotic Phase</h4>
<p>The <a name="BOOKMARK161"></a><strong>mitotic phase</strong> is portion of a <a href="#BOOKMARK74">cell</a>'s <a href="#BOOKMARK38">life</a> where it divides into 2 new daughter <a href="#BOOKMARK74">cell</a>s.</p>
<p><img src="/.datafile_edeba21c8999de8900b6c58e16754d9a/data.svg" alt="By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076" title="Animal cell cycle"></p><ol>
<li>
<p><a name="BOOKMARK162"></a><strong>Prophase</strong> / <a name="BOOKMARK163"></a><strong>Prometaphase</strong></p>
<p><img src="/.datafile_5cd9a36a3d7be3edd370d0de332fa346/data.svg" alt="By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076" title="Prophase showing the sister chromatid condensing and nucleus membrane breaking down"></p><p>The <a href="#BOOKMARK137">sister chromatid pair</a>s start to condense into an X shape, where the only part they remain attached at is their centromeres. The <a href="#BOOKMARK77">membrane</a> of the <a href="#BOOKMARK112">nucleus</a> goes away and the 2 centrosomes that were originally attached to that <a href="#BOOKMARK77">membrane</a> get moved to opposite ends of the <a href="#BOOKMARK74">cell</a>.</p>
<p>Microtubules are attached from the centrosomes to the <a href="#BOOKMARK164">kinetochore</a> of the <a href="#BOOKMARK137">sister chromatid pair</a>s. <a name="BOOKMARK164"></a><strong>Kinetochore</strong>s are <a href="#BOOKMARK55">protein</a>s located at the centromere of the <a href="#BOOKMARK136">sister chromatid</a>s.</p>
</li>
<li>
<p><a name="BOOKMARK165"></a><strong>Metaphase</strong></p>
<p><img src="/.datafile_6b5a96d596f60e86ea437af7d2efdf43/data.svg" alt="By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076" title="Metaphase showing the sister chromatid lining up"></p><p>The <a href="#BOOKMARK137">sister chromatid pair</a>s line up in the middle of the <a href="#BOOKMARK74">cell</a>.</p>
</li>
<li>
<p><a name="BOOKMARK166"></a><strong>Anaphase</strong></p>
<p><img src="/.datafile_66ec8ef74ffb0eb152f1f7492a5f8bf9/data.svg" alt="By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076" title="Anaphase showing the sister chromatid breaking apart"></p><p>As the <a href="#BOOKMARK74">cell</a> splits apart, the microtubules also split apart the <a href="#BOOKMARK137">sister chromatid pair</a>s. Once split, they're individually referred to as <a href="#BOOKMARK132">chromatin</a> again.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>I believe the labeling on the above diagram to be incorrect. Once split, they're no longer referred to as <a href="#BOOKMARK136">sister chromatid</a>s / <a href="#BOOKMARK134">chromatid</a>s. They're referred to as <a href="#BOOKMARK132">chromatin</a>.</p>
</div>
</li>
<li>
<p><a name="BOOKMARK167"></a><strong>Telophase</strong> and <a name="BOOKMARK168"></a><strong>Cytokinesis</strong></p>
<p><img src="/.datafile_03a8c8dd3074b7778014048dfefc359a/data.svg" alt="By Kelvinsong - Own work, CC0, https://commons.wikimedia.org/w/index.php?curid=22965076" title="Telophase / cytokinesis showing nucleus membranes re-forming and cell finally splitting into 2"></p><p>During <a href="#BOOKMARK167">telophase</a>, <a href="#BOOKMARK112">nucleus</a> <a href="#BOOKMARK77">membrane</a>s re-form around the newly split <a href="#BOOKMARK132">chromatin</a>. The <a href="#BOOKMARK132">chromatin</a> also start to un-condense.</p>
<p>During <a href="#BOOKMARK168">cytokinesis</a>, the middle of the <a href="#BOOKMARK74">cell</a> starts to push together and pinch off, forming 2 separate <a href="#BOOKMARK74">cell</a>s. <a href="#BOOKMARK168">Cytokinesis</a> technically happens in parallel with <a href="#BOOKMARK161">mitosis</a> (it isn't a part of <a href="#BOOKMARK161">mitosis</a>), but it starts near the end of <a href="#BOOKMARK161">mitosis</a> (around <a href="#BOOKMARK167">telophase</a>).</p>
</li>
</ol>
<a name="HEADREF39"></a><h3>Meiosis</h3>
<p>


</p>
<p><a name="BOOKMARK173"></a><strong>Meiosis</strong> is a type of <a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a> division that results in 4 daughter <a href="#BOOKMARK74">cell</a>s, where each daughter <a href="#BOOKMARK74">cell</a> has half the number of <a href="#BOOKMARK138">chromosome</a>s as the parent <a href="#BOOKMARK74">cell</a>.</p>
<p>The source <a href="#BOOKMARK74">cell</a> is known as a <a name="BOOKMARK174"></a><strong>germ</strong> <a href="#BOOKMARK74">cell</a> -- a <a href="#BOOKMARK74">cell</a> that can either go through <a href="#BOOKMARK161">mitosis</a> or <a href="#BOOKMARK173">meiosis</a>. <a name="BOOKMARK175"></a><strong>Somatic</strong> <a href="#BOOKMARK74">cell</a>s, on the other hand, are general body <a href="#BOOKMARK74">cell</a>s and can only go through <a href="#BOOKMARK161">mitosis</a>.</p>
<p>The resulting daughter <a href="#BOOKMARK74">cell</a>s are known as <a name="BOOKMARK176"></a><strong>gamete</strong> <a href="#BOOKMARK74">cell</a>s -- <a href="#BOOKMARK74">cell</a>s that have half the genetic information from the original parent (<a href="#BOOKMARK141">haploid</a> -- only 1 <a href="#BOOKMARK138">chromosome</a> of from each of the parent's <a href="#BOOKMARK139">homologous chromosome</a> pairs), and when they merge they mix that genetic material to create the new genetic material for the offspring. Male <a href="#BOOKMARK176">gamete</a> <a href="#BOOKMARK74">cell</a>s are called <a name="BOOKMARK177"></a><strong>sperm</strong>, while female <a href="#BOOKMARK176">gamete</a> <a href="#BOOKMARK74">cell</a>s are called <a name="BOOKMARK178"></a><strong>ova</strong> or <a name="BOOKMARK179"></a><strong>egg</strong>s.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p><a href="#BOOKMARK176">Gamete</a> <a href="#BOOKMARK74">cell</a>s no longer have the ability to divide (this is a terminal operation)</p>
</div>
<p><a href="#BOOKMARK173">Meiosis</a> has 2 major phases:</p>
<ol>
<li><a href="#BOOKMARK180">Meiosis I</a> phase: <a href="#BOOKMARK174">Germ</a> <a href="#BOOKMARK74">cell</a> mixes genetic information in itself and divides into 2.</li>
<li><a href="#BOOKMARK187">Meiosis II</a> phase: Each of the 2 <a href="#BOOKMARK74">cell</a>s created above get divided again. But, this time the resulting daughter <a href="#BOOKMARK74">cell</a>s only have half the generic information (<a href="#BOOKMARK176">gamete</a> <a href="#BOOKMARK74">cell</a>s).</li>
</ol>
<p><img src="/.datafile_a304dee6a09e47d40c9f36a175a178f7/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><a name="HEADREF40"></a><h4>Meiosis I</h4>
<p>During <a name="BOOKMARK180"></a><strong>meiosis I</strong>, a <a href="#BOOKMARK174">germ</a> <a href="#BOOKMARK74">cell</a> divides in such a way that, rather than exactly duplicating each <a href="#BOOKMARK138">chromosome</a>, it duplicates and mixes between each <a href="#BOOKMARK139">homologous chromosome</a> pair. That is, <a href="#BOOKMARK138">chromosome</a> pairs that are different versions of each other end up swapping segments (<a href="#BOOKMARK182">chromosomal crossover</a>).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Remember that most <a href="#BOOKMARK74">cell</a>s have a <a href="#BOOKMARK140">diploid</a> number of <a href="#BOOKMARK138">chromosome</a>s (2n), where each <a href="#BOOKMARK138">chromosome</a> is a different version of another <a href="#BOOKMARK138">chromosome</a> in that same <a href="#BOOKMARK74">cell</a> (<a href="#BOOKMARK139">homologous chromosome</a> pairs -- e.g. code for different versions of the same <a href="#BOOKMARK244">gene</a>). One comes from the mother and one comes from the father.</p>
</div>
<p><img src="/.datafile_243e72838f1b8fb76b2d4968a6540ca9/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><ol>
<li>
<p><a name="BOOKMARK181"></a><strong>Prophase I</strong></p>
<p><img src="/.datafile_cdf3ef1400707cf917f218d8da8e14f4/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><p>Each <a href="#BOOKMARK132">chromatin</a> duplicates to become a <a href="#BOOKMARK137">sister chromatid pair</a> (2 <a href="#BOOKMARK134">chromatid</a>s that are attached to each other). Then, each <a href="#BOOKMARK137">sister chromatid pair</a> condenses and forms an X shape where the only part that the <a href="#BOOKMARK136">sister chromatid</a>s remain attached at is their centromeres.</p>
<p>The <a href="#BOOKMARK77">membrane</a> of the <a href="#BOOKMARK112">nucleus</a> goes away and the 2 centrosomes that were originally attached to that <a href="#BOOKMARK77">membrane</a> get moved to opposite ends of the <a href="#BOOKMARK74">cell</a>. Microtubules are attached from the centrosomes to the <a href="#BOOKMARK164">kinetochore</a> of the <a href="#BOOKMARK137">sister chromatid pair</a>s.</p>
<p>While in their condensed form, <a href="#BOOKMARK137">sister chromatid pair</a>s go through a process called <a name="BOOKMARK182"></a><strong>chromosomal crossover</strong>: A <a href="#BOOKMARK134">chromatid</a> from a <a href="#BOOKMARK137">sister chromatid pair</a> swaps segments of itself with a <a href="#BOOKMARK134">chromatid</a> from another <a href="#BOOKMARK137">sister chromatid pair</a>. The <a href="#BOOKMARK137">sister chromatid pair</a>s doing the swapping must be <a href="#BOOKMARK356">homolog</a>s of each other (contains alternative versions of the same <a href="#BOOKMARK244">genes</a>).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Remember that the source <a href="#BOOKMARK74">cell</a> comes with 2n <a href="#BOOKMARK138">chromosome</a>s, where each <a href="#BOOKMARK138">chromosome</a> has a sibling that has alternate versions of the same <a href="#BOOKMARK244">genes</a>. One sibling came from the father and the other sibling came from the mother. The <a href="#BOOKMARK356">homolog</a> refers to this sibling -- each sibling contains different versions of the same set of <a href="#BOOKMARK244">genes</a>.</p>
</div>
<p><img src="/.datafile_249ab9ad90ba26509076b322643919d1/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Chromosome Crossover Diagram"></p><p>The end result is that the <a href="#BOOKMARK136">sister chromatid</a>s are no longer exact copies of each other.</p>
<p>The point at which a segment swap happens is called the <a name="BOOKMARK183"></a><strong>chiasma</strong>. The <a href="#BOOKMARK183">chiasma</a> is a well defined point on the <a href="#BOOKMARK134">chromatid</a>s (it isn't random).</p>
</li>
<li>
<p><a name="BOOKMARK184"></a><strong>Metaphase I</strong></p>
<p><img src="/.datafile_706ca4d4a1ebc9838add78e92a77bb4a/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><p>The <a href="#BOOKMARK137">sister chromatid pair</a>s line up in the middle of the <a href="#BOOKMARK74">cell</a>.</p>
</li>
<li>
<p><a name="BOOKMARK185"></a><strong>Anaphase I</strong></p>
<p><img src="/.datafile_4bb45082b30f9aeeb54c41e7f3eca928/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><p>As the <a href="#BOOKMARK74">cell</a> splits apart, microtubules move apart the <a href="#BOOKMARK356">homologous</a> pairs of <a href="#BOOKMARK137">sister chromatid pair</a>s to opposite ends of the <a href="#BOOKMARK74">cell</a>. Each side gets 1 of the of <a href="#BOOKMARK137">sister chromatid pair</a>s.</p>
<p>The term <em><a href="#BOOKMARK356">homologous</a> pairs of <a href="#BOOKMARK137">sister chromatid pair</a>s</em> here is confusing:</p>
<ul>
<li><em><a href="#BOOKMARK136">sister chromatid</a></em> means 1 of 2 <a href="#BOOKMARK134">chromatid</a>s that are attached to each other at their centromeres.</li>
<li><em><a href="#BOOKMARK137">sister chromatid pair</a></em> means 2 <em><a href="#BOOKMARK136">sister chromatid</a>s</em> that are attached to each other at their centromeres.</li>
<li><em>pair of <a href="#BOOKMARK137">sister chromatid pair</a>s</em> means 2 <em><a href="#BOOKMARK137">sister chromatid pair</a>s</em>.</li>
<li><em><a href="#BOOKMARK356">homologous</a> pair of <a href="#BOOKMARK137">sister chromatid pair</a>s</em> refers to a 2 <em><a href="#BOOKMARK137">sister chromatid pair</a>s</em> that were derived from 2 <a href="#BOOKMARK356">homologous</a> <a href="#BOOKMARK132">chromatin</a>/<a href="#BOOKMARK138">chromosome</a>s.</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>What does <a href="#BOOKMARK356">homologous</a> <a href="#BOOKMARK132">chromatin</a> mean? Remember that the source <a href="#BOOKMARK74">cell</a> comes with 2n <a href="#BOOKMARK132">chromatin</a>, where each <a href="#BOOKMARK132">chromatin</a> has a sibling that has alternate versions of the same <a href="#BOOKMARK244">genes</a>. One sibling came from the father and the other sibling came from the mother. <a href="#BOOKMARK356">Homologous</a> <a href="#BOOKMARK132">chromatin</a>/<a href="#BOOKMARK138">chromosome</a>s refers to those 2 siblings.</p>
</div>
</li>
<li>
<p><a name="BOOKMARK186"></a><strong>Telophase I</strong></p>
<p><img src="/.datafile_e94e7546ddae445ff93b7b95f7bd0b24/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><p><a href="#BOOKMARK112">Nucleus</a> <a href="#BOOKMARK77">membrane</a>s re-form around the newly moved apart <a href="#BOOKMARK137">sister chromatid pair</a>s. The <a href="#BOOKMARK137">sister chromatid pair</a>s also start to un-condense but still remain attached at their centromeres. As such, the resulting daughter <a href="#BOOKMARK74">cell</a>s are said to have a <a href="#BOOKMARK141">haploid</a> number of <a href="#BOOKMARK138">chromosome</a>s -- each <a href="#BOOKMARK137">sister chromatid pair</a> is technically considered to be 1 <a href="#BOOKMARK138">chromosome</a> until it splits apart.</p>
<p>During <a href="#BOOKMARK168">cytokinesis</a>, the middle of the <a href="#BOOKMARK74">cell</a> starts to push together and pinch off, forming 2 separate <a href="#BOOKMARK74">cell</a>s. <a href="#BOOKMARK168">Cytokinesis</a> technically happens in parallel with <a href="#BOOKMARK186">telophase I</a> (it isn't a part of <a href="#BOOKMARK186">telophase I</a>), but it starts around the same time.</p>
</li>
</ol>
<a name="HEADREF41"></a><h4>Meiosis II</h4>
<p>During <a name="BOOKMARK187"></a><strong>meiosis II</strong>, each daughter <a href="#BOOKMARK74">cell</a> from <a href="#BOOKMARK180">meiosis I</a> will divide again, but this time the <a href="#BOOKMARK137">sister chromatid pair</a>s split and each resulting daughter <a href="#BOOKMARK74">cell</a> keeps 1. In other words, each daughter <a href="#BOOKMARK74">cell</a> from <a href="#BOOKMARK180">meiosis I</a> will divide again to become 2 <a href="#BOOKMARK176">gamete</a> <a href="#BOOKMARK74">cell</a>s.</p>
<p>The steps in <a href="#BOOKMARK187">meiosis II</a> are almost exactly the same as the steps in the miotic phase -- the <a href="#BOOKMARK137">sister chromatid pair</a>s are essentially being split apart into 2 new <a href="#BOOKMARK74">cell</a>s. Similar to how there's a rest period between <a href="#BOOKMARK161">mitosis</a> iterations (called <a href="#BOOKMARK149">interphase</a>), there may be a rest period between <a href="#BOOKMARK180">meiosis I</a> and <a href="#BOOKMARK187">meiosis II</a> called <a name="BOOKMARK188"></a><strong>interphase II</strong>.</p>
<p><img src="/.datafile_c14749b2ebedaf51a21cef0a147687ad/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><ol>
<li>
<p><a name="BOOKMARK189"></a><strong>Prophase II</strong></p>
<p><img src="/.datafile_994b70a5a7f5177aeb82e0261a36641b/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><p>The <a href="#BOOKMARK137">sister chromatid pair</a>s start to condense into an X shape, where the only part they remain attached at is the centromere. The <a href="#BOOKMARK77">membrane</a> of the <a href="#BOOKMARK112">nucleus</a> goes away and the 2 centrosomes that were originally attached to that <a href="#BOOKMARK77">membrane</a> get moved to opposite ends of the <a href="#BOOKMARK74">cell</a>.</p>
<p>Microtubules are attached from the centrosomes to the <a href="#BOOKMARK164">kinetochore</a> of the <a href="#BOOKMARK137">sister chromatid pair</a>s.</p>
</li>
<li>
<p><a name="BOOKMARK190"></a><strong>Metaphase II</strong></p>
<p><img src="/.datafile_9afd49f7b2d739bef96567a1cd979834/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><p>The <a href="#BOOKMARK137">sister chromatid pair</a>s line up in the middle of the <a href="#BOOKMARK74">cell</a>.</p>
</li>
<li>
<p><a name="BOOKMARK191"></a><strong>Anaphase II</strong></p>
<p><img src="/.datafile_1c01ee0bd570d04573113681472aa22f/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><p>As the <a href="#BOOKMARK74">cell</a> splits apart, the microtubules also split apart the <a href="#BOOKMARK137">sister chromatid pair</a>s. Once split, they're individually referred to as <a href="#BOOKMARK132">chromatin</a> again.</p>
</li>
<li>
<p><a name="BOOKMARK192"></a><strong>Telophase II</strong></p>
<p><img src="/.datafile_e1666cd20536ffad9cdd2237f6322ad0/data.svg" alt="By Ali Zifan - Own work; Used information from Campbell Biology (10th Edition) by: Jane B. Reece &amp; Steven A. Wasserman., CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=49630204" title="Meiosis Stages"></p><p>During <a href="#BOOKMARK167">telophase</a>, <a href="#BOOKMARK112">nucleus</a> <a href="#BOOKMARK77">membrane</a>s re-form around the newly split <a href="#BOOKMARK132">chromatin</a>. The <a href="#BOOKMARK132">chromatin</a> also start to un-condense. The resulting daughter <a href="#BOOKMARK74">cell</a>s are <a href="#BOOKMARK176">gamete</a>s at this point (e.g. <a href="#BOOKMARK177">sperm</a> <a href="#BOOKMARK74">cell</a>s). They no longer divide.</p>
<p>During <a href="#BOOKMARK168">cytokinesis</a>, the middle of the <a href="#BOOKMARK74">cell</a> starts to push together and pinch off, forming 2 separate <a href="#BOOKMARK74">cell</a>s. <a href="#BOOKMARK168">Cytokinesis</a> technically happens in parallel with <a href="#BOOKMARK192">telophase II</a> (it isn't a part of <a href="#BOOKMARK192">telophase II</a>), but it starts around the same time.</p>
</li>
</ol>
<a name="HEADREF42"></a><h3>Cancer</h3>
<p><a name="BOOKMARK193"></a><strong>Cancer</strong> refers to the uncontrolled division of <a href="#BOOKMARK74">cell</a>s (<a href="#BOOKMARK161">mitosis</a>) in a multi<a href="#BOOKMARK74">cell</a>ular organism, typically brought on by one or more mutations.</p>
<p>Normally, a <a href="#BOOKMARK74">cell</a> has internal and external cues / regulators that signal when it should stop dividing (e.g. <a href="#BOOKMARK155">contact inhibition</a>). When a mutation occurs that causes one or more of these cues to be ignored, the <a href="#BOOKMARK74">cell</a> goes through a form a programmed suicide called <a href="#BOOKMARK208">apoptosis</a>. If the genetic mutations are so severe that <a href="#BOOKMARK208">apoptosis</a> no longer occurs, that's when <a href="#BOOKMARK193">cancer</a> occurs.</p>
<p>A group of <a href="#BOOKMARK74">cell</a>s that have been dividing unimpeded is known as a <a name="BOOKMARK194"></a><strong>neoplasm</strong> / <a name="BOOKMARK195"></a><strong>tumor</strong>. If the <a href="#BOOKMARK74">cell</a>s...</p>
<ul>
<li>eventually stop dividing on their own, that mass of <a href="#BOOKMARK74">cell</a>s is called a <a name="BOOKMARK196"></a><strong>benign neoplasm</strong> / <a name="BOOKMARK197"></a><strong>benign tumor</strong>.</li>
<li>continually divide without ever stopping, that mass of <a href="#BOOKMARK74">cell</a>s is called a <a name="BOOKMARK198"></a><strong>malignant neoplasm</strong> / <a name="BOOKMARK199"></a><strong>malignant tumor</strong>.</li>
</ul>
<p>Mutations of two types of <a href="#BOOKMARK148">cell cycle</a> regulators can promote the development of <a href="#BOOKMARK193">cancer</a>:</p>
<ul>
<li>Positive regulators, which normally promote <a href="#BOOKMARK74">cell</a> growth, may become hyperactive (oncogenic).</li>
<li>Negative regulators (<a href="#BOOKMARK195">tumor</a> suppressors), which prevent the formation of <a href="#BOOKMARK195">tumor</a>s, may become inactivated.</li>
</ul>
<p><a href="#BOOKMARK195">Tumor</a> <a href="#BOOKMARK74">cell</a>s may continue to mutate at a more rapid pace than other <a href="#BOOKMARK74">cell</a>s. If one of those mutations results in <a href="#BOOKMARK195">tumor</a> <a href="#BOOKMARK74">cell</a>s breaking off from the original mass of <a href="#BOOKMARK195">tumor</a> <a href="#BOOKMARK74">cell</a>s and floating around the body, the <a href="#BOOKMARK193">cancer</a> is said to have <a name="BOOKMARK200"></a><strong>metastasize</strong>d.</p>
<a name="HEADREF43"></a><h3>Fertilization</h3>
<p><a name="BOOKMARK201"></a><strong>Fertilization</strong> is the act of bringing together a male <a href="#BOOKMARK176">gamete</a> (e.g. <a href="#BOOKMARK177">sperm</a>) to a female <a href="#BOOKMARK176">gamete</a> (e.g. <a href="#BOOKMARK179">egg</a>) (<a href="#BOOKMARK74">cell</a> types produced through <a href="#BOOKMARK173">meiosis</a>). In humans, a <a href="#BOOKMARK201">fertilize</a>d <a href="#BOOKMARK179">egg</a> is referred to as a ...</p>
<ul>
<li>
<p><a name="BOOKMARK202"></a><strong>zygote</strong> upon creation. At this point the male <a href="#BOOKMARK176">gamete</a> and the female <a href="#BOOKMARK176">gamete</a> have fused together to become a single <a href="#BOOKMARK74">cell</a>, but their <a href="#BOOKMARK112">nuclei</a> will still be separate (will eventually fuse).</p>
<p><img src="/.datafile_9ff0fb156a020506bbddae478e18d39f/data.svg" alt="Zygote with 2 separate nuclei" title="By Nina Sesina - https://commons.wikimedia.org/wiki/File:Zygote.tif, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=67459911"></p></li>
<li>
<p><a name="BOOKMARK203"></a><strong>morula</strong> upon dividing to become 16 <a href="#BOOKMARK74">cell</a>s (around a few days).</p>
</li>
<li>
<p><a name="BOOKMARK204"></a><strong>blastocyst</strong> upon dividing to become 200-300 <a href="#BOOKMARK74">cell</a>s (around 5 to 9 days). At this point the group of <a href="#BOOKMARK74">cell</a>s form a hollowed out sphere where a mass is growing on inner top portion.</p>
<p><img src="/.datafile_0c55176f6ba4af2bf6c05b135b5d85b0/data.svg" alt="Blastocyst" title="By Mr. J. Conaghan - http://stemcells.nih.gov/info/scireport/pages/chapter3.aspx, Public Domain, https://commons.wikimedia.org/w/index.php?curid=32289210"></p></li>
<li>
<p><a name="BOOKMARK205"></a><strong>embryo</strong> after around 2 weeks. At this point the major internal organs and overall shape of the human are starting to emerge (e.g. the beginning of legs, arms, eyes, etc..).</p>
<p><img src="/.datafile_0d5ef275b9e416c2975c1b954555500d/data.svg" alt="Embryo" title="By Nina Sesina - https://commons.wikimedia.org/wiki/File:Zygote.tif, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=67459911"></p></li>
<li>
<p><a name="BOOKMARK206"></a><strong>fetus</strong> after around 8-10 weeks. At this point the features formed in the <a href="#BOOKMARK205">embryo</a>nic stage grow and <a href="#BOOKMARK74">cell</a>s begin to differentiate for their actual function / purpose (e.g. neuron <a href="#BOOKMARK74">cell</a>s, kidney <a href="#BOOKMARK74">cell</a>s, liver <a href="#BOOKMARK74">cell</a>s, etc..).</p>
<p><img src="/.datafile_5695fc4f5609c3aba6a3ddc120164565/data.svg" alt="Fetus" title="By National Museum of Health and Medicine - http://nmhm.washingtondc.museum/exhibits/single_cell/imgs/14_Fetus_3_months.jpg, Public Domain, https://commons.wikimedia.org/w/index.php?curid=9998241"></p></li>
</ul>
<p>The table below is a visualization of the phases. Note that the table starts count from the <a name="BOOKMARK207"></a><strong>gestational age</strong>: the point of last menstruation. The durations in the descriptions above are from the point of <a href="#BOOKMARK201">fertilization</a>.</p>
<p><img src="/.datafile_df68ba00df22129f0d71bd2f27edfcf1/data.svg" alt="Prenatal Development Table" title="By Mikael Häggström.When using this image in external works, it may be cited as:Häggström, Mikael (2014). &quot;Medical gallery of Mikael Häggström 2014&quot;. WikiJournal of Medicine 1 (2). DOI:10.15347/wjm/2014.008. ISSN 2002-4436. Public Domain.orBy Mikael Häggström, used with permission. - Own work, Public Domain, https://commons.wikimedia.org/w/index.php?curid=6843176"></p><a name="HEADREF44"></a><h3>Apoptosis</h3>
<p><a name="BOOKMARK208"></a><strong>Apoptosis</strong> is a form of programmed <a href="#BOOKMARK74">cell</a> death in multi<a href="#BOOKMARK74">cell</a>ular organisms. For example, <a href="#BOOKMARK208">apoptosis</a> can be triggered by the <a href="#BOOKMARK74">cell</a>...</p>
<ul>
<li>detecting a faulty mutation within its own <a href="#BOOKMARK241">DNA</a>, there by avoiding a potential problem (e.g. <a href="#BOOKMARK193">cancer</a>).</li>
<li>to sculpt body parts as part of normal development (e.g. <a href="#BOOKMARK74">cell</a>s in feet die during <a href="#BOOKMARK205">embryo</a>nic / <a href="#BOOKMARK206">fetal</a> development to create gaps between toes).</li>
</ul>
<p>In <a href="#BOOKMARK208">apoptosis</a>, the <a href="#BOOKMARK77">membrane</a> of the <a href="#BOOKMARK74">cell</a> begins to shrink and pinch inward. It breaks up its <a href="#BOOKMARK241">DNA</a> and <a href="#BOOKMARK111">organelle</a> and encircles them into miniature components that eventually bud off from the <a href="#BOOKMARK74">cell</a> and float away. <a href="#BOOKMARK74">Cell</a>s from the immune system will then come and ingest those buds and/or other <a href="#BOOKMARK74">cell</a>s can reuse the components in those buds for their own purposes without suffering any damage.</p>
<p><img src="/.datafile_ee3731d1fc5a3e3d5b4caa7d325a5eeb/data.svg" alt="Apoptotic Cell Disassembly" title="By Aaron Smith, Michael AF Parkes, Georgia K Atkin-Smith, Rochelle Tixeira, Ivan KH Poon - Wikiversity:Draft:WikiJournal of Medicine/Cell disassembly during cell death, CC BY 4.0, https://commons.wikimedia.org/w/index.php?curid=59865845"></p><p>In contrast, <a name="BOOKMARK209"></a><strong>necrosis</strong> is a form of <a href="#BOOKMARK74">cell</a> death that isn't programmed -- the <a href="#BOOKMARK74">cell</a> swells up and explodes. It isn't desirable and often occurs in extreme cases (e.g. exposed to a <a href="#BOOKMARK0">chemical</a> toxin, mechanical damage, etc..). The guts of the <a href="#BOOKMARK74">cell</a>, rather than being neatly encircled and <a href="#BOOKMARK360">budding</a> off, will burst outwards and potentially damage other <a href="#BOOKMARK74">cell</a>s in the vicinity.</p>
<p><img src="/.datafile_7bdb0616d92ca1836068e5df9308dbbd/data.svg" alt="Apoptosis vs Necrosis" title="By National institute on alcohol abuse and alcoholism (NIAAA) - File:Structural changes of cells undergoing necrosis or apoptosis.gif; (pubs.niaaa.nih.gov), Public Domain, https://commons.wikimedia.org/w/index.php?curid=24184862"></p><a name="HEADREF45"></a><h2>Prokaryotic Cell</h2>
<p><img src="/.datafile_242b9325cdba93304c756a88c3120529/data.svg" alt="Prokaryotic cell" title="By Mariana Ruiz Villarreal, LadyofHats - Own work (Source: Typical prokaryotic cell, Chapter 4: Mutagenicity of alkyl N-acetoxybenzohydroxamates, Concept 1: Common Features of All Cells, Cells - Structure and Function), Public Domain, https://commons.wikimedia.org/w/index.php?curid=3648821"></p><p><a name="BOOKMARK210"></a><strong>Prokaryotic</strong> <a href="#BOOKMARK74">cell</a>s: These <a href="#BOOKMARK74">cell</a>s are typically smaller and don't have <a href="#BOOKMARK111">organelle</a>s. For example, their <a href="#BOOKMARK241">DNA</a> is free-floating in the <a href="#BOOKMARK74">cell</a> (it's free floating but stays mostly in the center area called the  <a name="BOOKMARK211"></a><strong>nucleoid</strong>).</p>
<p><a href="#BOOKMARK210">Prokaryote</a>s have a single circular-strand of <a href="#BOOKMARK241">DNA</a>. They can only be single-<a href="#BOOKMARK74">cell</a>ular organisms (e.g. bacteria).</p>
<a name="HEADREF46"></a><h2>Enzyme</h2>
<p>An <a name="BOOKMARK212"></a><strong>enzyme</strong> is a <a href="#BOOKMARK18">molecule</a> that takes in a specific set of input <a href="#BOOKMARK18">molecule</a>s and transforms them into a specific set of output <a href="#BOOKMARK18">molecule</a>s. The transformation takes the inputs and either ...</p>
<ul>
<li>assembles them into new larger <a href="#BOOKMARK18">molecule</a>s (N to 1).</li>
<li>separates them into smaller <a href="#BOOKMARK18">molecule</a>s (1 to N).</li>
<li>transfers fragments between them (N to N).</li>
</ul>
<p><a href="#BOOKMARK212">Enzyme</a>s facilitate these transformations by lowering the <a name="BOOKMARK213"></a><strong>activation energy</strong> (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>E</mi><mi>A</mi></msub></mrow><annotation encoding="application/x-tex">E_A</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.05764em;">E</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.32833099999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.05764em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight">A</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>) required for the <a href="#BOOKMARK0">chemical</a> reactions to take place.  Normally this excess energy would come in the form of heat, but <a href="#BOOKMARK212">enzyme</a>s use different mechanisms such as...</p>
<ul>
<li>temporarily contorting input <a href="#BOOKMARK18">molecule</a>s</li>
<li>temporarily pulling away some charge from adjacent <a href="#BOOKMARK3">atom</a>s of input <a href="#BOOKMARK18">molecule</a>s</li>
</ul>
<p>... such that other <a href="#BOOKMARK3">atom</a>s can get close enough to bond.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>How does heat provide <a href="#BOOKMARK213">activation energy</a>? More heat = more <a href="#BOOKMARK18">molecule</a>s moving faster = more things bumping into each other faster. 2 <a href="#BOOKMARK18">molecule</a>s may have <a href="#BOOKMARK3">atom</a>s that want to bond but neighbouring <a href="#BOOKMARK3">atom</a>s on those <a href="#BOOKMARK18">molecule</a>s may be repelling away with stronger force. Increased speed means the repelling is less effective.</p>
</div>
<p>An <a href="#BOOKMARK212">enzyme</a> is almost always a <a href="#BOOKMARK55">protein</a> <a href="#BOOKMARK18">molecule</a> but can also be a <a href="#BOOKMARK247">RNA</a>-like <a href="#BOOKMARK18">molecule</a> called a <a name="BOOKMARK214"></a><strong>ribozyme</strong>.</p>
<p>The general terminology for <a href="#BOOKMARK212">enzyme</a>s are as follows:</p>
<ul>
<li><a name="BOOKMARK215"></a><strong>Substrate</strong> / <a name="BOOKMARK216"></a><strong>Reactant</strong>s refers to the input <a href="#BOOKMARK18">molecule</a>s of an <a href="#BOOKMARK212">enzyme</a>.</li>
<li><a name="BOOKMARK217"></a><strong>Product</strong> refers to the output <a href="#BOOKMARK18">molecule</a>s of an <a href="#BOOKMARK212">enzyme</a>.</li>
<li><a name="BOOKMARK218"></a><strong>Metabolite</strong> refers to either a <a href="#BOOKMARK215">substrate</a> or <a href="#BOOKMARK217">product</a>.</li>
<li><a name="BOOKMARK219"></a><strong>Metabolism</strong> is the total sum of all the <a href="#BOOKMARK0">chemical</a> reactions that take place inside of a <a href="#BOOKMARK74">cell</a>. These reactions happen via <a href="#BOOKMARK212">enzyme</a>s.</li>
<li><a name="BOOKMARK220"></a><strong>Active site</strong>s are the physical locations that <a href="#BOOKMARK215">substrate</a>s float into on an <a href="#BOOKMARK212">enzyme</a>.</li>
<li><a name="BOOKMARK221"></a><strong>Catalysis</strong> / <a name="BOOKMARK222"></a><strong>Catalyze</strong> is the term for increasing the rate / probability of a <a href="#BOOKMARK0">chemical</a> reaction. <a href="#BOOKMARK212">Enzyme</a>s perform <a href="#BOOKMARK221">catalysis</a>.</li>
<li><a name="BOOKMARK223"></a><strong>Catalyst</strong>s are entities that speed up <a href="#BOOKMARK0">chemical</a> reactions but don't change in the process (reaction happens but <a href="#BOOKMARK223">catalyst</a> remains). <a href="#BOOKMARK212">Enzyme</a>s are <a href="#BOOKMARK223">catalyst</a>s.</li>
</ul>
<p><a href="#BOOKMARK212">Enzyme</a>s have a limited set of <a href="#BOOKMARK215">substrate</a> types that they accept. A <a href="#BOOKMARK215">substrate</a> will bind to the <a href="#BOOKMARK220">active site</a> of the <a href="#BOOKMARK212">enzyme</a> only if it fits into the <a href="#BOOKMARK220">active site</a>. For example, the following diagram shows 2 <a href="#BOOKMARK215">substrate</a>s binding to an <a href="#BOOKMARK212">enzyme</a>, the <a href="#BOOKMARK212">enzyme</a> facilitating their their assembly, then releasing back out.</p>
<p><img src="/.datafile_20b8ab8d4097402f09dc4276f1d960cc/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Anabolism process."></p><p>It was previously thought that <a href="#BOOKMARK212">enzyme</a>s had a “lock-and-key” model, similar to how puzzle pieces fit together. Later on it was found out that an <a href="#BOOKMARK212">enzyme</a>s actually <a name="BOOKMARK224"></a><strong>induce fit</strong> by changing shape slightly when they bind with <a href="#BOOKMARK215">substrate</a>s, such that they can better hold on to those <a href="#BOOKMARK215">substrate</a>s.</p>
<p>Examples of <a href="#BOOKMARK212">enzyme</a>s and what they do:</p>
<ul>
<li>protease → breaks up <a href="#BOOKMARK55">protein</a> <a href="#BOOKMARK18">molecule</a>s to smaller <a href="#BOOKMARK55">protein</a> <a href="#BOOKMARK18">molecule</a>s.</li>
<li>sucrase → breaks up sucrose <a href="#BOOKMARK18">molecule</a>s to simpler <a href="#BOOKMARK51">sugar</a> <a href="#BOOKMARK18">molecule</a>s.</li>
<li><a href="#BOOKMARK247">RNA</a> <a href="#BOOKMARK21">polymer</a>ase → scans over <a href="#BOOKMARK241">DNA</a>... splitting it, generating <a href="#BOOKMARK248">mRNA</a>, and stitching it back up.</li>
</ul>
<p><img src="/.datafile_6ebaf9cda3408f77cb3406a91b515da0/data.svg" alt="RNA Polymerase II" title="By JWSchmidt - http://www.rcsb.org/pdb/static.do?p=general_information/news_publications/newsletters/2003q2/mom.htmlhttps://en.wikiversity.org/wiki/File:Label_RNA_pol_II.png, Public Domain, https://commons.wikimedia.org/w/index.php?curid=32033313"></p><a name="HEADREF47"></a><h3>Metabolic Pathway</h3>
<p>A <a name="BOOKMARK225"></a><strong>metabolic pathway</strong> is a network/graph of <a href="#BOOKMARK212">enzyme</a>s that produces a final resulting <a href="#BOOKMARK18">molecule</a>. Each <a href="#BOOKMARK212">enzyme</a> produces <a href="#BOOKMARK18">molecule</a>s that feed into other <a href="#BOOKMARK212">enzyme</a>s in the pathway, eventually forming the final <a href="#BOOKMARK18">molecule</a>. The term <a name="BOOKMARK226"></a><strong>intermediate</strong> refers to an output of one <a href="#BOOKMARK212">enzyme</a> that’s used as an input by another.</p>
<p>For example, the following graph is the <a href="#BOOKMARK225">metabolic pathway</a> for gamma-hydroxybutyric <a href="#BOOKMARK33">acid</a>...</p>
<p><img src="/.datafile_e2567f6ebd6b3e35eb02b1633c7de2b9/data.svg" alt="By Anypodetos - Own work, vectorised version of File:GHB metab path.png by User:Meodipt, Public Domain, https://commons.wikimedia.org/w/index.php?curid=8988213" title="Example metabolic pathway"></p><a name="HEADREF48"></a><h3>Anabolism</h3>
<p><a href="#BOOKMARK219">Metabolism</a> can be broken down into 2 categories: <a href="#BOOKMARK227">anabolism</a> (building-up) and <a href="#BOOKMARK229">catabolism</a> (breaking-down).</p>
<p>The process that builds up a <a href="#BOOKMARK18">molecule</a> from smaller <a href="#BOOKMARK18">molecule</a>s is called <a name="BOOKMARK227"></a><strong>anabolism</strong>. An <a href="#BOOKMARK212">enzyme</a> takes in the <a href="#BOOKMARK18">molecule</a>s and creates bonds between them via an <a name="BOOKMARK228"></a><strong>endergonic reaction</strong>s: energy is stored as bonds between the smaller <a href="#BOOKMARK18">molecule</a>s thereby forming the larger <a href="#BOOKMARK18">molecule</a>.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>A good way to remember the reaction types... In <a href="#BOOKMARK228">ENDergonic reaction</a>s, the energy ENDs up in a bond. In <a href="#BOOKMARK230">EXergonic reaction</a>s, the energy EXplodes out thereby breaking the bond.</p>
</div>
<p><img src="/.datafile_20b8ab8d4097402f09dc4276f1d960cc/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Anabolism process."></p><p>An example of <a href="#BOOKMARK227">anabolism</a> is <a href="#BOOKMARK313">photosynthesis</a>: plants will bond carbon dioxide gas (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><msub><mi>O</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">CO_2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>) with <a href="#BOOKMARK59">water</a> (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>2</mn></msub><mi>O</mi></mrow><annotation encoding="application/x-tex">H_2O</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathdefault" style="margin-right:0.02778em;">O</span></span></span></span>) using energy from the sun, creating <a href="#BOOKMARK51">sugar</a> (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>C</mi><mn>6</mn></msub><msub><mi>H</mi><mn>12</mn></msub><msub><mi>O</mi><mn>6</mn></msub></mrow><annotation encoding="application/x-tex">C_6H_{12}O_6</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.07153em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>)</p>
<p><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>6</mn><mi>C</mi><msub><mi>O</mi><mn>2</mn></msub><mo>+</mo><mn>6</mn><msub><mi>H</mi><mn>2</mn></msub><mi>O</mi><mo>+</mo><mi>e</mi><mi>n</mi><mi>e</mi><mi>r</mi><mi>g</mi><mi>y</mi><mo>→</mo><msub><mi>C</mi><mn>6</mn></msub><msub><mi>H</mi><mn>12</mn></msub><msub><mi>O</mi><mn>6</mn></msub><mo>+</mo><mn>6</mn><msub><mi>O</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">6CO_2 + 6H_2O + energy \to C_6H_{12}O_6 + 6O_2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.19444em;"></span><span class="mord mathdefault">e</span><span class="mord mathdefault">n</span><span class="mord mathdefault">e</span><span class="mord mathdefault" style="margin-right:0.02778em;">r</span><span class="mord mathdefault" style="margin-right:0.03588em;">g</span><span class="mord mathdefault" style="margin-right:0.03588em;">y</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">→</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.07153em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span></p>
<a name="HEADREF49"></a><h3>Catabolism</h3>
<p><a href="#BOOKMARK219">Metabolism</a> can be broken down into 2 categories: <a href="#BOOKMARK227">anabolism</a> (building-up) and <a href="#BOOKMARK229">catabolism</a> (breaking-down).</p>
<p>The process that breaks down a large <a href="#BOOKMARK18">molecule</a> into smaller <a href="#BOOKMARK18">molecule</a>s is called <a name="BOOKMARK229"></a><strong>catabolism</strong>. An <a href="#BOOKMARK212">enzyme</a> takes in a larger <a href="#BOOKMARK18">molecule</a> breaks up some of its bonds via <a name="BOOKMARK230"></a><strong>exergonic reaction</strong>s: energy used as bonds in the <a href="#BOOKMARK18">molecule</a> are release thereby breaking it up into smaller <a href="#BOOKMARK18">molecule</a>s.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>A good way to remember the reaction types... In <a href="#BOOKMARK228">ENDergonic reaction</a>s, the energy ENDs up in a bond. In <a href="#BOOKMARK230">EXergonic reaction</a>s, the energy EXplodes out thereby breaking the bond.</p>
</div>
<p><img src="/.datafile_06ea10671ecbbec31a17b4e680e70c50/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Catabolism process."></p><p>An example of <a href="#BOOKMARK229">catabolism</a> is <a href="#BOOKMARK317">cellular respiration</a>: <a href="#BOOKMARK74">cell</a>s will break down the bonds in glucose (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>C</mi><mn>6</mn></msub><msub><mi>H</mi><mn>12</mn></msub><msub><mi>O</mi><mn>6</mn></msub><mo>+</mo><mn>6</mn><msub><mi>O</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">C_6H_{12}O_6 + 6O_2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.07153em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>) to release energy, splitting into carbon dioxide (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><msub><mi>O</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">CO_2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>) and <a href="#BOOKMARK59">water</a> (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>2</mn></msub><mi>O</mi></mrow><annotation encoding="application/x-tex">H_2O</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathdefault" style="margin-right:0.02778em;">O</span></span></span></span>)</p>
<p><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>C</mi><mn>6</mn></msub><msub><mi>H</mi><mn>12</mn></msub><msub><mi>O</mi><mn>6</mn></msub><mo>+</mo><mn>6</mn><msub><mi>O</mi><mn>2</mn></msub><mo>→</mo><mn>6</mn><mi>C</mi><msub><mi>O</mi><mn>2</mn></msub><mo>+</mo><mn>6</mn><msub><mi>H</mi><mn>2</mn></msub><mi>O</mi><mo>+</mo><mi>e</mi><mi>n</mi><mi>e</mi><mi>r</mi><mi>g</mi><mi>y</mi></mrow><annotation encoding="application/x-tex">C_6H_{12}O_6 + 6O_2 \to 6CO_2 + 6H_2O + energy</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.07153em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">→</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.19444em;"></span><span class="mord mathdefault">e</span><span class="mord mathdefault">n</span><span class="mord mathdefault">e</span><span class="mord mathdefault" style="margin-right:0.02778em;">r</span><span class="mord mathdefault" style="margin-right:0.03588em;">g</span><span class="mord mathdefault" style="margin-right:0.03588em;">y</span></span></span></span></p>
<a name="HEADREF50"></a><h2>Molecular Genetics</h2>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<p>TODO: Fill me in after learning enough <a href="#BOOKMARK0">chemistry</a></p>
<a name="HEADREF51"></a><h3>Nucleotides</h3>
<p><a name="BOOKMARK231"></a><strong>Nucleic Acid</strong> is a <a href="#BOOKMARK18">molecule</a> (<a href="#BOOKMARK23">heteropolymer</a>) that's built up from other <a href="#BOOKMARK18">molecule</a>s called <a name="BOOKMARK232"></a><strong>nucleotide</strong>s (<a href="#BOOKMARK19">monomer</a>s). <a href="#BOOKMARK231">Nucleic acid</a> comes in 2 flavours: <a href="#BOOKMARK241">DNA</a> and <a href="#BOOKMARK247">RNA</a>. Each <a href="#BOOKMARK232">nucleotide</a> consists of a <a href="#BOOKMARK51">sugar</a> <a href="#BOOKMARK18">molecule</a> (ribose in <a href="#BOOKMARK247">RNA</a> / deoxyribose in <a href="#BOOKMARK241">DNA</a>) attached to a phosphate group and a nitrogen-containing <a href="#BOOKMARK233">base</a>.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>It's called <a href="#BOOKMARK231">nucleic acid</a> because it has some <a href="#BOOKMARK33">acidic</a> properties to it and <a href="#BOOKMARK241">DNA</a> is found in the <a href="#BOOKMARK112">nucleus</a> of a <a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a>. But <a href="#BOOKMARK241">DNA</a> also in <a href="#BOOKMARK210">prokaryotic</a> <a href="#BOOKMARK74">cell</a>s and some <a href="#BOOKMARK111">organelle</a>s -- those don't have a <a href="#BOOKMARK112">nucleus</a>.</p>
</div>
<p><img src="/.datafile_d6637fda8a33640c3aad2a7fd7ff86c8/data.svg" alt="By OpenStax - https://cnx.org/contents/FPtK1zmh@8.25:fEI3C8Ot@10/Preface, CC BY 4.0, https://commons.wikimedia.org/w/index.php?curid=30131206" title="DNA Nucleotides"></p><p>There are 5 <a name="BOOKMARK233"></a><strong>base</strong>.:</p>
<ul>
<li>A (<a name="BOOKMARK234"></a><strong>adenine</strong>)</li>
<li>C (<a name="BOOKMARK235"></a><strong>cytosine</strong>)</li>
<li>G (<a name="BOOKMARK236"></a><strong>guanine</strong>)</li>
<li>T (<a name="BOOKMARK237"></a><strong>thymine</strong>) / U (<a name="BOOKMARK238"></a><strong>uracil</strong>)</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>T only appears in <a href="#BOOKMARK241">DNA</a>. In <a href="#BOOKMARK247">RNA</a>, T is replaced by U.</p>
</div>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The <a href="#BOOKMARK233">base</a> is what distinguishes the types of <a href="#BOOKMARK232">nucleotide</a>s from each other. The term <a href="#BOOKMARK232">nucleotide</a> and <a href="#BOOKMARK233">base</a> are often used interchangeably.</p>
</div>
<p>Two <a href="#BOOKMARK232">nucleotide</a>s connected together are called a <a name="BOOKMARK239"></a><strong>base pair</strong>. The rules to <a href="#BOOKMARK239">base pair</a>s are:</p>
<ul>
<li>A only ever binds to T/U (e.g. AT or TA in <a href="#BOOKMARK241">DNA</a>, AU or UA in <a href="#BOOKMARK247">RNA</a>)</li>
<li>G only ever binds to C (e.g. GC or CG)</li>
</ul>
<a name="HEADREF52"></a><h3>DNA</h3>
<p><a name="BOOKMARK240"></a><strong>Deoxyribonucleic acid</strong> (<a name="BOOKMARK241"></a><strong>DNA</strong>) is a <a href="#BOOKMARK231">nucleic acid</a> <a href="#BOOKMARK18">molecule</a> that contains the instructions needed for the growth/functioning/maintenance of an organism. Depending on the type of organism, <a href="#BOOKMARK241">DNA</a> is located in different parts fo the <a href="#BOOKMARK74">cell</a>.</p>
<ul>
<li>For <a href="#BOOKMARK210">prokaryote</a>s, <a href="#BOOKMARK241">DNA</a> is free-floating in the <a href="#BOOKMARK75">cytoplasm</a>.</li>
<li>For <a href="#BOOKMARK109">eukaryote</a>s, <a href="#BOOKMARK241">DNA</a> is isolated in the <a href="#BOOKMARK112">nucleus</a> (<a name="BOOKMARK242"></a><strong>nuclear DNA</strong>). Their <a href="#BOOKMARK117">mitochondria</a> and <a href="#BOOKMARK123">chloroplast</a> also have their own <a href="#BOOKMARK241">DNA</a>.</li>
</ul>
<p><img src="/.datafile_dba3bbf4270e2a23662c2e15cdf773f4/data.svg" alt="By Zephyris - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=15027555" title="DNA"></p><p><a href="#BOOKMARK241">DNA</a> is composed of two strands of <a href="#BOOKMARK232">nucleotide</a>s that connect at various points in between. The order these <a href="#BOOKMARK232">nucleotide</a>s appear in defines the genetic information/instructions of the organism. For example, a string/sequence of <a href="#BOOKMARK241">DNA</a> <a href="#BOOKMARK233">bases</a>: ATATTTTCGATATCCACCA.</p>
<p><a href="#BOOKMARK241">DNA</a> strands can be made up of 4 different <a href="#BOOKMARK232">nucleotide</a> types (<a href="#BOOKMARK233">bases</a>):</p>
<ul>
<li>A (<a href="#BOOKMARK234">adenine</a>)</li>
<li>C (<a href="#BOOKMARK235">cytosine</a>)</li>
<li>G (<a href="#BOOKMARK236">guanine</a>)</li>
<li>T (<a href="#BOOKMARK237">thymine</a>)</li>
</ul>
<p>The two <a href="#BOOKMARK232">nucleotide</a>s that make up a connection are called a <a href="#BOOKMARK239">base pair</a>. In <a href="#BOOKMARK241">DNA</a>, the rules to <a href="#BOOKMARK239">base pair</a>s are...</p>
<ul>
<li>A only ever binds to T (e.g. AT or TA)</li>
<li>G only ever binds to C (e.g. GC or CG)</li>
</ul>
<p>Terminology specific to <a href="#BOOKMARK241">DNA</a>:</p>
<ul>
<li><a name="BOOKMARK243"></a><strong>genome</strong> - The entire set of <a href="#BOOKMARK241">DNA</a> for an organism. For <a href="#BOOKMARK109">eukaryote</a>s, this refers to <a href="#BOOKMARK242">nuclear DNA</a>, not the independent <a href="#BOOKMARK241">DNA</a> carried by <a href="#BOOKMARK111">organelle</a>s such as <a href="#BOOKMARK117">mitochondria</a>.</li>
<li><a name="BOOKMARK244"></a><strong>gene</strong> - A section of an organism's <a href="#BOOKMARK241">DNA</a> that contains instructions for some functionality, typically for building some <a href="#BOOKMARK55">protein</a>.</li>
<li><a name="BOOKMARK245"></a><strong>allele</strong> - A version of a <a href="#BOOKMARK244">gene</a> (different coding at the some position of <a href="#BOOKMARK241">DNA</a>).</li>
</ul>
<a name="HEADREF53"></a><h3>RNA</h3>
<p><a name="BOOKMARK246"></a><strong>Ribonucleic acid</strong> (<a name="BOOKMARK247"></a><strong>RNA</strong>) is a <a href="#BOOKMARK231">nucleic acid</a> <a href="#BOOKMARK18">molecule</a> used in various ways to facilitate building <a href="#BOOKMARK55">protein</a>s. It can also act as an <a href="#BOOKMARK212">enzyme</a> (<a href="#BOOKMARK214">ribozyme</a>) or contain the genetic information for some viruses.</p>
<p><a href="#BOOKMARK247">RNA</a> is commonly composed of a single strand that folds over onto itself.</p>
<p><img src="/.datafile_bf93d98db3f4f9a5eb45f67184f597e9/data.svg" alt="By Vossman - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=7115139" title="RNA"></p><p><a href="#BOOKMARK247">RNA</a> strands can be made up of 4 different <a href="#BOOKMARK232">nucleotide</a> types (<a href="#BOOKMARK233">bases</a>):</p>
<ul>
<li>A (<a href="#BOOKMARK234">adenine</a>)</li>
<li>C (<a href="#BOOKMARK235">cytosine</a>)</li>
<li>G (<a href="#BOOKMARK236">guanine</a>)</li>
<li>U (<a href="#BOOKMARK238">uracil</a>)</li>
</ul>
<p>The two <a href="#BOOKMARK232">nucleotide</a>s that make up a connection are called a <a href="#BOOKMARK239">base pair</a>. In <a href="#BOOKMARK247">RNA</a>, the rules to <a href="#BOOKMARK239">base pair</a>s are...</p>
<ul>
<li>A only ever binds to U (e.g. AU or UA)</li>
<li>G only ever binds to C (e.g. GC or CG)</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The rules are similar to <a href="#BOOKMARK241">DNA</a>, except T is replaced by U. <a href="#BOOKMARK241">DNA</a> can't have U and <a href="#BOOKMARK247">RNA</a> can't have T.</p>
</div>
<p>Unlike <a href="#BOOKMARK241">DNA</a>, <a href="#BOOKMARK247">RNA</a> is transient (lasts for minutes) and comes in multiple flavours:</p>
<ul>
<li><a name="BOOKMARK248"></a><strong>mRNA</strong> or <a name="BOOKMARK249"></a><strong>messenger RNA</strong> - Carries information from <a href="#BOOKMARK241">DNA</a> to the <a href="#BOOKMARK81">ribosome</a>s detailing a <a href="#BOOKMARK55">protein</a> to build. <a href="#BOOKMARK81">Ribosome</a>s are the parts of <a href="#BOOKMARK74">cell</a>s that synthesize <a href="#BOOKMARK55">protein</a>s.</li>
<li><a name="BOOKMARK250"></a><strong>tRNA</strong> or <a name="BOOKMARK251"></a><strong>transfer RNA</strong> - Brings to the <a href="#BOOKMARK81">ribosome</a> the <a href="#BOOKMARK56">amino acid</a>s required to build the <a href="#BOOKMARK55">protein</a> that the <a href="#BOOKMARK248">mRNA</a> is requesting.</li>
<li><a name="BOOKMARK252"></a><strong>rRNA</strong> or <a name="BOOKMARK253"></a><strong>ribosomal RNA</strong> - Makes up a large portion of the <a href="#BOOKMARK81">ribosome</a> (60%). The rest of the <a href="#BOOKMARK81">ribosome</a> is <a href="#BOOKMARK55">protein</a>.</li>
</ul>
<a name="HEADREF54"></a><h2>Classical Genetics</h2>
<p><a name="BOOKMARK254"></a><strong>Classical genetics</strong> is study of the probability of which <a href="#BOOKMARK244">genes</a> get passed down from parents to offspring and the probability of which <a href="#BOOKMARK244">genes</a> express themselves into some physically observable trait(s). It's focused exclusively on <a href="#BOOKMARK363">sexual</a>ly reproducing <a href="#BOOKMARK109">eukaryotic</a> organisms. The term...</p>
<ul>
<li><a name="BOOKMARK255"></a><strong>genotype</strong> refers to the <a href="#BOOKMARK245">allele</a>s present for a set of one or more <a href="#BOOKMARK244">genes</a>.</li>
<li><a name="BOOKMARK256"></a><strong>phenotype</strong> refers to the observable trait(s) expressed by a <a href="#BOOKMARK255">genotype</a>.</li>
</ul>
<p>Essentially, <a href="#BOOKMARK254">classical genetics</a> is the idea that the coding / <a href="#BOOKMARK245">allele</a>s for a specific set of <a href="#BOOKMARK244">genes</a> (<a href="#BOOKMARK255">genotype</a>) results in some change in the organism that's observable (<a href="#BOOKMARK256">phenotype</a>), and the probability that an organism can end up with a specific <a href="#BOOKMARK255">genotype</a>/<a href="#BOOKMARK256">phenotype</a> can be calculated from the <a href="#BOOKMARK255">genotype</a> of its parents. For example, a certain set of <a href="#BOOKMARK245">allele</a>s (<a href="#BOOKMARK255">genotype</a>) may be responsible for blue eyes (<a href="#BOOKMARK256">phenotype</a>).</p>
<p>If a single <a href="#BOOKMARK244">gene</a> is responsible for contributing to multiple unrelated <a href="#BOOKMARK256">phenotype</a>s, it's referred to as <a name="BOOKMARK257"></a><strong>pleiotropy</strong>. For example, a specific <a href="#BOOKMARK244">gene</a> is responsible for generating melanin. If the <a href="#BOOKMARK245">allele</a> for that <a href="#BOOKMARK244">gene</a> has a specific mutation in it, melanin production stops / is drastically reduced. A lack of melanin results in albinism: a condition that effects the color of eyes, hair, and skin (3 separate <a href="#BOOKMARK256">phenotype</a>s).</p>
<p>If multiple <a href="#BOOKMARK244">genes</a> are responsible for contribution to a single <a href="#BOOKMARK256">phenotype</a>, it's referred to as <a name="BOOKMARK258"></a><strong>polygene</strong>. For example, many <a href="#BOOKMARK244">genes</a> are responsible for contributing to a person's height (around 400).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Don't be fooled by the name <a href="#BOOKMARK254">classical genetics</a>. The term classical doesn't mean that it's deprecated.</p>
</div>
<p><a href="#BOOKMARK109">Eukaryotic</a> organisms that <a href="#BOOKMARK363">sexual</a>ly reproduce hold <a href="#BOOKMARK356">homologous</a> pairs of <a href="#BOOKMARK138">chromosome</a>s -- they have multiple copies of <a href="#BOOKMARK138">chromosome</a>s, each of which codes for the same <a href="#BOOKMARK244">genes</a> but different versions of those <a href="#BOOKMARK244">genes</a> (<a href="#BOOKMARK245">allele</a>s). In humans, the number of <a href="#BOOKMARK138">chromosome</a>s that make up a single set of <a href="#BOOKMARK139">homologous chromosome</a>s is 2 (<a href="#BOOKMARK140">diploid</a>). That means that a human has 2 <a href="#BOOKMARK245">allele</a>s for each <a href="#BOOKMARK244">gene</a> -- 1 <a href="#BOOKMARK245">allele</a> is randomly chosen from the father and the other <a href="#BOOKMARK245">allele</a> is randomly chosen from the mother.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>This isn't the case for <a href="#BOOKMARK144">XY chromosome</a>s. See <a href="#BOOKMARK288">sex-linked gene</a>s section for more information on how <a href="#BOOKMARK144">XY chromosome</a>s (which are not <a href="#BOOKMARK139">homologous chromosome</a>s) are treated.</p>
</div>
<p>Depending on which <a href="#BOOKMARK245">allele</a> combinations are present, different observable traits may be expressed. For example, imagine the color of a flower is determined by a single <a href="#BOOKMARK244">gene</a>. If a flower had both a red <a href="#BOOKMARK245">allele</a> and a white <a href="#BOOKMARK245">allele</a> for that <a href="#BOOKMARK244">gene</a>, the observable trait would be that it'd have a pink color.</p>
<p><img src="/.datafile_84433433b2be815fdf9e3a134d3250fc/diagram.svg" alt="Graphviz Dot Diagram"></p><p>If an organism has...</p>
<ul>
<li>the same <a href="#BOOKMARK245">allele</a> for each <a href="#BOOKMARK244">gene</a>, it's said to be <a name="BOOKMARK259"></a><strong>homozygous</strong>.</li>
<li>different <a href="#BOOKMARK245">allele</a>s for each <a href="#BOOKMARK244">gene</a>, it's said to be <a name="BOOKMARK260"></a><strong>heterozygous</strong>.</li>
</ul>
<p>In the above example, the...</p>
<ul>
<li>male parent's <a href="#BOOKMARK255">genotype</a> is 2 red <a href="#BOOKMARK245">allele</a>s (<a href="#BOOKMARK259">homozygous</a>) while the <a href="#BOOKMARK256">phenotype</a> is red.</li>
<li>female parent's <a href="#BOOKMARK255">genotype</a> is 2 white <a href="#BOOKMARK245">allele</a>s (<a href="#BOOKMARK259">homozygous</a>) while the <a href="#BOOKMARK256">phenotype</a> is white.</li>
<li>child's <a href="#BOOKMARK255">genotype</a> is 1 red <a href="#BOOKMARK245">allele</a> and 1 white <a href="#BOOKMARK245">allele</a> (<a href="#BOOKMARK260">heterozygous</a>) while the <a href="#BOOKMARK256">phenotype</a> is pink.</li>
</ul>
<p>For each <a href="#BOOKMARK244">gene</a>, offspring get a random <a href="#BOOKMARK245">allele</a> from each parent. The odds of which <a href="#BOOKMARK245">allele</a>s the offspring ends up with can be visualized using a <a name="BOOKMARK261"></a><strong>Punnett square</strong> diagram: a table where the <a href="#BOOKMARK245">allele</a>s for a <a href="#BOOKMARK244">gene</a> are written across the top axis (male parent) and down the left axis (female parent), and each cell maps to the <a href="#BOOKMARK245">allele</a>s in the top/left coordinate that it's in.</p>
<p>For example, a <a href="#BOOKMARK244">gene</a> that controls the color of a flower has 2 <a href="#BOOKMARK245">allele</a>s: red and white. The male parent has 2 red <a href="#BOOKMARK245">allele</a>s while the female parent has both a red <a href="#BOOKMARK245">allele</a> and a white <a href="#BOOKMARK245">allele</a>. The <a href="#BOOKMARK261">Punnett square</a> describing the probability offspring's <a href="#BOOKMARK245">allele</a>s:</p>
<table><tbody><tr><td></td><td> R </td><td> R</td></tr><tr><td>R</td><td> RR</td><td> RR</td></tr><tr><td>W</td><td> WR</td><td> WR</td></tr></tbody></table><div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Unsure what the convention for <a href="#BOOKMARK261">Punnett square</a>s is -- I'm always putting the <a href="#BOOKMARK245">allele</a> on the left axis (female parent) first.</p>
</div>
<p>The above diagram visualizes that the probability of the offspring having...</p>
<ul>
<li>2 red <a href="#BOOKMARK245">allele</a>s is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mo stretchy="false">(</mo><mi>R</mi><mi>R</mi><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>2</mn><mn>4</mn></mfrac></mrow><annotation encoding="application/x-tex">P(RR)=\frac{2}{4}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathdefault" style="margin-right:0.13889em;">P</span><span class="mopen">(</span><span class="mord mathdefault" style="margin-right:0.00773em;">R</span><span class="mord mathdefault" style="margin-right:0.00773em;">R</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">4</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</li>
<li>2 white <a href="#BOOKMARK245">allele</a>s is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mo stretchy="false">(</mo><mi>W</mi><mi>W</mi><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>0</mn><mn>4</mn></mfrac></mrow><annotation encoding="application/x-tex">P(WW)=\frac{0}{4}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathdefault" style="margin-right:0.13889em;">P</span><span class="mopen">(</span><span class="mord mathdefault" style="margin-right:0.13889em;">W</span><span class="mord mathdefault" style="margin-right:0.13889em;">W</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">4</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">0</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</li>
<li>a red <a href="#BOOKMARK245">allele</a> and a white <a href="#BOOKMARK245">allele</a> is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mo stretchy="false">(</mo><mi>R</mi><mi>W</mi><mi mathvariant="normal">∣</mi><mi>W</mi><mi>R</mi><mo stretchy="false">)</mo><mo>=</mo><mfrac><mn>2</mn><mn>4</mn></mfrac></mrow><annotation encoding="application/x-tex">P(RW|WR)=\frac{2}{4}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathdefault" style="margin-right:0.13889em;">P</span><span class="mopen">(</span><span class="mord mathdefault" style="margin-right:0.00773em;">R</span><span class="mord mathdefault" style="margin-right:0.13889em;">W</span><span class="mord">∣</span><span class="mord mathdefault" style="margin-right:0.13889em;">W</span><span class="mord mathdefault" style="margin-right:0.00773em;">R</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">4</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</li>
</ul>
<p><a href="#BOOKMARK261">Punnett square</a>s can be extended to cover multiple <a href="#BOOKMARK244">genes</a> so long as those <a href="#BOOKMARK244">genes</a> are independently assorted. That is, each <a href="#BOOKMARK244">gene</a> in the list must be on a different <a href="#BOOKMARK138">chromosome</a>. If they were on the same <a href="#BOOKMARK138">chromosome</a>, the chance of you getting <a href="#BOOKMARK244">gene</a> A may be dependent on you getting <a href="#BOOKMARK244">gene</a> B (see <a href="#BOOKMARK267">linked gene</a>s).</p>
<p>For example, a <a href="#BOOKMARK244">gene</a> that controls...</p>
<ul>
<li>the color of a flower has 2 <a href="#BOOKMARK245">allele</a>s: red and white.</li>
<li>whether the flower has thorns has 2 <a href="#BOOKMARK245">allele</a>s: true and false.</li>
</ul>
<p>Both parents have 1 red <a href="#BOOKMARK245">allele</a> and 1 white <a href="#BOOKMARK245">allele</a> for color, and 1 true <a href="#BOOKMARK245">allele</a> and 1 false <a href="#BOOKMARK245">allele</a> for thorns. The <a href="#BOOKMARK261">Punnett square</a> describing the odds of the offspring:</p>
<table><tbody><tr><td></td><td> [R][T]  </td><td> [R][F]  </td><td> [W][T]  </td><td> [W][F]</td></tr><tr><td>[R][T]</td><td> [RR][TT]</td><td> [RR][RF]</td><td> [RW][TT]</td><td> [RW][TF]</td></tr><tr><td>[R][F]</td><td> [RR][FT]</td><td> [RR][FF]</td><td> [RW][FT]</td><td> [RW][FF]</td></tr><tr><td>[W][T]</td><td> [WR][TT]</td><td> [WR][TF]</td><td> [WW][TT]</td><td> [WW][TF]</td></tr><tr><td>[W][F]</td><td> [WR][FT]</td><td> [WR][FF]</td><td> [WW][FT]</td><td> [WW][FF]</td></tr></tbody></table><a name="HEADREF55"></a><h3>Laws</h3>
<p><a href="#BOOKMARK254">Classical genetics</a> was started in the 1800s by a scientist monk named <a name="BOOKMARK262"></a><strong>Gregor Mendel</strong> (prior to the discovery of <a href="#BOOKMARK241">DNA</a>) with his model called <a name="BOOKMARK263"></a><strong>Mendelian inheritance</strong>. <a href="#BOOKMARK262">Gregor Mendel</a> wasn't aware of <a href="#BOOKMARK241">DNA</a> and <a href="#BOOKMARK244">genes</a> (hadn't been discovered yet) and instead created his model based on some abstract idea of "inheritable characteristics." <a href="#BOOKMARK254">Classical genetics</a> corrects and extends the model introduced by <a href="#BOOKMARK262">Gregor Mendel</a> by taking <a href="#BOOKMARK241">DNA</a> / <a href="#BOOKMARK138">chromosome</a>s / <a href="#BOOKMARK244">genes</a> / <a href="#BOOKMARK245">allele</a>s / etc.. into account.</p>
<p>Mendel originally had 3 laws:</p>
<ol>
<li><a name="BOOKMARK264"></a><strong>law of segregation</strong></li>
<li><a name="BOOKMARK265"></a><strong>law of independent assortment</strong></li>
<li><a name="BOOKMARK266"></a><strong>law of dominance</strong></li>
</ol>
<p>The subsections below describe these laws in their modern / corrected form.</p>
<a name="HEADREF56"></a><h4>Segregation</h4>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Segregation was originally discovered by <a href="#BOOKMARK262">Gregor Mendel</a> as <a href="#BOOKMARK263">Mendelian inheritance</a>'s first law / <a href="#BOOKMARK264">law of segregation</a>.</p>
</div>
<p>Segregation is the idea every <a href="#BOOKMARK363">sexual</a>ly reproducing organism has 2 <a href="#BOOKMARK245">allele</a>s for each <a href="#BOOKMARK244">gene</a>, but when it comes time to reproduce only 1 <a href="#BOOKMARK245">allele</a> is kept by the reproductive <a href="#BOOKMARK74">cell</a> (<a href="#BOOKMARK176">gamete</a>). As such, the resulting offspring gets 1 <a href="#BOOKMARK245">allele</a> from its father and 1 <a href="#BOOKMARK245">allele</a> from its mother.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Unsure how this extends to organisms that are <a href="#BOOKMARK142">polyploid</a> (have more than 2 <a href="#BOOKMARK139">homologous chromosome</a>s / more than 2 <a href="#BOOKMARK245">allele</a>s for each <a href="#BOOKMARK244">gene</a>).</p>
</div>
<p><img src="/.datafile_ff56c94c6d146245e98048dba20b8f0c/diagram.svg" alt="Graphviz Dot Diagram"></p><p>The reason for this is that <a href="#BOOKMARK363">sexual</a>ly reproducing <a href="#BOOKMARK109">eukaryotic</a> species (e.g. humans) carry pairs of <a href="#BOOKMARK139">homologous chromosome</a>s. That is, <a href="#BOOKMARK138">chromosome</a>s come in pairs where each <a href="#BOOKMARK138">chromosome</a> in the pair has the same set of <a href="#BOOKMARK244">genes</a> as the other but different versions of those <a href="#BOOKMARK244">genes</a> (<a href="#BOOKMARK245">allele</a>s). One <a href="#BOOKMARK138">chromosome</a> comes from the mother and the other from the father.</p>
<p>During <a href="#BOOKMARK173">meiosis</a>, each <a href="#BOOKMARK176">gamete</a> <a href="#BOOKMARK74">cell</a> gets 1 <a href="#BOOKMARK138">chromosome</a> from each <a href="#BOOKMARK356">homologous</a> pair (1 set of <a href="#BOOKMARK245">allele</a>s). When <a href="#BOOKMARK176">gamete</a>s meet to form an offspring, the <a href="#BOOKMARK138">chromosome</a>s from each <a href="#BOOKMARK176">gamete</a> match up to to their <a href="#BOOKMARK356">homolog</a> to create that offspring's set of <a href="#BOOKMARK139">homologous chromosome</a> pairs.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>There are cases where a <a href="#BOOKMARK176">gamete</a> gets 2 <a href="#BOOKMARK138">chromosome</a>s instead of 1. Depending on the <a href="#BOOKMARK138">chromosome</a>, it may end up being fatal or cause disease (or be benign). See <a href="#BOOKMARK147">aneuploid</a> section.</p>
</div>
<a name="HEADREF57"></a><h4>Independent Assortment</h4>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Independent assortment was originally discovered by <a href="#BOOKMARK262">Gregor Mendel</a> as <a href="#BOOKMARK263">Mendelian inheritance</a>'s second law / <a href="#BOOKMARK265">law of independent assortment</a>. It states that inherited attributes are passed down from parent to child independently of each other. For example, the odds that a baby ends up with green eyes isn't linked to the odds that it'll have dark hair. It turns out that this isn't entirely correctly.</p>
</div>
<p>The high-level algorithm for how a parent passes down its genetic information is as follows:</p>
<ol>
<li>
<p>For each <a href="#BOOKMARK139">homologous chromosome</a> pair, randomly swap segments between <a href="#BOOKMARK138">chromosome</a>s (<a href="#BOOKMARK182">chromosomal crossover</a>).</p>
<p><img src="/.datafile_83e7e01a3e8b775123923487d076fae0/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Homologous chromosomes segment swap diagram"></p></li>
<li>
<p>For each <a href="#BOOKMARK139">homologous chromosome</a> pair, randomly pick 1 <a href="#BOOKMARK138">chromosome</a>.</p>
</li>
<li>
<p>The picked <a href="#BOOKMARK138">chromosome</a>s from parent A combine with the picked <a href="#BOOKMARK138">chromosome</a>s from parent B to form the <a href="#BOOKMARK139">homologous chromosome</a> pairs of the new child.</p>
<p><img src="/.datafile_be39f58f53950f69fb7599f6a3d0fd3b/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Diagram showing a random chromosome being selected (after segment swap) from each parent's homologous chromosome pair to make up the child's homologous chromosome pair."></p></li>
</ol>
<p>Essentially, what the above says / shows is that <a href="#BOOKMARK245">allele</a>s get passed down from parent to child in groups, where each group is a segment of one of the <a href="#BOOKMARK138">chromosome</a>s. In the example above, the segments passed down are...</p>
<ul>
<li>CHILD_C1 = PARENT1_C2[1,3] + PARENT1_C1[4,28]</li>
<li>CHILD_C2 = PARENT2_C2[1,5] + PARENT2_C1[6,20] + PARENT2_C2[21,28]</li>
</ul>
<p><a href="#BOOKMARK244">Genes</a> located close to each other are more likely to get grouped together as part of the same segment being passed down -- it's very unlikely that a <a href="#BOOKMARK182">chromosomal crossover</a> boundary will be at a point which splits them. These <a href="#BOOKMARK244">genes</a> are referred to as <a name="BOOKMARK267"></a><strong>linked gene</strong>s. The <a href="#BOOKMARK256">phenotype</a>s they express are often referred to as <a name="BOOKMARK268"></a><strong>linked trait</strong>s. For example, if diagram above were for a plant, and <a href="#BOOKMARK244">gene</a> 11 controlled height while <a href="#BOOKMARK244">gene</a> 12 controlled color, there would likely be a correlation between the height and color across that plant species.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>See section on <a href="#BOOKMARK173">meiosis</a> for more in depth description of how this steps happen.</p>
</div>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Unsure how this extends to organisms that are <a href="#BOOKMARK142">polyploid</a> (have more than 2 <a href="#BOOKMARK139">homologous chromosome</a>s / more than 2 <a href="#BOOKMARK245">allele</a>s for each <a href="#BOOKMARK244">gene</a>).</p>
</div>
<a name="HEADREF58"></a><h4>Dominance</h4>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Dominance was originally discovered by <a href="#BOOKMARK262">Gregor Mendel</a> as <a href="#BOOKMARK263">Mendelian inheritance</a>'s third law / <a href="#BOOKMARK266">law of dominance</a>. It states in a cross of parents that are pure for contrasting traits, only one form of the trait will appear in the next generation. Offspring that are hybrid for a trait will have only the <a href="#BOOKMARK270">dominant trait</a> in the <a href="#BOOKMARK256">phenotype</a> while the <a href="#BOOKMARK272">recessive trait</a> remains dormant. It turns out that this isn't entirely correct (it's close) -- more scenarios have come up: <a href="#BOOKMARK274">incomplete dominance</a>, <a href="#BOOKMARK275">co-dominance</a>, etc..</p>
</div>
<p>In an organism that is <a href="#BOOKMARK260">heterozygous</a> for some <a href="#BOOKMARK244">gene</a>, some <a href="#BOOKMARK245">allele</a>s may take precedence in expressing themselves over others. How these <a href="#BOOKMARK245">allele</a>s are expressed define the type of dominance they have over other <a href="#BOOKMARK245">allele</a>s.</p>
<p><a href="#BOOKMARK245">Allele</a>s that ...</p>
<ul>
<li>take precedence over others are referred to as <a name="BOOKMARK269"></a><strong>dominant allele</strong>s. The <a href="#BOOKMARK256">phenotype</a>s they express are often referred to as <a name="BOOKMARK270"></a><strong>dominant trait</strong>s.</li>
<li>surrender precedence to others are referred to as <a name="BOOKMARK271"></a><strong>recessive allele</strong>s. The <a href="#BOOKMARK256">phenotype</a>s they express are often referred to as <a name="BOOKMARK272"></a><strong>recessive trait</strong>s.</li>
</ul>
<p>For example, given the <a href="#BOOKMARK269">dominant allele</a> D and the <a href="#BOOKMARK271">recessive allele</a> d, ...</p>
<ul>
<li>[Dd] results in the expression of the <a href="#BOOKMARK256">phenotype</a> for D -- D takes precedence / d surrenders precedence.</li>
<li>[DD] results in the expression of the <a href="#BOOKMARK256">phenotype</a> for D -- no other types to contend with.</li>
<li>[dd] results in the expression of the <a href="#BOOKMARK256">phenotype</a> for d -- no other types to surrender to.</li>
</ul>
<p>In cases where both <a href="#BOOKMARK245">allele</a>s are <a href="#BOOKMARK269">dominant allele</a>s, there's several different ways <a href="#BOOKMARK256">phenotype</a> may be expressed. For example, given the <a href="#BOOKMARK269">dominant allele</a>s A and B, if the <a href="#BOOKMARK256">phenotype</a> expressed by ...</p>
<ul>
<li>[AA] is the same as [AB]/[BA] but not [BB], A is said to have <a name="BOOKMARK273"></a><strong>complete dominance</strong> -- one <a href="#BOOKMARK245">allele</a>'s <a href="#BOOKMARK256">phenotype</a> is expressed while others are hidden.</li>
<li>[AB]/[BA] is a blend of [AA] and [BB], it's said to have <a name="BOOKMARK274"></a><strong>incomplete dominance</strong> -- more than one <a href="#BOOKMARK245">allele</a> is dominant, resulting in a blended <a href="#BOOKMARK256">phenotype</a> expression.</li>
<li>[AB]/[BA] is the <a href="#BOOKMARK256">phenotype</a> expressed by both [AA] and [BB], it's said to have <a name="BOOKMARK275"></a><strong>co-dominance</strong> -- more than one <a href="#BOOKMARK245">allele</a> is dominant, resulting in a both <a href="#BOOKMARK256">phenotype</a>s being expressed.</li>
</ul>
<p>For example, in the diagram below the offspring ends up with red color <a href="#BOOKMARK245">allele</a> and a white color <a href="#BOOKMARK245">allele</a>.</p>
<p><img src="/.datafile_84433433b2be815fdf9e3a134d3250fc/diagram.svg" alt="Graphviz Dot Diagram"></p><ul>
<li>
<p>In a <a href="#BOOKMARK273">complete dominance</a> scenario, the child's color may end up being totally red because the red <a href="#BOOKMARK245">allele</a> is the only one that gets expressed.</p>
<p><img src="/.datafile_3213808c1a1c16708e310ae8e3d09380/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Diagram showing a red flower offspring."></p></li>
<li>
<p>In a <a href="#BOOKMARK274">incomplete dominance</a> scenario, the child's color may end up being pink because both the red and the white <a href="#BOOKMARK245">allele</a> express together and blend.</p>
<p><img src="/.datafile_45eeaf8a17fbeeb094c44dbe68db4095/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Diagram showing a pink flower offspring."></p></li>
<li>
<p>In a <a href="#BOOKMARK275">co-dominance</a> scenario, the child may end up having blotches of red / white as its colors because both the red and the white <a href="#BOOKMARK245">allele</a> express but are discrete.</p>
<p><img src="/.datafile_071a8151a7cc865d1d1c45e62186013c/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Diagram showing a blotchy flower offspring."></p><div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Another likely scenario for <a href="#BOOKMARK275">co-dominance</a> is that some petals will be red while others are white.</p>
</div>
</li>
</ul>
<p>Rarely ever is there only 2 <a href="#BOOKMARK245">allele</a>s for a <a href="#BOOKMARK244">gene</a>. A real-<a href="#BOOKMARK38">life</a> example of dominance is blood type. 3 <a href="#BOOKMARK245">allele</a>s are present: A, B, and O. A and B have <a href="#BOOKMARK274">incomplete dominance</a> while O is recessive. If the <a href="#BOOKMARK245">allele</a>s for a person's blood type is...</p>
<ul>
<li>[AB]/[BA], their blood type (<a href="#BOOKMARK256">phenotype</a>) is AB.</li>
<li>[AA], their blood type (<a href="#BOOKMARK256">phenotype</a>) is A.</li>
<li>[BB], their blood type (<a href="#BOOKMARK256">phenotype</a>) is B.</li>
<li>[AO]/[OA], their blood type (<a href="#BOOKMARK256">phenotype</a>) is A.</li>
<li>[BO]/[OB], their blood type (<a href="#BOOKMARK256">phenotype</a>) is B.</li>
<li>[OO], their blood type (<a href="#BOOKMARK256">phenotype</a>) is O.</li>
</ul>
<p>If the <a href="#BOOKMARK245">allele</a>s present for a <a href="#BOOKMARK244">gene</a> are...</p>
<ul>
<li>are both dominant, it's said to be <a name="BOOKMARK276"></a><strong>homozygous dominant</strong>. In the example above, [AB]/[BA]/[AA]/[BB] are all <a href="#BOOKMARK276">homozygous dominant</a> (both <a href="#BOOKMARK245">allele</a>s are <a href="#BOOKMARK269">dominant allele</a>s).</li>
<li>are both recessive, it's said to be <a name="BOOKMARK277"></a><strong>homozygous recessive</strong>. In the example above, [OO] is <a href="#BOOKMARK260">heterozygous</a> recessive (both <a href="#BOOKMARK245">allele</a>s are <a href="#BOOKMARK271">recessive allele</a>s).</li>
</ul>
<a name="HEADREF59"></a><h3>Lethality</h3>
<p><a name="BOOKMARK278"></a><strong>Lethal allele</strong>s are <a href="#BOOKMARK245">allele</a>s that cause the death of an organism that carries them. These <a href="#BOOKMARK245">allele</a>s may cause death during development (e.g. gestation in the womb) or possibly after having fully matured.</p>
<p><a href="#BOOKMARK278">Lethal allele</a>s may fall into one of many categories. A ...</p>
<ul>
<li>
<p><a name="BOOKMARK279"></a><strong>recessive lethal allele</strong> is a condition where 2 of 2 <a href="#BOOKMARK245">allele</a>s for a <a href="#BOOKMARK244">gene</a> need to be the <a href="#BOOKMARK278">lethal allele</a> for it to cause death (needs to be <a href="#BOOKMARK259">homozygous</a>). For example, Tay-Sachs disease appears and leads to death only if 2 copies of the <a href="#BOOKMARK278">lethal allele</a> are present.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>If only 1 copy of a <a href="#BOOKMARK279">recessive lethal</a> <a href="#BOOKMARK245">allele</a> is present, the organism doesn't die but may live on in a diseased form.</p>
</div>
</li>
<li>
<p><a name="BOOKMARK280"></a><strong>dominant lethal allele</strong> is a condition where only 1 of 2 <a href="#BOOKMARK245">allele</a>s for a <a href="#BOOKMARK244">gene</a> needs to be the <a href="#BOOKMARK278">lethal allele</a> for it to cause death (can be <a href="#BOOKMARK260">heterozygous</a> or <a href="#BOOKMARK259">homozygous</a>). For example, Huntington's disease leads to death and only requires a single <a href="#BOOKMARK245">allele</a> copy of an <a href="#BOOKMARK245">allele</a> to be present.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>If 2 copies of a <a href="#BOOKMARK280">dominant lethal</a> <a href="#BOOKMARK245">allele</a> are present, the organism still dies.</p>
</div>
</li>
<li>
<p><a name="BOOKMARK281"></a><strong>conditional lethal allele</strong> is a condition where an <a href="#BOOKMARK245">allele</a> can only be fatal in response to some environmental factor. For example, favism is a disease that causes the organism carrying it to die when fava beans are consumed.</p>
</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Don't get confused. <a href="#BOOKMARK280">Dominant lethal</a>s / <a href="#BOOKMARK279">recessive lethal</a>s have nothing to do with <a href="#BOOKMARK269">dominant allele</a>s / <a href="#BOOKMARK271">recessive allele</a>s. The only thing they state is how many copies of the <a href="#BOOKMARK245">allele</a> are needed for it to be lethal (1 / <a href="#BOOKMARK260">heterozygous</a> or 2 / <a href="#BOOKMARK259">homozygous</a>). Whether that <a href="#BOOKMARK278">lethal allele</a> is a <a href="#BOOKMARK269">dominant allele</a> / <a href="#BOOKMARK271">recessive allele</a> isn't a requirement.</p>
</div>
<a name="HEADREF60"></a><h3>Expressivity</h3>
<p><a name="BOOKMARK282"></a><strong>Expressivity</strong> is the concept that, even if a known <a href="#BOOKMARK255">genotype</a> is responsible for a <a href="#BOOKMARK256">phenotype</a>, organisms having that <a href="#BOOKMARK255">genotype</a> will show a <a href="#BOOKMARK284">variable expression</a> of that <a href="#BOOKMARK256">phenotype</a>. For example, in certain genetic disorders, the same <a href="#BOOKMARK255">genotype</a> may result in stronger or weaker forms of that disorder (maybe even to the point of not developing at all).</p>
<ul>
<li>
<p><a name="BOOKMARK283"></a><strong>Narrow expressivity</strong> refers to low variability in the expression of the <a href="#BOOKMARK256">phenotype</a>. That is, the <a href="#BOOKMARK256">phenotype</a> is likely to be expressed to the same degrees across organisms.</p>
<p>Example of expression rates for <a href="#BOOKMARK283">narrow expressivity</a>:</p>
<table><tbody><tr><td>0.99</td><td> 0.95</td><td> 0.98</td><td> 0.99</td></tr><tr><td>0.91</td><td> 0.94</td><td> 0.95</td><td> 0.93</td></tr></tbody></table></li>
<li>
<p><a name="BOOKMARK284"></a><strong>Variable expressivity</strong> refers to high variability in the expression of the <a href="#BOOKMARK256">phenotype</a>. That is, the <a href="#BOOKMARK256">phenotype</a> is likely to be expressed to different degrees across organisms.</p>
<p>Example of expression rates for <a href="#BOOKMARK284">variable expressivity</a>:</p>
<table><tbody><tr><td>0.21</td><td> 0.55</td><td> 0.19</td><td> 0.78</td></tr><tr><td>0.91</td><td> 0.43</td><td> 0.67</td><td> 0.31</td></tr></tbody></table></li>
</ul>
<p><a name="BOOKMARK285"></a><strong>Penetrance</strong> is the concept that, even if a known <a href="#BOOKMARK255">genotype</a> is responsible for a <a href="#BOOKMARK256">phenotype</a>, organisms having that <a href="#BOOKMARK255">genotype</a> may not express the related <a href="#BOOKMARK256">phenotype</a>. or example, in certain genetic disorders, the same <a href="#BOOKMARK255">genotype</a> may not result in the disorder.</p>
<ul>
<li>
<p><a name="BOOKMARK286"></a><strong>Complete penetrance</strong> refers to the <a href="#BOOKMARK256">phenotype</a> always being expressed when the <a href="#BOOKMARK255">genotype</a> is present.</p>
<p>Example of expression rates for <a href="#BOOKMARK286">complete penetrance</a>:</p>
<table><tbody><tr><td>1.00</td><td> 1.00</td><td> 1.00</td><td> 1.00</td></tr><tr><td>1.00</td><td> 1.00</td><td> 1.00</td><td> 1.00</td></tr></tbody></table></li>
<li>
<p><a name="BOOKMARK287"></a><strong>Incomplete penetrance</strong> refers to the <a href="#BOOKMARK256">phenotype</a> maybe being expressed when the <a href="#BOOKMARK255">genotype</a> is present.</p>
<p>Example of expression rates for <a href="#BOOKMARK287">incomplete penetrance</a>:</p>
<table><tbody><tr><td>1.00</td><td> 0.00</td><td> 0.00</td><td> 1.00</td></tr><tr><td>1.00</td><td> 1.00</td><td> 1.00</td><td> 0.00</td></tr></tbody></table></li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>It was never really expressed why this happens. My only guess is environmental factors or maybe some type of intrinsic built-in genetic randomness (e.g. not enough of the specific <a href="#BOOKMARK245">allele</a> for a disease randomly collided with whatever other <a href="#BOOKMARK18">molecule</a>(s) were required to express the <a href="#BOOKMARK256">phenotype</a>).</p>
</div>
<a name="HEADREF61"></a><h3>Sex-linkage</h3>
<p>In humans / mammals, sex is determined by the <a href="#BOOKMARK144">XY chromosome</a>s. In...</p>
<ul>
<li>females, the <a href="#BOOKMARK144">XY chromosome</a> pair is a <a href="#BOOKMARK139">homologous chromosome</a> pair -- X and X <a href="#BOOKMARK138">chromosome</a>s.</li>
<li>males, the <a href="#BOOKMARK144">XY chromosome</a> pair is NOT a <a href="#BOOKMARK139">homologous chromosome</a> pair -- X and Y <a href="#BOOKMARK138">chromosome</a>s.</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>How different are the X and Y <a href="#BOOKMARK138">chromosome</a>s? Each codes for a completely different set of <a href="#BOOKMARK244">genes</a>.</p>
<ul>
<li>X <a href="#BOOKMARK138">chromosome</a> codes for ~1500 <a href="#BOOKMARK244">genes</a>.</li>
<li>Y <a href="#BOOKMARK138">chromosome</a> codes for ~78 <a href="#BOOKMARK244">genes</a>.</li>
</ul>
<p>For example, the SRY <a href="#BOOKMARK244">gene</a> contributes to development of testicles and is only found on the Y <a href="#BOOKMARK138">chromosome</a>.</p>
</div>
<p><a href="#BOOKMARK244">Genes</a> on the <a href="#BOOKMARK144">XY chromosome</a>s are called <a name="BOOKMARK288"></a><strong>sex-linked gene</strong>s.</p>
<a name="HEADREF62"></a><h4>X-linkage</h4>
<p>Since males only have a single copy of both X and Y <a href="#BOOKMARK138">chromosome</a>s, standard <a href="#BOOKMARK269">dominant allele</a> / <a href="#BOOKMARK271">recessive allele</a> rules DO NOT apply to these <a href="#BOOKMARK138">chromosome</a>s in males. For example, regardless of if an <a href="#BOOKMARK245">allele</a> on the X <a href="#BOOKMARK138">chromosome</a> is a <a href="#BOOKMARK269">dominant allele</a> or <a href="#BOOKMARK271">recessive allele</a>, it will always express in males. There is no second X <a href="#BOOKMARK138">chromosome</a> to provide a second <a href="#BOOKMARK245">allele</a> for that <a href="#BOOKMARK244">gene</a>.</p>
<p><a name="BOOKMARK289"></a><strong></strong>Since a human / mammal must have at least a single X <a href="#BOOKMARK138">chromosome</a>, ...</p>
<ul>
<li><a href="#BOOKMARK271">recessive allele</a>s on the X <a href="#BOOKMARK138">chromosome</a> are called <a name="BOOKMARK290"></a><strong>X-linked recessive allele</strong>s.</li>
<li><a href="#BOOKMARK272">recessive trait</a>s caused by <a href="#BOOKMARK245">allele</a>s on the X <a href="#BOOKMARK138">chromosome</a> are called <a name="BOOKMARK291"></a><strong>X-linked recessive trait</strong>s.</li>
</ul>
<p>For example, hemophilia is caused by an <a href="#BOOKMARK290">x-linked recessive allele</a>. If there's a 2nd X <a href="#BOOKMARK138">chromosome</a> that has a <a href="#BOOKMARK269">dominant allele</a> for that <a href="#BOOKMARK244">gene</a>, it'll suppress the hemophilia <a href="#BOOKMARK245">allele</a> from expressing. For...</p>
<ul>
<li>females (X and X), there IS a 2nd X <a href="#BOOKMARK138">chromosome</a>. If that 2nd X <a href="#BOOKMARK138">chromosome</a> contains the same recessive hemophilia <a href="#BOOKMARK245">allele</a> (<a href="#BOOKMARK259">homozygous</a>), the hemophilia <a href="#BOOKMARK256">phenotype</a> will express.</li>
<li>males (X and Y), there ISN'T a 2nd X <a href="#BOOKMARK138">chromosome</a>. If the <a href="#BOOKMARK245">allele</a> on the single X <a href="#BOOKMARK138">chromosome</a> is the hemophilia <a href="#BOOKMARK245">allele</a>, the hemophilia <a href="#BOOKMARK256">phenotype</a> will express.</li>
</ul>
<p>If the probability of having the hemophilia <a href="#BOOKMARK245">allele</a> on one X <a href="#BOOKMARK138">chromosome</a> were <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mn>1</mn><mn>7000</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{1}{7000}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">7</span><span class="mord mtight">0</span><span class="mord mtight">0</span><span class="mord mtight">0</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>, the probability of expressing the hemophilia <a href="#BOOKMARK256">phenotype</a> / trait for...</p>
<ul>
<li>
<p>females would be <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mn>1</mn><mn>7000</mn></mfrac><mo>⋅</mo><mfrac><mn>1</mn><mn>7000</mn></mfrac><mo>=</mo><mfrac><mn>1</mn><mn>49000000</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{1}{7000} \cdot \frac{1}{7000} = \frac{1}{49000000}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">7</span><span class="mord mtight">0</span><span class="mord mtight">0</span><span class="mord mtight">0</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">⋅</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">7</span><span class="mord mtight">0</span><span class="mord mtight">0</span><span class="mord mtight">0</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">4</span><span class="mord mtight">9</span><span class="mord mtight">0</span><span class="mord mtight">0</span><span class="mord mtight">0</span><span class="mord mtight">0</span><span class="mord mtight">0</span><span class="mord mtight">0</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Since there's 2 X <a href="#BOOKMARK138">chromosome</a>s and it's a <a href="#BOOKMARK271">recessive allele</a>, both <a href="#BOOKMARK138">chromosome</a>s need that <a href="#BOOKMARK245">allele</a> (<a href="#BOOKMARK259">homozygous</a>).</p>
</div>
</li>
<li>
<p>males would be <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mn>1</mn><mn>7000</mn></mfrac></mrow><annotation encoding="application/x-tex">\frac{1}{7000}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.190108em;vertical-align:-0.345em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.845108em;"><span style="top:-2.6550000000000002em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">7</span><span class="mord mtight">0</span><span class="mord mtight">0</span><span class="mord mtight">0</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span>.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Since there's only 1 X <a href="#BOOKMARK138">chromosome</a>, only that <a href="#BOOKMARK138">chromosome</a> needs the <a href="#BOOKMARK245">allele</a>.</p>
</div>
</li>
</ul>
<p>As such, <a href="#BOOKMARK291">X-linked recessive trait</a>s are much more common in men than they are women -- women need 2 copies of the <a href="#BOOKMARK271">recessive allele</a>(s) while men only need 1.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>To find out the exact mechanism around <a href="#BOOKMARK291">X-linked recessive trait</a>s, see the section on <a href="#BOOKMARK292">x-linked inactivation</a> below.</p>
</div>
<a name="HEADREF63"></a><h4>X-inactivation</h4>
<p>The reason why males can function with only 1 copy of an X <a href="#BOOKMARK138">chromosome</a> is because of <a name="BOOKMARK292"></a><strong>X-linked inactivation</strong>. A single X <a href="#BOOKMARK138">chromosome</a> is all that's needed for a human / certain mammals to operate normally. If more than 1 copy of an X <a href="#BOOKMARK138">chromosome</a> is present in a <a href="#BOOKMARK74">cell</a>, the <a href="#BOOKMARK74">cell</a> chooses 1 at random to use while the others are made inactive. Inactive X <a href="#BOOKMARK138">chromosome</a>s are compacted into small <a href="#BOOKMARK342">dense</a> structures called <a name="BOOKMARK293"></a><strong>Barr bodies</strong> that prevent most of their <a href="#BOOKMARK244">genes</a> from expressing.</p>
<p>The above applies to both females (XX) and males/females with cases of allosomal <a href="#BOOKMARK147">aneuploid</a>y (e.g. XXX or XXXY).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>How is it possible for there to be more than 2 X <a href="#BOOKMARK138">chromosome</a>s? For example, the XX pair for a female may fail to separate during <a href="#BOOKMARK173">meiosis</a>, meaning one of the <a href="#BOOKMARK179">egg</a>s will contain 2 X <a href="#BOOKMARK138">chromosome</a>s while the other will contain 0 X <a href="#BOOKMARK138">chromosome</a>s.</p>
<p>Because of <a href="#BOOKMARK292">x-linked inactivation</a>, allosomal <a href="#BOOKMARK147">aneuploid</a>y isn't fatal but may lead to mild forms of disease. For example, Klinefelter syndrome (XXY) leads to infertility and may lead to learning disabilities / low testosterone.</p>
</div>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The process of <a href="#BOOKMARK292">X-linked inactivation</a> is called <a name="BOOKMARK294"></a><strong>lyonization</strong>.</p>
</div>
<p>How is it that <a href="#BOOKMARK291">x-linked recessive trait</a>s happen if only 1 X <a href="#BOOKMARK138">chromosome</a> is ever active? Aren't 2 copies of the <a href="#BOOKMARK271">recessive allele</a> needed for the <a href="#BOOKMARK256">phenotype</a> to be expressed? It turns out that if some <a href="#BOOKMARK74">cell</a>s choose the X <a href="#BOOKMARK138">chromosome</a> that contains the non <a href="#BOOKMARK271">recessive allele</a>, those <a href="#BOOKMARK74">cell</a>s overpower the <a href="#BOOKMARK74">cell</a>s with the <a href="#BOOKMARK271">recessive allele</a>.</p>
<p>For example, color blindness is an <a href="#BOOKMARK291">x-linked recessive trait</a>. All copies of the X <a href="#BOOKMARK138">chromosome</a> have to have the color-blind <a href="#BOOKMARK245">allele</a> for human to be color blind. If one of the X <a href="#BOOKMARK138">chromosome</a>s doesn't, some of the eye <a href="#BOOKMARK74">cell</a>s will be able to see color. The <a href="#BOOKMARK74">cell</a>s that do see color will relay color information back to the brain (the person will see color).</p>
<a name="HEADREF64"></a><h4>Punnett Square Examples</h4>
<p>The following are examples showing the probability that offspring end up with the recessive <a href="#BOOKMARK289">X-linked disease</a> known as hemophilia.</p>
<hr>
<p>Given a hemophiliac mother and a hemophiliac father, the odds that ...</p>
<table><tbody><tr><td></td><td> H (X chromosome)</td><td> N/A (Y chromosome)</td></tr><tr><td>H (X chromosome)</td><td> H (X) H (X) ✓   </td><td> H (X) N/A (Y) ✓</td></tr><tr><td>H (X chromosome)</td><td> H (X) H (X) ✓   </td><td> H (X) N/A (Y) ✓</td></tr></tbody></table><ul>
<li>the child will have hemophilia are 100%.</li>
</ul>
<hr>
<p>Given a carrier mother and a hemophiliac father, the odds that ...</p>
<table><tbody><tr><td></td><td> H (X chromosome)</td><td> N/A (Y chromosome)</td></tr><tr><td>H (X chromosome)</td><td> H (X) H (X) ✓   </td><td> H (X) N/A (Y) ✓</td></tr><tr><td>h (X chromosome)</td><td> h (X) H (X)     </td><td> h (X) N/A (Y)</td></tr></tbody></table><ul>
<li>
<p>the child will be a female with hemophilia is 25%.</p>
</li>
<li>
<p>the child will be a male with hemophilia is 25%.</p>
</li>
<li>
<p>a daughter will have hemophilia is 50%.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>In other words, out of all possible female children (XX) what are the chances of hemophilia. <a href="#BOOKMARK261">Punnett square</a> below shows 2 female children, only one of which has HH (2 hemophilia <a href="#BOOKMARK245">allele</a>s).</p>
</div>
</li>
<li>
<p>a son will have hemophilia is 50%.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>In other words, out of all possible male children (XY) what are the chances of hemophilia. <a href="#BOOKMARK261">Punnett square</a> below shows 2 male children, only one of which has HH (2 hemophilia <a href="#BOOKMARK245">allele</a>s).</p>
</div>
</li>
</ul>
<hr>
<p>Given a carrier mother and a non-hemophiliac father, the odds that ...</p>
<table><tbody><tr><td></td><td> h (X chromosome)</td><td> N/A (Y chromosome)</td></tr><tr><td>H (X chromosome)</td><td> H (X) h (X)     </td><td> H (X) N/A (Y) ✓</td></tr><tr><td>h (X chromosome)</td><td> h (X) H (X)     </td><td> h (X) N/A (Y)</td></tr></tbody></table><ul>
<li>
<p>the child will be a female with hemophilia is 0%.</p>
</li>
<li>
<p>the child will be a male with hemophilia is 25%.</p>
</li>
<li>
<p>a daughter will have hemophilia is 0%.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>In other words, out of all possible female children (XX) what are the chances of hemophilia. <a href="#BOOKMARK261">Punnett square</a> below shows 2 female children, none of which has HH (2 hemophilia <a href="#BOOKMARK245">allele</a>s).</p>
</div>
</li>
<li>
<p>a son will have hemophilia is 50%.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>In other words, out of all possible male children (XY) what are the chances of hemophilia. <a href="#BOOKMARK261">Punnett square</a> below shows 2 male children, only one of which has H (1 hemophilia <a href="#BOOKMARK245">allele</a>s).</p>
</div>
</li>
</ul>
<a name="HEADREF65"></a><h3>Pedigree</h3>
<p>A <a name="BOOKMARK295"></a><strong>pedigree</strong> chart is a hierarchy diagram that shows the appearance of a specific <a href="#BOOKMARK256">phenotype</a> in an organism and its ancestors.</p>
<p><img src="/.datafile_16a57fae658c443a28e268e0eb962d3f/data.svg" alt="Example pedigree chart." title="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License."></p><p>Each individual in the diagram is identified by a shape:</p>
<ul>
<li>Squares represent males.</li>
<li>Circles represent females.</li>
</ul>
<p>The male (square) always comes first in a mating couple.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Is a different shape necessary? The species that <a href="#BOOKMARK295">pedigree</a> charts are used for (dogs, horses, humans, etc..) always have 2 sexes: male and female. Just starting off each mating couple with the same sex is enough to convey sex.</p>
</div>
<p><img src="/.datafile_db34ae969ea69fc4abf4628550588b75/data.svg" alt="Example pedigree chart with explicitly labeled sexes." title="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License."></p><p>If the shape is filled in, it means that the specific <a href="#BOOKMARK256">phenotype</a> was expressed in that organism.</p>
<p><img src="/.datafile_5bec3fa980e83060bcc08d1c4a3d6250/data.svg" alt="Example pedigree chart with explicitly labeled sexes and phenotype expression." title="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License."></p><p>For each mating pair, the ...</p>
<ul>
<li>horizontal line connecting a male and a female is called a <a name="BOOKMARK296"></a><strong>marriage line</strong>.</li>
<li>vertical line that goes down from the <a href="#BOOKMARK296">marriage line</a> is known as the <a name="BOOKMARK297"></a><strong>line of descent</strong>.</li>
<li>horizontal line connected to the bottom of that <a href="#BOOKMARK297">line of descent</a> is known as a <a name="BOOKMARK298"></a><strong>sibling line</strong>.</li>
</ul>
<p><img src="/.datafile_078ba15f5d52a5a021099a247e80837d/data.svg" alt="Example pedigree chart close-up with highlighted marriage line, line of descent, and sibling line." title="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License."></p><p><a href="#BOOKMARK295">Pedigree</a> charts can be used to infer / deduce the odds that a child will end up with the specific <a href="#BOOKMARK256">phenotype</a> that the chart is for. For example, if the <a href="#BOOKMARK256">phenotype</a> is caused by a single <a href="#BOOKMARK245">allele</a> (<a href="#BOOKMARK257">pleiotropic</a>) and you know that the <a href="#BOOKMARK245">allele</a> needs to be <a href="#BOOKMARK277">homozygous recessive</a> for the <a href="#BOOKMARK256">phenotype</a> to be expressed (<a href="#BOOKMARK271">recessive allele</a>), you can infer the <a href="#BOOKMARK255">genotype</a>s of the parents...</p>
<p><img src="/.datafile_88c07dab787ff8897aaa308ed25c8d75/data.svg" alt="Example pedigree chart mating pair close-up genotype possibilities." title="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License."></p><div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Since it's a <a href="#BOOKMARK271">recessive allele</a>, both <a href="#BOOKMARK245">allele</a>s need to be that <a href="#BOOKMARK271">recessive allele</a> for the <a href="#BOOKMARK256">phenotype</a> to express. That's why we know the female is [tt] (expresses <a href="#BOOKMARK256">phenotype</a>) and the male is either [TT] or [Tt] (doesn't express <a href="#BOOKMARK256">phenotype</a>).</p>
</div>
<p>From there you can use <a href="#BOOKMARK261">Punnett square</a>s to visualize child <a href="#BOOKMARK256">phenotype</a> outcomes in the different scenarios...</p>
<table><tbody><tr><td></td><td> T </td><td> T</td></tr><tr><td>t</td><td> tT</td><td> tT</td></tr><tr><td>t</td><td> tT</td><td> tT</td></tr></tbody></table><p>0% chance that the child ends up with [tt] in this scenario, meaning the <a href="#BOOKMARK256">phenotype</a> will never express in any of the children.</p>
<table><tbody><tr><td></td><td> T </td><td> t</td></tr><tr><td>t</td><td> tT</td><td> tt ✓</td></tr><tr><td>t</td><td> tT</td><td> tt ✓</td></tr></tbody></table><p>50% chance that the child ends up with [tt] in this scenario, meaning the <a href="#BOOKMARK256">phenotype</a> has a 50/50 chance of showing up in each child.</p>
<p>In the example above, if the 2 parents already had a few children and those children were present in the <a href="#BOOKMARK295">pedigree</a> chart, you may be able to narrow the male down to just one of the <a href="#BOOKMARK255">genotype</a> possibilities...</p>
<p><img src="/.datafile_ca1c17589b1bfcc0e1f162b9726c6bdc/data.svg" alt="Example pedigree chart mating pair and children close-up genotype possibilities." title="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License."></p><p>Since we can see children exhibiting the <a href="#BOOKMARK256">phenotype</a>, we know that that the [TT] <a href="#BOOKMARK255">genotype</a> of the male can't be the case. If it were, no children would end up with the <a href="#BOOKMARK256">phenotype</a>. The only other option is the [Tt] <a href="#BOOKMARK255">genotype</a>.</p>
<a name="HEADREF66"></a><h2>Adenosine Triphosphate</h2>
<p><a name="BOOKMARK299"></a><strong></strong> <a name="BOOKMARK300"></a><strong>Adenosine Triphosphate</strong> (<a href="#BOOKMARK299">ATP</a>) is a <a href="#BOOKMARK18">molecule</a> that provides energy to drive various <a href="#BOOKMARK37">biological</a> processes (e.g. muscle contractions). The third phosphoral group at the very end has a high-energy bond. When broken, energy is released and the resulting <a href="#BOOKMARK18">molecule</a>s are the broken up phosphoral group and <a name="BOOKMARK301"></a><strong></strong> <a name="BOOKMARK302"></a><strong>Adenosine Diphosphate</strong> (<a href="#BOOKMARK301">ADP</a>).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>High-energy bonds are actually a thing: A <a href="#BOOKMARK0">chemical</a> bond whose hydrolysis results in the generation of 30kJ (7kcal) of energy or, if coupled to an energetically unfavourable reaction, can drive that reaction forward. (https://www.genscript.com/<a href="#BOOKMARK16">mole</a>cular-<a href="#BOOKMARK37">biology</a>-glossary/1364/high-energy-bond)</p>
</div>
<p><img src="/.datafile_a3ea0e1ff55e5b118739d2f8eff34c54/data.svg" alt="By Muessig - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=27614630" title="ADP ATP cycle"></p><p><a href="#BOOKMARK299">ATP</a> is produced in the <a href="#BOOKMARK117">mitochondria</a>. Similar to how the <a href="#BOOKMARK117">mitochondria</a> is referred to as the powerhouse of the <a href="#BOOKMARK74">cell</a>, <a href="#BOOKMARK299">ATP</a> is often referred to as the energy currency of the <a href="#BOOKMARK74">cell</a> / energy store for the <a href="#BOOKMARK74">cell</a>.</p>
<a name="HEADREF67"></a><h2>Transport</h2>
<p>There are 2 different types of mechanism used to transport <a href="#BOOKMARK18">molecule</a>s in and out of a <a href="#BOOKMARK74">cell</a>: <a href="#BOOKMARK303">passive transport</a> and <a href="#BOOKMARK307">active transport</a>.</p>
<p><a name="BOOKMARK303"></a><strong>Passive transport</strong> is when <a href="#BOOKMARK18">molecule</a>s naturally move towards the gradient. In this context, gradient refers to the natural direction in which things are supposed to go -- no explicit energy is needed to move/push it along, it just moves in that direction by virtue of some implicit property.</p>
<ul>
<li>
<p>A <a name="BOOKMARK304"></a><strong>concentration gradient</strong> is when the concentration of a <a href="#BOOKMARK18">molecule</a> evens out in a volume just by virtue of the <a href="#BOOKMARK18">molecule</a>s randomly bouncing around (<a href="#BOOKMARK346">diffusion</a>). For example, gas pumped into a vacuum will end up filling the vacuum evenly (more-or-less) -- traveling from areas of high concentration to areas of low concentration.</p>
<p><img src="/.datafile_468034a584a29c4b20b460b435da0e6b/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Concentration Gradient"></p><div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>All this is saying is that if you have a group of items clustered in a single area, each traveling at a different speed / direction, will spread out over time.</p>
</div>
</li>
<li>
<p>An <a name="BOOKMARK305"></a><strong>electrical gradient</strong> is when <a href="#BOOKMARK18">molecule</a>s flow in some direction because their electrical charges attract. For example, a <a href="#BOOKMARK29">negatively charged</a> <a href="#BOOKMARK18">molecule</a> will gravitate towards a <a href="#BOOKMARK27">positively charged</a> <a href="#BOOKMARK18">molecule</a>. Similarly, a negative <a href="#BOOKMARK18">molecule</a> will gravitate away from from other negative <a href="#BOOKMARK18">molecule</a>s / a positive <a href="#BOOKMARK18">molecule</a> will gravitate away from other positive <a href="#BOOKMARK18">molecule</a>s.</p>
<p><img src="/.datafile_15af1c88025984e607011c2a9a42fd2a/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Electrical Gradient"></p></li>
<li>
<p>An <a name="BOOKMARK306"></a><strong>electrochemical gradient</strong> is a combination of both a <a href="#BOOKMARK304">concentration gradient</a> and an <a href="#BOOKMARK305">electrical gradient</a>.</p>
</li>
</ul>
<p><a name="BOOKMARK307"></a><strong>Active transport</strong> is when <a href="#BOOKMARK18">molecule</a>s use energy (e.g. <a href="#BOOKMARK299">ATP</a>) to move against their gradient. It's the opposite of <a href="#BOOKMARK303">passive transport</a> -- energy is explicitly used to drive a <a href="#BOOKMARK18">molecule</a> to where it naturally / normally wouldn't go. An example of <a href="#BOOKMARK307">active transport</a> is the "<a href="#BOOKMARK328">sodium potassium pump</a>" <a href="#BOOKMARK212">enzyme</a>: <a href="#BOOKMARK299">ATP</a> is used to force open/close the ends of the <a href="#BOOKMARK212">enzyme</a>, which allow sodium and potassium to be exchanged across the <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK77">membrane</a>.</p>
<p>Note that the <a href="#BOOKMARK307">active transport</a> in the example above is the opening/closing of the <a href="#BOOKMARK212">enzyme</a> ends, not the exchange of sodium and potassium. Energy (<a href="#BOOKMARK299">ATP</a>) is being used to shape-shift the <a href="#BOOKMARK212">enzyme</a> to open/close (<a href="#BOOKMARK307">active transport</a>) while the sodium and potassium are passively entering and exiting the gates (<a href="#BOOKMARK303">passive transport</a> via <a href="#BOOKMARK101">facilitated diffusion</a>).</p>
<a name="HEADREF68"></a><h2>Osmosis</h2>
<p><a name="BOOKMARK308"></a><strong>Osmosis</strong> is the <a href="#BOOKMARK303">passive transport</a> of <a href="#BOOKMARK66">solvent</a> <a href="#BOOKMARK18">molecule</a>s (typically <a href="#BOOKMARK59">water</a>), across a <a href="#BOOKMARK349">semipermeable</a> <a href="#BOOKMARK77">membrane</a>, from areas where <a href="#BOOKMARK70">solute</a>s are less concentrated to areas where <a href="#BOOKMARK70">solute</a>s are more concentrated.</p>
<p>For example, imagine you have a semipermeabl <a href="#BOOKMARK77">membrane</a> that allows <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>s (<a href="#BOOKMARK66">solvent</a>) to pass but not sodium (<a href="#BOOKMARK70">solute</a>). That <a href="#BOOKMARK77">membrane</a> is separating 2 regions, where the ...</p>
<ul>
<li>left region is a solution of 25% sodium (<a href="#BOOKMARK70">solute</a>) and 75% <a href="#BOOKMARK59">water</a> (<a href="#BOOKMARK66">solvent</a>).</li>
<li>right region is a solution of 75% sodium (<a href="#BOOKMARK70">solute</a>) and 25% <a href="#BOOKMARK59">water</a> (<a href="#BOOKMARK66">solvent</a>).</li>
</ul>
<p>There will be a net movement of some <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a>s from the left region (lower <a href="#BOOKMARK70">solute</a> concentration) to the right region (higher <a href="#BOOKMARK70">solute</a> concentration).</p>
<p><img src="/.datafile_9fe60340ab8f0787489cb03e5ef87853/data.svg" alt="By OpenStax - https://cnx.org/contents/FPtK1zmh@8.25:fEI3C8Ot@10/Preface, CC BY 4.0, https://commons.wikimedia.org/w/index.php?curid=30131189" title="Osmosis is the diffusion of water through a semipermeable membrane down its concentration gradient. If a membrane is permeable to water, though not to a solute, water will equalize its own concentration by diffusing to the side of lower water concentration (and thus the side of higher solute concentration). In the beaker on the left, the solution on the right side of the membrane is hypertonic."></p><div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Another diagram that may make more sense conceptually: https://commons.wikimedia.org/wiki/File:<a href="#BOOKMARK308">Osmosis</a>_diagram.svg</p>
</div>
<p>There are 2 reasons why <a href="#BOOKMARK308">osmosis</a> happens. The first is that the <a href="#BOOKMARK349">semipermeable</a> <a href="#BOOKMARK77">membrane</a> will only allow certain types of <a href="#BOOKMARK18">molecule</a>s to pass through. If the <a href="#BOOKMARK349">semipermeable</a> <a href="#BOOKMARK77">membrane</a> is gated by ...</p>
<ul>
<li>size, the <a href="#BOOKMARK70">solute</a> <a href="#BOOKMARK18">molecule</a>s may be too large to pass through.</li>
<li>charge, the <a href="#BOOKMARK70">solute</a> <a href="#BOOKMARK18">molecule</a>s may not have the required charge to pass through.</li>
</ul>
<p>The higher the concentration of <a href="#BOOKMARK70">solute</a> <a href="#BOOKMARK18">molecule</a>s, the less likely it is for the <a href="#BOOKMARK66">solvent</a> <a href="#BOOKMARK18">molecule</a>s to reach a pore in the <a href="#BOOKMARK349">semipermeable</a> <a href="#BOOKMARK77">membrane</a>. The side with the lower concentration of <a href="#BOOKMARK70">solute</a> <a href="#BOOKMARK18">molecule</a>s is more likely to have a <a href="#BOOKMARK66">solvent</a> <a href="#BOOKMARK18">molecule</a> reach a pore than the other way around.</p>
<p><img src="/.datafile_e924a9ba9ed7fb2a09dbadf9391800b6/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Semipermeable membrane with pores big enough for water but not sodium."></p><p>The second reason is that, depending on the charge of <a href="#BOOKMARK66">solvent</a> and charge of <a href="#BOOKMARK70">solute</a>, the <a href="#BOOKMARK66">solvent</a> may be attracted to the <a href="#BOOKMARK70">solute</a>. More <a href="#BOOKMARK70">solute</a> = more chance that a <a href="#BOOKMARK66">solvent</a> gets attracted to it instead of crossing a pore in the <a href="#BOOKMARK77">membrane</a>. For example, if the <a href="#BOOKMARK66">solvent</a> is <a href="#BOOKMARK59">water</a> and the <a href="#BOOKMARK70">solute</a> is sodium, the <a href="#BOOKMARK62">weakly negative</a> charge of the oxygen <a href="#BOOKMARK3">atom</a> in a <a href="#BOOKMARK59">water</a> <a href="#BOOKMARK18">molecule</a> may get attracted to the <a href="#BOOKMARK27">positive charge</a> of the sodium <a href="#BOOKMARK26">ion</a>.</p>
<p><img src="/.datafile_49cb89a0d183ba53bce669fe661307de/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Semipermeable membrane but solvent attracted to solute."></p><p><a name="BOOKMARK309"></a><strong>Tonicity</strong> is the amount of pressure applied to a <a href="#BOOKMARK349">semipermeable</a> <a href="#BOOKMARK77">membrane</a> due to <a href="#BOOKMARK308">osmosis</a>. In other words, it's the amount of <a href="#BOOKMARK59">water</a> that flows in or out of a <a href="#BOOKMARK74">cell</a> due to the type of solution it's put in. A ...</p>
<ul>
<li><a name="BOOKMARK310"></a><strong>hypertonic</strong> solution is one that has more <a href="#BOOKMARK70">solute</a>s than the <a href="#BOOKMARK74">cell</a>, meaning the inside of the <a href="#BOOKMARK74">cell</a> loses <a href="#BOOKMARK59">water</a> (<a href="#BOOKMARK74">cell</a> shrivels).</li>
<li><a name="BOOKMARK311"></a><strong>hypotonic</strong> solution is one that has less <a href="#BOOKMARK70">solute</a>s than the <a href="#BOOKMARK74">cell</a>, meaning the inside of the <a href="#BOOKMARK74">cell</a> gains <a href="#BOOKMARK59">water</a> (<a href="#BOOKMARK74">cell</a> swells).</li>
<li><a name="BOOKMARK312"></a><strong>isotonic</strong> solution is one that has a roughly equal amount of <a href="#BOOKMARK70">solute</a>s, meaning the <a href="#BOOKMARK74">cell</a>'s <a href="#BOOKMARK59">water</a> content remains the same.</li>
</ul>
<p><img src="/.datafile_7a1118b9f947f899a002624c15f625e2/data.svg" alt="This work by Kasra Faghihi is licensed under a Creative Commons Attribution 4.0 International License." title="Different types of tonicity."></p><div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The prefix is referring to the amount of <a href="#BOOKMARK70">solute</a> in the solution relative to the <a href="#BOOKMARK74">cell</a>. Hyper = more. Hypo = less. Iso = same.</p>
</div>
<p>The following is a <a href="#BOOKMARK331">micrograph</a> of the red blood <a href="#BOOKMARK74">cell</a>s in solutions of different <a href="#BOOKMARK309">tonicity</a>. Notice how they shrivel in a <a href="#BOOKMARK310">hypertonic</a> solution (lose <a href="#BOOKMARK59">water</a>) and expand in a <a href="#BOOKMARK311">hypotonic</a> solution (gain <a href="#BOOKMARK59">water</a>).</p>
<p><img src="/.datafile_00f295e405d0fef1cf9562aee48d0265/data.svg" alt="By Zephyris - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=18401754" title="Human erythrocytes (red blood cells) viewed by phase contrast light microscopy. Three conditions are shown: hypertonic conditions (where the erythrocytes contract and appear &quot;spiky&quot;), isotonic conditions (where the erythrocytes appear normal) and hypotonic conditions (where the etrythrocytes expand and become more round)."></p><a name="HEADREF69"></a><h2>Photosynthesis</h2>
<p><a name="BOOKMARK313"></a><strong>Photosynthesis</strong> is the process by which certain organisms convert light energy (photons) to <a href="#BOOKMARK0">chemical</a> energy (<a href="#BOOKMARK51">sugar</a>s). These organisms are called <a name="BOOKMARK314"></a><strong>Photoautotroph</strong>s, and they include ...</p>
<ul>
<li>plants and algae (<a href="#BOOKMARK109">eukaryotic</a>), which perform <a href="#BOOKMARK313">photosynthesis</a> via their <a href="#BOOKMARK123">chloroplast</a>.</li>
<li><a href="#BOOKMARK130">cyanobacterium</a>, which are bacteria that can generate their own food.</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Another way to think of <a href="#BOOKMARK313">photosynthesis</a> is that it uses light energy (<a href="#BOOKMARK313">PHOTOsynthesis</a>) to synthesize (<a href="#BOOKMARK313">photoSYNTHESIS</a>) <a href="#BOOKMARK51">sugar</a>s.</p>
</div>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p><a href="#BOOKMARK123">Chloroplast</a> and <a href="#BOOKMARK130">cyanobacterium</a> share a similar structure. It's speculated that they have the same parent organism: that parent formed an <a href="#BOOKMARK345">endosymbiotic</a> relationship with a larger <a href="#BOOKMARK74">cell</a> and eventually became the <a href="#BOOKMARK123">chloroplast</a> <a href="#BOOKMARK111">organelle</a>.</p>
</div>
<p>The overall <a href="#BOOKMARK0">chemical</a> reaction for this is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>6</mn><mi>C</mi><msub><mi>O</mi><mn>2</mn></msub><mo>+</mo><mn>6</mn><msub><mi>H</mi><mn>2</mn></msub><mi>O</mi><mo>+</mo><mi>e</mi><mi>n</mi><mi>e</mi><mi>r</mi><mi>g</mi><mi>y</mi><mo>→</mo><msub><mi>C</mi><mn>6</mn></msub><msub><mi>H</mi><mn>12</mn></msub><msub><mi>O</mi><mn>6</mn></msub><mo>+</mo><mn>6</mn><msub><mi>O</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">6CO_2 + 6H_2O + energy \to C_6H_{12}O_6 + 6O_2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.19444em;"></span><span class="mord mathdefault">e</span><span class="mord mathdefault">n</span><span class="mord mathdefault">e</span><span class="mord mathdefault" style="margin-right:0.02778em;">r</span><span class="mord mathdefault" style="margin-right:0.03588em;">g</span><span class="mord mathdefault" style="margin-right:0.03588em;">y</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">→</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.07153em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>. Carbon dioxide gas (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><msub><mi>O</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">CO_2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>) bonds with <a href="#BOOKMARK59">water</a> (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>2</mn></msub><mi>O</mi></mrow><annotation encoding="application/x-tex">H_2O</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathdefault" style="margin-right:0.02778em;">O</span></span></span></span>) using energy from the sun (photons), creating glucose (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>C</mi><mn>6</mn></msub><msub><mi>H</mi><mn>12</mn></msub><msub><mi>O</mi><mn>6</mn></msub></mrow><annotation encoding="application/x-tex">C_6H_{12}O_6</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.07153em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>).</p>
<p>This reaction happens in 2 steps:</p>
<ol>
<li>
<p><a name="BOOKMARK315"></a><strong>Light-dependent reaction</strong>s: Energy <a href="#BOOKMARK18">molecule</a>s are created from <a href="#BOOKMARK59">water</a> and photos, with oxygen being a byproduct.</p>
<p>This occurs in a <a href="#BOOKMARK126">thylakoid</a> <a href="#BOOKMARK77">membrane</a>.</p>
<p><img src="/.datafile_5df9f7aa48e3ada56c37f1b2282f9c59/data.svg" alt="Ultrastructure of a chloroplast." title="By Kelvinsong - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=26147364"></p></li>
<li>
<p><a name="BOOKMARK316"></a><strong>Calvin cycle</strong>: It's a cyclical process that requires multiple iterations (3?) to produce a single glucose <a href="#BOOKMARK18">molecule</a>. Each cycle, <a href="#BOOKMARK299">ATP</a> and NADPH are used for energy (producing <a href="#BOOKMARK301">ADP</a> and NADP+ respectively), while the carbon dioxide (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><msub><mi>O</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">CO_2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>) is used as a source of carbons for the resulting glucose.</p>
<p>This occurs in the <a href="#BOOKMARK125">stroma</a>.</p>
<p><img src="/.datafile_bc7d0036db24b6aeaaeed50a84c9e75f/data.svg" alt="Ultrastructure of a chloroplast." title="By Kelvinsong - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=26147364"></p></li>
</ol>
<p>The following workflow diagram provides a ultra-simple high-level overview of the processes that take place. Note that this doesn't specify how many of each <a href="#BOOKMARK18">molecule</a> get input / output, nor does it provide a complete set of a input / output <a href="#BOOKMARK18">molecule</a>s for each reaction.</p>
<p><img src="/.datafile_83a73c5e549372c452be2d9a0a674b6a/diagram.svg" alt="Graphviz Dot Diagram"></p><a name="HEADREF70"></a><h2>Cellular Respiration</h2>
<p><a name="BOOKMARK317"></a><strong>Cellular respiration</strong> is the process by which certain organisms convert glucose (<a href="#BOOKMARK51">sugar</a>) to energy. These organisms include ...</p>
<ul>
<li><a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a>s, which perform <a href="#BOOKMARK317">cellular respiration</a> via their <a href="#BOOKMARK117">mitochondria</a>.</li>
<li>bacteria which share their ancestry with <a href="#BOOKMARK117">mitochondria</a>.</li>
</ul>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Remember that all <a href="#BOOKMARK109">eukaryote</a>s have <a href="#BOOKMARK117">mitochondria</a> -- both plant and animal <a href="#BOOKMARK74">cell</a>s. Unsure if all bacteria can perform <a href="#BOOKMARK317">cellular respiration</a>?</p>
</div>
<p>The overall <a href="#BOOKMARK0">chemical</a> reaction for this is <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>C</mi><mn>6</mn></msub><msub><mi>H</mi><mn>12</mn></msub><msub><mi>O</mi><mn>6</mn></msub><mo>+</mo><mn>6</mn><msub><mi>O</mi><mn>2</mn></msub><mo>→</mo><mn>6</mn><mi>C</mi><msub><mi>O</mi><mn>2</mn></msub><mo>+</mo><mn>6</mn><msub><mi>H</mi><mn>2</mn></msub><mi>O</mi><mo>+</mo><mi>e</mi><mi>n</mi><mi>e</mi><mi>r</mi><mi>g</mi><mi>y</mi></mrow><annotation encoding="application/x-tex">C_6H_{12}O_6 + 6O_2 \to 6CO_2 + 6H_2O + energy</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.07153em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">→</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord">6</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.19444em;"></span><span class="mord mathdefault">e</span><span class="mord mathdefault">n</span><span class="mord mathdefault">e</span><span class="mord mathdefault" style="margin-right:0.02778em;">r</span><span class="mord mathdefault" style="margin-right:0.03588em;">g</span><span class="mord mathdefault" style="margin-right:0.03588em;">y</span></span></span></span>. Glucose (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>C</mi><mn>6</mn></msub><msub><mi>H</mi><mn>12</mn></msub><msub><mi>O</mi><mn>6</mn></msub></mrow><annotation encoding="application/x-tex">C_6H_{12}O_6</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.07153em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">2</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">6</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>) and oxygen (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>O</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">O_2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>) break down into carbon dioxide gas (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><msub><mi>O</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">CO_2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord mathdefault" style="margin-right:0.07153em;">C</span><span class="mord"><span class="mord mathdefault" style="margin-right:0.02778em;">O</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.02778em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span>), <a href="#BOOKMARK59">water</a> (<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>2</mn></msub><mi>O</mi></mrow><annotation encoding="application/x-tex">H_2O</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.83333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathdefault" style="margin-right:0.08125em;">H</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.30110799999999993em;"><span style="top:-2.5500000000000003em;margin-left:-0.08125em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathdefault" style="margin-right:0.02778em;">O</span></span></span></span>), and energy (roughly 38 <a href="#BOOKMARK299">ATP</a> <a href="#BOOKMARK18">molecule</a>s and some heat).</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The number of <a href="#BOOKMARK299">ATP</a> actually generated is variable and dependent on many factors, but 38 is the generally agreed upon number.</p>
</div>
<p>The reaction happens in 3 steps:</p>
<ol>
<li>
<p><a name="BOOKMARK318"></a><strong>Glycolysis</strong>: The carbon backbone of the glucose <a href="#BOOKMARK18">molecule</a> is split, creating 2 <a name="BOOKMARK319"></a><strong>Pyruvate</strong> <a href="#BOOKMARK18">molecule</a>s along with <a href="#BOOKMARK59">water</a> and several other <a href="#BOOKMARK18">molecule</a>s. This is an <a href="#BOOKMARK353">anaerobic</a> process (no oxygen needed) that nets 2 <a href="#BOOKMARK299">ATP</a>. This happens in the <a href="#BOOKMARK75">cytoplasm</a> of <a href="#BOOKMARK74">cell</a>s.</p>
<p><img src="/.datafile_244513ad5578c85145249d5c965f8c85/data.svg" alt="Eukaryote with cell ribosomes highlighted" title="By No machine-readable author provided. Chb assumed (based on copyright claims). - No  machine-readable source provided. Own work assumed (based on copyright claims)., Public Domain, https://commons.wikimedia.org/w/index.php?curid=688296"></p></li>
<li>
<p><a name="BOOKMARK320"></a><strong>Krebs cycle</strong>: The <a href="#BOOKMARK319">pyruvate</a> <a href="#BOOKMARK18">molecule</a>s get further sliced and diced with other <a href="#BOOKMARK18">molecule</a>s.  This is an <a href="#BOOKMARK352">aerobic</a> process (requires oxygen) that nets 2 <a href="#BOOKMARK299">ATP</a>. This happens in the <a href="#BOOKMARK119">matrix</a> of the <a href="#BOOKMARK117">mitochondria</a>.</p>
<p><img src="/.datafile_76f2c0d4c2de2f4a2033a022646f1dac/data.svg" alt="Mitochondria" title="By Mariana Ruiz Villarreal LadyofHats - the diagram i made myself using adobe illustrator. as a source for the information i used the diagrams found here:[1], [2], [3], [4], [5], [6] and [7]., Public Domain, https://commons.wikimedia.org/w/index.php?curid=8152599"></p></li>
<li>
<p><a name="BOOKMARK321"></a><strong>Oxidative phosphorylation</strong>: Produces around 34 <a href="#BOOKMARK299">ATP</a>. This is an <a href="#BOOKMARK352">aerobic</a> process (requires oxygen) that nets roughly 34 <a href="#BOOKMARK299">ATP</a> (bulk of conversions). This happens in the <a name="BOOKMARK322"></a><strong>electron transport chain</strong> section of the <a href="#BOOKMARK117">mitochondria</a> (inner <a href="#BOOKMARK77">membrane</a>).</p>
<p><img src="/.datafile_226de1850916c6ac2bdfc073a2ec172e/data.svg" alt="Mitochondria" title="By Mariana Ruiz Villarreal LadyofHats - the diagram i made myself using adobe illustrator. as a source for the information i used the diagrams found here:[1], [2], [3], [4], [5], [6] and [7]., Public Domain, https://commons.wikimedia.org/w/index.php?curid=8152599"></p></li>
</ol>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>The material says that, technically step 3 doesn't have to happen after step 2 but it usually does.</p>
</div>
<p>Because the <a href="#BOOKMARK320">Krebs cycle</a> and the <a href="#BOOKMARK321">oxidative phosphorylation</a> are <a href="#BOOKMARK352">aerobic</a> processes (require oxygen), if no oxygen is present the output of <a href="#BOOKMARK318">glycolysis</a> goes through a process called <a name="BOOKMARK323"></a><strong>fermentation</strong>. <a href="#BOOKMARK323">Ferment</a>ation is an <a href="#BOOKMARK353">anaerobic</a> process (no oxygen required). Depending on the organism, it'll end up producing either...</p>
<ul>
<li><a name="BOOKMARK324"></a><strong>alcohol</strong> via <a name="BOOKMARK325"></a><strong>alcohol fermentation</strong> (e.g. yeast).</li>
<li><a name="BOOKMARK326"></a><strong>lactic acid</strong> via <a name="BOOKMARK327"></a><strong>lactic acid fermentation</strong> (e.g. humans and mammals).</li>
</ul>
<p><a href="#BOOKMARK323">Ferment</a>ation does produce <a href="#BOOKMARK299">ATP</a>, but much less so than Kerbs cycle + <a href="#BOOKMARK321">oxidative phosphorylation</a>.</p>
<p>For example, if a human is vigorously running, that human may not have enough oxygen available to trigger the <a href="#BOOKMARK320">Krebs cycle</a> / <a href="#BOOKMARK321">oxidative phosphorylation</a> (steps 2 and 3). As such, the output from <a href="#BOOKMARK318">glycolysis</a> (step 1) will end up going through <a href="#BOOKMARK327">lactic acid fermentation</a> instead.</p>
<p>The following workflow diagram provides a ultra-simple high-level overview of the processes that take place. Note that this doesn't specify how many of each <a href="#BOOKMARK18">molecule</a> get input / output, nor does it provide a complete set of a input / output <a href="#BOOKMARK18">molecule</a>s for each reaction.</p>
<p><img src="/.datafile_2333bcb32f80bfd59f5b494cce7d8ad6/diagram.svg" alt="Graphviz Dot Diagram"></p><a name="HEADREF71"></a><h2>Sodium Potassium Pump</h2>
<p>The <a name="BOOKMARK328"></a><strong>sodium potassium pump</strong> is an <a href="#BOOKMARK96">transmembrane enzyme</a> that allows the exchange of sodium <a href="#BOOKMARK26">ions</a> and potassium <a href="#BOOKMARK26">ions</a> across the <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK77">membrane</a> by opening and closing its ends.</p>
<p><img src="/.datafile_a158f71086caaadf3bf5053437571adc/data.svg" alt="By LadyofHats Mariana Ruiz Villarreal - Own work. Image renamed from Image:Sodium-Potassium_pump.svg, Public Domain, https://commons.wikimedia.org/w/index.php?curid=3981038" title="Example of primary active transport, where energy from hydrolysis of ATP is directly coupled to the movement of a specific substance across a membrane independent of any other species."></p><p>Only 1 end of the <a href="#BOOKMARK212">enzyme</a> is open at a time. When the...</p>
<ul>
<li><a href="#BOOKMARK350">intracellular</a> end is open, the <a href="#BOOKMARK212">enzyme</a> has an affinity for 3 sodium <a href="#BOOKMARK26">ions</a> followed by an affinity for <a href="#BOOKMARK299">ATP</a>.</li>
<li><a href="#BOOKMARK351">extracellular</a> end is open, the <a href="#BOOKMARK212">enzyme</a> has an affinity for 2 potassium <a href="#BOOKMARK26">ions</a>.</li>
</ul>
<p>Since both potassium and sodium have a <a href="#BOOKMARK27">positive charge</a> and an an unequal number are being exchanged each cycle (3 sodium out vs 2 potassium in), the <a href="#BOOKMARK350">intracellular</a> space will be more positive than the <a href="#BOOKMARK351">extracellular</a> space.</p>
<p>This charge difference is further reinforced by <a href="#BOOKMARK77">membrane</a> <a href="#BOOKMARK97">channel protein</a>s which allow potassium <a href="#BOOKMARK26">ions</a> to flow across the <a href="#BOOKMARK77">membrane</a> (potassium channels). Since there's a higher concentration of potassium <a href="#BOOKMARK26">ions</a> inside the <a href="#BOOKMARK74">cell</a>, those potassium <a href="#BOOKMARK26">ions</a> have a higher chance of flowing through the channel to the outside. Some percentage may be impeded by the slightly more <a href="#BOOKMARK27">positive charge</a> on the outside, but overall more will make it to the outside than stay on the inside.</p>
<p>This charge difference is referred to as the <a name="BOOKMARK329"></a><strong>resting membrane potential</strong> for a <a href="#BOOKMARK74">cell</a>.</p>
<a name="HEADREF72"></a><h2>Microscopy</h2>
<p><a name="BOOKMARK330"></a><strong>Microscope</strong>s are devices used to magnify (zoom in) on objects, such that you can see things that you normally would be too small to see on your own. The term <a href="#BOOKMARK330">microscope</a> comes from the words...</p>
<ul>
<li>micro, meaning small.</li>
<li>scope, meaning to look at.</li>
</ul>
<p>A picture taken through a <a href="#BOOKMARK330">microscope</a> is called a <a name="BOOKMARK331"></a><strong>micrograph</strong>. The distinguishing factors for most <a href="#BOOKMARK330">microscope</a>s or the amount of magnification and the resolution of the output image.</p>
<p>There are different types of <a href="#BOOKMARK330">microscope</a>s:</p>
<ul>
<li><a name="BOOKMARK332"></a><strong>Simple microscope</strong>s have just one lense (e.g. magnifying glass).</li>
<li><a name="BOOKMARK333"></a><strong>Compound microscope</strong>s have 2 ore more lenses (e.g. traditional <a href="#BOOKMARK330">microscope</a>s used in labs).
<ul>
<li><a name="BOOKMARK334"></a><strong>Light microscope</strong>s / <a name="BOOKMARK335"></a><strong>Brightfield microscope</strong>s work by shining a light through the specimen (which can be alive) and into the lens system that eventually reaches your eye. They're the type of <a href="#BOOKMARK330">microscope</a>s typically used in student labs.
<ul>
<li><a name="BOOKMARK336"></a><strong>Fluorescence microscope</strong>s are <a href="#BOOKMARK330">microscope</a>s that require the specimen being viewed is treated in some way to make it glow, allowing you to highlight parts of the specimen.</li>
<li><a name="BOOKMARK337"></a><strong>Confocal microscope</strong>s are a special form of <a href="#BOOKMARK336">fluorescence microscope</a>s that uses a laser to excite a sample and get it to glow fluorescent.</li>
</ul>
</li>
</ul>
</li>
<li><a name="BOOKMARK338"></a><strong>Electron microscope</strong>s use beams of <a href="#BOOKMARK8">electron</a>s rather than beams of light. They produce very magnified very high resolutions images. But, unlike <a href="#BOOKMARK334">light microscope</a>s, the specimens being observed are dead because <a href="#BOOKMARK338">electron microscope</a>s operate in a vacuum.
<ul>
<li><a name="BOOKMARK339"></a><strong>scanning electron microscope</strong>s (SEM) work by scanning a beam of <a href="#BOOKMARK8">electron</a>s over the surface of the specimen to produce an image of the 3D surface.</li>
<li><a name="BOOKMARK340"></a><strong>transmission electron microscope</strong>s (TEM) work by slicing up the specimen into ultra thin slices and passing an <a href="#BOOKMARK8">electron</a> bean through those slices to produce an image of the insides.</li>
</ul>
</li>
</ul>
<a name="HEADREF73"></a><h1>Other Terminology</h1>
<p>Terminology that's relevant but doesn't fit in any other section goes here.</p>
<p>
<a name="BOOKMARK342"></a><strong>Density</strong> - The mass per unit volume of a substance.</p>
<p><a name="BOOKMARK343"></a><strong>Specific heat capacity</strong> - The amount of heat needed to raise the temperature of one gram of a substance by one degree Celsius.</p>
<p><a name="BOOKMARK344"></a><strong>Heat of vaporization</strong> - The amount of energy needed to change one gram of a liquid substance to a gas at constant temperature.</p>
<p><a name="BOOKMARK345"></a><strong>Endosymbiosis</strong> - A form of symbiosis where one organism lives inside of of the other (e.g. gut bacteria lives in our colon). The prefix <em>endo</em> means <em>within</em>.</p>
<p><a name="BOOKMARK346"></a><strong>Diffusion</strong> - A physical process where <a href="#BOOKMARK18">molecule</a>s of a material move from an area of high concentration (where there are many <a href="#BOOKMARK18">molecule</a>s) to an area of low concentration (where there are fewer <a href="#BOOKMARK18">molecule</a>s) until it has reached <a href="#BOOKMARK347">equilibrium</a> (<a href="#BOOKMARK18">molecule</a>s evenly spread). <a href="https://simple.wikipedia.org/wiki/Diffusion">See more</a>.</p>
<p><a name="BOOKMARK347"></a><strong>Equilibrium</strong> - A state in which opposing forces / influences are balanced. In the context of a <a href="#BOOKMARK304">concentration gradient</a>, it means the state at which a substance is equally distributed throughout the volume that it's in (roughly).</p>
<p><a name="BOOKMARK348"></a><strong>Permeability</strong> - The state or quality of a material or <a href="#BOOKMARK77">membrane</a> that causes it to allow liquids/gases to pass through it.</p>
<p><a name="BOOKMARK349"></a><strong>Semipermeable</strong> - The state or quality of a material or <a href="#BOOKMARK77">membrane</a> that causes it to allow certain types of <a href="#BOOKMARK18">molecule</a>s to pass through it.</p>
<p><a name="BOOKMARK350"></a><strong>Intracellular</strong> - The fluid inside of the <a href="#BOOKMARK74">cell</a>, which is technically on the inside of the <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK77">membrane</a> (<a href="#BOOKMARK75">cytoplasm</a>).</p>
<p><a name="BOOKMARK351"></a><strong>Extracellular</strong> - The fluid outside of the <a href="#BOOKMARK74">cell</a>.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>Technically unsure at which layer the <a href="#BOOKMARK351">extracellular</a> region begins. Is it outside of the <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK77">membrane</a>? <a href="#BOOKMARK78">cell wall</a>? <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK80">capsule</a>? I'm pretty sure any fluid outside of the <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK77">membrane</a> qualifies as <a href="#BOOKMARK351">extracellular</a>, while any fluid inside of the <a href="#BOOKMARK74">cell</a> <a href="#BOOKMARK77">membrane</a> qualifies as <a href="#BOOKMARK350">intracellular</a> (<a href="#BOOKMARK75">cytoplasm</a>).</p>
</div>
<p><a name="BOOKMARK352"></a><strong>Aerobic</strong> - A <a href="#BOOKMARK37">biological</a> process that requires oxygen.</p>
<p><a name="BOOKMARK353"></a><strong>Anaerobic</strong> - A <a href="#BOOKMARK37">biological</a> process that doesn't require oxygen.</p>
<p><a name="BOOKMARK354"></a><strong>Fission</strong> - The act of dividing or splitting something into two or more parts.</p>
<p><a name="BOOKMARK355"></a><strong>Tetrad</strong> - A group / set of 4.</p>
<p><a name="BOOKMARK356"></a><strong>Homologous</strong> - Having the same relation, relative position, or structure. Particularly in <a href="#BOOKMARK37">biology</a>, it s the existence of a shared ancestry between a pair of structures or <a href="#BOOKMARK244">genes</a>.</p>
<p><a name="BOOKMARK357"></a><strong>Karyotype</strong> - <a href="#BOOKMARK331">Micrograph</a> image of <a href="#BOOKMARK140">diploid</a> set of <a href="#BOOKMARK138">chromosome</a>s, grouped in pairs.</p>
<p><a name="BOOKMARK358"></a><strong>Asexual</strong> - When offspring is created using the genetic material from 1 parent. The offspring are essentially copies of the parent in terms of their genetic material (clone). Examples of <a href="#BOOKMARK358">asexual</a> reproduction include:</p>
<ul>
<li><a name="BOOKMARK359"></a><strong>Binary fission</strong> - Process of cloning a <a href="#BOOKMARK210">prokaryotic</a> <a href="#BOOKMARK74">cell</a> and certain <a href="#BOOKMARK111">organelle</a> inside of <a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a> (e.g. <a href="#BOOKMARK117">mitochondria</a>).</li>
<li><a href="#BOOKMARK161">Mitosis</a> - Process of cloning a <a href="#BOOKMARK109">eukaryotic</a> <a href="#BOOKMARK74">cell</a>.</li>
<li><a name="BOOKMARK360"></a><strong>Budding</strong> - When a growth on a plant or animal breaks off, that broken piece is the offspring.</li>
<li><a name="BOOKMARK361"></a><strong>Fragmentation</strong> - When an organism breaks into two or more pieces, each fragment grows back into a whole.</li>
<li><a name="BOOKMARK362"></a><strong>Parthenogenesis</strong> - When an animal lays an <a href="#BOOKMARK179">egg</a> but that <a href="#BOOKMARK179">egg</a> already has all the genetic information needed to develop (it doesn't need to be <a href="#BOOKMARK201">fertilize</a>d).</li>
</ul>
<p><a name="BOOKMARK363"></a><strong>Sexual</strong> - When offspring is created by fusing genetic material from 2 parents. The offspring has a mixture of genetic material from both parents. An example of <a href="#BOOKMARK363">sexual</a> reproduction is when a <a href="#BOOKMARK176">gamete</a> <a href="#BOOKMARK74">cell</a>s merge to create the offspring.</p>
<div style="margin: 2em; background-color: #e0e0e0">
<strong>
⚠️NOTE️️️⚠️</strong>
<p>In some cases, the genetic material being fused in <a href="#BOOKMARK363">sexual</a> reproduction may be from the same parent. Answer to a question on the site... it is still <a href="#BOOKMARK363">sexual</a>, because <a href="#BOOKMARK363">sexual</a> reproduction means fusion of male and female <a href="#BOOKMARK176">gamete</a>s, doesn't <a href="#BOOKMARK1">matter</a> if they're from the same plant.</p>
</div>
<p><a name="BOOKMARK364"></a><strong>Model system</strong> - A system with a reduced set of parameters/complexity that makes it easy for a researcher to investigate a particular scientific question. For example, <a href="#BOOKMARK262">Gregor Mendel</a> used pea pods to research the theory of <a href="#BOOKMARK254">classical genetics</a> / inheritance (simple, grows and matures quickly, inbreeding okay)</p>
<p>
</p>
<p>

</p>
<p></p>

        </body></html>